{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pandas.api.types import CategoricalDtype\n",
    "import polars as pl\n",
    "import seaborn as sns\n",
    "import glob\n",
    "import os\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from tqdm import tqdm\n",
    "import importlib\n",
    "\n",
    "\n",
    "pedigree_file_loc = 'resources/pedigrees/gatk_1kgp.ped'\n",
    "summary_statistics_folder='intermediate_data'\n",
    "\n",
    "if os.getcwd()!='phasing_T2T_project':\n",
    "    os.chdir('..') # Change to the parent directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['intermediate_data/rolling_stats_500k_window.parquet',\n",
       " 'intermediate_data/flips_and_switches.parquet',\n",
       " 'intermediate_data/samples.parquet',\n",
       " 'intermediate_data/compressed_ideogram_100000_window.parquet',\n",
       " 'intermediate_data/variants.parquet',\n",
       " 'intermediate_data/compressed_ideogram_1000000_window.parquet',\n",
       " 'intermediate_data/per_genome_cnv_regions_with_slop.parquet',\n",
       " 'intermediate_data/chroms.parquet',\n",
       " 'intermediate_data/variant_frequency_stats',\n",
       " 'intermediate_data/switch_errors.parquet',\n",
       " 'intermediate_data/MAF_performance_variants.parquet',\n",
       " 'intermediate_data/sample_genotype_concordance.parquet',\n",
       " 'intermediate_data/per_sample_imputation_performance.parquet',\n",
       " 'intermediate_data/rolling_stats_250k_window.parquet',\n",
       " 'intermediate_data/bcftools_query_variant_data.parquet',\n",
       " 'intermediate_data/per_genome_cnv_regions.parquet',\n",
       " 'intermediate_data/per_variant_category_imputation_performance.parquet',\n",
       " 'intermediate_data/per_MAF_bin.parquet',\n",
       " 'intermediate_data/methods.parquet',\n",
       " 'intermediate_data/filter_summary_stats.parquet',\n",
       " 'intermediate_data/frequency.parquet',\n",
       " 'intermediate_data/variants.nozeros.parquet',\n",
       " 'intermediate_data/ancestries.parquet',\n",
       " 'intermediate_data/block_length.parquet',\n",
       " 'intermediate_data/all_variant_data.parquet',\n",
       " 'intermediate_data/rolling_stats_1m_window.parquet',\n",
       " 'intermediate_data/binned_maf_data.parquet',\n",
       " 'intermediate_data/type.parquet',\n",
       " 'intermediate_data/per_cytoband_variant_data.parquet']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import glob; glob.glob(f\"{summary_statistics_folder}/*\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "type",
         "rawType": "category",
         "type": "unknown"
        },
        {
         "name": "rounded_MAF",
         "rawType": "category",
         "type": "unknown"
        },
        {
         "name": "genome",
         "rawType": "category",
         "type": "unknown"
        },
        {
         "name": "method_of_phasing",
         "rawType": "category",
         "type": "unknown"
        },
        {
         "name": "ground_truth_data_source",
         "rawType": "category",
         "type": "unknown"
        },
        {
         "name": "n_switch_errors",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "n_checked",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "n_gt_errors",
         "rawType": "uint32",
         "type": "integer"
        },
        {
         "name": "n_gt_checked",
         "rawType": "uint32",
         "type": "integer"
        },
        {
         "name": "MAC",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "AN",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "switch_error_rate",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "gt_error_rate",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "MAF",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "syntenic",
         "rawType": "object",
         "type": "string"
        }
       ],
       "ref": "7191938b-7341-4a16-8203-a7c938a1fa5d",
       "rows": [
        [
         "862",
         "SNPs + Indels",
         "singleton",
         "CHM13v2.0",
         "phased_with_parents_and_pedigree",
         "HGSVC_samples",
         "389.0",
         "1505.0",
         "3071",
         "142515",
         "2312",
         "14806048",
         "25.84717607973422",
         "2.154860891835947",
         "0.01561524047470331",
         "All"
        ],
        [
         "1032",
         "SNPs + Indels",
         "singleton",
         "CHM13v2.0",
         "phased_with_parents_and_pedigree",
         "HPRC_HGSVC_probands",
         "4.0",
         "109.0",
         "3581",
         "52836",
         "777",
         "4975908",
         "3.669724770642202",
         "6.777575895222954",
         "0.01561524047470331",
         "All"
        ],
        [
         "978",
         "SNPs + Indels",
         "singleton",
         "CHM13v2.0",
         "phased_with_parents_and_pedigree",
         "HGSVC_samples_nontrios_only",
         "369.0",
         "1123.0",
         "2261",
         "31863",
         "1677",
         "10739508",
         "32.85841495992876",
         "7.0960047704233755",
         "0.01561524047470331",
         "All"
        ],
        [
         "901",
         "SNPs + Indels",
         "singleton",
         "CHM13v2.0",
         "phased_with_parents_and_pedigree",
         "HGSVC_probands",
         "4.0",
         "109.0",
         "1607",
         "20646",
         "666",
         "4265064",
         "3.669724770642202",
         "7.783590041654557",
         "0.01561524047470331",
         "All"
        ],
        [
         "1101",
         "SNPs + Indels",
         "singleton",
         "CHM13v2.0",
         "phased_with_parents_and_pedigree",
         "HGSVC_parents",
         "20.0",
         "274.0",
         "1176",
         "10192",
         "784",
         "5020736",
         "7.2992700729927",
         "11.538461538461538",
         "0.01561524047470331",
         "All"
        ],
        [
         "893",
         "SNPs + Indels",
         "singleton",
         "CHM13v2.0",
         "phased_with_parents_and_pedigree",
         "HPRC_samples",
         null,
         null,
         "1046",
         "32658",
         "851",
         "5449804",
         null,
         "3.2028905628023763",
         "0.01561524047470331",
         "All"
        ],
        [
         "748",
         "SNPs + Indels",
         "singleton",
         "CHM13v2.0",
         "phased_with_parents_and_pedigree",
         "HPRC_HGSVC_all_samples",
         "393.0",
         "1506.0",
         "9450",
         "267100",
         "2671",
         "17105084",
         "26.095617529880478",
         "3.5380007487832272",
         "0.01561524047470331",
         "All"
        ]
       ],
       "shape": {
        "columns": 15,
        "rows": 7
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>rounded_MAF</th>\n",
       "      <th>genome</th>\n",
       "      <th>method_of_phasing</th>\n",
       "      <th>ground_truth_data_source</th>\n",
       "      <th>n_switch_errors</th>\n",
       "      <th>n_checked</th>\n",
       "      <th>n_gt_errors</th>\n",
       "      <th>n_gt_checked</th>\n",
       "      <th>MAC</th>\n",
       "      <th>AN</th>\n",
       "      <th>switch_error_rate</th>\n",
       "      <th>gt_error_rate</th>\n",
       "      <th>MAF</th>\n",
       "      <th>syntenic</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>862</th>\n",
       "      <td>SNPs + Indels</td>\n",
       "      <td>singleton</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HGSVC_samples</td>\n",
       "      <td>389.0</td>\n",
       "      <td>1505.0</td>\n",
       "      <td>3071</td>\n",
       "      <td>142515</td>\n",
       "      <td>2312</td>\n",
       "      <td>14806048</td>\n",
       "      <td>25.847176</td>\n",
       "      <td>2.154861</td>\n",
       "      <td>0.015615</td>\n",
       "      <td>All</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1032</th>\n",
       "      <td>SNPs + Indels</td>\n",
       "      <td>singleton</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HPRC_HGSVC_probands</td>\n",
       "      <td>4.0</td>\n",
       "      <td>109.0</td>\n",
       "      <td>3581</td>\n",
       "      <td>52836</td>\n",
       "      <td>777</td>\n",
       "      <td>4975908</td>\n",
       "      <td>3.669725</td>\n",
       "      <td>6.777576</td>\n",
       "      <td>0.015615</td>\n",
       "      <td>All</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>978</th>\n",
       "      <td>SNPs + Indels</td>\n",
       "      <td>singleton</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HGSVC_samples_nontrios_only</td>\n",
       "      <td>369.0</td>\n",
       "      <td>1123.0</td>\n",
       "      <td>2261</td>\n",
       "      <td>31863</td>\n",
       "      <td>1677</td>\n",
       "      <td>10739508</td>\n",
       "      <td>32.858415</td>\n",
       "      <td>7.096005</td>\n",
       "      <td>0.015615</td>\n",
       "      <td>All</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>901</th>\n",
       "      <td>SNPs + Indels</td>\n",
       "      <td>singleton</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HGSVC_probands</td>\n",
       "      <td>4.0</td>\n",
       "      <td>109.0</td>\n",
       "      <td>1607</td>\n",
       "      <td>20646</td>\n",
       "      <td>666</td>\n",
       "      <td>4265064</td>\n",
       "      <td>3.669725</td>\n",
       "      <td>7.783590</td>\n",
       "      <td>0.015615</td>\n",
       "      <td>All</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1101</th>\n",
       "      <td>SNPs + Indels</td>\n",
       "      <td>singleton</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HGSVC_parents</td>\n",
       "      <td>20.0</td>\n",
       "      <td>274.0</td>\n",
       "      <td>1176</td>\n",
       "      <td>10192</td>\n",
       "      <td>784</td>\n",
       "      <td>5020736</td>\n",
       "      <td>7.299270</td>\n",
       "      <td>11.538462</td>\n",
       "      <td>0.015615</td>\n",
       "      <td>All</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>893</th>\n",
       "      <td>SNPs + Indels</td>\n",
       "      <td>singleton</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HPRC_samples</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1046</td>\n",
       "      <td>32658</td>\n",
       "      <td>851</td>\n",
       "      <td>5449804</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.202891</td>\n",
       "      <td>0.015615</td>\n",
       "      <td>All</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>748</th>\n",
       "      <td>SNPs + Indels</td>\n",
       "      <td>singleton</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HPRC_HGSVC_all_samples</td>\n",
       "      <td>393.0</td>\n",
       "      <td>1506.0</td>\n",
       "      <td>9450</td>\n",
       "      <td>267100</td>\n",
       "      <td>2671</td>\n",
       "      <td>17105084</td>\n",
       "      <td>26.095618</td>\n",
       "      <td>3.538001</td>\n",
       "      <td>0.015615</td>\n",
       "      <td>All</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               type rounded_MAF     genome                 method_of_phasing  \\\n",
       "862   SNPs + Indels   singleton  CHM13v2.0  phased_with_parents_and_pedigree   \n",
       "1032  SNPs + Indels   singleton  CHM13v2.0  phased_with_parents_and_pedigree   \n",
       "978   SNPs + Indels   singleton  CHM13v2.0  phased_with_parents_and_pedigree   \n",
       "901   SNPs + Indels   singleton  CHM13v2.0  phased_with_parents_and_pedigree   \n",
       "1101  SNPs + Indels   singleton  CHM13v2.0  phased_with_parents_and_pedigree   \n",
       "893   SNPs + Indels   singleton  CHM13v2.0  phased_with_parents_and_pedigree   \n",
       "748   SNPs + Indels   singleton  CHM13v2.0  phased_with_parents_and_pedigree   \n",
       "\n",
       "         ground_truth_data_source  n_switch_errors  n_checked  n_gt_errors  \\\n",
       "862                 HGSVC_samples            389.0     1505.0         3071   \n",
       "1032          HPRC_HGSVC_probands              4.0      109.0         3581   \n",
       "978   HGSVC_samples_nontrios_only            369.0     1123.0         2261   \n",
       "901                HGSVC_probands              4.0      109.0         1607   \n",
       "1101                HGSVC_parents             20.0      274.0         1176   \n",
       "893                  HPRC_samples              NaN        NaN         1046   \n",
       "748        HPRC_HGSVC_all_samples            393.0     1506.0         9450   \n",
       "\n",
       "      n_gt_checked   MAC        AN  switch_error_rate  gt_error_rate  \\\n",
       "862         142515  2312  14806048          25.847176       2.154861   \n",
       "1032         52836   777   4975908           3.669725       6.777576   \n",
       "978          31863  1677  10739508          32.858415       7.096005   \n",
       "901          20646   666   4265064           3.669725       7.783590   \n",
       "1101         10192   784   5020736           7.299270      11.538462   \n",
       "893          32658   851   5449804                NaN       3.202891   \n",
       "748         267100  2671  17105084          26.095618       3.538001   \n",
       "\n",
       "           MAF syntenic  \n",
       "862   0.015615      All  \n",
       "1032  0.015615      All  \n",
       "978   0.015615      All  \n",
       "901   0.015615      All  \n",
       "1101  0.015615      All  \n",
       "893   0.015615      All  \n",
       "748   0.015615      All  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = pd.read_parquet('intermediate_data/per_MAF_bin.parquet')\n",
    "q = q[(q.rounded_MAF=='singleton')&(q.method_of_phasing=='phased_with_parents_and_pedigree')&(q.syntenic=='All')&(q.type=='SNPs + Indels')]\n",
    "# q.type=='SNPs + Indels'\n",
    "q"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "MAF_performance_variant_df = pl.read_parquet(f'{summary_statistics_folder}/MAF_performance_variants.parquet')\n",
    "all_chroms = pd.read_parquet(f'{summary_statistics_folder}/chroms.parquet')\n",
    "all_samples = pd.read_parquet(f'{summary_statistics_folder}/samples.parquet')\n",
    "all_ancestries = pd.read_parquet(f'{summary_statistics_folder}/ancestries.parquet')\n",
    "all_methods = pd.read_parquet(f'{summary_statistics_folder}/methods.parquet')\n",
    "all_variants = pl.read_parquet(f'{summary_statistics_folder}/variants.parquet')\n",
    "\n",
    "per_variant_category_imputation_performance = pd.read_parquet(f'{summary_statistics_folder}/per_variant_category_imputation_performance.parquet')\n",
    "per_sample_imputation_performance = pd.read_parquet(f'{summary_statistics_folder}/per_sample_imputation_performance.parquet')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Variant Filtering Stats \n",
    "Stats in paragraph after \"Full length, telomere to telomere maps of human genetic recombination rates are largely consistent with deCODE recombination rates\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Our VQSLOD filter resulted in the loss of an additional \n",
      "14494016 variants compared to the NYGCâ€™s GRCh38 1KGP variant filtering methods, but the inclusion of singleton alleles added \n",
      "111805 unique variants to our dataset (Supplemental Figure 1).\n"
     ]
    },
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "genome",
         "rawType": "category",
         "type": "unknown"
        },
        {
         "name": "Syntenic",
         "rawType": "bool",
         "type": "boolean"
        },
        {
         "name": "singleton",
         "rawType": "bool",
         "type": "boolean"
        },
        {
         "name": "VQSLOD_filter",
         "rawType": "bool",
         "type": "boolean"
        },
        {
         "name": "MERR_filter",
         "rawType": "bool",
         "type": "boolean"
        },
        {
         "name": "HWE_pop_filter",
         "rawType": "bool",
         "type": "boolean"
        },
        {
         "name": "MAC_filter",
         "rawType": "bool",
         "type": "boolean"
        },
        {
         "name": "AC_filter",
         "rawType": "bool",
         "type": "boolean"
        },
        {
         "name": "f_missing_filter",
         "rawType": "bool",
         "type": "boolean"
        },
        {
         "name": "var_len_filter",
         "rawType": "bool",
         "type": "boolean"
        },
        {
         "name": "alt_star_filter",
         "rawType": "bool",
         "type": "boolean"
        },
        {
         "name": "pass_filter",
         "rawType": "bool",
         "type": "boolean"
        },
        {
         "name": "CHM13_filtered",
         "rawType": "bool",
         "type": "boolean"
        },
        {
         "name": "GRCh38_filtered",
         "rawType": "bool",
         "type": "boolean"
        },
        {
         "name": "GRCh38_criteria_fail",
         "rawType": "bool",
         "type": "boolean"
        },
        {
         "name": "CHM13_criteria_fail",
         "rawType": "bool",
         "type": "boolean"
        },
        {
         "name": "len",
         "rawType": "uint64",
         "type": "integer"
        },
        {
         "name": "gatk_filter",
         "rawType": "bool",
         "type": "boolean"
        }
       ],
       "ref": "e05e7c9f-c13b-4a98-95ca-0d4f68e7f9da",
       "rows": [
        [
         "7",
         "CHM13v2.0",
         "False",
         "True",
         "False",
         "False",
         "False",
         "False",
         "True",
         "False",
         "False",
         "False",
         "False",
         "True",
         "True",
         "True",
         "False",
         "106488",
         "False"
        ],
        [
         "12",
         "CHM13v2.0",
         "True",
         "True",
         "False",
         "False",
         "False",
         "False",
         "True",
         "True",
         "True",
         "False",
         "False",
         "True",
         "True",
         "True",
         "True",
         "1360",
         "False"
        ],
        [
         "40",
         "CHM13v2.0",
         "False",
         "True",
         "True",
         "False",
         "False",
         "False",
         "True",
         "False",
         "False",
         "False",
         "False",
         "True",
         "True",
         "True",
         "True",
         "589443",
         "True"
        ],
        [
         "55",
         "CHM13v2.0",
         "True",
         "True",
         "False",
         "False",
         "False",
         "False",
         "False",
         "True",
         "False",
         "False",
         "False",
         "True",
         "True",
         "True",
         "True",
         "3",
         "False"
        ],
        [
         "90",
         "CHM13v2.0",
         "True",
         "True",
         "False",
         "False",
         "False",
         "False",
         "False",
         "False",
         "False",
         "True",
         "False",
         "True",
         "True",
         "True",
         "True",
         "1",
         "False"
        ],
        [
         "95",
         "CHM13v2.0",
         "True",
         "True",
         "True",
         "False",
         "False",
         "False",
         "False",
         "False",
         "False",
         "True",
         "False",
         "True",
         "True",
         "True",
         "True",
         "12",
         "True"
        ],
        [
         "104",
         "CHM13v2.0",
         "True",
         "True",
         "True",
         "True",
         "False",
         "False",
         "True",
         "True",
         "False",
         "True",
         "False",
         "True",
         "True",
         "True",
         "True",
         "2",
         "True"
        ],
        [
         "109",
         "CHM13v2.0",
         "True",
         "True",
         "False",
         "False",
         "False",
         "False",
         "True",
         "False",
         "False",
         "True",
         "False",
         "True",
         "True",
         "True",
         "True",
         "441729",
         "False"
        ],
        [
         "113",
         "CHM13v2.0",
         "True",
         "True",
         "False",
         "False",
         "False",
         "False",
         "True",
         "True",
         "False",
         "False",
         "False",
         "True",
         "True",
         "True",
         "True",
         "31253",
         "False"
        ],
        [
         "122",
         "CHM13v2.0",
         "True",
         "True",
         "True",
         "False",
         "False",
         "False",
         "True",
         "True",
         "False",
         "True",
         "False",
         "True",
         "True",
         "True",
         "True",
         "10378",
         "True"
        ],
        [
         "124",
         "CHM13v2.0",
         "True",
         "True",
         "False",
         "False",
         "False",
         "False",
         "True",
         "False",
         "True",
         "True",
         "False",
         "True",
         "True",
         "True",
         "True",
         "2318",
         "False"
        ],
        [
         "137",
         "CHM13v2.0",
         "True",
         "True",
         "True",
         "False",
         "False",
         "False",
         "True",
         "True",
         "True",
         "True",
         "False",
         "True",
         "True",
         "True",
         "True",
         "535",
         "True"
        ],
        [
         "166",
         "CHM13v2.0",
         "False",
         "True",
         "False",
         "False",
         "False",
         "False",
         "True",
         "True",
         "False",
         "False",
         "False",
         "True",
         "True",
         "True",
         "True",
         "15800",
         "False"
        ],
        [
         "167",
         "CHM13v2.0",
         "True",
         "True",
         "True",
         "False",
         "False",
         "False",
         "False",
         "False",
         "False",
         "False",
         "False",
         "True",
         "True",
         "False",
         "True",
         "533",
         "True"
        ],
        [
         "180",
         "CHM13v2.0",
         "True",
         "True",
         "False",
         "False",
         "False",
         "False",
         "True",
         "False",
         "False",
         "False",
         "False",
         "False",
         "True",
         "True",
         "False",
         "111601",
         "False"
        ],
        [
         "181",
         "CHM13v2.0",
         "False",
         "True",
         "True",
         "False",
         "False",
         "False",
         "True",
         "True",
         "False",
         "False",
         "False",
         "True",
         "True",
         "True",
         "True",
         "429834",
         "True"
        ],
        [
         "205",
         "CHM13v2.0",
         "False",
         "True",
         "True",
         "False",
         "False",
         "False",
         "False",
         "False",
         "False",
         "False",
         "False",
         "True",
         "True",
         "False",
         "True",
         "11",
         "True"
        ],
        [
         "209",
         "CHM13v2.0",
         "False",
         "True",
         "False",
         "False",
         "False",
         "False",
         "True",
         "False",
         "True",
         "False",
         "False",
         "True",
         "True",
         "True",
         "True",
         "1097",
         "False"
        ],
        [
         "219",
         "CHM13v2.0",
         "False",
         "True",
         "True",
         "False",
         "False",
         "False",
         "True",
         "True",
         "False",
         "True",
         "False",
         "True",
         "True",
         "True",
         "True",
         "47924",
         "True"
        ],
        [
         "224",
         "CHM13v2.0",
         "True",
         "True",
         "True",
         "False",
         "False",
         "False",
         "True",
         "False",
         "True",
         "True",
         "False",
         "True",
         "True",
         "True",
         "True",
         "4862",
         "True"
        ],
        [
         "235",
         "CHM13v2.0",
         "True",
         "True",
         "True",
         "False",
         "False",
         "False",
         "True",
         "True",
         "False",
         "False",
         "False",
         "True",
         "True",
         "True",
         "True",
         "210022",
         "True"
        ],
        [
         "238",
         "CHM13v2.0",
         "True",
         "True",
         "False",
         "False",
         "False",
         "False",
         "True",
         "False",
         "False",
         "False",
         "False",
         "False",
         "False",
         "True",
         "False",
         "122",
         "False"
        ],
        [
         "261",
         "CHM13v2.0",
         "True",
         "True",
         "True",
         "False",
         "False",
         "False",
         "True",
         "False",
         "False",
         "True",
         "False",
         "True",
         "True",
         "True",
         "True",
         "203660",
         "True"
        ],
        [
         "266",
         "CHM13v2.0",
         "True",
         "True",
         "True",
         "False",
         "False",
         "False",
         "True",
         "False",
         "False",
         "False",
         "False",
         "True",
         "True",
         "True",
         "True",
         "7816298",
         "True"
        ],
        [
         "269",
         "CHM13v2.0",
         "True",
         "True",
         "False",
         "False",
         "False",
         "False",
         "True",
         "True",
         "True",
         "True",
         "False",
         "True",
         "True",
         "True",
         "True",
         "44",
         "False"
        ],
        [
         "276",
         "CHM13v2.0",
         "True",
         "True",
         "False",
         "False",
         "False",
         "False",
         "False",
         "False",
         "False",
         "False",
         "False",
         "False",
         "True",
         "False",
         "False",
         "3",
         "False"
        ],
        [
         "289",
         "CHM13v2.0",
         "True",
         "True",
         "True",
         "False",
         "False",
         "False",
         "False",
         "True",
         "False",
         "False",
         "False",
         "True",
         "True",
         "True",
         "True",
         "16",
         "True"
        ],
        [
         "292",
         "CHM13v2.0",
         "False",
         "True",
         "True",
         "False",
         "False",
         "False",
         "True",
         "True",
         "True",
         "False",
         "False",
         "True",
         "True",
         "True",
         "True",
         "11841",
         "True"
        ],
        [
         "301",
         "CHM13v2.0",
         "False",
         "True",
         "False",
         "False",
         "False",
         "False",
         "True",
         "False",
         "True",
         "True",
         "False",
         "True",
         "True",
         "True",
         "True",
         "532",
         "False"
        ],
        [
         "311",
         "CHM13v2.0",
         "True",
         "True",
         "True",
         "False",
         "False",
         "False",
         "True",
         "False",
         "True",
         "False",
         "False",
         "True",
         "True",
         "True",
         "True",
         "68335",
         "True"
        ],
        [
         "315",
         "CHM13v2.0",
         "True",
         "True",
         "False",
         "False",
         "False",
         "False",
         "True",
         "True",
         "False",
         "True",
         "False",
         "True",
         "True",
         "True",
         "True",
         "1133",
         "False"
        ],
        [
         "326",
         "CHM13v2.0",
         "True",
         "True",
         "False",
         "False",
         "False",
         "False",
         "False",
         "False",
         "False",
         "False",
         "False",
         "True",
         "True",
         "False",
         "False",
         "1545",
         "False"
        ],
        [
         "328",
         "CHM13v2.0",
         "False",
         "True",
         "True",
         "False",
         "False",
         "False",
         "True",
         "False",
         "True",
         "True",
         "False",
         "True",
         "True",
         "True",
         "True",
         "2747",
         "True"
        ],
        [
         "343",
         "CHM13v2.0",
         "True",
         "True",
         "False",
         "False",
         "False",
         "False",
         "True",
         "False",
         "True",
         "False",
         "False",
         "True",
         "True",
         "True",
         "True",
         "14518",
         "False"
        ],
        [
         "355",
         "CHM13v2.0",
         "False",
         "True",
         "False",
         "False",
         "False",
         "False",
         "True",
         "False",
         "False",
         "False",
         "False",
         "False",
         "True",
         "True",
         "False",
         "82",
         "False"
        ],
        [
         "382",
         "CHM13v2.0",
         "False",
         "True",
         "False",
         "False",
         "False",
         "False",
         "True",
         "True",
         "False",
         "True",
         "False",
         "True",
         "True",
         "True",
         "True",
         "1245",
         "False"
        ],
        [
         "393",
         "CHM13v2.0",
         "False",
         "True",
         "False",
         "False",
         "False",
         "False",
         "True",
         "True",
         "True",
         "False",
         "False",
         "True",
         "True",
         "True",
         "True",
         "354",
         "False"
        ],
        [
         "405",
         "CHM13v2.0",
         "True",
         "True",
         "False",
         "False",
         "False",
         "False",
         "True",
         "False",
         "False",
         "False",
         "False",
         "True",
         "False",
         "True",
         "False",
         "66",
         "False"
        ],
        [
         "417",
         "CHM13v2.0",
         "False",
         "True",
         "False",
         "False",
         "False",
         "False",
         "True",
         "True",
         "True",
         "True",
         "False",
         "True",
         "True",
         "True",
         "True",
         "140",
         "False"
        ],
        [
         "422",
         "CHM13v2.0",
         "False",
         "True",
         "False",
         "False",
         "False",
         "False",
         "False",
         "False",
         "False",
         "False",
         "False",
         "True",
         "True",
         "False",
         "False",
         "3",
         "False"
        ],
        [
         "426",
         "CHM13v2.0",
         "False",
         "True",
         "True",
         "False",
         "False",
         "False",
         "True",
         "False",
         "False",
         "True",
         "False",
         "True",
         "True",
         "True",
         "True",
         "52949",
         "True"
        ],
        [
         "431",
         "CHM13v2.0",
         "True",
         "True",
         "True",
         "False",
         "False",
         "False",
         "True",
         "True",
         "True",
         "False",
         "False",
         "True",
         "True",
         "True",
         "True",
         "18167",
         "True"
        ],
        [
         "436",
         "CHM13v2.0",
         "False",
         "True",
         "True",
         "False",
         "False",
         "False",
         "False",
         "True",
         "False",
         "False",
         "False",
         "True",
         "True",
         "True",
         "True",
         "5",
         "True"
        ],
        [
         "443",
         "CHM13v2.0",
         "False",
         "True",
         "True",
         "False",
         "False",
         "False",
         "True",
         "False",
         "True",
         "False",
         "False",
         "True",
         "True",
         "True",
         "True",
         "11239",
         "True"
        ],
        [
         "446",
         "CHM13v2.0",
         "True",
         "True",
         "False",
         "False",
         "False",
         "False",
         "True",
         "False",
         "False",
         "False",
         "False",
         "True",
         "True",
         "True",
         "False",
         "40008264",
         "False"
        ],
        [
         "464",
         "CHM13v2.0",
         "False",
         "True",
         "True",
         "False",
         "False",
         "False",
         "True",
         "True",
         "True",
         "True",
         "False",
         "True",
         "True",
         "True",
         "True",
         "2418",
         "True"
        ],
        [
         "469",
         "CHM13v2.0",
         "False",
         "True",
         "False",
         "False",
         "False",
         "False",
         "True",
         "False",
         "False",
         "True",
         "False",
         "True",
         "True",
         "True",
         "True",
         "6834",
         "False"
        ],
        [
         "490",
         "CHM13v2.0",
         "False",
         "True",
         "True",
         "False",
         "False",
         "False",
         "True",
         "False",
         "False",
         "False",
         "False",
         "True",
         "False",
         "True",
         "True",
         "3",
         "True"
        ],
        [
         "497",
         "CHM13v2.0",
         "True",
         "True",
         "True",
         "False",
         "False",
         "False",
         "True",
         "False",
         "False",
         "False",
         "False",
         "True",
         "False",
         "True",
         "True",
         "26",
         "True"
        ]
       ],
       "shape": {
        "columns": 18,
        "rows": 49
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome</th>\n",
       "      <th>Syntenic</th>\n",
       "      <th>singleton</th>\n",
       "      <th>VQSLOD_filter</th>\n",
       "      <th>MERR_filter</th>\n",
       "      <th>HWE_pop_filter</th>\n",
       "      <th>MAC_filter</th>\n",
       "      <th>AC_filter</th>\n",
       "      <th>f_missing_filter</th>\n",
       "      <th>var_len_filter</th>\n",
       "      <th>alt_star_filter</th>\n",
       "      <th>pass_filter</th>\n",
       "      <th>CHM13_filtered</th>\n",
       "      <th>GRCh38_filtered</th>\n",
       "      <th>GRCh38_criteria_fail</th>\n",
       "      <th>CHM13_criteria_fail</th>\n",
       "      <th>len</th>\n",
       "      <th>gatk_filter</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>106488</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1360</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>589443</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>441729</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>31253</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>10378</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>2318</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>535</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>15800</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>533</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>111601</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>429834</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>11</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1097</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>47924</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>224</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>4862</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>210022</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>238</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>122</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>203660</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>266</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>7816298</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>269</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>44</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>289</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>16</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>292</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>11841</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>301</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>532</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>68335</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>315</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1133</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>326</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1545</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>2747</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>343</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>14518</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>82</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>382</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1245</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>393</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>354</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>405</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>66</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>417</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>140</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>422</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>426</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>52949</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>431</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>18167</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>436</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>5</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>443</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>11239</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>446</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>40008264</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>464</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>2418</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>469</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>6834</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>490</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>497</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>26</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        genome  Syntenic  singleton  VQSLOD_filter  MERR_filter  \\\n",
       "7    CHM13v2.0     False       True          False        False   \n",
       "12   CHM13v2.0      True       True          False        False   \n",
       "40   CHM13v2.0     False       True           True        False   \n",
       "55   CHM13v2.0      True       True          False        False   \n",
       "90   CHM13v2.0      True       True          False        False   \n",
       "95   CHM13v2.0      True       True           True        False   \n",
       "104  CHM13v2.0      True       True           True         True   \n",
       "109  CHM13v2.0      True       True          False        False   \n",
       "113  CHM13v2.0      True       True          False        False   \n",
       "122  CHM13v2.0      True       True           True        False   \n",
       "124  CHM13v2.0      True       True          False        False   \n",
       "137  CHM13v2.0      True       True           True        False   \n",
       "166  CHM13v2.0     False       True          False        False   \n",
       "167  CHM13v2.0      True       True           True        False   \n",
       "180  CHM13v2.0      True       True          False        False   \n",
       "181  CHM13v2.0     False       True           True        False   \n",
       "205  CHM13v2.0     False       True           True        False   \n",
       "209  CHM13v2.0     False       True          False        False   \n",
       "219  CHM13v2.0     False       True           True        False   \n",
       "224  CHM13v2.0      True       True           True        False   \n",
       "235  CHM13v2.0      True       True           True        False   \n",
       "238  CHM13v2.0      True       True          False        False   \n",
       "261  CHM13v2.0      True       True           True        False   \n",
       "266  CHM13v2.0      True       True           True        False   \n",
       "269  CHM13v2.0      True       True          False        False   \n",
       "276  CHM13v2.0      True       True          False        False   \n",
       "289  CHM13v2.0      True       True           True        False   \n",
       "292  CHM13v2.0     False       True           True        False   \n",
       "301  CHM13v2.0     False       True          False        False   \n",
       "311  CHM13v2.0      True       True           True        False   \n",
       "315  CHM13v2.0      True       True          False        False   \n",
       "326  CHM13v2.0      True       True          False        False   \n",
       "328  CHM13v2.0     False       True           True        False   \n",
       "343  CHM13v2.0      True       True          False        False   \n",
       "355  CHM13v2.0     False       True          False        False   \n",
       "382  CHM13v2.0     False       True          False        False   \n",
       "393  CHM13v2.0     False       True          False        False   \n",
       "405  CHM13v2.0      True       True          False        False   \n",
       "417  CHM13v2.0     False       True          False        False   \n",
       "422  CHM13v2.0     False       True          False        False   \n",
       "426  CHM13v2.0     False       True           True        False   \n",
       "431  CHM13v2.0      True       True           True        False   \n",
       "436  CHM13v2.0     False       True           True        False   \n",
       "443  CHM13v2.0     False       True           True        False   \n",
       "446  CHM13v2.0      True       True          False        False   \n",
       "464  CHM13v2.0     False       True           True        False   \n",
       "469  CHM13v2.0     False       True          False        False   \n",
       "490  CHM13v2.0     False       True           True        False   \n",
       "497  CHM13v2.0      True       True           True        False   \n",
       "\n",
       "     HWE_pop_filter  MAC_filter  AC_filter  f_missing_filter  var_len_filter  \\\n",
       "7             False       False       True             False           False   \n",
       "12            False       False       True              True            True   \n",
       "40            False       False       True             False           False   \n",
       "55            False       False      False              True           False   \n",
       "90            False       False      False             False           False   \n",
       "95            False       False      False             False           False   \n",
       "104           False       False       True              True           False   \n",
       "109           False       False       True             False           False   \n",
       "113           False       False       True              True           False   \n",
       "122           False       False       True              True           False   \n",
       "124           False       False       True             False            True   \n",
       "137           False       False       True              True            True   \n",
       "166           False       False       True              True           False   \n",
       "167           False       False      False             False           False   \n",
       "180           False       False       True             False           False   \n",
       "181           False       False       True              True           False   \n",
       "205           False       False      False             False           False   \n",
       "209           False       False       True             False            True   \n",
       "219           False       False       True              True           False   \n",
       "224           False       False       True             False            True   \n",
       "235           False       False       True              True           False   \n",
       "238           False       False       True             False           False   \n",
       "261           False       False       True             False           False   \n",
       "266           False       False       True             False           False   \n",
       "269           False       False       True              True            True   \n",
       "276           False       False      False             False           False   \n",
       "289           False       False      False              True           False   \n",
       "292           False       False       True              True            True   \n",
       "301           False       False       True             False            True   \n",
       "311           False       False       True             False            True   \n",
       "315           False       False       True              True           False   \n",
       "326           False       False      False             False           False   \n",
       "328           False       False       True             False            True   \n",
       "343           False       False       True             False            True   \n",
       "355           False       False       True             False           False   \n",
       "382           False       False       True              True           False   \n",
       "393           False       False       True              True            True   \n",
       "405           False       False       True             False           False   \n",
       "417           False       False       True              True            True   \n",
       "422           False       False      False             False           False   \n",
       "426           False       False       True             False           False   \n",
       "431           False       False       True              True            True   \n",
       "436           False       False      False              True           False   \n",
       "443           False       False       True             False            True   \n",
       "446           False       False       True             False           False   \n",
       "464           False       False       True              True            True   \n",
       "469           False       False       True             False           False   \n",
       "490           False       False       True             False           False   \n",
       "497           False       False       True             False           False   \n",
       "\n",
       "     alt_star_filter  pass_filter  CHM13_filtered  GRCh38_filtered  \\\n",
       "7              False        False            True             True   \n",
       "12             False        False            True             True   \n",
       "40             False        False            True             True   \n",
       "55             False        False            True             True   \n",
       "90              True        False            True             True   \n",
       "95              True        False            True             True   \n",
       "104             True        False            True             True   \n",
       "109             True        False            True             True   \n",
       "113            False        False            True             True   \n",
       "122             True        False            True             True   \n",
       "124             True        False            True             True   \n",
       "137             True        False            True             True   \n",
       "166            False        False            True             True   \n",
       "167            False        False            True             True   \n",
       "180            False        False           False             True   \n",
       "181            False        False            True             True   \n",
       "205            False        False            True             True   \n",
       "209            False        False            True             True   \n",
       "219             True        False            True             True   \n",
       "224             True        False            True             True   \n",
       "235            False        False            True             True   \n",
       "238            False        False           False            False   \n",
       "261             True        False            True             True   \n",
       "266            False        False            True             True   \n",
       "269             True        False            True             True   \n",
       "276            False        False           False             True   \n",
       "289            False        False            True             True   \n",
       "292            False        False            True             True   \n",
       "301             True        False            True             True   \n",
       "311            False        False            True             True   \n",
       "315             True        False            True             True   \n",
       "326            False        False            True             True   \n",
       "328             True        False            True             True   \n",
       "343            False        False            True             True   \n",
       "355            False        False           False             True   \n",
       "382             True        False            True             True   \n",
       "393            False        False            True             True   \n",
       "405            False        False            True            False   \n",
       "417             True        False            True             True   \n",
       "422            False        False            True             True   \n",
       "426             True        False            True             True   \n",
       "431            False        False            True             True   \n",
       "436            False        False            True             True   \n",
       "443            False        False            True             True   \n",
       "446            False        False            True             True   \n",
       "464             True        False            True             True   \n",
       "469             True        False            True             True   \n",
       "490            False        False            True            False   \n",
       "497            False        False            True            False   \n",
       "\n",
       "     GRCh38_criteria_fail  CHM13_criteria_fail       len  gatk_filter  \n",
       "7                    True                False    106488        False  \n",
       "12                   True                 True      1360        False  \n",
       "40                   True                 True    589443         True  \n",
       "55                   True                 True         3        False  \n",
       "90                   True                 True         1        False  \n",
       "95                   True                 True        12         True  \n",
       "104                  True                 True         2         True  \n",
       "109                  True                 True    441729        False  \n",
       "113                  True                 True     31253        False  \n",
       "122                  True                 True     10378         True  \n",
       "124                  True                 True      2318        False  \n",
       "137                  True                 True       535         True  \n",
       "166                  True                 True     15800        False  \n",
       "167                 False                 True       533         True  \n",
       "180                  True                False    111601        False  \n",
       "181                  True                 True    429834         True  \n",
       "205                 False                 True        11         True  \n",
       "209                  True                 True      1097        False  \n",
       "219                  True                 True     47924         True  \n",
       "224                  True                 True      4862         True  \n",
       "235                  True                 True    210022         True  \n",
       "238                  True                False       122        False  \n",
       "261                  True                 True    203660         True  \n",
       "266                  True                 True   7816298         True  \n",
       "269                  True                 True        44        False  \n",
       "276                 False                False         3        False  \n",
       "289                  True                 True        16         True  \n",
       "292                  True                 True     11841         True  \n",
       "301                  True                 True       532        False  \n",
       "311                  True                 True     68335         True  \n",
       "315                  True                 True      1133        False  \n",
       "326                 False                False      1545        False  \n",
       "328                  True                 True      2747         True  \n",
       "343                  True                 True     14518        False  \n",
       "355                  True                False        82        False  \n",
       "382                  True                 True      1245        False  \n",
       "393                  True                 True       354        False  \n",
       "405                  True                False        66        False  \n",
       "417                  True                 True       140        False  \n",
       "422                 False                False         3        False  \n",
       "426                  True                 True     52949         True  \n",
       "431                  True                 True     18167         True  \n",
       "436                  True                 True         5         True  \n",
       "443                  True                 True     11239         True  \n",
       "446                  True                False  40008264        False  \n",
       "464                  True                 True      2418         True  \n",
       "469                  True                 True      6834        False  \n",
       "490                  True                 True         3         True  \n",
       "497                  True                 True        26         True  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_filter_stats = pd.read_parquet(f\"{summary_statistics_folder}/filter_summary_stats.parquet\")\n",
    "variant_filter_stats['gatk_filter'] = variant_filter_stats.VQSLOD_filter #| variant_filter_stats.neg_train_site_filter\n",
    "chm13_variant_stats = variant_filter_stats.loc[(variant_filter_stats.genome=='CHM13v2.0')]\n",
    "singletons = chm13_variant_stats.loc[chm13_variant_stats.singleton]\n",
    "not_singletons = chm13_variant_stats.loc[~chm13_variant_stats.singleton]\n",
    "\n",
    "num_affected_gatk = chm13_variant_stats.loc[~(chm13_variant_stats.GRCh38_criteria_fail) &\n",
    "                                            # chm13_variant_stats.CHM13_criteria_fail &\n",
    "                                            chm13_variant_stats.gatk_filter].len.sum()\n",
    "\n",
    "\n",
    "singletons_gained = singletons.loc[chm13_variant_stats.GRCh38_criteria_fail & \n",
    "                                   ~(chm13_variant_stats.CHM13_criteria_fail | \n",
    "                                   variant_filter_stats.CHM13_filtered)].len.sum()\n",
    "print(f\"Our VQSLOD filter resulted in the loss of an additional \\n\\\n",
    "{num_affected_gatk} \\\n",
    "variants compared to the NYGCâ€™s GRCh38 1KGP variant filtering methods, but the inclusion of singleton alleles added \\n\\\n",
    "{singletons_gained} \\\n",
    "unique variants to our dataset (Supplemental Figure 1).\")\n",
    "singletons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "('CHM13_filtered', 'CHM13_criteria_fail', 'gatk_filter')",
         "rawType": "object",
         "type": "unknown"
        },
        {
         "name": "len",
         "rawType": "uint64",
         "type": "integer"
        }
       ],
       "ref": "56012b04-4828-4f04-a2a0-d417f1b1327f",
       "rows": [
        [
         "(np.False_, np.False_, np.False_)",
         "213333"
        ],
        [
         "(np.True_, np.False_, np.False_)",
         "60511946"
        ],
        [
         "(np.True_, np.True_, np.False_)",
         "2032062"
        ],
        [
         "(np.True_, np.True_, np.True_)",
         "17508360"
        ]
       ],
       "shape": {
        "columns": 1,
        "rows": 4
       }
      },
      "text/plain": [
       "CHM13_filtered  CHM13_criteria_fail  gatk_filter\n",
       "False           False                False            213333\n",
       "True            False                False          60511946\n",
       "                True                 False           2032062\n",
       "                                     True           17508360\n",
       "Name: len, dtype: uint64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "not_singletons.loc[(not_singletons.Syntenic)].groupby(['CHM13_filtered', 'CHM13_criteria_fail','gatk_filter']).len.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3427642/2682831855.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  variant_filter_stats.groupby(['genome','gatk_filter','GRCh38_criteria_fail'])['len'].sum().reset_index()\n"
     ]
    },
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "genome",
         "rawType": "category",
         "type": "unknown"
        },
        {
         "name": "gatk_filter",
         "rawType": "bool",
         "type": "boolean"
        },
        {
         "name": "GRCh38_criteria_fail",
         "rawType": "bool",
         "type": "boolean"
        },
        {
         "name": "len",
         "rawType": "uint64",
         "type": "integer"
        }
       ],
       "ref": "3f9521ce-24e1-4d41-870c-f816bb3ba43e",
       "rows": [
        [
         "0",
         "GRCh38",
         "False",
         "False",
         "61243112"
        ],
        [
         "1",
         "GRCh38",
         "False",
         "True",
         "42784742"
        ],
        [
         "2",
         "GRCh38",
         "True",
         "False",
         "13365695"
        ],
        [
         "3",
         "GRCh38",
         "True",
         "True",
         "25729879"
        ],
        [
         "4",
         "CHM13v2.0",
         "False",
         "False",
         "60934254"
        ],
        [
         "5",
         "CHM13v2.0",
         "False",
         "True",
         "42867140"
        ],
        [
         "6",
         "CHM13v2.0",
         "True",
         "False",
         "14494016"
        ],
        [
         "7",
         "CHM13v2.0",
         "True",
         "True",
         "16177227"
        ]
       ],
       "shape": {
        "columns": 4,
        "rows": 8
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome</th>\n",
       "      <th>gatk_filter</th>\n",
       "      <th>GRCh38_criteria_fail</th>\n",
       "      <th>len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>61243112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>42784742</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>13365695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GRCh38</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>25729879</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>60934254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>42867140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>14494016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>16177227</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      genome  gatk_filter  GRCh38_criteria_fail       len\n",
       "0     GRCh38        False                 False  61243112\n",
       "1     GRCh38        False                  True  42784742\n",
       "2     GRCh38         True                 False  13365695\n",
       "3     GRCh38         True                  True  25729879\n",
       "4  CHM13v2.0        False                 False  60934254\n",
       "5  CHM13v2.0        False                  True  42867140\n",
       "6  CHM13v2.0         True                 False  14494016\n",
       "7  CHM13v2.0         True                  True  16177227"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_filter_stats.groupby(['genome','gatk_filter','GRCh38_criteria_fail'])['len'].sum().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3427642/1937395110.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  variant_filter_stats.loc[(~variant_filter_stats.CHM13_criteria_fail)&(variant_filter_stats.GRCh38_criteria_fail)].groupby(['GRCh38_criteria_fail','gatk_filter','genome','singleton'])['len'].sum().reset_index()\n"
     ]
    },
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "GRCh38_criteria_fail",
         "rawType": "bool",
         "type": "boolean"
        },
        {
         "name": "gatk_filter",
         "rawType": "bool",
         "type": "boolean"
        },
        {
         "name": "genome",
         "rawType": "category",
         "type": "unknown"
        },
        {
         "name": "singleton",
         "rawType": "bool",
         "type": "boolean"
        },
        {
         "name": "len",
         "rawType": "uint64",
         "type": "integer"
        }
       ],
       "ref": "35656de1-45cc-4d71-af41-710a559f2864",
       "rows": [
        [
         "0",
         "True",
         "False",
         "GRCh38",
         "True",
         "39599697"
        ],
        [
         "1",
         "True",
         "False",
         "CHM13v2.0",
         "True",
         "40226623"
        ]
       ],
       "shape": {
        "columns": 5,
        "rows": 2
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GRCh38_criteria_fail</th>\n",
       "      <th>gatk_filter</th>\n",
       "      <th>genome</th>\n",
       "      <th>singleton</th>\n",
       "      <th>len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>True</td>\n",
       "      <td>39599697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>40226623</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   GRCh38_criteria_fail  gatk_filter     genome  singleton       len\n",
       "0                  True        False     GRCh38       True  39599697\n",
       "1                  True        False  CHM13v2.0       True  40226623"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_filter_stats.loc[(~variant_filter_stats.CHM13_criteria_fail)&(variant_filter_stats.GRCh38_criteria_fail)].groupby(['GRCh38_criteria_fail','gatk_filter','genome','singleton'])['len'].sum().reset_index()    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Table 1 and supplemental table XX"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "best_phased_variants = MAF_performance_variant_df.filter((pl.col('ground_truth_data_source')!='trios')&(pl.col('method_of_phasing')=='phased_with_parents_and_pedigree'))\n",
    "catVarType = pl.Enum(['SNP','Indel','SNPs + Indels'])\n",
    "best_phased_variants = best_phased_variants.with_columns(type=pl.col('type').cast(catVarType))\n",
    "best_phased_variants = pl.concat([best_phased_variants, best_phased_variants.with_columns(type=pl.lit('SNPs + Indels').cast(catVarType))])\n",
    "best_phased_variants_by_syntenic = best_phased_variants.group_by(['type','Syntenic','method_of_phasing','ground_truth_data_source','genome']\n",
    "                                                                 ).sum(\n",
    "                                                                 ).with_columns(gt_error_rate = pl.col('n_gt_errors')/pl.col('n_gt_checked')*100,\n",
    "                                                                                switch_error_rate = pl.col('n_switch_errors')/pl.col('n_checked')*100\n",
    "                                                                 ).to_pandas()\n",
    "best_phased_variants_overall = best_phased_variants.group_by(['type','method_of_phasing','ground_truth_data_source','genome']\n",
    "                                                                 ).sum(\n",
    "                                                                 ).with_columns(gt_error_rate = pl.col('n_gt_errors')/pl.col('n_gt_checked')*100,\n",
    "                                                                                switch_error_rate = pl.col('n_switch_errors')/pl.col('n_checked')*100\n",
    "                                                                 ).to_pandas()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "type",
         "rawType": "category",
         "type": "unknown"
        },
        {
         "name": "method_of_phasing",
         "rawType": "category",
         "type": "unknown"
        },
        {
         "name": "ground_truth_data_source",
         "rawType": "category",
         "type": "unknown"
        },
        {
         "name": "genome",
         "rawType": "category",
         "type": "unknown"
        },
        {
         "name": "variant_id",
         "rawType": "object",
         "type": "unknown"
        },
        {
         "name": "n_switch_errors",
         "rawType": "uint32",
         "type": "integer"
        },
        {
         "name": "n_checked",
         "rawType": "uint32",
         "type": "integer"
        },
        {
         "name": "n_gt_errors",
         "rawType": "uint32",
         "type": "integer"
        },
        {
         "name": "n_gt_checked",
         "rawType": "uint32",
         "type": "integer"
        },
        {
         "name": "MAC",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "AN",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "Syntenic",
         "rawType": "uint64",
         "type": "integer"
        },
        {
         "name": "multiallelic",
         "rawType": "uint64",
         "type": "integer"
        },
        {
         "name": "in_platinum_STRs",
         "rawType": "uint64",
         "type": "integer"
        },
        {
         "name": "in_segdups",
         "rawType": "uint64",
         "type": "integer"
        },
        {
         "name": "in_STRs",
         "rawType": "uint64",
         "type": "integer"
        },
        {
         "name": "MAF",
         "rawType": "float32",
         "type": "float"
        },
        {
         "name": "rounded_MAF",
         "rawType": "category",
         "type": "unknown"
        },
        {
         "name": "gt_error_rate",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "switch_error_rate",
         "rawType": "float64",
         "type": "float"
        }
       ],
       "ref": "7f0aff92-bfc6-461f-b272-4d9d449a5e00",
       "rows": [
        [
         "19",
         "SNP",
         "phased_with_parents_and_pedigree",
         "HPRC_samples",
         "GRCh38",
         null,
         "3664",
         "368771",
         "55269",
         "2690863",
         "41946142",
         "444066168",
         "69221",
         "5601",
         "7803",
         "22995",
         "3083",
         "654836.5",
         null,
         "2.053950721385667",
         "0.9935705356440718"
        ],
        [
         "16",
         "Indel",
         "phased_with_parents_and_pedigree",
         "HPRC_samples",
         "GRCh38",
         null,
         "1466",
         "70966",
         "33755",
         "531857",
         "9134280",
         "87709184",
         "13683",
         "9941",
         "9679",
         "3451",
         "2011",
         "142638.36",
         null,
         "6.346630767292711",
         "2.0657779781867376"
        ],
        [
         "8",
         "SNPs + Indels",
         "phased_with_parents_and_pedigree",
         "HPRC_samples",
         "GRCh38",
         null,
         "5130",
         "439737",
         "89024",
         "3222720",
         "51080422",
         "531775352",
         "82904",
         "15542",
         "17482",
         "26446",
         "5094",
         "797406.06",
         null,
         "2.762387051931288",
         "1.1666064033729253"
        ],
        [
         "38",
         "SNP",
         "phased_with_parents_and_pedigree",
         "HPRC_samples",
         "CHM13v2.0",
         null,
         "2170",
         "354018",
         "12172",
         "2499336",
         "39847161",
         "412936324",
         "64480",
         "4688",
         "6145",
         "7788",
         "2267",
         "622073.75",
         null,
         "0.48700934968327586",
         "0.6129631826630284"
        ],
        [
         "33",
         "Indel",
         "phased_with_parents_and_pedigree",
         "HPRC_samples",
         "CHM13v2.0",
         null,
         "1193",
         "79594",
         "23808",
         "563401",
         "9924572",
         "93120564",
         "14534",
         "10155",
         "9844",
         "2301",
         "2584",
         "154979.97",
         null,
         "4.2257645975069265",
         "1.4988566977410358"
        ],
        [
         "9",
         "SNPs + Indels",
         "phased_with_parents_and_pedigree",
         "HPRC_samples",
         "CHM13v2.0",
         null,
         "3363",
         "433612",
         "35980",
         "3062737",
         "49771733",
         "506056888",
         "79014",
         "14843",
         "15989",
         "10089",
         "4851",
         "776970.94",
         null,
         "1.1747662303358075",
         "0.7755781666559044"
        ],
        [
         "27",
         "SNP",
         "phased_with_parents_and_pedigree",
         "HGSVC_samples",
         "GRCh38",
         null,
         "6682",
         "532570",
         "75542",
         "4308072",
         "38507201",
         "445878500",
         "69489",
         "5524",
         "8023",
         "22003",
         "3179",
         "601137.4",
         null,
         "1.7534990130155672",
         "1.25467074750737"
        ],
        [
         "13",
         "Indel",
         "phased_with_parents_and_pedigree",
         "HGSVC_samples",
         "GRCh38",
         null,
         "2465",
         "104443",
         "51501",
         "855247",
         "8512412",
         "88432836",
         "13792",
         "10062",
         "9783",
         "3216",
         "2018",
         "132927.28",
         null,
         "6.021769149731013",
         "2.360139023199257"
        ],
        [
         "31",
         "SNPs + Indels",
         "phased_with_parents_and_pedigree",
         "HGSVC_samples",
         "GRCh38",
         null,
         "9147",
         "637013",
         "127043",
         "5163319",
         "47019613",
         "534311336",
         "83281",
         "15586",
         "17806",
         "25219",
         "5197",
         "733991.1",
         null,
         "2.4604910136290243",
         "1.435920460022009"
        ],
        [
         "0",
         "SNP",
         "phased_with_parents_and_pedigree",
         "HGSVC_samples",
         "CHM13v2.0",
         null,
         "4688",
         "529875",
         "46334",
         "4278024",
         "37923130",
         "442337088",
         "69069",
         "5047",
         "6718",
         "8337",
         "2512",
         "592002.56",
         null,
         "1.0830701277038184",
         "0.8847369662656287"
        ],
        [
         "4",
         "Indel",
         "phased_with_parents_and_pedigree",
         "HGSVC_samples",
         "CHM13v2.0",
         null,
         "2592",
         "118246",
         "48364",
         "948753",
         "9459864",
         "98051644",
         "15275",
         "10694",
         "10343",
         "2354",
         "2655",
         "147723.48",
         null,
         "5.097638689943537",
         "2.1920403227170473"
        ],
        [
         "15",
         "SNPs + Indels",
         "phased_with_parents_and_pedigree",
         "HGSVC_samples",
         "CHM13v2.0",
         null,
         "7280",
         "648121",
         "94698",
         "5226777",
         "47382994",
         "540388732",
         "84344",
         "15741",
         "17061",
         "10691",
         "5167",
         "739630.8",
         null,
         "1.8117857333496339",
         "1.123247048004925"
        ],
        [
         "20",
         "SNP",
         "phased_with_parents_and_pedigree",
         "HPRC_HGSVC_all_samples",
         "GRCh38",
         null,
         "12237",
         "860644",
         "239404",
         "8314500",
         "39412191",
         "532460580",
         "82986",
         "6336",
         "9313",
         "25414",
         "3635",
         "615215.7",
         null,
         "2.8793553430753502",
         "1.4218422483628539"
        ],
        [
         "7",
         "Indel",
         "phased_with_parents_and_pedigree",
         "HPRC_HGSVC_all_samples",
         "GRCh38",
         null,
         "3735",
         "166310",
         "109060",
         "1570600",
         "8797182",
         "100581224",
         "15686",
         "11178",
         "10917",
         "3802",
         "2309",
         "137374.73",
         null,
         "6.94384311728002",
         "2.245806024893272"
        ],
        [
         "29",
         "SNPs + Indels",
         "phased_with_parents_and_pedigree",
         "HPRC_HGSVC_all_samples",
         "GRCh38",
         null,
         "15972",
         "1026954",
         "348464",
         "9885100",
         "48209373",
         "633041804",
         "98672",
         "17514",
         "20230",
         "29216",
         "5944",
         "752485.1",
         null,
         "3.5251439034506475",
         "1.5552790095758915"
        ],
        [
         "23",
         "SNP",
         "phased_with_parents_and_pedigree",
         "HPRC_HGSVC_all_samples",
         "CHM13v2.0",
         null,
         "8412",
         "854138",
         "171416",
         "8197200",
         "38603424",
         "524948688",
         "81967",
         "5861",
         "7735",
         "9270",
         "2847",
         "602568.4",
         null,
         "2.091153076660323",
         "0.9848525648080286"
        ],
        [
         "5",
         "Indel",
         "phased_with_parents_and_pedigree",
         "HPRC_HGSVC_all_samples",
         "CHM13v2.0",
         null,
         "3661",
         "190066",
         "101392",
         "1734000",
         "9762226",
         "111045360",
         "17298",
         "11858",
         "11522",
         "2804",
         "2989",
         "152445.73",
         null,
         "5.847289504036909",
         "1.9261730135847546"
        ],
        [
         "35",
         "SNPs + Indels",
         "phased_with_parents_and_pedigree",
         "HPRC_HGSVC_all_samples",
         "CHM13v2.0",
         null,
         "12073",
         "1044204",
         "272808",
         "9931200",
         "48365650",
         "635994048",
         "99265",
         "17719",
         "19257",
         "12074",
         "5836",
         "754892.7",
         null,
         "2.74697921701305",
         "1.156191702004589"
        ],
        [
         "32",
         "SNP",
         "phased_with_parents_and_pedigree",
         "HPRC_HGSVC_probands",
         "GRCh38",
         null,
         "7041",
         "595191",
         "94047",
         "4968420",
         "39115048",
         "467908260",
         "72917",
         "5653",
         "8214",
         "22287",
         "3234",
         "610620.6",
         null,
         "1.8928955281558322",
         "1.1829815975039946"
        ],
        [
         "30",
         "Indel",
         "phased_with_parents_and_pedigree",
         "HPRC_HGSVC_probands",
         "GRCh38",
         null,
         "2188",
         "113913",
         "58298",
         "965600",
         "8594170",
         "90936800",
         "14182",
         "10202",
         "9933",
         "3351",
         "2079",
         "134204.1",
         null,
         "6.037489643744822",
         "1.9207640918946915"
        ],
        [
         "26",
         "SNPs + Indels",
         "phased_with_parents_and_pedigree",
         "HPRC_HGSVC_probands",
         "GRCh38",
         null,
         "9229",
         "709104",
         "152345",
         "5934020",
         "47709218",
         "558845060",
         "87099",
         "15855",
         "18147",
         "25638",
         "5313",
         "744732.2",
         null,
         "2.5673152432920685",
         "1.3015016133035493"
        ],
        [
         "11",
         "SNP",
         "phased_with_parents_and_pedigree",
         "HPRC_HGSVC_probands",
         "CHM13v2.0",
         null,
         "4921",
         "586088",
         "70223",
         "4803044",
         "38331543",
         "452333732",
         "70628",
         "5171",
         "6693",
         "8015",
         "2490",
         "598381.56",
         null,
         "1.4620519820347262",
         "0.8396350036172043"
        ],
        [
         "34",
         "Indel",
         "phased_with_parents_and_pedigree",
         "HPRC_HGSVC_probands",
         "CHM13v2.0",
         null,
         "2265",
         "130287",
         "55686",
         "1059984",
         "9543504",
         "99825552",
         "15548",
         "10745",
         "10432",
         "2467",
         "2708",
         "149029.7",
         null,
         "5.253475524158855",
         "1.7384696861544129"
        ],
        [
         "18",
         "SNPs + Indels",
         "phased_with_parents_and_pedigree",
         "HPRC_HGSVC_probands",
         "CHM13v2.0",
         null,
         "7186",
         "716375",
         "125909",
         "5863028",
         "47875047",
         "552159284",
         "86176",
         "15916",
         "17125",
         "10482",
         "5198",
         "747309.8",
         null,
         "2.147508079442909",
         "1.0031059151980457"
        ],
        [
         "39",
         "SNP",
         "phased_with_parents_and_pedigree",
         "HGSVC_probands",
         "GRCh38",
         null,
         "3319",
         "270023",
         "38096",
         "1731660",
         "38364549",
         "357727440",
         "55741",
         "4507",
         "6342",
         "17574",
         "2534",
         "598971.2",
         null,
         "2.1999699710104754",
         "1.229154553500998"
        ],
        [
         "24",
         "Indel",
         "phased_with_parents_and_pedigree",
         "HGSVC_probands",
         "GRCh38",
         null,
         "996",
         "51477",
         "23736",
         "362266",
         "8249435",
         "74837144",
         "11672",
         "8635",
         "8403",
         "2696",
         "1723",
         "128819.96",
         null,
         "6.552091557032677",
         "1.9348446879188765"
        ],
        [
         "1",
         "SNPs + Indels",
         "phased_with_parents_and_pedigree",
         "HGSVC_probands",
         "GRCh38",
         null,
         "4315",
         "321500",
         "61832",
         "2093926",
         "46613984",
         "432564584",
         "67413",
         "13142",
         "14745",
         "20270",
         "4257",
         "727732.4",
         null,
         "2.9529219275179734",
         "1.3421461897356144"
        ],
        [
         "25",
         "SNP",
         "phased_with_parents_and_pedigree",
         "HGSVC_probands",
         "CHM13v2.0",
         null,
         "2237",
         "263892",
         "36644",
         "1672481",
         "37624081",
         "345502204",
         "53947",
         "4145",
         "5174",
         "6339",
         "1967",
         "587404.06",
         null,
         "2.19099648964622",
         "0.8476952692768254"
        ],
        [
         "3",
         "Indel",
         "phased_with_parents_and_pedigree",
         "HGSVC_probands",
         "CHM13v2.0",
         null,
         "1049",
         "58636",
         "24185",
         "394661",
         "9186869",
         "81529324",
         "12699",
         "9032",
         "8749",
         "1920",
         "2264",
         "143459.36",
         null,
         "6.128044068200303",
         "1.7890033426563885"
        ],
        [
         "6",
         "SNPs + Indels",
         "phased_with_parents_and_pedigree",
         "HGSVC_probands",
         "CHM13v2.0",
         null,
         "3286",
         "322528",
         "60829",
         "2067142",
         "46810950",
         "427031528",
         "66646",
         "13177",
         "13923",
         "8259",
         "4231",
         "730796.6",
         null,
         "2.942661897441008",
         "1.0188262724476636"
        ],
        [
         "10",
         "SNP",
         "phased_with_parents_and_pedigree",
         "HGSVC_parents",
         "GRCh38",
         null,
         "1812",
         "115944",
         "33585",
         "574613",
         "37052464",
         "283063204",
         "44089",
         "3622",
         "4948",
         "14165",
         "1971",
         "578524.7",
         null,
         "5.844803372008639",
         "1.562823432001656"
        ],
        [
         "37",
         "Indel",
         "phased_with_parents_and_pedigree",
         "HGSVC_parents",
         "GRCh38",
         null,
         "568",
         "22762",
         "12698",
         "121420",
         "7753885",
         "59813360",
         "9330",
         "6888",
         "6700",
         "2108",
         "1397",
         "121080.87",
         null,
         "10.457914676330093",
         "2.495387048589755"
        ],
        [
         "28",
         "SNPs + Indels",
         "phased_with_parents_and_pedigree",
         "HGSVC_parents",
         "GRCh38",
         null,
         "2380",
         "138706",
         "46283",
         "696033",
         "44806349",
         "342876564",
         "53419",
         "10510",
         "11648",
         "16273",
         "3368",
         "699571.9",
         null,
         "6.649541041875888",
         "1.7158594437154846"
        ],
        [
         "36",
         "SNP",
         "phased_with_parents_and_pedigree",
         "HGSVC_parents",
         "CHM13v2.0",
         null,
         "1044",
         "118176",
         "21603",
         "558467",
         "36412859",
         "275109436",
         "42958",
         "3365",
         "4108",
         "5190",
         "1554",
         "568530.8",
         null,
         "3.8682679549552614",
         "0.8834281072298944"
        ],
        [
         "2",
         "Indel",
         "phased_with_parents_and_pedigree",
         "HGSVC_parents",
         "CHM13v2.0",
         null,
         "560",
         "26409",
         "11560",
         "133666",
         "8759995",
         "65845928",
         "10259",
         "7310",
         "7076",
         "1517",
         "1890",
         "136792.19",
         null,
         "8.648422186644321",
         "2.1204892271574085"
        ],
        [
         "40",
         "SNPs + Indels",
         "phased_with_parents_and_pedigree",
         "HGSVC_parents",
         "CHM13v2.0",
         null,
         "1604",
         "144585",
         "33163",
         "692133",
         "45172854",
         "340955364",
         "53217",
         "10675",
         "11184",
         "6707",
         "3444",
         "705290.1",
         null,
         "4.791420146127984",
         "1.1093820244147041"
        ],
        [
         "41",
         "SNP",
         "phased_with_parents_and_pedigree",
         "HGSVC_samples_nontrios_only",
         "GRCh38",
         null,
         "3386",
         "149192",
         "44905",
         "874437",
         "37640950",
         "294731292",
         "45927",
         "3830",
         "5192",
         "14599",
         "2077",
         "587707.7",
         null,
         "5.135304201446187",
         "2.2695586894739663"
        ],
        [
         "22",
         "Indel",
         "phased_with_parents_and_pedigree",
         "HGSVC_samples_nontrios_only",
         "GRCh38",
         null,
         "938",
         "29739",
         "18356",
         "188157",
         "7982317",
         "63418812",
         "9894",
         "7421",
         "7220",
         "2163",
         "1448",
         "124648.086",
         null,
         "9.755682754295615",
         "3.1541074010558527"
        ],
        [
         "14",
         "SNPs + Indels",
         "phased_with_parents_and_pedigree",
         "HGSVC_samples_nontrios_only",
         "GRCh38",
         null,
         "4324",
         "178931",
         "63261",
         "1062594",
         "45623267",
         "358150104",
         "55821",
         "11251",
         "12412",
         "16762",
         "3525",
         "712320.0",
         null,
         "5.9534497653854626",
         "2.416573986620541"
        ],
        [
         "12",
         "SNP",
         "phased_with_parents_and_pedigree",
         "HGSVC_samples_nontrios_only",
         "CHM13v2.0",
         null,
         "2440",
         "149650",
         "29222",
         "864823",
         "37005573",
         "291490868",
         "45515",
         "3620",
         "4369",
         "5424",
         "1657",
         "577773.44",
         null,
         "3.3789573126524157",
         "1.6304710992315403"
        ],
        [
         "17",
         "Indel",
         "phased_with_parents_and_pedigree",
         "HGSVC_samples_nontrios_only",
         "CHM13v2.0",
         null,
         "793",
         "33490",
         "16321",
         "207309",
         "8979034",
         "69874044",
         "10888",
         "7862",
         "7609",
         "1581",
         "1954",
         "140212.9",
         null,
         "7.8727889286041615",
         "2.3678710062705286"
        ],
        [
         "21",
         "SNPs + Indels",
         "phased_with_parents_and_pedigree",
         "HGSVC_samples_nontrios_only",
         "CHM13v2.0",
         null,
         "3233",
         "183140",
         "45543",
         "1072132",
         "45984607",
         "361364912",
         "56403",
         "11482",
         "11978",
         "7005",
         "3611",
         "717948.7",
         null,
         "4.2478911178847385",
         "1.7653161515780276"
        ]
       ],
       "shape": {
        "columns": 20,
        "rows": 42
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>method_of_phasing</th>\n",
       "      <th>ground_truth_data_source</th>\n",
       "      <th>genome</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>n_switch_errors</th>\n",
       "      <th>n_checked</th>\n",
       "      <th>n_gt_errors</th>\n",
       "      <th>n_gt_checked</th>\n",
       "      <th>MAC</th>\n",
       "      <th>AN</th>\n",
       "      <th>Syntenic</th>\n",
       "      <th>multiallelic</th>\n",
       "      <th>in_platinum_STRs</th>\n",
       "      <th>in_segdups</th>\n",
       "      <th>in_STRs</th>\n",
       "      <th>MAF</th>\n",
       "      <th>rounded_MAF</th>\n",
       "      <th>gt_error_rate</th>\n",
       "      <th>switch_error_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>SNP</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HPRC_samples</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>3664</td>\n",
       "      <td>368771</td>\n",
       "      <td>55269</td>\n",
       "      <td>2690863</td>\n",
       "      <td>41946142</td>\n",
       "      <td>444066168</td>\n",
       "      <td>69221</td>\n",
       "      <td>5601</td>\n",
       "      <td>7803</td>\n",
       "      <td>22995</td>\n",
       "      <td>3083</td>\n",
       "      <td>654836.500000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.053951</td>\n",
       "      <td>0.993571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Indel</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HPRC_samples</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>1466</td>\n",
       "      <td>70966</td>\n",
       "      <td>33755</td>\n",
       "      <td>531857</td>\n",
       "      <td>9134280</td>\n",
       "      <td>87709184</td>\n",
       "      <td>13683</td>\n",
       "      <td>9941</td>\n",
       "      <td>9679</td>\n",
       "      <td>3451</td>\n",
       "      <td>2011</td>\n",
       "      <td>142638.359375</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.346631</td>\n",
       "      <td>2.065778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SNPs + Indels</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HPRC_samples</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>5130</td>\n",
       "      <td>439737</td>\n",
       "      <td>89024</td>\n",
       "      <td>3222720</td>\n",
       "      <td>51080422</td>\n",
       "      <td>531775352</td>\n",
       "      <td>82904</td>\n",
       "      <td>15542</td>\n",
       "      <td>17482</td>\n",
       "      <td>26446</td>\n",
       "      <td>5094</td>\n",
       "      <td>797406.062500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.762387</td>\n",
       "      <td>1.166606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>SNP</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HPRC_samples</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>None</td>\n",
       "      <td>2170</td>\n",
       "      <td>354018</td>\n",
       "      <td>12172</td>\n",
       "      <td>2499336</td>\n",
       "      <td>39847161</td>\n",
       "      <td>412936324</td>\n",
       "      <td>64480</td>\n",
       "      <td>4688</td>\n",
       "      <td>6145</td>\n",
       "      <td>7788</td>\n",
       "      <td>2267</td>\n",
       "      <td>622073.750000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.487009</td>\n",
       "      <td>0.612963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Indel</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HPRC_samples</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>None</td>\n",
       "      <td>1193</td>\n",
       "      <td>79594</td>\n",
       "      <td>23808</td>\n",
       "      <td>563401</td>\n",
       "      <td>9924572</td>\n",
       "      <td>93120564</td>\n",
       "      <td>14534</td>\n",
       "      <td>10155</td>\n",
       "      <td>9844</td>\n",
       "      <td>2301</td>\n",
       "      <td>2584</td>\n",
       "      <td>154979.968750</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.225765</td>\n",
       "      <td>1.498857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SNPs + Indels</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HPRC_samples</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>None</td>\n",
       "      <td>3363</td>\n",
       "      <td>433612</td>\n",
       "      <td>35980</td>\n",
       "      <td>3062737</td>\n",
       "      <td>49771733</td>\n",
       "      <td>506056888</td>\n",
       "      <td>79014</td>\n",
       "      <td>14843</td>\n",
       "      <td>15989</td>\n",
       "      <td>10089</td>\n",
       "      <td>4851</td>\n",
       "      <td>776970.937500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.174766</td>\n",
       "      <td>0.775578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>SNP</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HGSVC_samples</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>6682</td>\n",
       "      <td>532570</td>\n",
       "      <td>75542</td>\n",
       "      <td>4308072</td>\n",
       "      <td>38507201</td>\n",
       "      <td>445878500</td>\n",
       "      <td>69489</td>\n",
       "      <td>5524</td>\n",
       "      <td>8023</td>\n",
       "      <td>22003</td>\n",
       "      <td>3179</td>\n",
       "      <td>601137.375000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.753499</td>\n",
       "      <td>1.254671</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Indel</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HGSVC_samples</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>2465</td>\n",
       "      <td>104443</td>\n",
       "      <td>51501</td>\n",
       "      <td>855247</td>\n",
       "      <td>8512412</td>\n",
       "      <td>88432836</td>\n",
       "      <td>13792</td>\n",
       "      <td>10062</td>\n",
       "      <td>9783</td>\n",
       "      <td>3216</td>\n",
       "      <td>2018</td>\n",
       "      <td>132927.281250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.021769</td>\n",
       "      <td>2.360139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>SNPs + Indels</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HGSVC_samples</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>9147</td>\n",
       "      <td>637013</td>\n",
       "      <td>127043</td>\n",
       "      <td>5163319</td>\n",
       "      <td>47019613</td>\n",
       "      <td>534311336</td>\n",
       "      <td>83281</td>\n",
       "      <td>15586</td>\n",
       "      <td>17806</td>\n",
       "      <td>25219</td>\n",
       "      <td>5197</td>\n",
       "      <td>733991.125000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.460491</td>\n",
       "      <td>1.435920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SNP</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HGSVC_samples</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>None</td>\n",
       "      <td>4688</td>\n",
       "      <td>529875</td>\n",
       "      <td>46334</td>\n",
       "      <td>4278024</td>\n",
       "      <td>37923130</td>\n",
       "      <td>442337088</td>\n",
       "      <td>69069</td>\n",
       "      <td>5047</td>\n",
       "      <td>6718</td>\n",
       "      <td>8337</td>\n",
       "      <td>2512</td>\n",
       "      <td>592002.562500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.083070</td>\n",
       "      <td>0.884737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Indel</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HGSVC_samples</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>None</td>\n",
       "      <td>2592</td>\n",
       "      <td>118246</td>\n",
       "      <td>48364</td>\n",
       "      <td>948753</td>\n",
       "      <td>9459864</td>\n",
       "      <td>98051644</td>\n",
       "      <td>15275</td>\n",
       "      <td>10694</td>\n",
       "      <td>10343</td>\n",
       "      <td>2354</td>\n",
       "      <td>2655</td>\n",
       "      <td>147723.484375</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.097639</td>\n",
       "      <td>2.192040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>SNPs + Indels</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HGSVC_samples</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>None</td>\n",
       "      <td>7280</td>\n",
       "      <td>648121</td>\n",
       "      <td>94698</td>\n",
       "      <td>5226777</td>\n",
       "      <td>47382994</td>\n",
       "      <td>540388732</td>\n",
       "      <td>84344</td>\n",
       "      <td>15741</td>\n",
       "      <td>17061</td>\n",
       "      <td>10691</td>\n",
       "      <td>5167</td>\n",
       "      <td>739630.812500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.811786</td>\n",
       "      <td>1.123247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>SNP</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HPRC_HGSVC_all_samples</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>12237</td>\n",
       "      <td>860644</td>\n",
       "      <td>239404</td>\n",
       "      <td>8314500</td>\n",
       "      <td>39412191</td>\n",
       "      <td>532460580</td>\n",
       "      <td>82986</td>\n",
       "      <td>6336</td>\n",
       "      <td>9313</td>\n",
       "      <td>25414</td>\n",
       "      <td>3635</td>\n",
       "      <td>615215.687500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.879355</td>\n",
       "      <td>1.421842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Indel</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HPRC_HGSVC_all_samples</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>3735</td>\n",
       "      <td>166310</td>\n",
       "      <td>109060</td>\n",
       "      <td>1570600</td>\n",
       "      <td>8797182</td>\n",
       "      <td>100581224</td>\n",
       "      <td>15686</td>\n",
       "      <td>11178</td>\n",
       "      <td>10917</td>\n",
       "      <td>3802</td>\n",
       "      <td>2309</td>\n",
       "      <td>137374.734375</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.943843</td>\n",
       "      <td>2.245806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>SNPs + Indels</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HPRC_HGSVC_all_samples</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>15972</td>\n",
       "      <td>1026954</td>\n",
       "      <td>348464</td>\n",
       "      <td>9885100</td>\n",
       "      <td>48209373</td>\n",
       "      <td>633041804</td>\n",
       "      <td>98672</td>\n",
       "      <td>17514</td>\n",
       "      <td>20230</td>\n",
       "      <td>29216</td>\n",
       "      <td>5944</td>\n",
       "      <td>752485.125000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.525144</td>\n",
       "      <td>1.555279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>SNP</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HPRC_HGSVC_all_samples</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>None</td>\n",
       "      <td>8412</td>\n",
       "      <td>854138</td>\n",
       "      <td>171416</td>\n",
       "      <td>8197200</td>\n",
       "      <td>38603424</td>\n",
       "      <td>524948688</td>\n",
       "      <td>81967</td>\n",
       "      <td>5861</td>\n",
       "      <td>7735</td>\n",
       "      <td>9270</td>\n",
       "      <td>2847</td>\n",
       "      <td>602568.375000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.091153</td>\n",
       "      <td>0.984853</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Indel</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HPRC_HGSVC_all_samples</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>None</td>\n",
       "      <td>3661</td>\n",
       "      <td>190066</td>\n",
       "      <td>101392</td>\n",
       "      <td>1734000</td>\n",
       "      <td>9762226</td>\n",
       "      <td>111045360</td>\n",
       "      <td>17298</td>\n",
       "      <td>11858</td>\n",
       "      <td>11522</td>\n",
       "      <td>2804</td>\n",
       "      <td>2989</td>\n",
       "      <td>152445.734375</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.847290</td>\n",
       "      <td>1.926173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>SNPs + Indels</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HPRC_HGSVC_all_samples</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>None</td>\n",
       "      <td>12073</td>\n",
       "      <td>1044204</td>\n",
       "      <td>272808</td>\n",
       "      <td>9931200</td>\n",
       "      <td>48365650</td>\n",
       "      <td>635994048</td>\n",
       "      <td>99265</td>\n",
       "      <td>17719</td>\n",
       "      <td>19257</td>\n",
       "      <td>12074</td>\n",
       "      <td>5836</td>\n",
       "      <td>754892.687500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.746979</td>\n",
       "      <td>1.156192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>SNP</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HPRC_HGSVC_probands</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>7041</td>\n",
       "      <td>595191</td>\n",
       "      <td>94047</td>\n",
       "      <td>4968420</td>\n",
       "      <td>39115048</td>\n",
       "      <td>467908260</td>\n",
       "      <td>72917</td>\n",
       "      <td>5653</td>\n",
       "      <td>8214</td>\n",
       "      <td>22287</td>\n",
       "      <td>3234</td>\n",
       "      <td>610620.625000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.892896</td>\n",
       "      <td>1.182982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Indel</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HPRC_HGSVC_probands</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>2188</td>\n",
       "      <td>113913</td>\n",
       "      <td>58298</td>\n",
       "      <td>965600</td>\n",
       "      <td>8594170</td>\n",
       "      <td>90936800</td>\n",
       "      <td>14182</td>\n",
       "      <td>10202</td>\n",
       "      <td>9933</td>\n",
       "      <td>3351</td>\n",
       "      <td>2079</td>\n",
       "      <td>134204.093750</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.037490</td>\n",
       "      <td>1.920764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>SNPs + Indels</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HPRC_HGSVC_probands</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>9229</td>\n",
       "      <td>709104</td>\n",
       "      <td>152345</td>\n",
       "      <td>5934020</td>\n",
       "      <td>47709218</td>\n",
       "      <td>558845060</td>\n",
       "      <td>87099</td>\n",
       "      <td>15855</td>\n",
       "      <td>18147</td>\n",
       "      <td>25638</td>\n",
       "      <td>5313</td>\n",
       "      <td>744732.187500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.567315</td>\n",
       "      <td>1.301502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SNP</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HPRC_HGSVC_probands</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>None</td>\n",
       "      <td>4921</td>\n",
       "      <td>586088</td>\n",
       "      <td>70223</td>\n",
       "      <td>4803044</td>\n",
       "      <td>38331543</td>\n",
       "      <td>452333732</td>\n",
       "      <td>70628</td>\n",
       "      <td>5171</td>\n",
       "      <td>6693</td>\n",
       "      <td>8015</td>\n",
       "      <td>2490</td>\n",
       "      <td>598381.562500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.462052</td>\n",
       "      <td>0.839635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Indel</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HPRC_HGSVC_probands</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>None</td>\n",
       "      <td>2265</td>\n",
       "      <td>130287</td>\n",
       "      <td>55686</td>\n",
       "      <td>1059984</td>\n",
       "      <td>9543504</td>\n",
       "      <td>99825552</td>\n",
       "      <td>15548</td>\n",
       "      <td>10745</td>\n",
       "      <td>10432</td>\n",
       "      <td>2467</td>\n",
       "      <td>2708</td>\n",
       "      <td>149029.703125</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.253476</td>\n",
       "      <td>1.738470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>SNPs + Indels</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HPRC_HGSVC_probands</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>None</td>\n",
       "      <td>7186</td>\n",
       "      <td>716375</td>\n",
       "      <td>125909</td>\n",
       "      <td>5863028</td>\n",
       "      <td>47875047</td>\n",
       "      <td>552159284</td>\n",
       "      <td>86176</td>\n",
       "      <td>15916</td>\n",
       "      <td>17125</td>\n",
       "      <td>10482</td>\n",
       "      <td>5198</td>\n",
       "      <td>747309.812500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.147508</td>\n",
       "      <td>1.003106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>SNP</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HGSVC_probands</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>3319</td>\n",
       "      <td>270023</td>\n",
       "      <td>38096</td>\n",
       "      <td>1731660</td>\n",
       "      <td>38364549</td>\n",
       "      <td>357727440</td>\n",
       "      <td>55741</td>\n",
       "      <td>4507</td>\n",
       "      <td>6342</td>\n",
       "      <td>17574</td>\n",
       "      <td>2534</td>\n",
       "      <td>598971.187500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.199970</td>\n",
       "      <td>1.229155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Indel</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HGSVC_probands</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>996</td>\n",
       "      <td>51477</td>\n",
       "      <td>23736</td>\n",
       "      <td>362266</td>\n",
       "      <td>8249435</td>\n",
       "      <td>74837144</td>\n",
       "      <td>11672</td>\n",
       "      <td>8635</td>\n",
       "      <td>8403</td>\n",
       "      <td>2696</td>\n",
       "      <td>1723</td>\n",
       "      <td>128819.960938</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.552092</td>\n",
       "      <td>1.934845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SNPs + Indels</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HGSVC_probands</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>4315</td>\n",
       "      <td>321500</td>\n",
       "      <td>61832</td>\n",
       "      <td>2093926</td>\n",
       "      <td>46613984</td>\n",
       "      <td>432564584</td>\n",
       "      <td>67413</td>\n",
       "      <td>13142</td>\n",
       "      <td>14745</td>\n",
       "      <td>20270</td>\n",
       "      <td>4257</td>\n",
       "      <td>727732.375000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.952922</td>\n",
       "      <td>1.342146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>SNP</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HGSVC_probands</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>None</td>\n",
       "      <td>2237</td>\n",
       "      <td>263892</td>\n",
       "      <td>36644</td>\n",
       "      <td>1672481</td>\n",
       "      <td>37624081</td>\n",
       "      <td>345502204</td>\n",
       "      <td>53947</td>\n",
       "      <td>4145</td>\n",
       "      <td>5174</td>\n",
       "      <td>6339</td>\n",
       "      <td>1967</td>\n",
       "      <td>587404.062500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.190996</td>\n",
       "      <td>0.847695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Indel</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HGSVC_probands</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>None</td>\n",
       "      <td>1049</td>\n",
       "      <td>58636</td>\n",
       "      <td>24185</td>\n",
       "      <td>394661</td>\n",
       "      <td>9186869</td>\n",
       "      <td>81529324</td>\n",
       "      <td>12699</td>\n",
       "      <td>9032</td>\n",
       "      <td>8749</td>\n",
       "      <td>1920</td>\n",
       "      <td>2264</td>\n",
       "      <td>143459.359375</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.128044</td>\n",
       "      <td>1.789003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SNPs + Indels</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HGSVC_probands</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>None</td>\n",
       "      <td>3286</td>\n",
       "      <td>322528</td>\n",
       "      <td>60829</td>\n",
       "      <td>2067142</td>\n",
       "      <td>46810950</td>\n",
       "      <td>427031528</td>\n",
       "      <td>66646</td>\n",
       "      <td>13177</td>\n",
       "      <td>13923</td>\n",
       "      <td>8259</td>\n",
       "      <td>4231</td>\n",
       "      <td>730796.625000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.942662</td>\n",
       "      <td>1.018826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SNP</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HGSVC_parents</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>1812</td>\n",
       "      <td>115944</td>\n",
       "      <td>33585</td>\n",
       "      <td>574613</td>\n",
       "      <td>37052464</td>\n",
       "      <td>283063204</td>\n",
       "      <td>44089</td>\n",
       "      <td>3622</td>\n",
       "      <td>4948</td>\n",
       "      <td>14165</td>\n",
       "      <td>1971</td>\n",
       "      <td>578524.687500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.844803</td>\n",
       "      <td>1.562823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Indel</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HGSVC_parents</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>568</td>\n",
       "      <td>22762</td>\n",
       "      <td>12698</td>\n",
       "      <td>121420</td>\n",
       "      <td>7753885</td>\n",
       "      <td>59813360</td>\n",
       "      <td>9330</td>\n",
       "      <td>6888</td>\n",
       "      <td>6700</td>\n",
       "      <td>2108</td>\n",
       "      <td>1397</td>\n",
       "      <td>121080.867188</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.457915</td>\n",
       "      <td>2.495387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>SNPs + Indels</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HGSVC_parents</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>2380</td>\n",
       "      <td>138706</td>\n",
       "      <td>46283</td>\n",
       "      <td>696033</td>\n",
       "      <td>44806349</td>\n",
       "      <td>342876564</td>\n",
       "      <td>53419</td>\n",
       "      <td>10510</td>\n",
       "      <td>11648</td>\n",
       "      <td>16273</td>\n",
       "      <td>3368</td>\n",
       "      <td>699571.875000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.649541</td>\n",
       "      <td>1.715859</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>SNP</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HGSVC_parents</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>None</td>\n",
       "      <td>1044</td>\n",
       "      <td>118176</td>\n",
       "      <td>21603</td>\n",
       "      <td>558467</td>\n",
       "      <td>36412859</td>\n",
       "      <td>275109436</td>\n",
       "      <td>42958</td>\n",
       "      <td>3365</td>\n",
       "      <td>4108</td>\n",
       "      <td>5190</td>\n",
       "      <td>1554</td>\n",
       "      <td>568530.812500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.868268</td>\n",
       "      <td>0.883428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Indel</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HGSVC_parents</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>None</td>\n",
       "      <td>560</td>\n",
       "      <td>26409</td>\n",
       "      <td>11560</td>\n",
       "      <td>133666</td>\n",
       "      <td>8759995</td>\n",
       "      <td>65845928</td>\n",
       "      <td>10259</td>\n",
       "      <td>7310</td>\n",
       "      <td>7076</td>\n",
       "      <td>1517</td>\n",
       "      <td>1890</td>\n",
       "      <td>136792.187500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.648422</td>\n",
       "      <td>2.120489</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>SNPs + Indels</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HGSVC_parents</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>None</td>\n",
       "      <td>1604</td>\n",
       "      <td>144585</td>\n",
       "      <td>33163</td>\n",
       "      <td>692133</td>\n",
       "      <td>45172854</td>\n",
       "      <td>340955364</td>\n",
       "      <td>53217</td>\n",
       "      <td>10675</td>\n",
       "      <td>11184</td>\n",
       "      <td>6707</td>\n",
       "      <td>3444</td>\n",
       "      <td>705290.125000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.791420</td>\n",
       "      <td>1.109382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>SNP</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HGSVC_samples_nontrios_only</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>3386</td>\n",
       "      <td>149192</td>\n",
       "      <td>44905</td>\n",
       "      <td>874437</td>\n",
       "      <td>37640950</td>\n",
       "      <td>294731292</td>\n",
       "      <td>45927</td>\n",
       "      <td>3830</td>\n",
       "      <td>5192</td>\n",
       "      <td>14599</td>\n",
       "      <td>2077</td>\n",
       "      <td>587707.687500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.135304</td>\n",
       "      <td>2.269559</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Indel</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HGSVC_samples_nontrios_only</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>938</td>\n",
       "      <td>29739</td>\n",
       "      <td>18356</td>\n",
       "      <td>188157</td>\n",
       "      <td>7982317</td>\n",
       "      <td>63418812</td>\n",
       "      <td>9894</td>\n",
       "      <td>7421</td>\n",
       "      <td>7220</td>\n",
       "      <td>2163</td>\n",
       "      <td>1448</td>\n",
       "      <td>124648.085938</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.755683</td>\n",
       "      <td>3.154107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SNPs + Indels</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HGSVC_samples_nontrios_only</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>4324</td>\n",
       "      <td>178931</td>\n",
       "      <td>63261</td>\n",
       "      <td>1062594</td>\n",
       "      <td>45623267</td>\n",
       "      <td>358150104</td>\n",
       "      <td>55821</td>\n",
       "      <td>11251</td>\n",
       "      <td>12412</td>\n",
       "      <td>16762</td>\n",
       "      <td>3525</td>\n",
       "      <td>712320.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.953450</td>\n",
       "      <td>2.416574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SNP</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HGSVC_samples_nontrios_only</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>None</td>\n",
       "      <td>2440</td>\n",
       "      <td>149650</td>\n",
       "      <td>29222</td>\n",
       "      <td>864823</td>\n",
       "      <td>37005573</td>\n",
       "      <td>291490868</td>\n",
       "      <td>45515</td>\n",
       "      <td>3620</td>\n",
       "      <td>4369</td>\n",
       "      <td>5424</td>\n",
       "      <td>1657</td>\n",
       "      <td>577773.437500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.378957</td>\n",
       "      <td>1.630471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Indel</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HGSVC_samples_nontrios_only</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>None</td>\n",
       "      <td>793</td>\n",
       "      <td>33490</td>\n",
       "      <td>16321</td>\n",
       "      <td>207309</td>\n",
       "      <td>8979034</td>\n",
       "      <td>69874044</td>\n",
       "      <td>10888</td>\n",
       "      <td>7862</td>\n",
       "      <td>7609</td>\n",
       "      <td>1581</td>\n",
       "      <td>1954</td>\n",
       "      <td>140212.906250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.872789</td>\n",
       "      <td>2.367871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>SNPs + Indels</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HGSVC_samples_nontrios_only</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>None</td>\n",
       "      <td>3233</td>\n",
       "      <td>183140</td>\n",
       "      <td>45543</td>\n",
       "      <td>1072132</td>\n",
       "      <td>45984607</td>\n",
       "      <td>361364912</td>\n",
       "      <td>56403</td>\n",
       "      <td>11482</td>\n",
       "      <td>11978</td>\n",
       "      <td>7005</td>\n",
       "      <td>3611</td>\n",
       "      <td>717948.687500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.247891</td>\n",
       "      <td>1.765316</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             type                 method_of_phasing  \\\n",
       "19            SNP  phased_with_parents_and_pedigree   \n",
       "16          Indel  phased_with_parents_and_pedigree   \n",
       "8   SNPs + Indels  phased_with_parents_and_pedigree   \n",
       "38            SNP  phased_with_parents_and_pedigree   \n",
       "33          Indel  phased_with_parents_and_pedigree   \n",
       "9   SNPs + Indels  phased_with_parents_and_pedigree   \n",
       "27            SNP  phased_with_parents_and_pedigree   \n",
       "13          Indel  phased_with_parents_and_pedigree   \n",
       "31  SNPs + Indels  phased_with_parents_and_pedigree   \n",
       "0             SNP  phased_with_parents_and_pedigree   \n",
       "4           Indel  phased_with_parents_and_pedigree   \n",
       "15  SNPs + Indels  phased_with_parents_and_pedigree   \n",
       "20            SNP  phased_with_parents_and_pedigree   \n",
       "7           Indel  phased_with_parents_and_pedigree   \n",
       "29  SNPs + Indels  phased_with_parents_and_pedigree   \n",
       "23            SNP  phased_with_parents_and_pedigree   \n",
       "5           Indel  phased_with_parents_and_pedigree   \n",
       "35  SNPs + Indels  phased_with_parents_and_pedigree   \n",
       "32            SNP  phased_with_parents_and_pedigree   \n",
       "30          Indel  phased_with_parents_and_pedigree   \n",
       "26  SNPs + Indels  phased_with_parents_and_pedigree   \n",
       "11            SNP  phased_with_parents_and_pedigree   \n",
       "34          Indel  phased_with_parents_and_pedigree   \n",
       "18  SNPs + Indels  phased_with_parents_and_pedigree   \n",
       "39            SNP  phased_with_parents_and_pedigree   \n",
       "24          Indel  phased_with_parents_and_pedigree   \n",
       "1   SNPs + Indels  phased_with_parents_and_pedigree   \n",
       "25            SNP  phased_with_parents_and_pedigree   \n",
       "3           Indel  phased_with_parents_and_pedigree   \n",
       "6   SNPs + Indels  phased_with_parents_and_pedigree   \n",
       "10            SNP  phased_with_parents_and_pedigree   \n",
       "37          Indel  phased_with_parents_and_pedigree   \n",
       "28  SNPs + Indels  phased_with_parents_and_pedigree   \n",
       "36            SNP  phased_with_parents_and_pedigree   \n",
       "2           Indel  phased_with_parents_and_pedigree   \n",
       "40  SNPs + Indels  phased_with_parents_and_pedigree   \n",
       "41            SNP  phased_with_parents_and_pedigree   \n",
       "22          Indel  phased_with_parents_and_pedigree   \n",
       "14  SNPs + Indels  phased_with_parents_and_pedigree   \n",
       "12            SNP  phased_with_parents_and_pedigree   \n",
       "17          Indel  phased_with_parents_and_pedigree   \n",
       "21  SNPs + Indels  phased_with_parents_and_pedigree   \n",
       "\n",
       "       ground_truth_data_source     genome variant_id  n_switch_errors  \\\n",
       "19                 HPRC_samples     GRCh38       None             3664   \n",
       "16                 HPRC_samples     GRCh38       None             1466   \n",
       "8                  HPRC_samples     GRCh38       None             5130   \n",
       "38                 HPRC_samples  CHM13v2.0       None             2170   \n",
       "33                 HPRC_samples  CHM13v2.0       None             1193   \n",
       "9                  HPRC_samples  CHM13v2.0       None             3363   \n",
       "27                HGSVC_samples     GRCh38       None             6682   \n",
       "13                HGSVC_samples     GRCh38       None             2465   \n",
       "31                HGSVC_samples     GRCh38       None             9147   \n",
       "0                 HGSVC_samples  CHM13v2.0       None             4688   \n",
       "4                 HGSVC_samples  CHM13v2.0       None             2592   \n",
       "15                HGSVC_samples  CHM13v2.0       None             7280   \n",
       "20       HPRC_HGSVC_all_samples     GRCh38       None            12237   \n",
       "7        HPRC_HGSVC_all_samples     GRCh38       None             3735   \n",
       "29       HPRC_HGSVC_all_samples     GRCh38       None            15972   \n",
       "23       HPRC_HGSVC_all_samples  CHM13v2.0       None             8412   \n",
       "5        HPRC_HGSVC_all_samples  CHM13v2.0       None             3661   \n",
       "35       HPRC_HGSVC_all_samples  CHM13v2.0       None            12073   \n",
       "32          HPRC_HGSVC_probands     GRCh38       None             7041   \n",
       "30          HPRC_HGSVC_probands     GRCh38       None             2188   \n",
       "26          HPRC_HGSVC_probands     GRCh38       None             9229   \n",
       "11          HPRC_HGSVC_probands  CHM13v2.0       None             4921   \n",
       "34          HPRC_HGSVC_probands  CHM13v2.0       None             2265   \n",
       "18          HPRC_HGSVC_probands  CHM13v2.0       None             7186   \n",
       "39               HGSVC_probands     GRCh38       None             3319   \n",
       "24               HGSVC_probands     GRCh38       None              996   \n",
       "1                HGSVC_probands     GRCh38       None             4315   \n",
       "25               HGSVC_probands  CHM13v2.0       None             2237   \n",
       "3                HGSVC_probands  CHM13v2.0       None             1049   \n",
       "6                HGSVC_probands  CHM13v2.0       None             3286   \n",
       "10                HGSVC_parents     GRCh38       None             1812   \n",
       "37                HGSVC_parents     GRCh38       None              568   \n",
       "28                HGSVC_parents     GRCh38       None             2380   \n",
       "36                HGSVC_parents  CHM13v2.0       None             1044   \n",
       "2                 HGSVC_parents  CHM13v2.0       None              560   \n",
       "40                HGSVC_parents  CHM13v2.0       None             1604   \n",
       "41  HGSVC_samples_nontrios_only     GRCh38       None             3386   \n",
       "22  HGSVC_samples_nontrios_only     GRCh38       None              938   \n",
       "14  HGSVC_samples_nontrios_only     GRCh38       None             4324   \n",
       "12  HGSVC_samples_nontrios_only  CHM13v2.0       None             2440   \n",
       "17  HGSVC_samples_nontrios_only  CHM13v2.0       None              793   \n",
       "21  HGSVC_samples_nontrios_only  CHM13v2.0       None             3233   \n",
       "\n",
       "    n_checked  n_gt_errors  n_gt_checked       MAC         AN  Syntenic  \\\n",
       "19     368771        55269       2690863  41946142  444066168     69221   \n",
       "16      70966        33755        531857   9134280   87709184     13683   \n",
       "8      439737        89024       3222720  51080422  531775352     82904   \n",
       "38     354018        12172       2499336  39847161  412936324     64480   \n",
       "33      79594        23808        563401   9924572   93120564     14534   \n",
       "9      433612        35980       3062737  49771733  506056888     79014   \n",
       "27     532570        75542       4308072  38507201  445878500     69489   \n",
       "13     104443        51501        855247   8512412   88432836     13792   \n",
       "31     637013       127043       5163319  47019613  534311336     83281   \n",
       "0      529875        46334       4278024  37923130  442337088     69069   \n",
       "4      118246        48364        948753   9459864   98051644     15275   \n",
       "15     648121        94698       5226777  47382994  540388732     84344   \n",
       "20     860644       239404       8314500  39412191  532460580     82986   \n",
       "7      166310       109060       1570600   8797182  100581224     15686   \n",
       "29    1026954       348464       9885100  48209373  633041804     98672   \n",
       "23     854138       171416       8197200  38603424  524948688     81967   \n",
       "5      190066       101392       1734000   9762226  111045360     17298   \n",
       "35    1044204       272808       9931200  48365650  635994048     99265   \n",
       "32     595191        94047       4968420  39115048  467908260     72917   \n",
       "30     113913        58298        965600   8594170   90936800     14182   \n",
       "26     709104       152345       5934020  47709218  558845060     87099   \n",
       "11     586088        70223       4803044  38331543  452333732     70628   \n",
       "34     130287        55686       1059984   9543504   99825552     15548   \n",
       "18     716375       125909       5863028  47875047  552159284     86176   \n",
       "39     270023        38096       1731660  38364549  357727440     55741   \n",
       "24      51477        23736        362266   8249435   74837144     11672   \n",
       "1      321500        61832       2093926  46613984  432564584     67413   \n",
       "25     263892        36644       1672481  37624081  345502204     53947   \n",
       "3       58636        24185        394661   9186869   81529324     12699   \n",
       "6      322528        60829       2067142  46810950  427031528     66646   \n",
       "10     115944        33585        574613  37052464  283063204     44089   \n",
       "37      22762        12698        121420   7753885   59813360      9330   \n",
       "28     138706        46283        696033  44806349  342876564     53419   \n",
       "36     118176        21603        558467  36412859  275109436     42958   \n",
       "2       26409        11560        133666   8759995   65845928     10259   \n",
       "40     144585        33163        692133  45172854  340955364     53217   \n",
       "41     149192        44905        874437  37640950  294731292     45927   \n",
       "22      29739        18356        188157   7982317   63418812      9894   \n",
       "14     178931        63261       1062594  45623267  358150104     55821   \n",
       "12     149650        29222        864823  37005573  291490868     45515   \n",
       "17      33490        16321        207309   8979034   69874044     10888   \n",
       "21     183140        45543       1072132  45984607  361364912     56403   \n",
       "\n",
       "    multiallelic  in_platinum_STRs  in_segdups  in_STRs            MAF  \\\n",
       "19          5601              7803       22995     3083  654836.500000   \n",
       "16          9941              9679        3451     2011  142638.359375   \n",
       "8          15542             17482       26446     5094  797406.062500   \n",
       "38          4688              6145        7788     2267  622073.750000   \n",
       "33         10155              9844        2301     2584  154979.968750   \n",
       "9          14843             15989       10089     4851  776970.937500   \n",
       "27          5524              8023       22003     3179  601137.375000   \n",
       "13         10062              9783        3216     2018  132927.281250   \n",
       "31         15586             17806       25219     5197  733991.125000   \n",
       "0           5047              6718        8337     2512  592002.562500   \n",
       "4          10694             10343        2354     2655  147723.484375   \n",
       "15         15741             17061       10691     5167  739630.812500   \n",
       "20          6336              9313       25414     3635  615215.687500   \n",
       "7          11178             10917        3802     2309  137374.734375   \n",
       "29         17514             20230       29216     5944  752485.125000   \n",
       "23          5861              7735        9270     2847  602568.375000   \n",
       "5          11858             11522        2804     2989  152445.734375   \n",
       "35         17719             19257       12074     5836  754892.687500   \n",
       "32          5653              8214       22287     3234  610620.625000   \n",
       "30         10202              9933        3351     2079  134204.093750   \n",
       "26         15855             18147       25638     5313  744732.187500   \n",
       "11          5171              6693        8015     2490  598381.562500   \n",
       "34         10745             10432        2467     2708  149029.703125   \n",
       "18         15916             17125       10482     5198  747309.812500   \n",
       "39          4507              6342       17574     2534  598971.187500   \n",
       "24          8635              8403        2696     1723  128819.960938   \n",
       "1          13142             14745       20270     4257  727732.375000   \n",
       "25          4145              5174        6339     1967  587404.062500   \n",
       "3           9032              8749        1920     2264  143459.359375   \n",
       "6          13177             13923        8259     4231  730796.625000   \n",
       "10          3622              4948       14165     1971  578524.687500   \n",
       "37          6888              6700        2108     1397  121080.867188   \n",
       "28         10510             11648       16273     3368  699571.875000   \n",
       "36          3365              4108        5190     1554  568530.812500   \n",
       "2           7310              7076        1517     1890  136792.187500   \n",
       "40         10675             11184        6707     3444  705290.125000   \n",
       "41          3830              5192       14599     2077  587707.687500   \n",
       "22          7421              7220        2163     1448  124648.085938   \n",
       "14         11251             12412       16762     3525  712320.000000   \n",
       "12          3620              4369        5424     1657  577773.437500   \n",
       "17          7862              7609        1581     1954  140212.906250   \n",
       "21         11482             11978        7005     3611  717948.687500   \n",
       "\n",
       "   rounded_MAF  gt_error_rate  switch_error_rate  \n",
       "19         NaN       2.053951           0.993571  \n",
       "16         NaN       6.346631           2.065778  \n",
       "8          NaN       2.762387           1.166606  \n",
       "38         NaN       0.487009           0.612963  \n",
       "33         NaN       4.225765           1.498857  \n",
       "9          NaN       1.174766           0.775578  \n",
       "27         NaN       1.753499           1.254671  \n",
       "13         NaN       6.021769           2.360139  \n",
       "31         NaN       2.460491           1.435920  \n",
       "0          NaN       1.083070           0.884737  \n",
       "4          NaN       5.097639           2.192040  \n",
       "15         NaN       1.811786           1.123247  \n",
       "20         NaN       2.879355           1.421842  \n",
       "7          NaN       6.943843           2.245806  \n",
       "29         NaN       3.525144           1.555279  \n",
       "23         NaN       2.091153           0.984853  \n",
       "5          NaN       5.847290           1.926173  \n",
       "35         NaN       2.746979           1.156192  \n",
       "32         NaN       1.892896           1.182982  \n",
       "30         NaN       6.037490           1.920764  \n",
       "26         NaN       2.567315           1.301502  \n",
       "11         NaN       1.462052           0.839635  \n",
       "34         NaN       5.253476           1.738470  \n",
       "18         NaN       2.147508           1.003106  \n",
       "39         NaN       2.199970           1.229155  \n",
       "24         NaN       6.552092           1.934845  \n",
       "1          NaN       2.952922           1.342146  \n",
       "25         NaN       2.190996           0.847695  \n",
       "3          NaN       6.128044           1.789003  \n",
       "6          NaN       2.942662           1.018826  \n",
       "10         NaN       5.844803           1.562823  \n",
       "37         NaN      10.457915           2.495387  \n",
       "28         NaN       6.649541           1.715859  \n",
       "36         NaN       3.868268           0.883428  \n",
       "2          NaN       8.648422           2.120489  \n",
       "40         NaN       4.791420           1.109382  \n",
       "41         NaN       5.135304           2.269559  \n",
       "22         NaN       9.755683           3.154107  \n",
       "14         NaN       5.953450           2.416574  \n",
       "12         NaN       3.378957           1.630471  \n",
       "17         NaN       7.872789           2.367871  \n",
       "21         NaN       4.247891           1.765316  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "best_phased_variants_overall.sort_values(['ground_truth_data_source','genome','type'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "type",
         "rawType": "category",
         "type": "unknown"
        },
        {
         "name": "Syntenic",
         "rawType": "bool",
         "type": "boolean"
        },
        {
         "name": "method_of_phasing",
         "rawType": "category",
         "type": "unknown"
        },
        {
         "name": "ground_truth_data_source",
         "rawType": "category",
         "type": "unknown"
        },
        {
         "name": "genome",
         "rawType": "category",
         "type": "unknown"
        },
        {
         "name": "variant_id",
         "rawType": "object",
         "type": "unknown"
        },
        {
         "name": "n_switch_errors",
         "rawType": "uint32",
         "type": "integer"
        },
        {
         "name": "n_checked",
         "rawType": "uint32",
         "type": "integer"
        },
        {
         "name": "n_gt_errors",
         "rawType": "uint32",
         "type": "integer"
        },
        {
         "name": "n_gt_checked",
         "rawType": "uint32",
         "type": "integer"
        },
        {
         "name": "MAC",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "AN",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "multiallelic",
         "rawType": "uint64",
         "type": "integer"
        },
        {
         "name": "in_platinum_STRs",
         "rawType": "uint64",
         "type": "integer"
        },
        {
         "name": "in_segdups",
         "rawType": "uint64",
         "type": "integer"
        },
        {
         "name": "in_STRs",
         "rawType": "uint64",
         "type": "integer"
        },
        {
         "name": "MAF",
         "rawType": "float32",
         "type": "float"
        },
        {
         "name": "rounded_MAF",
         "rawType": "category",
         "type": "unknown"
        },
        {
         "name": "gt_error_rate",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "switch_error_rate",
         "rawType": "float64",
         "type": "float"
        }
       ],
       "ref": "c1f8c327-af93-428b-8bf8-b460e7ac80b1",
       "rows": [
        [
         "4",
         "SNP",
         "False",
         "phased_with_parents_and_pedigree",
         "HPRC_samples",
         "GRCh38",
         null,
         "0",
         "882",
         "145",
         "4716",
         "97434",
         "774884",
         "9",
         "7",
         "2",
         "3",
         "1521.4553",
         null,
         "3.0746395250212046",
         "0.0"
        ],
        [
         "30",
         "Indel",
         "False",
         "phased_with_parents_and_pedigree",
         "HPRC_samples",
         "GRCh38",
         null,
         "0",
         "47",
         "22",
         "507",
         "7212",
         "83252",
         "8",
         "7",
         "2",
         "1",
         "112.61711",
         null,
         "4.339250493096647",
         "0.0"
        ],
        [
         "28",
         "SNPs + Indels",
         "False",
         "phased_with_parents_and_pedigree",
         "HPRC_samples",
         "GRCh38",
         null,
         "0",
         "929",
         "167",
         "5223",
         "104646",
         "858136",
         "17",
         "14",
         "4",
         "4",
         "1634.0728",
         null,
         "3.1973961324909053",
         "0.0"
        ],
        [
         "53",
         "SNP",
         "False",
         "phased_with_parents_and_pedigree",
         "HPRC_samples",
         "CHM13v2.0",
         null,
         "0",
         "0",
         "3",
         "39",
         "109",
         "6404",
         "1",
         "0",
         "1",
         "0",
         "1.7020613",
         null,
         "7.6923076923076925",
         null
        ],
        [
         "73",
         "Indel",
         "False",
         "phased_with_parents_and_pedigree",
         "HPRC_samples",
         "CHM13v2.0",
         null,
         "3",
         "22",
         "60",
         "273",
         "6347",
         "44828",
         "5",
         "5",
         "7",
         "3",
         "99.109924",
         null,
         "21.978021978021978",
         "13.636363636363635"
        ],
        [
         "82",
         "SNPs + Indels",
         "False",
         "phased_with_parents_and_pedigree",
         "HPRC_samples",
         "CHM13v2.0",
         null,
         "3",
         "22",
         "63",
         "312",
         "6456",
         "51232",
         "6",
         "5",
         "8",
         "3",
         "100.81199",
         null,
         "20.192307692307693",
         "13.636363636363635"
        ],
        [
         "79",
         "SNP",
         "True",
         "phased_with_parents_and_pedigree",
         "HPRC_samples",
         "GRCh38",
         null,
         "3664",
         "367889",
         "55124",
         "2686147",
         "41848708",
         "443291284",
         "5592",
         "7796",
         "22993",
         "3080",
         "653316.6",
         null,
         "2.05215872400133",
         "0.9959525835238345"
        ],
        [
         "26",
         "Indel",
         "True",
         "phased_with_parents_and_pedigree",
         "HPRC_samples",
         "GRCh38",
         null,
         "1466",
         "70919",
         "33733",
         "531350",
         "9127068",
         "87625932",
         "9933",
         "9672",
         "3449",
         "2010",
         "142525.72",
         null,
         "6.348546156017691",
         "2.067147026889832"
        ],
        [
         "75",
         "SNPs + Indels",
         "True",
         "phased_with_parents_and_pedigree",
         "HPRC_samples",
         "GRCh38",
         null,
         "5130",
         "438808",
         "88857",
         "3217497",
         "50975776",
         "530917216",
         "15525",
         "17468",
         "26442",
         "5090",
         "795773.6",
         null,
         "2.7616808966721647",
         "1.1690762246814095"
        ],
        [
         "59",
         "SNP",
         "True",
         "phased_with_parents_and_pedigree",
         "HPRC_samples",
         "CHM13v2.0",
         null,
         "2170",
         "354018",
         "12169",
         "2499297",
         "39847052",
         "412929920",
         "4687",
         "6145",
         "7787",
         "2267",
         "622072.06",
         null,
         "0.48689691541261404",
         "0.6129631826630284"
        ],
        [
         "10",
         "Indel",
         "True",
         "phased_with_parents_and_pedigree",
         "HPRC_samples",
         "CHM13v2.0",
         null,
         "1190",
         "79572",
         "23748",
         "563128",
         "9918225",
         "93075736",
         "10150",
         "9839",
         "2294",
         "2581",
         "154880.88",
         null,
         "4.217158443551022",
         "1.495500929975368"
        ],
        [
         "77",
         "SNPs + Indels",
         "True",
         "phased_with_parents_and_pedigree",
         "HPRC_samples",
         "CHM13v2.0",
         null,
         "3360",
         "433590",
         "35917",
         "3062425",
         "49765277",
         "506005656",
         "14837",
         "15984",
         "10081",
         "4848",
         "776870.1",
         null,
         "1.1728287223360572",
         "0.7749256209783436"
        ]
       ],
       "shape": {
        "columns": 20,
        "rows": 12
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>Syntenic</th>\n",
       "      <th>method_of_phasing</th>\n",
       "      <th>ground_truth_data_source</th>\n",
       "      <th>genome</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>n_switch_errors</th>\n",
       "      <th>n_checked</th>\n",
       "      <th>n_gt_errors</th>\n",
       "      <th>n_gt_checked</th>\n",
       "      <th>MAC</th>\n",
       "      <th>AN</th>\n",
       "      <th>multiallelic</th>\n",
       "      <th>in_platinum_STRs</th>\n",
       "      <th>in_segdups</th>\n",
       "      <th>in_STRs</th>\n",
       "      <th>MAF</th>\n",
       "      <th>rounded_MAF</th>\n",
       "      <th>gt_error_rate</th>\n",
       "      <th>switch_error_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SNP</td>\n",
       "      <td>False</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HPRC_samples</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>882</td>\n",
       "      <td>145</td>\n",
       "      <td>4716</td>\n",
       "      <td>97434</td>\n",
       "      <td>774884</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>1521.455322</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.074640</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Indel</td>\n",
       "      <td>False</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HPRC_samples</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>47</td>\n",
       "      <td>22</td>\n",
       "      <td>507</td>\n",
       "      <td>7212</td>\n",
       "      <td>83252</td>\n",
       "      <td>8</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>112.617111</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.339250</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>SNPs + Indels</td>\n",
       "      <td>False</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HPRC_samples</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>929</td>\n",
       "      <td>167</td>\n",
       "      <td>5223</td>\n",
       "      <td>104646</td>\n",
       "      <td>858136</td>\n",
       "      <td>17</td>\n",
       "      <td>14</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>1634.072754</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.197396</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>SNP</td>\n",
       "      <td>False</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HPRC_samples</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>39</td>\n",
       "      <td>109</td>\n",
       "      <td>6404</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1.702061</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.692308</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>Indel</td>\n",
       "      <td>False</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HPRC_samples</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>None</td>\n",
       "      <td>3</td>\n",
       "      <td>22</td>\n",
       "      <td>60</td>\n",
       "      <td>273</td>\n",
       "      <td>6347</td>\n",
       "      <td>44828</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>99.109924</td>\n",
       "      <td>NaN</td>\n",
       "      <td>21.978022</td>\n",
       "      <td>13.636364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>SNPs + Indels</td>\n",
       "      <td>False</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HPRC_samples</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>None</td>\n",
       "      <td>3</td>\n",
       "      <td>22</td>\n",
       "      <td>63</td>\n",
       "      <td>312</td>\n",
       "      <td>6456</td>\n",
       "      <td>51232</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>100.811989</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20.192308</td>\n",
       "      <td>13.636364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>SNP</td>\n",
       "      <td>True</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HPRC_samples</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>3664</td>\n",
       "      <td>367889</td>\n",
       "      <td>55124</td>\n",
       "      <td>2686147</td>\n",
       "      <td>41848708</td>\n",
       "      <td>443291284</td>\n",
       "      <td>5592</td>\n",
       "      <td>7796</td>\n",
       "      <td>22993</td>\n",
       "      <td>3080</td>\n",
       "      <td>653316.625000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.052159</td>\n",
       "      <td>0.995953</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Indel</td>\n",
       "      <td>True</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HPRC_samples</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>1466</td>\n",
       "      <td>70919</td>\n",
       "      <td>33733</td>\n",
       "      <td>531350</td>\n",
       "      <td>9127068</td>\n",
       "      <td>87625932</td>\n",
       "      <td>9933</td>\n",
       "      <td>9672</td>\n",
       "      <td>3449</td>\n",
       "      <td>2010</td>\n",
       "      <td>142525.718750</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.348546</td>\n",
       "      <td>2.067147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>SNPs + Indels</td>\n",
       "      <td>True</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HPRC_samples</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>None</td>\n",
       "      <td>5130</td>\n",
       "      <td>438808</td>\n",
       "      <td>88857</td>\n",
       "      <td>3217497</td>\n",
       "      <td>50975776</td>\n",
       "      <td>530917216</td>\n",
       "      <td>15525</td>\n",
       "      <td>17468</td>\n",
       "      <td>26442</td>\n",
       "      <td>5090</td>\n",
       "      <td>795773.625000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.761681</td>\n",
       "      <td>1.169076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>SNP</td>\n",
       "      <td>True</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HPRC_samples</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>None</td>\n",
       "      <td>2170</td>\n",
       "      <td>354018</td>\n",
       "      <td>12169</td>\n",
       "      <td>2499297</td>\n",
       "      <td>39847052</td>\n",
       "      <td>412929920</td>\n",
       "      <td>4687</td>\n",
       "      <td>6145</td>\n",
       "      <td>7787</td>\n",
       "      <td>2267</td>\n",
       "      <td>622072.062500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.486897</td>\n",
       "      <td>0.612963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Indel</td>\n",
       "      <td>True</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HPRC_samples</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>None</td>\n",
       "      <td>1190</td>\n",
       "      <td>79572</td>\n",
       "      <td>23748</td>\n",
       "      <td>563128</td>\n",
       "      <td>9918225</td>\n",
       "      <td>93075736</td>\n",
       "      <td>10150</td>\n",
       "      <td>9839</td>\n",
       "      <td>2294</td>\n",
       "      <td>2581</td>\n",
       "      <td>154880.875000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.217158</td>\n",
       "      <td>1.495501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>SNPs + Indels</td>\n",
       "      <td>True</td>\n",
       "      <td>phased_with_parents_and_pedigree</td>\n",
       "      <td>HPRC_samples</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>None</td>\n",
       "      <td>3360</td>\n",
       "      <td>433590</td>\n",
       "      <td>35917</td>\n",
       "      <td>3062425</td>\n",
       "      <td>49765277</td>\n",
       "      <td>506005656</td>\n",
       "      <td>14837</td>\n",
       "      <td>15984</td>\n",
       "      <td>10081</td>\n",
       "      <td>4848</td>\n",
       "      <td>776870.125000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.172829</td>\n",
       "      <td>0.774926</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             type  Syntenic                 method_of_phasing  \\\n",
       "4             SNP     False  phased_with_parents_and_pedigree   \n",
       "30          Indel     False  phased_with_parents_and_pedigree   \n",
       "28  SNPs + Indels     False  phased_with_parents_and_pedigree   \n",
       "53            SNP     False  phased_with_parents_and_pedigree   \n",
       "73          Indel     False  phased_with_parents_and_pedigree   \n",
       "82  SNPs + Indels     False  phased_with_parents_and_pedigree   \n",
       "79            SNP      True  phased_with_parents_and_pedigree   \n",
       "26          Indel      True  phased_with_parents_and_pedigree   \n",
       "75  SNPs + Indels      True  phased_with_parents_and_pedigree   \n",
       "59            SNP      True  phased_with_parents_and_pedigree   \n",
       "10          Indel      True  phased_with_parents_and_pedigree   \n",
       "77  SNPs + Indels      True  phased_with_parents_and_pedigree   \n",
       "\n",
       "   ground_truth_data_source     genome variant_id  n_switch_errors  n_checked  \\\n",
       "4              HPRC_samples     GRCh38       None                0        882   \n",
       "30             HPRC_samples     GRCh38       None                0         47   \n",
       "28             HPRC_samples     GRCh38       None                0        929   \n",
       "53             HPRC_samples  CHM13v2.0       None                0          0   \n",
       "73             HPRC_samples  CHM13v2.0       None                3         22   \n",
       "82             HPRC_samples  CHM13v2.0       None                3         22   \n",
       "79             HPRC_samples     GRCh38       None             3664     367889   \n",
       "26             HPRC_samples     GRCh38       None             1466      70919   \n",
       "75             HPRC_samples     GRCh38       None             5130     438808   \n",
       "59             HPRC_samples  CHM13v2.0       None             2170     354018   \n",
       "10             HPRC_samples  CHM13v2.0       None             1190      79572   \n",
       "77             HPRC_samples  CHM13v2.0       None             3360     433590   \n",
       "\n",
       "    n_gt_errors  n_gt_checked       MAC         AN  multiallelic  \\\n",
       "4           145          4716     97434     774884             9   \n",
       "30           22           507      7212      83252             8   \n",
       "28          167          5223    104646     858136            17   \n",
       "53            3            39       109       6404             1   \n",
       "73           60           273      6347      44828             5   \n",
       "82           63           312      6456      51232             6   \n",
       "79        55124       2686147  41848708  443291284          5592   \n",
       "26        33733        531350   9127068   87625932          9933   \n",
       "75        88857       3217497  50975776  530917216         15525   \n",
       "59        12169       2499297  39847052  412929920          4687   \n",
       "10        23748        563128   9918225   93075736         10150   \n",
       "77        35917       3062425  49765277  506005656         14837   \n",
       "\n",
       "    in_platinum_STRs  in_segdups  in_STRs            MAF rounded_MAF  \\\n",
       "4                  7           2        3    1521.455322         NaN   \n",
       "30                 7           2        1     112.617111         NaN   \n",
       "28                14           4        4    1634.072754         NaN   \n",
       "53                 0           1        0       1.702061         NaN   \n",
       "73                 5           7        3      99.109924         NaN   \n",
       "82                 5           8        3     100.811989         NaN   \n",
       "79              7796       22993     3080  653316.625000         NaN   \n",
       "26              9672        3449     2010  142525.718750         NaN   \n",
       "75             17468       26442     5090  795773.625000         NaN   \n",
       "59              6145        7787     2267  622072.062500         NaN   \n",
       "10              9839        2294     2581  154880.875000         NaN   \n",
       "77             15984       10081     4848  776870.125000         NaN   \n",
       "\n",
       "    gt_error_rate  switch_error_rate  \n",
       "4        3.074640           0.000000  \n",
       "30       4.339250           0.000000  \n",
       "28       3.197396           0.000000  \n",
       "53       7.692308                NaN  \n",
       "73      21.978022          13.636364  \n",
       "82      20.192308          13.636364  \n",
       "79       2.052159           0.995953  \n",
       "26       6.348546           2.067147  \n",
       "75       2.761681           1.169076  \n",
       "59       0.486897           0.612963  \n",
       "10       4.217158           1.495501  \n",
       "77       1.172829           0.774926  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "best_phased_variants_by_syntenic.loc[best_phased_variants_by_syntenic.ground_truth_data_source=='HPRC_samples'].sort_values(['ground_truth_data_source','Syntenic','genome','type'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Percent concordance in different panels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "overall_subsetted = MAF_performance_variant_df.filter((pl.col('ground_truth_data_source')!='trios')&(pl.col('method_of_phasing')=='phased_with_parents_and_pedigree')).group_by(['type','Syntenic','method_of_phasing','ground_truth_data_source','genome']).sum().with_columns(gt_error_rate = pl.col('n_gt_errors')/pl.col('n_gt_checked'))\n",
    "overall           = MAF_performance_variant_df.filter((pl.col('ground_truth_data_source')!='trios')&(pl.col('method_of_phasing')=='phased_with_parents_and_pedigree')).group_by(['method_of_phasing','ground_truth_data_source','genome']).sum().with_columns(gt_error_rate = pl.col('n_gt_errors')/pl.col('n_gt_checked'))\n",
    "overall_syntenic  = MAF_performance_variant_df.filter((pl.col('ground_truth_data_source')!='trios')&(pl.col('Syntenic'))&(pl.col('method_of_phasing')=='phased_with_parents_and_pedigree')).group_by(['method_of_phasing','ground_truth_data_source','genome']).sum().with_columns(gt_error_rate = pl.col('n_gt_errors')/pl.col('n_gt_checked'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "Syntenic",
         "rawType": "uint64",
         "type": "integer"
        },
        {
         "name": "genome",
         "rawType": "category",
         "type": "unknown"
        },
        {
         "name": "ground_truth_data_source",
         "rawType": "category",
         "type": "unknown"
        },
        {
         "name": "gt_error_rate",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "Non-reference true positive rate",
         "rawType": "float64",
         "type": "float"
        }
       ],
       "ref": "6ec39355-4232-4ff4-8f2d-4363f16071fe",
       "rows": [
        [
         "4",
         "82904",
         "GRCh38",
         "HPRC_samples",
         "2.7616808966721647",
         "97.23831910332783"
        ],
        [
         "5",
         "83281",
         "GRCh38",
         "HGSVC_samples",
         "2.4631944050994248",
         "97.53680559490057"
        ],
        [
         "1",
         "98672",
         "GRCh38",
         "HPRC_HGSVC_all_samples",
         "3.5226609372466355",
         "96.47733906275336"
        ],
        [
         "2",
         "79014",
         "CHM13v2.0",
         "HPRC_samples",
         "1.1728287223360572",
         "98.82717127766394"
        ],
        [
         "13",
         "84344",
         "CHM13v2.0",
         "HGSVC_samples",
         "1.8060575040440336",
         "98.19394249595597"
        ],
        [
         "3",
         "99265",
         "CHM13v2.0",
         "HPRC_HGSVC_all_samples",
         "2.7405832871606304",
         "97.25941671283937"
        ]
       ],
       "shape": {
        "columns": 5,
        "rows": 6
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Syntenic</th>\n",
       "      <th>genome</th>\n",
       "      <th>ground_truth_data_source</th>\n",
       "      <th>gt_error_rate</th>\n",
       "      <th>Non-reference true positive rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>82904</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>HPRC_samples</td>\n",
       "      <td>2.761681</td>\n",
       "      <td>97.238319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>83281</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>HGSVC_samples</td>\n",
       "      <td>2.463194</td>\n",
       "      <td>97.536806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>98672</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>HPRC_HGSVC_all_samples</td>\n",
       "      <td>3.522661</td>\n",
       "      <td>96.477339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>79014</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>HPRC_samples</td>\n",
       "      <td>1.172829</td>\n",
       "      <td>98.827171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>84344</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>HGSVC_samples</td>\n",
       "      <td>1.806058</td>\n",
       "      <td>98.193942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>99265</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>HPRC_HGSVC_all_samples</td>\n",
       "      <td>2.740583</td>\n",
       "      <td>97.259417</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Syntenic     genome ground_truth_data_source  gt_error_rate  \\\n",
       "4      82904     GRCh38             HPRC_samples       2.761681   \n",
       "5      83281     GRCh38            HGSVC_samples       2.463194   \n",
       "1      98672     GRCh38   HPRC_HGSVC_all_samples       3.522661   \n",
       "2      79014  CHM13v2.0             HPRC_samples       1.172829   \n",
       "13     84344  CHM13v2.0            HGSVC_samples       1.806058   \n",
       "3      99265  CHM13v2.0   HPRC_HGSVC_all_samples       2.740583   \n",
       "\n",
       "    Non-reference true positive rate  \n",
       "4                          97.238319  \n",
       "5                          97.536806  \n",
       "1                          96.477339  \n",
       "2                          98.827171  \n",
       "13                         98.193942  \n",
       "3                          97.259417  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overall_syntenic = overall_syntenic.to_pandas()[['Syntenic','genome','ground_truth_data_source','gt_error_rate']].sort_values(['genome','Syntenic','gt_error_rate'])\n",
    "overall_syntenic['Non-reference true positive rate'] = (1-overall_syntenic.gt_error_rate) * 100\n",
    "overall_syntenic['gt_error_rate'] *= 100\n",
    "overall_syntenic.loc[overall_syntenic.ground_truth_data_source.isin(('HPRC_samples', 'HGSVC_samples','HPRC_HGSVC_all_samples'))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "Syntenic",
         "rawType": "uint64",
         "type": "integer"
        },
        {
         "name": "genome",
         "rawType": "category",
         "type": "unknown"
        },
        {
         "name": "ground_truth_data_source",
         "rawType": "category",
         "type": "unknown"
        },
        {
         "name": "gt_error_rate",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "Non-reference true positive rate",
         "rawType": "float64",
         "type": "float"
        }
       ],
       "ref": "f25aeed8-1f98-4e2a-ab9a-b7b89939b6b8",
       "rows": [
        [
         "2",
         "82904",
         "GRCh38",
         "HPRC_samples",
         "2.762387051931288",
         "97.2376129480687"
        ],
        [
         "6",
         "83281",
         "GRCh38",
         "HGSVC_samples",
         "2.4604910136290243",
         "97.53950898637098"
        ],
        [
         "1",
         "98672",
         "GRCh38",
         "HPRC_HGSVC_all_samples",
         "3.5251439034506475",
         "96.47485609654936"
        ],
        [
         "4",
         "79014",
         "CHM13v2.0",
         "HPRC_samples",
         "1.1747662303358075",
         "98.8252337696642"
        ],
        [
         "7",
         "84344",
         "CHM13v2.0",
         "HGSVC_samples",
         "1.8117857333496339",
         "98.18821426665036"
        ],
        [
         "8",
         "99265",
         "CHM13v2.0",
         "HPRC_HGSVC_all_samples",
         "2.74697921701305",
         "97.25302078298695"
        ]
       ],
       "shape": {
        "columns": 5,
        "rows": 6
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Syntenic</th>\n",
       "      <th>genome</th>\n",
       "      <th>ground_truth_data_source</th>\n",
       "      <th>gt_error_rate</th>\n",
       "      <th>Non-reference true positive rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>82904</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>HPRC_samples</td>\n",
       "      <td>2.762387</td>\n",
       "      <td>97.237613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>83281</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>HGSVC_samples</td>\n",
       "      <td>2.460491</td>\n",
       "      <td>97.539509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>98672</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>HPRC_HGSVC_all_samples</td>\n",
       "      <td>3.525144</td>\n",
       "      <td>96.474856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>79014</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>HPRC_samples</td>\n",
       "      <td>1.174766</td>\n",
       "      <td>98.825234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>84344</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>HGSVC_samples</td>\n",
       "      <td>1.811786</td>\n",
       "      <td>98.188214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>99265</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>HPRC_HGSVC_all_samples</td>\n",
       "      <td>2.746979</td>\n",
       "      <td>97.253021</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Syntenic     genome ground_truth_data_source  gt_error_rate  \\\n",
       "2     82904     GRCh38             HPRC_samples       2.762387   \n",
       "6     83281     GRCh38            HGSVC_samples       2.460491   \n",
       "1     98672     GRCh38   HPRC_HGSVC_all_samples       3.525144   \n",
       "4     79014  CHM13v2.0             HPRC_samples       1.174766   \n",
       "7     84344  CHM13v2.0            HGSVC_samples       1.811786   \n",
       "8     99265  CHM13v2.0   HPRC_HGSVC_all_samples       2.746979   \n",
       "\n",
       "   Non-reference true positive rate  \n",
       "2                         97.237613  \n",
       "6                         97.539509  \n",
       "1                         96.474856  \n",
       "4                         98.825234  \n",
       "7                         98.188214  \n",
       "8                         97.253021  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overall = overall.to_pandas()[['Syntenic','genome','ground_truth_data_source','gt_error_rate']].sort_values(['genome','Syntenic','gt_error_rate'])\n",
    "overall['Non-reference true positive rate'] = (1-overall.gt_error_rate) * 100\n",
    "overall['gt_error_rate'] *= 100\n",
    "overall.loc[overall.ground_truth_data_source.isin(('HPRC_samples', 'HGSVC_samples','HPRC_HGSVC_all_samples'))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "type",
         "rawType": "category",
         "type": "unknown"
        },
        {
         "name": "Syntenic",
         "rawType": "bool",
         "type": "boolean"
        },
        {
         "name": "genome",
         "rawType": "category",
         "type": "unknown"
        },
        {
         "name": "ground_truth_data_source",
         "rawType": "category",
         "type": "unknown"
        },
        {
         "name": "gt_error_rate",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "Non-reference true positive rate",
         "rawType": "float64",
         "type": "float"
        }
       ],
       "ref": "65d170a9-a5b8-4387-aa8b-30c119654a7b",
       "rows": [
        [
         "17",
         "SNP",
         "False",
         "GRCh38",
         "HGSVC_samples",
         "0.771055753262159",
         "99.22894424673784"
        ],
        [
         "14",
         "SNP",
         "False",
         "GRCh38",
         "HPRC_samples",
         "3.0746395250212046",
         "96.92536047497879"
        ],
        [
         "54",
         "SNP",
         "False",
         "GRCh38",
         "HPRC_HGSVC_all_samples",
         "5.0",
         "95.0"
        ],
        [
         "55",
         "SNP",
         "True",
         "GRCh38",
         "HGSVC_samples",
         "1.7554252191228945",
         "98.2445747808771"
        ],
        [
         "24",
         "SNP",
         "True",
         "GRCh38",
         "HPRC_samples",
         "2.05215872400133",
         "97.94784127599867"
        ],
        [
         "33",
         "SNP",
         "True",
         "GRCh38",
         "HPRC_HGSVC_all_samples",
         "2.875292217964476",
         "97.12470778203553"
        ],
        [
         "7",
         "Indel",
         "False",
         "GRCh38",
         "HGSVC_samples",
         "2.7540360873694207",
         "97.24596391263059"
        ],
        [
         "1",
         "Indel",
         "False",
         "GRCh38",
         "HPRC_HGSVC_all_samples",
         "4.05",
         "95.95"
        ],
        [
         "39",
         "Indel",
         "False",
         "GRCh38",
         "HPRC_samples",
         "4.339250493096647",
         "95.66074950690336"
        ],
        [
         "46",
         "Indel",
         "True",
         "GRCh38",
         "HGSVC_samples",
         "6.025797418385051",
         "93.97420258161495"
        ],
        [
         "37",
         "Indel",
         "True",
         "GRCh38",
         "HPRC_samples",
         "6.348546156017691",
         "93.65145384398231"
        ],
        [
         "51",
         "Indel",
         "True",
         "GRCh38",
         "HPRC_HGSVC_all_samples",
         "6.9475328318245575",
         "93.05246716817544"
        ],
        [
         "28",
         "SNP",
         "False",
         "CHM13v2.0",
         "HPRC_samples",
         "7.6923076923076925",
         "92.3076923076923"
        ],
        [
         "10",
         "SNP",
         "False",
         "CHM13v2.0",
         "HGSVC_samples",
         "15.591397849462366",
         "84.40860215053763"
        ],
        [
         "42",
         "SNP",
         "False",
         "CHM13v2.0",
         "HPRC_HGSVC_all_samples",
         "17.2",
         "82.80000000000001"
        ],
        [
         "21",
         "SNP",
         "True",
         "CHM13v2.0",
         "HPRC_samples",
         "0.48689691541261404",
         "99.51310308458739"
        ],
        [
         "19",
         "SNP",
         "True",
         "CHM13v2.0",
         "HGSVC_samples",
         "1.082439306958328",
         "98.91756069304168"
        ],
        [
         "22",
         "SNP",
         "True",
         "CHM13v2.0",
         "HPRC_HGSVC_all_samples",
         "2.09023143460173",
         "97.90976856539827"
        ],
        [
         "2",
         "Indel",
         "False",
         "CHM13v2.0",
         "HGSVC_samples",
         "14.093357271095153",
         "85.90664272890484"
        ],
        [
         "44",
         "Indel",
         "False",
         "CHM13v2.0",
         "HPRC_HGSVC_all_samples",
         "16.142857142857142",
         "83.85714285714285"
        ],
        [
         "25",
         "Indel",
         "False",
         "CHM13v2.0",
         "HPRC_samples",
         "21.978021978021978",
         "78.02197802197803"
        ],
        [
         "0",
         "Indel",
         "True",
         "CHM13v2.0",
         "HPRC_samples",
         "4.217158443551022",
         "95.78284155644899"
        ],
        [
         "5",
         "Indel",
         "True",
         "CHM13v2.0",
         "HGSVC_samples",
         "5.076463907450939",
         "94.92353609254907"
        ],
        [
         "12",
         "Indel",
         "True",
         "CHM13v2.0",
         "HPRC_HGSVC_all_samples",
         "5.822291594403977",
         "94.17770840559602"
        ]
       ],
       "shape": {
        "columns": 6,
        "rows": 24
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>Syntenic</th>\n",
       "      <th>genome</th>\n",
       "      <th>ground_truth_data_source</th>\n",
       "      <th>gt_error_rate</th>\n",
       "      <th>Non-reference true positive rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>SNP</td>\n",
       "      <td>False</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>HGSVC_samples</td>\n",
       "      <td>0.771056</td>\n",
       "      <td>99.228944</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SNP</td>\n",
       "      <td>False</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>HPRC_samples</td>\n",
       "      <td>3.074640</td>\n",
       "      <td>96.925360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>SNP</td>\n",
       "      <td>False</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>HPRC_HGSVC_all_samples</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>95.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>SNP</td>\n",
       "      <td>True</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>HGSVC_samples</td>\n",
       "      <td>1.755425</td>\n",
       "      <td>98.244575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>SNP</td>\n",
       "      <td>True</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>HPRC_samples</td>\n",
       "      <td>2.052159</td>\n",
       "      <td>97.947841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>SNP</td>\n",
       "      <td>True</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>HPRC_HGSVC_all_samples</td>\n",
       "      <td>2.875292</td>\n",
       "      <td>97.124708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Indel</td>\n",
       "      <td>False</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>HGSVC_samples</td>\n",
       "      <td>2.754036</td>\n",
       "      <td>97.245964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Indel</td>\n",
       "      <td>False</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>HPRC_HGSVC_all_samples</td>\n",
       "      <td>4.050000</td>\n",
       "      <td>95.950000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Indel</td>\n",
       "      <td>False</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>HPRC_samples</td>\n",
       "      <td>4.339250</td>\n",
       "      <td>95.660750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>Indel</td>\n",
       "      <td>True</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>HGSVC_samples</td>\n",
       "      <td>6.025797</td>\n",
       "      <td>93.974203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Indel</td>\n",
       "      <td>True</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>HPRC_samples</td>\n",
       "      <td>6.348546</td>\n",
       "      <td>93.651454</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>Indel</td>\n",
       "      <td>True</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>HPRC_HGSVC_all_samples</td>\n",
       "      <td>6.947533</td>\n",
       "      <td>93.052467</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>SNP</td>\n",
       "      <td>False</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>HPRC_samples</td>\n",
       "      <td>7.692308</td>\n",
       "      <td>92.307692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SNP</td>\n",
       "      <td>False</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>HGSVC_samples</td>\n",
       "      <td>15.591398</td>\n",
       "      <td>84.408602</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>SNP</td>\n",
       "      <td>False</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>HPRC_HGSVC_all_samples</td>\n",
       "      <td>17.200000</td>\n",
       "      <td>82.800000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>SNP</td>\n",
       "      <td>True</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>HPRC_samples</td>\n",
       "      <td>0.486897</td>\n",
       "      <td>99.513103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>SNP</td>\n",
       "      <td>True</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>HGSVC_samples</td>\n",
       "      <td>1.082439</td>\n",
       "      <td>98.917561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>SNP</td>\n",
       "      <td>True</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>HPRC_HGSVC_all_samples</td>\n",
       "      <td>2.090231</td>\n",
       "      <td>97.909769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Indel</td>\n",
       "      <td>False</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>HGSVC_samples</td>\n",
       "      <td>14.093357</td>\n",
       "      <td>85.906643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>Indel</td>\n",
       "      <td>False</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>HPRC_HGSVC_all_samples</td>\n",
       "      <td>16.142857</td>\n",
       "      <td>83.857143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Indel</td>\n",
       "      <td>False</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>HPRC_samples</td>\n",
       "      <td>21.978022</td>\n",
       "      <td>78.021978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Indel</td>\n",
       "      <td>True</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>HPRC_samples</td>\n",
       "      <td>4.217158</td>\n",
       "      <td>95.782842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Indel</td>\n",
       "      <td>True</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>HGSVC_samples</td>\n",
       "      <td>5.076464</td>\n",
       "      <td>94.923536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Indel</td>\n",
       "      <td>True</td>\n",
       "      <td>CHM13v2.0</td>\n",
       "      <td>HPRC_HGSVC_all_samples</td>\n",
       "      <td>5.822292</td>\n",
       "      <td>94.177708</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     type  Syntenic     genome ground_truth_data_source  gt_error_rate  \\\n",
       "17    SNP     False     GRCh38            HGSVC_samples       0.771056   \n",
       "14    SNP     False     GRCh38             HPRC_samples       3.074640   \n",
       "54    SNP     False     GRCh38   HPRC_HGSVC_all_samples       5.000000   \n",
       "55    SNP      True     GRCh38            HGSVC_samples       1.755425   \n",
       "24    SNP      True     GRCh38             HPRC_samples       2.052159   \n",
       "33    SNP      True     GRCh38   HPRC_HGSVC_all_samples       2.875292   \n",
       "7   Indel     False     GRCh38            HGSVC_samples       2.754036   \n",
       "1   Indel     False     GRCh38   HPRC_HGSVC_all_samples       4.050000   \n",
       "39  Indel     False     GRCh38             HPRC_samples       4.339250   \n",
       "46  Indel      True     GRCh38            HGSVC_samples       6.025797   \n",
       "37  Indel      True     GRCh38             HPRC_samples       6.348546   \n",
       "51  Indel      True     GRCh38   HPRC_HGSVC_all_samples       6.947533   \n",
       "28    SNP     False  CHM13v2.0             HPRC_samples       7.692308   \n",
       "10    SNP     False  CHM13v2.0            HGSVC_samples      15.591398   \n",
       "42    SNP     False  CHM13v2.0   HPRC_HGSVC_all_samples      17.200000   \n",
       "21    SNP      True  CHM13v2.0             HPRC_samples       0.486897   \n",
       "19    SNP      True  CHM13v2.0            HGSVC_samples       1.082439   \n",
       "22    SNP      True  CHM13v2.0   HPRC_HGSVC_all_samples       2.090231   \n",
       "2   Indel     False  CHM13v2.0            HGSVC_samples      14.093357   \n",
       "44  Indel     False  CHM13v2.0   HPRC_HGSVC_all_samples      16.142857   \n",
       "25  Indel     False  CHM13v2.0             HPRC_samples      21.978022   \n",
       "0   Indel      True  CHM13v2.0             HPRC_samples       4.217158   \n",
       "5   Indel      True  CHM13v2.0            HGSVC_samples       5.076464   \n",
       "12  Indel      True  CHM13v2.0   HPRC_HGSVC_all_samples       5.822292   \n",
       "\n",
       "    Non-reference true positive rate  \n",
       "17                         99.228944  \n",
       "14                         96.925360  \n",
       "54                         95.000000  \n",
       "55                         98.244575  \n",
       "24                         97.947841  \n",
       "33                         97.124708  \n",
       "7                          97.245964  \n",
       "1                          95.950000  \n",
       "39                         95.660750  \n",
       "46                         93.974203  \n",
       "37                         93.651454  \n",
       "51                         93.052467  \n",
       "28                         92.307692  \n",
       "10                         84.408602  \n",
       "42                         82.800000  \n",
       "21                         99.513103  \n",
       "19                         98.917561  \n",
       "22                         97.909769  \n",
       "2                          85.906643  \n",
       "44                         83.857143  \n",
       "25                         78.021978  \n",
       "0                          95.782842  \n",
       "5                          94.923536  \n",
       "12                         94.177708  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overall_subsetted = overall_subsetted.to_pandas()[['type','Syntenic','genome','ground_truth_data_source','gt_error_rate']].sort_values(['genome','type','Syntenic','gt_error_rate'])\n",
    "overall_subsetted['Non-reference true positive rate'] = (1-overall_subsetted.gt_error_rate) * 100\n",
    "overall_subsetted['gt_error_rate'] *= 100\n",
    "overall_subsetted.loc[overall_subsetted.ground_truth_data_source.isin(('HPRC_samples', 'HGSVC_samples','HPRC_HGSVC_all_samples'))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# wide_genome_ideogram['SER_diff'] = wide_genome_ideogram['SER_t2t'] - wide_genome_ideogram['SER_grch38']\n",
    "# region_table = (wide_genome_ideogram.loc[(wide_genome_ideogram.ground_truth_data_source=='HPRC_samples') & \n",
    "#                          (wide_genome_ideogram.method_of_phasing=='HPRC_variation_phased_with_reference_panel') #& \n",
    "#                         & (wide_genome_ideogram.contextual_MAC_grch38 > 1e6)\n",
    "#                         & (wide_genome_ideogram.contextual_MAC_t2t > 1e6)\n",
    "#                         ])\n",
    "# decipher_regions = ['1p36.33','1p36.32','1p36.31','1p36.23','1p36.22','1p36.21','1q21.1','1q21.2','2p21','2p16.1','2p15','2q32.3','2q33.1','2q33.2','2q33.3','2q37.3','3q29','4p16.3','5p15.33','5p15.32','5p15.31','5p15.2','5q22.3','5q23.2','5q35.2','5q35.3','7q11.22','7q21.3','8p23.1','8q21.13','9q34.3','11p13','11p11.2','12q14.2','12q14.3','12q15','15q11.2','15q12','15q13.1','15q13.2','15q13.3','15q24.1','15q24.2','15q26.3','16p13.3','16p13.11','16p12.2','16p12.1','16p11.2','17p13.3','17p12','17p11.2','17q11.2','17q12','17q21.31','21q21.3','22p13','22p12','22p11.2','22p11.1','22q11.1','22q11.21','22q11.22','22q11.23','22q13.33','Xp22.33','Xp22.31','Xp11.23','Xp11.22','Xq22.2','Xq28']\n",
    "# region_table.interval_names = region_table.chrom.str.replace('chr','') + region_table.interval_names\n",
    "# region_table['decipher_region'] = region_table.interval_names.isin(decipher_regions)\n",
    "\n",
    "# region_table = (region_table.sort_values('SER_diff').iloc[:].reset_index(drop=True)\n",
    "#                         [['chrom','interval_names','decipher_region','n_switch_errors_grch38','n_checked_grch38','n_switch_errors_t2t','n_checked_t2t','SER_grch38','SER_t2t','SER_diff']])\n",
    "\n",
    "# region_table = region_table.drop(columns='chrom').rename(columns={'interval_names':'Genetic Region',\n",
    "#                                                    'n_switch_errors_grch38': 'Switch Errors (GRCh38 panel)',\n",
    "#                                                    'n_switch_errors_t2t': 'Switch Errors (CHM13 panel)',\n",
    "#                                                    'n_checked_grch38': 'Heterozygous sites (GRCh38 panel)',\n",
    "#                                                    'n_checked_t2t': 'Heterozygous sites (CHM13 panel)',\n",
    "#                                                    'SER_grch38': 'SER % (GRCh38 panel)',\n",
    "#                                                    'SER_t2t': 'SER % (CHM13 panel)',\n",
    "#                                                    'SER_diff': 'Absolute reduction in SER (% points)'})\n",
    "\n",
    "# pd.set_option('display.float_format', lambda x: f'{x:,.2f}%')\n",
    "# region_table.set_index('Genetic Region')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Singletons in GRCh38: 14.39873417721519\n",
      "Singletons in CHM13v2.0: 9.825145711906744\n",
      "\n",
      "Rare in CHM13v2.0: 6.018179918503814\n",
      "Common in CHM13v2.0: 1.1376264478260545\n",
      "\n",
      "Rare in GRCh38: 7.419862621637092\n",
      "Common in GRCh38: 1.6314417692446646\n",
      "\n",
      "Common Syntenic in GRCh38: 1.635230560250521\n",
      "Common Nonsyntenic GRCh38: 0.798175598631699\n",
      "\n",
      "Common Syntenic in CHM13v2.0: 1.1374146006742445\n",
      "Common Nonsyntenic CHM13v2.0: 7.6923076923076925\n"
     ]
    }
   ],
   "source": [
    "rephase=all_variants.filter((pl.col('method_of_phasing') == '1kgp_variation_phased_with_reference_panel') & \n",
    "                         (pl.col('ground_truth_data_source') == 'HPRC_samples'))\n",
    "\n",
    "singletons_rephase_sum = rephase.filter(pl.col('contextual_MAC') == 1).group_by('genome').sum().to_pandas().reset_index()\n",
    "singletons_rephase_sum['SER'] = singletons_rephase_sum.n_switch_errors/singletons_rephase_sum.n_checked*100\n",
    "for idx, (genome, ser) in singletons_rephase_sum[['genome','SER']].iterrows():\n",
    "    print(f'Singletons in {genome}: {ser}')\n",
    "\n",
    "rare_rephase = rephase.filter(pl.col('contextual_MAC')/pl.col('contextual_AN') < 0.01)\n",
    "common_rephase = rephase.filter(pl.col('contextual_MAC')/pl.col('contextual_AN') >= 0.01)\n",
    "\n",
    "rare_rephase_sum = rare_rephase.group_by('genome').sum().to_pandas().reset_index()\n",
    "common_rephase_sum = common_rephase.group_by('genome').sum().to_pandas().reset_index()\n",
    "for idx, (genome, n_switch_errors, n_checked) in rare_rephase_sum[['genome','n_switch_errors', 'n_checked']].iterrows():\n",
    "    print (f\"\\nRare in {genome}: {n_switch_errors/n_checked*100}\")\n",
    "    print (f\"Common in {genome}: {common_rephase_sum.iloc[idx].n_switch_errors/common_rephase_sum.iloc[idx].n_checked*100}\")\n",
    "syntenic = common_rephase.filter(pl.col('Syntenic') == True)\n",
    "nonsyntenic = common_rephase.filter(pl.col('Syntenic') == False)\n",
    "\n",
    "syntenic_sum = syntenic.group_by('genome').sum().to_pandas().set_index('genome')\n",
    "nonsyntenic_sum = nonsyntenic.group_by('genome').sum().to_pandas().set_index('genome')\n",
    "\n",
    "for genome, row in syntenic_sum.iterrows():\n",
    "    print (f\"\\nCommon Syntenic in {genome}: {row.n_switch_errors/row.n_checked*100}\")\n",
    "    print (f\"Common Nonsyntenic {genome}: {nonsyntenic_sum.loc[genome].n_switch_errors/nonsyntenic_sum.loc[genome].n_checked*100}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "synEnum = pl.Enum(categories=['All_regions', 'Syntenic', 'Nonsyntenic'])\n",
    "concordance = all_variants.filter((pl.col('method_of_phasing') == 'phased_with_parents_and_pedigree') &\n",
    "                                  (pl.col('ground_truth_data_source') == 'HPRC_samples')).with_columns(Syntenic=pl.col('Syntenic').cast(str).replace('true','Syntenic').replace('false','Nonsyntenic').cast(synEnum))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "concordance_summary = concordance.clone().with_columns(Syntenic = pl.lit('All_regions').cast(synEnum))\n",
    "concordance_summary = pl.concat((concordance, concordance_summary)).group_by(['Syntenic','genome','type']).sum().to_pandas().reset_index()\n",
    "concordance_summary = pd.concat((concordance_summary, concordance_summary.copy().assign(type = 'SNPs + Indels').groupby(['Syntenic','genome','type'], observed=True).sum(numeric_only=True).reset_index()))\n",
    "concordance_summary['gt_error_rate'] = concordance_summary.n_gt_errors/concordance_summary.n_gt_checked*100\n",
    "concordance_summary['gt_accuracy_rate'] = 100-concordance_summary['gt_error_rate'] \n",
    "concordance_summary['switch_error_rate'] = concordance_summary.n_switch_errors/concordance_summary.n_checked*100\n",
    "concordance_summary['Syntenic'] = concordance_summary.Syntenic.replace('false', 'Nonsyntenic').replace('true','Syntenic')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "('Syntenic', 'genome', 'type')",
         "rawType": "object",
         "type": "unknown"
        },
        {
         "name": "switch_error_rate",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "gt_error_rate",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "n_gt_errors",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "n_gt_checked",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "n_switch_errors",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "n_checked",
         "rawType": "float64",
         "type": "float"
        }
       ],
       "ref": "ebc75457-8a88-4b5a-b092-56f892e7c9bb",
       "rows": [
        [
         "('All_regions', 'GRCh38', 'Indel')",
         "2.0657779781867376",
         "6.346630767292711",
         "33755.0",
         "531857.0",
         "1466.0",
         "70966.0"
        ],
        [
         "('All_regions', 'GRCh38', 'SNP')",
         "0.9935705356440718",
         "2.053950721385667",
         "55269.0",
         "2690863.0",
         "3664.0",
         "368771.0"
        ],
        [
         "('All_regions', 'GRCh38', 'SNPs + Indels')",
         "1.1666064033729253",
         "2.762387051931288",
         "89024.0",
         "3222720.0",
         "5130.0",
         "439737.0"
        ],
        [
         "('All_regions', 'CHM13v2.0', 'Indel')",
         "1.4988566977410358",
         "4.2257645975069265",
         "23808.0",
         "563401.0",
         "1193.0",
         "79594.0"
        ],
        [
         "('All_regions', 'CHM13v2.0', 'SNP')",
         "0.6129631826630284",
         "0.48700934968327586",
         "12172.0",
         "2499336.0",
         "2170.0",
         "354018.0"
        ],
        [
         "('All_regions', 'CHM13v2.0', 'SNPs + Indels')",
         "0.7755781666559044",
         "1.1747662303358075",
         "35980.0",
         "3062737.0",
         "3363.0",
         "433612.0"
        ],
        [
         "('Syntenic', 'GRCh38', 'Indel')",
         "2.067147026889832",
         "6.348546156017691",
         "33733.0",
         "531350.0",
         "1466.0",
         "70919.0"
        ],
        [
         "('Syntenic', 'GRCh38', 'SNP')",
         "0.9959525835238345",
         "2.05215872400133",
         "55124.0",
         "2686147.0",
         "3664.0",
         "367889.0"
        ],
        [
         "('Syntenic', 'GRCh38', 'SNPs + Indels')",
         "1.1690762246814095",
         "2.7616808966721647",
         "88857.0",
         "3217497.0",
         "5130.0",
         "438808.0"
        ],
        [
         "('Syntenic', 'CHM13v2.0', 'Indel')",
         "1.495500929975368",
         "4.217158443551022",
         "23748.0",
         "563128.0",
         "1190.0",
         "79572.0"
        ],
        [
         "('Syntenic', 'CHM13v2.0', 'SNP')",
         "0.6129631826630284",
         "0.48689691541261404",
         "12169.0",
         "2499297.0",
         "2170.0",
         "354018.0"
        ],
        [
         "('Syntenic', 'CHM13v2.0', 'SNPs + Indels')",
         "0.7749256209783436",
         "1.1728287223360572",
         "35917.0",
         "3062425.0",
         "3360.0",
         "433590.0"
        ],
        [
         "('Nonsyntenic', 'GRCh38', 'Indel')",
         "0.0",
         "4.339250493096647",
         "22.0",
         "507.0",
         "0.0",
         "47.0"
        ],
        [
         "('Nonsyntenic', 'GRCh38', 'SNP')",
         "0.0",
         "3.0746395250212046",
         "145.0",
         "4716.0",
         "0.0",
         "882.0"
        ],
        [
         "('Nonsyntenic', 'GRCh38', 'SNPs + Indels')",
         "0.0",
         "3.1973961324909053",
         "167.0",
         "5223.0",
         "0.0",
         "929.0"
        ],
        [
         "('Nonsyntenic', 'CHM13v2.0', 'Indel')",
         "13.636363636363635",
         "21.978021978021978",
         "60.0",
         "273.0",
         "3.0",
         "22.0"
        ],
        [
         "('Nonsyntenic', 'CHM13v2.0', 'SNP')",
         null,
         "7.6923076923076925",
         "3.0",
         "39.0",
         "0.0",
         "0.0"
        ],
        [
         "('Nonsyntenic', 'CHM13v2.0', 'SNPs + Indels')",
         "13.636363636363635",
         "20.192307692307693",
         "63.0",
         "312.0",
         "3.0",
         "22.0"
        ]
       ],
       "shape": {
        "columns": 6,
        "rows": 18
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>switch_error_rate</th>\n",
       "      <th>gt_error_rate</th>\n",
       "      <th>n_gt_errors</th>\n",
       "      <th>n_gt_checked</th>\n",
       "      <th>n_switch_errors</th>\n",
       "      <th>n_checked</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Syntenic</th>\n",
       "      <th>genome</th>\n",
       "      <th>type</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"6\" valign=\"top\">All_regions</th>\n",
       "      <th rowspan=\"3\" valign=\"top\">GRCh38</th>\n",
       "      <th>Indel</th>\n",
       "      <td>2.066</td>\n",
       "      <td>6.347</td>\n",
       "      <td>33755.0</td>\n",
       "      <td>5.319e+05</td>\n",
       "      <td>1466.0</td>\n",
       "      <td>70966.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP</th>\n",
       "      <td>0.994</td>\n",
       "      <td>2.054</td>\n",
       "      <td>55269.0</td>\n",
       "      <td>2.691e+06</td>\n",
       "      <td>3664.0</td>\n",
       "      <td>368771.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNPs + Indels</th>\n",
       "      <td>1.167</td>\n",
       "      <td>2.762</td>\n",
       "      <td>89024.0</td>\n",
       "      <td>3.223e+06</td>\n",
       "      <td>5130.0</td>\n",
       "      <td>439737.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">CHM13v2.0</th>\n",
       "      <th>Indel</th>\n",
       "      <td>1.499</td>\n",
       "      <td>4.226</td>\n",
       "      <td>23808.0</td>\n",
       "      <td>5.634e+05</td>\n",
       "      <td>1193.0</td>\n",
       "      <td>79594.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP</th>\n",
       "      <td>0.613</td>\n",
       "      <td>0.487</td>\n",
       "      <td>12172.0</td>\n",
       "      <td>2.499e+06</td>\n",
       "      <td>2170.0</td>\n",
       "      <td>354018.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNPs + Indels</th>\n",
       "      <td>0.776</td>\n",
       "      <td>1.175</td>\n",
       "      <td>35980.0</td>\n",
       "      <td>3.063e+06</td>\n",
       "      <td>3363.0</td>\n",
       "      <td>433612.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"6\" valign=\"top\">Syntenic</th>\n",
       "      <th rowspan=\"3\" valign=\"top\">GRCh38</th>\n",
       "      <th>Indel</th>\n",
       "      <td>2.067</td>\n",
       "      <td>6.349</td>\n",
       "      <td>33733.0</td>\n",
       "      <td>5.314e+05</td>\n",
       "      <td>1466.0</td>\n",
       "      <td>70919.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP</th>\n",
       "      <td>0.996</td>\n",
       "      <td>2.052</td>\n",
       "      <td>55124.0</td>\n",
       "      <td>2.686e+06</td>\n",
       "      <td>3664.0</td>\n",
       "      <td>367889.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNPs + Indels</th>\n",
       "      <td>1.169</td>\n",
       "      <td>2.762</td>\n",
       "      <td>88857.0</td>\n",
       "      <td>3.217e+06</td>\n",
       "      <td>5130.0</td>\n",
       "      <td>438808.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">CHM13v2.0</th>\n",
       "      <th>Indel</th>\n",
       "      <td>1.496</td>\n",
       "      <td>4.217</td>\n",
       "      <td>23748.0</td>\n",
       "      <td>5.631e+05</td>\n",
       "      <td>1190.0</td>\n",
       "      <td>79572.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP</th>\n",
       "      <td>0.613</td>\n",
       "      <td>0.487</td>\n",
       "      <td>12169.0</td>\n",
       "      <td>2.499e+06</td>\n",
       "      <td>2170.0</td>\n",
       "      <td>354018.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNPs + Indels</th>\n",
       "      <td>0.775</td>\n",
       "      <td>1.173</td>\n",
       "      <td>35917.0</td>\n",
       "      <td>3.062e+06</td>\n",
       "      <td>3360.0</td>\n",
       "      <td>433590.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"6\" valign=\"top\">Nonsyntenic</th>\n",
       "      <th rowspan=\"3\" valign=\"top\">GRCh38</th>\n",
       "      <th>Indel</th>\n",
       "      <td>0.000</td>\n",
       "      <td>4.339</td>\n",
       "      <td>22.0</td>\n",
       "      <td>5.070e+02</td>\n",
       "      <td>0.0</td>\n",
       "      <td>47.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP</th>\n",
       "      <td>0.000</td>\n",
       "      <td>3.075</td>\n",
       "      <td>145.0</td>\n",
       "      <td>4.716e+03</td>\n",
       "      <td>0.0</td>\n",
       "      <td>882.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNPs + Indels</th>\n",
       "      <td>0.000</td>\n",
       "      <td>3.197</td>\n",
       "      <td>167.0</td>\n",
       "      <td>5.223e+03</td>\n",
       "      <td>0.0</td>\n",
       "      <td>929.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">CHM13v2.0</th>\n",
       "      <th>Indel</th>\n",
       "      <td>13.636</td>\n",
       "      <td>21.978</td>\n",
       "      <td>60.0</td>\n",
       "      <td>2.730e+02</td>\n",
       "      <td>3.0</td>\n",
       "      <td>22.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP</th>\n",
       "      <td>NaN</td>\n",
       "      <td>7.692</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.900e+01</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNPs + Indels</th>\n",
       "      <td>13.636</td>\n",
       "      <td>20.192</td>\n",
       "      <td>63.0</td>\n",
       "      <td>3.120e+02</td>\n",
       "      <td>3.0</td>\n",
       "      <td>22.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     switch_error_rate  gt_error_rate  \\\n",
       "Syntenic    genome    type                                              \n",
       "All_regions GRCh38    Indel                      2.066          6.347   \n",
       "                      SNP                        0.994          2.054   \n",
       "                      SNPs + Indels              1.167          2.762   \n",
       "            CHM13v2.0 Indel                      1.499          4.226   \n",
       "                      SNP                        0.613          0.487   \n",
       "                      SNPs + Indels              0.776          1.175   \n",
       "Syntenic    GRCh38    Indel                      2.067          6.349   \n",
       "                      SNP                        0.996          2.052   \n",
       "                      SNPs + Indels              1.169          2.762   \n",
       "            CHM13v2.0 Indel                      1.496          4.217   \n",
       "                      SNP                        0.613          0.487   \n",
       "                      SNPs + Indels              0.775          1.173   \n",
       "Nonsyntenic GRCh38    Indel                      0.000          4.339   \n",
       "                      SNP                        0.000          3.075   \n",
       "                      SNPs + Indels              0.000          3.197   \n",
       "            CHM13v2.0 Indel                     13.636         21.978   \n",
       "                      SNP                          NaN          7.692   \n",
       "                      SNPs + Indels             13.636         20.192   \n",
       "\n",
       "                                     n_gt_errors  n_gt_checked  \\\n",
       "Syntenic    genome    type                                       \n",
       "All_regions GRCh38    Indel              33755.0     5.319e+05   \n",
       "                      SNP                55269.0     2.691e+06   \n",
       "                      SNPs + Indels      89024.0     3.223e+06   \n",
       "            CHM13v2.0 Indel              23808.0     5.634e+05   \n",
       "                      SNP                12172.0     2.499e+06   \n",
       "                      SNPs + Indels      35980.0     3.063e+06   \n",
       "Syntenic    GRCh38    Indel              33733.0     5.314e+05   \n",
       "                      SNP                55124.0     2.686e+06   \n",
       "                      SNPs + Indels      88857.0     3.217e+06   \n",
       "            CHM13v2.0 Indel              23748.0     5.631e+05   \n",
       "                      SNP                12169.0     2.499e+06   \n",
       "                      SNPs + Indels      35917.0     3.062e+06   \n",
       "Nonsyntenic GRCh38    Indel                 22.0     5.070e+02   \n",
       "                      SNP                  145.0     4.716e+03   \n",
       "                      SNPs + Indels        167.0     5.223e+03   \n",
       "            CHM13v2.0 Indel                 60.0     2.730e+02   \n",
       "                      SNP                    3.0     3.900e+01   \n",
       "                      SNPs + Indels         63.0     3.120e+02   \n",
       "\n",
       "                                     n_switch_errors  n_checked  \n",
       "Syntenic    genome    type                                       \n",
       "All_regions GRCh38    Indel                   1466.0    70966.0  \n",
       "                      SNP                     3664.0   368771.0  \n",
       "                      SNPs + Indels           5130.0   439737.0  \n",
       "            CHM13v2.0 Indel                   1193.0    79594.0  \n",
       "                      SNP                     2170.0   354018.0  \n",
       "                      SNPs + Indels           3363.0   433612.0  \n",
       "Syntenic    GRCh38    Indel                   1466.0    70919.0  \n",
       "                      SNP                     3664.0   367889.0  \n",
       "                      SNPs + Indels           5130.0   438808.0  \n",
       "            CHM13v2.0 Indel                   1190.0    79572.0  \n",
       "                      SNP                     2170.0   354018.0  \n",
       "                      SNPs + Indels           3360.0   433590.0  \n",
       "Nonsyntenic GRCh38    Indel                      0.0       47.0  \n",
       "                      SNP                        0.0      882.0  \n",
       "                      SNPs + Indels              0.0      929.0  \n",
       "            CHM13v2.0 Indel                      3.0       22.0  \n",
       "                      SNP                        0.0        0.0  \n",
       "                      SNPs + Indels              3.0       22.0  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.set_option(\"display.precision\", 3)\n",
    "concordance_summary.sort_values(['Syntenic','genome','type']).set_index(['Syntenic','genome','type'])[['switch_error_rate','gt_error_rate','n_gt_errors','n_gt_checked', 'n_switch_errors', 'n_checked']].astype(float)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
