<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.34">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Variant Filtering Stats</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="calc_figures_for_paper_files/libs/clipboard/clipboard.min.js"></script>
<script src="calc_figures_for_paper_files/libs/quarto-html/quarto.js" type="module"></script>
<script src="calc_figures_for_paper_files/libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="calc_figures_for_paper_files/libs/quarto-html/popper.min.js"></script>
<script src="calc_figures_for_paper_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="calc_figures_for_paper_files/libs/quarto-html/anchor.min.js"></script>
<link href="calc_figures_for_paper_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="calc_figures_for_paper_files/libs/quarto-html/quarto-syntax-highlighting-c8ad9e5dbd60b7b70b38521ab19b7da4.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="calc_figures_for_paper_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="calc_figures_for_paper_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="calc_figures_for_paper_files/libs/bootstrap/bootstrap-996b3f8b7236a2c8b17e0060dfcdce54.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


</head>

<body class="quarto-light">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#number-of-reference-samples-that-are-probandsparentsunrelateds" id="toc-number-of-reference-samples-that-are-probandsparentsunrelateds" class="nav-link active" data-scroll-target="#number-of-reference-samples-that-are-probandsparentsunrelateds"><span class="header-section-number">0.1</span> Number of reference samples that are probands/parents/unrelateds</a></li>
  <li><a href="#export-per-sample-statistics-for-supplemental-table-1" id="toc-export-per-sample-statistics-for-supplemental-table-1" class="nav-link" data-scroll-target="#export-per-sample-statistics-for-supplemental-table-1"><span class="header-section-number">0.2</span> Export per-sample statistics for supplemental table 1</a></li>
  <li><a href="#get-number-of-probandsparents-in-1kgp" id="toc-get-number-of-probandsparents-in-1kgp" class="nav-link" data-scroll-target="#get-number-of-probandsparents-in-1kgp"><span class="header-section-number">1</span> Get number of probands/parents in 1KGP</a></li>
  </ul>
</nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Variant Filtering Stats</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<div id="cell-1" class="cell" data-execution_count="1">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> polars <span class="im">as</span> pl</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pandas.api.types <span class="im">import</span> CategoricalDtype</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> glob</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> tqdm <span class="im">import</span> tqdm</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> importlib</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>plt.rcParams[<span class="st">"figure.dpi"</span>] <span class="op">=</span> <span class="dv">120</span></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>summary_statistics_folder <span class="op">=</span> <span class="st">'intermediate_data_fullgenome'</span></span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> os.getcwd()<span class="op">!=</span><span class="st">'phasing_T2T_project'</span>:</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a>    os.chdir(<span class="st">'..'</span>) <span class="co"># Change to the parent directory</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-2" class="cell" data-execution_count="2">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>binned_maf_data <span class="op">=</span> pd.read_parquet(<span class="ss">f"</span><span class="sc">{</span>summary_statistics_folder<span class="sc">}</span><span class="ss">/binned_maf_data.parquet"</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>all_chroms <span class="op">=</span> pd.read_parquet(<span class="ss">f'</span><span class="sc">{</span>summary_statistics_folder<span class="sc">}</span><span class="ss">/chroms.parquet'</span>)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>all_samples <span class="op">=</span> pd.read_parquet(<span class="ss">f'</span><span class="sc">{</span>summary_statistics_folder<span class="sc">}</span><span class="ss">/samples.parquet'</span>)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>all_ancestries <span class="op">=</span> pd.read_parquet(<span class="ss">f'</span><span class="sc">{</span>summary_statistics_folder<span class="sc">}</span><span class="ss">/ancestries.parquet'</span>)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>all_methods <span class="op">=</span> pd.read_parquet(<span class="ss">f'</span><span class="sc">{</span>summary_statistics_folder<span class="sc">}</span><span class="ss">/methods.parquet'</span>)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>per_variant_category_imputation_performance <span class="op">=</span> pd.read_parquet(<span class="ss">f'</span><span class="sc">{</span>summary_statistics_folder<span class="sc">}</span><span class="ss">/per_variant_category_imputation_performance.parquet'</span>)</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>per_sample_imputation_performance <span class="op">=</span> pd.read_parquet(<span class="ss">f'</span><span class="sc">{</span>summary_statistics_folder<span class="sc">}</span><span class="ss">/per_sample_imputation_performance.parquet'</span>)</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>rolling_window_ideogram_250k <span class="op">=</span> pd.read_parquet(<span class="ss">f'</span><span class="sc">{</span>summary_statistics_folder<span class="sc">}</span><span class="ss">/rolling_stats_250k_window.parquet'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-3" class="cell" data-execution_count="3">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Post hoc fixes - one off events that are tricky to deal with when compiling stats</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="co"># there is one variant (I have not investigated, but likely a denovo SNP) that is a singleton variant in the HPRC dataset. </span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="co"># That variant passed all filters, but makes for a very odd figure (correctly phased: 0% error rate, incorrect: 100% error rate)</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="co"># We will trim it here</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>binned_maf_data.loc[(binned_maf_data.ground_truth_data_source <span class="op">==</span> <span class="st">'HPRC_samples'</span>)<span class="op">&amp;</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>                    (binned_maf_data.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)<span class="op">&amp;</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>                    (binned_maf_data.rounded_MAF <span class="op">==</span> <span class="st">'singleton'</span>), [<span class="st">'n_switch_errors'</span>,<span class="st">'n_checked'</span>,<span class="st">'switch_error_rate'</span>]] <span class="op">=</span> np.nan</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Isolate syntenic vs nonsyntenic regions</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>regional_binned_maf_data <span class="op">=</span> binned_maf_data.copy()</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>binned_maf_data <span class="op">=</span> (binned_maf_data.loc[binned_maf_data.region.isin([<span class="st">'syntenic'</span>,<span class="st">'nonsyntenic'</span>,<span class="st">'All'</span>])]</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>                                  .replace(<span class="st">'syntenic'</span>,<span class="st">'Syntenic'</span>)</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>                                  .replace(<span class="st">'nonsyntenic'</span>,<span class="st">'Nonsyntenic'</span>)</span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>                                  .rename(columns<span class="op">=</span>{<span class="st">'region'</span>:<span class="st">'Syntenic'</span>}))</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Nonsyntenic GRCh38 regions are simply GRCh38 errors, and we are not interested in the performance in these regions</span></span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a>binned_maf_data <span class="op">=</span> binned_maf_data.loc[<span class="op">~</span>((binned_maf_data.genome<span class="op">==</span><span class="st">'GRCh38'</span>)<span class="op">&amp;</span>(binned_maf_data.Syntenic<span class="op">==</span><span class="st">'Nonsyntenic'</span>))]</span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a><span class="co"># NA20355 is a duo member, and as such was not part of duohmm trio correction for the GRCh38 2022 effort.</span></span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a>all_chroms.loc[(all_chroms.genome<span class="op">==</span><span class="st">'GRCh38'</span>)<span class="op">&amp;</span>(all_chroms[<span class="st">'sample_id'</span>]<span class="op">==</span><span class="st">'NA20355'</span>), <span class="st">'trio_phased'</span>] <span class="op">=</span> <span class="va">False</span></span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a>all_samples.loc[(all_samples.genome<span class="op">==</span><span class="st">'GRCh38'</span>)<span class="op">&amp;</span>(all_samples[<span class="st">'sample_id'</span>]<span class="op">==</span><span class="st">'NA20355'</span>), <span class="st">'trio_phased'</span>] <span class="op">=</span> <span class="va">False</span></span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a>all_chroms <span class="op">=</span> all_chroms.loc[<span class="op">~</span>((all_chroms.genome<span class="op">==</span><span class="st">'GRCh38'</span>)<span class="op">&amp;</span>(all_chroms[<span class="st">'sample_id'</span>]<span class="op">==</span><span class="st">'NA20355'</span>)<span class="op">&amp;</span>(all_chroms.ground_truth_data_source.isin([<span class="st">'HPRC_HGSVC_probands'</span>,<span class="st">'HGSVC_probands'</span>])))]</span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a>all_samples <span class="op">=</span> all_samples.loc[<span class="op">~</span>((all_samples.genome<span class="op">==</span><span class="st">'GRCh38'</span>)<span class="op">&amp;</span>(all_samples[<span class="st">'sample_id'</span>]<span class="op">==</span><span class="st">'NA20355'</span>)<span class="op">&amp;</span></span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a>                                (all_samples.ground_truth_data_source.isin([<span class="st">'HPRC_HGSVC_probands'</span>,<span class="st">'HGSVC_probands'</span>])))]</span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a><span class="co"># very high variance stats when n_checked is less than 10. Replace switch error rate at these sites with np.nan</span></span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true" tabindex="-1"></a>binned_maf_data.loc[(binned_maf_data.n_checked<span class="op">&gt;</span><span class="dv">10</span>), <span class="st">'switch_error_rate'</span>] <span class="op">=</span> binned_maf_data.loc[(binned_maf_data.n_checked<span class="op">&gt;</span><span class="dv">10</span>), <span class="st">'switch_error_rate'</span>].replace(<span class="dv">0</span>, np.nan)</span>
<span id="cb3-29"><a href="#cb3-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-30"><a href="#cb3-30" aria-hidden="true" tabindex="-1"></a><span class="co"># bins with a switch error rate of 0 make for insane figures when plotted on a log scale. Replace with np.nan, and maybe note</span></span>
<span id="cb3-31"><a href="#cb3-31" aria-hidden="true" tabindex="-1"></a>binned_maf_data[<span class="st">'switch_error_rate'</span>] <span class="op">=</span> binned_maf_data[<span class="st">'switch_error_rate'</span>].replace(<span class="dv">0</span>, np.nan)</span>
<span id="cb3-32"><a href="#cb3-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-33"><a href="#cb3-33" aria-hidden="true" tabindex="-1"></a><span class="co"># A reviewer suggests combining chrX, PAR1, and PAR2 statistics. I will do that, and set male chrX switch error rates to np.nan.</span></span>
<span id="cb3-34"><a href="#cb3-34" aria-hidden="true" tabindex="-1"></a><span class="co">## Calculate percent discordance in imputation performance</span></span>
<span id="cb3-35"><a href="#cb3-35" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="st">'perc_discordance'</span> <span class="kw">not</span> <span class="kw">in</span> per_variant_category_imputation_performance.columns:</span>
<span id="cb3-36"><a href="#cb3-36" aria-hidden="true" tabindex="-1"></a>    per_variant_category_imputation_performance <span class="op">=</span> per_variant_category_imputation_performance.loc[per_variant_category_imputation_performance.num_AA <span class="op">&gt;</span> <span class="dv">0</span>]</span>
<span id="cb3-37"><a href="#cb3-37" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="bu">max</span>(per_variant_category_imputation_performance.mean_AF) <span class="op">&lt;</span> <span class="dv">1</span>:</span>
<span id="cb3-38"><a href="#cb3-38" aria-hidden="true" tabindex="-1"></a>        per_variant_category_imputation_performance.mean_AF <span class="op">*=</span> <span class="dv">100</span></span>
<span id="cb3-39"><a href="#cb3-39" aria-hidden="true" tabindex="-1"></a>    per_variant_category_imputation_performance[<span class="st">'num_alt_mismatches'</span>] <span class="op">=</span> per_variant_category_imputation_performance.num_Aa_mismatches <span class="op">+</span> per_variant_category_imputation_performance.num_aa_mismatches</span>
<span id="cb3-40"><a href="#cb3-40" aria-hidden="true" tabindex="-1"></a>    per_variant_category_imputation_performance[<span class="st">'num_alt_variants'</span>] <span class="op">=</span> per_variant_category_imputation_performance.num_Aa <span class="op">+</span> per_variant_category_imputation_performance.num_aa</span>
<span id="cb3-41"><a href="#cb3-41" aria-hidden="true" tabindex="-1"></a>    per_variant_category_imputation_performance[<span class="st">'perc_discordance'</span>] <span class="op">=</span> per_variant_category_imputation_performance[<span class="st">'num_alt_mismatches'</span>]<span class="op">/</span>per_variant_category_imputation_performance[<span class="st">'num_alt_variants'</span>]<span class="op">*</span><span class="dv">100</span></span>
<span id="cb3-42"><a href="#cb3-42" aria-hidden="true" tabindex="-1"></a>    per_variant_category_imputation_performance[<span class="st">'perc_concordance'</span>] <span class="op">=</span> <span class="dv">100</span><span class="op">-</span>per_variant_category_imputation_performance[<span class="st">'perc_discordance'</span>]</span>
<span id="cb3-43"><a href="#cb3-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-44"><a href="#cb3-44" aria-hidden="true" tabindex="-1"></a><span class="co"># imputed df needs to have CHM13v2.0 as genome name instead of T2T</span></span>
<span id="cb3-45"><a href="#cb3-45" aria-hidden="true" tabindex="-1"></a>per_variant_category_imputation_performance <span class="op">=</span> per_variant_category_imputation_performance.replace(<span class="st">'T2T'</span>,<span class="st">'CHM13v2.0'</span>)</span>
<span id="cb3-46"><a href="#cb3-46" aria-hidden="true" tabindex="-1"></a>per_sample_imputation_performance <span class="op">=</span> per_sample_imputation_performance.replace(<span class="st">'T2T'</span>,<span class="st">'CHM13v2.0'</span>)</span>
<span id="cb3-47"><a href="#cb3-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-48"><a href="#cb3-48" aria-hidden="true" tabindex="-1"></a><span class="co"># A supplemental calculates SER by % of chromosome that is new. we calc that here.</span></span>
<span id="cb3-49"><a href="#cb3-49" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="st">'percent_new'</span> <span class="kw">not</span> <span class="kw">in</span> all_chroms.columns:</span>
<span id="cb3-50"><a href="#cb3-50" aria-hidden="true" tabindex="-1"></a>    syntenic_bed <span class="op">=</span> pd.read_csv(<span class="st">'resources/hg38.GCA_009914755.4.synNet.summary.bed.gz'</span>, sep<span class="op">=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>, header<span class="op">=</span><span class="va">None</span>)</span>
<span id="cb3-51"><a href="#cb3-51" aria-hidden="true" tabindex="-1"></a>    syntenic_bed.columns <span class="op">=</span> [<span class="st">'chrom'</span>,<span class="st">'start'</span>,<span class="st">'end'</span>,<span class="st">'status'</span>]</span>
<span id="cb3-52"><a href="#cb3-52" aria-hidden="true" tabindex="-1"></a>    starts <span class="op">=</span> syntenic_bed.groupby(<span class="st">'chrom'</span>).first().reset_index()</span>
<span id="cb3-53"><a href="#cb3-53" aria-hidden="true" tabindex="-1"></a>    syntenic_bed[<span class="st">'length'</span>] <span class="op">=</span> syntenic_bed.end<span class="op">-</span>syntenic_bed.start</span>
<span id="cb3-54"><a href="#cb3-54" aria-hidden="true" tabindex="-1"></a>    syntenic_bed[<span class="st">'dif'</span>] <span class="op">=</span> syntenic_bed.end.diff()</span>
<span id="cb3-55"><a href="#cb3-55" aria-hidden="true" tabindex="-1"></a>    syntenic_bed[<span class="st">'nonsyn_len'</span>] <span class="op">=</span> syntenic_bed.dif <span class="op">-</span> syntenic_bed.length</span>
<span id="cb3-56"><a href="#cb3-56" aria-hidden="true" tabindex="-1"></a>    syntenic_bed.loc[syntenic_bed.nonsyn_len <span class="op">&lt;</span> <span class="dv">0</span>, <span class="st">'nonsyn_len'</span>] <span class="op">=</span> syntenic_bed.loc[syntenic_bed.nonsyn_len <span class="op">&lt;</span> <span class="dv">0</span>, <span class="st">'start'</span>]</span>
<span id="cb3-57"><a href="#cb3-57" aria-hidden="true" tabindex="-1"></a>    syntenic_bed.loc[syntenic_bed.nonsyn_len.isna(), <span class="st">'nonsyn_len'</span>] <span class="op">=</span> syntenic_bed.loc[syntenic_bed.nonsyn_len.isna(), <span class="st">'start'</span>]</span>
<span id="cb3-58"><a href="#cb3-58" aria-hidden="true" tabindex="-1"></a>    rough_length <span class="op">=</span> syntenic_bed.groupby(<span class="st">'chrom'</span>).end.<span class="bu">max</span>().reset_index()</span>
<span id="cb3-59"><a href="#cb3-59" aria-hidden="true" tabindex="-1"></a>    chr_nonsyn_lengths <span class="op">=</span> syntenic_bed.groupby(<span class="st">'chrom'</span>).<span class="bu">sum</span>(numeric_only<span class="op">=</span><span class="va">True</span>).reset_index()[[<span class="st">'chrom'</span>,<span class="st">'nonsyn_len'</span>]]</span>
<span id="cb3-60"><a href="#cb3-60" aria-hidden="true" tabindex="-1"></a>    chr_nonsyn_lengths <span class="op">=</span> chr_nonsyn_lengths.merge(rough_length, on<span class="op">=</span><span class="st">'chrom'</span>)</span>
<span id="cb3-61"><a href="#cb3-61" aria-hidden="true" tabindex="-1"></a>    chr_nonsyn_lengths[<span class="st">'percent_new'</span>] <span class="op">=</span> chr_nonsyn_lengths.nonsyn_len<span class="op">/</span>chr_nonsyn_lengths.end <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb3-62"><a href="#cb3-62" aria-hidden="true" tabindex="-1"></a>    all_chroms <span class="op">=</span> all_chroms.merge(chr_nonsyn_lengths[[<span class="st">'chrom'</span>, <span class="st">'percent_new'</span>]], on<span class="op">=</span>[<span class="st">'chrom'</span>],how<span class="op">=</span><span class="st">'left'</span>)</span>
<span id="cb3-63"><a href="#cb3-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-64"><a href="#cb3-64" aria-hidden="true" tabindex="-1"></a>all_chroms_chrX <span class="op">=</span> all_chroms.loc[all_chroms.chrom.isin((<span class="st">'chrX'</span>,<span class="st">'PAR1'</span>,<span class="st">'PAR2'</span>))].assign(chrom<span class="op">=</span><span class="st">'chrX'</span>)</span>
<span id="cb3-65"><a href="#cb3-65" aria-hidden="true" tabindex="-1"></a>all_chroms_chrX <span class="op">=</span> all_chroms_chrX.groupby([<span class="st">'chrom'</span>,<span class="st">'genome'</span>,<span class="st">'sample_id'</span>,<span class="st">'method_of_phasing'</span>,<span class="st">'ground_truth_data_source'</span>,<span class="st">'population'</span>, <span class="st">'superpopulation'</span>, <span class="st">'sex'</span>, <span class="st">'trio_phased'</span>],observed<span class="op">=</span><span class="va">True</span>).<span class="bu">sum</span>(numeric_only<span class="op">=</span><span class="va">True</span>).reset_index()</span>
<span id="cb3-66"><a href="#cb3-66" aria-hidden="true" tabindex="-1"></a>all_chroms_chrX.loc[all_chroms_chrX.sex<span class="op">==</span><span class="st">'Male'</span>, [<span class="st">'n_switch_errors'</span>,<span class="st">'n_total_switch_errors'</span>,<span class="st">'n_flips'</span>,<span class="st">'n_consecutive_flips'</span>,<span class="st">'n_true_switch_errors'</span>,<span class="st">'num_correct_phased_hets'</span>,</span>
<span id="cb3-67"><a href="#cb3-67" aria-hidden="true" tabindex="-1"></a>                                                  <span class="st">'n_total_hets'</span>,<span class="st">'flip_and_switches_SER'</span>, <span class="st">'flip_error_rate'</span>, <span class="st">'true_switch_error_rate'</span>,<span class="st">'accurately_phased_rate'</span>]] <span class="op">=</span> np.nan</span>
<span id="cb3-68"><a href="#cb3-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-69"><a href="#cb3-69" aria-hidden="true" tabindex="-1"></a>all_chroms_chrX[<span class="st">'switch_error_rate'</span>] <span class="op">=</span> all_chroms_chrX.n_switch_errors<span class="op">/</span>all_chroms_chrX.n_checked <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb3-70"><a href="#cb3-70" aria-hidden="true" tabindex="-1"></a>all_chroms_chrX[<span class="st">'gt_error_rate'</span>] <span class="op">=</span> all_chroms_chrX.n_gt_errors<span class="op">/</span>all_chroms_chrX.n_gt_checked <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb3-71"><a href="#cb3-71" aria-hidden="true" tabindex="-1"></a>all_chroms_chrX[<span class="st">'flip_and_switches_SER'</span>] <span class="op">=</span> all_chroms_chrX.n_total_switch_errors<span class="op">/</span>all_chroms_chrX.n_total_hets <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb3-72"><a href="#cb3-72" aria-hidden="true" tabindex="-1"></a>all_chroms_chrX[<span class="st">'flip_error_rate'</span>] <span class="op">=</span> all_chroms_chrX.n_flips<span class="op">/</span>all_chroms_chrX.n_total_hets <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb3-73"><a href="#cb3-73" aria-hidden="true" tabindex="-1"></a>all_chroms_chrX[<span class="st">'true_switch_error_rate'</span>] <span class="op">=</span> all_chroms_chrX.n_true_switch_errors<span class="op">/</span>all_chroms_chrX.n_total_hets <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb3-74"><a href="#cb3-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-75"><a href="#cb3-75" aria-hidden="true" tabindex="-1"></a>all_chroms_chrX <span class="op">=</span> pd.concat((all_chroms_chrX, all_chroms.loc[<span class="op">~</span>all_chroms.chrom.isin((<span class="st">'chrX'</span>,<span class="st">'PAR1'</span>,<span class="st">'PAR2'</span>))])).drop(columns<span class="op">=</span>[<span class="st">'L50'</span>,<span class="st">'N50'</span>,<span class="st">'percent_new'</span>])</span>
<span id="cb3-76"><a href="#cb3-76" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-77"><a href="#cb3-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-78"><a href="#cb3-78" aria-hidden="true" tabindex="-1"></a>binned_maf_data <span class="op">=</span> binned_maf_data.sort_values(<span class="st">'type'</span>, key<span class="op">=</span><span class="kw">lambda</span> col: col.<span class="bu">apply</span>(<span class="kw">lambda</span> x: {<span class="st">'Indel'</span>:<span class="dv">0</span>, <span class="st">'SNP'</span>:<span class="dv">1</span>, <span class="st">'SNPs + Indels'</span>:<span class="dv">2</span>}.get(x,x)))</span>
<span id="cb3-79"><a href="#cb3-79" aria-hidden="true" tabindex="-1"></a>binned_maf_data <span class="op">=</span> binned_maf_data.sort_values(<span class="st">'genome'</span>, key<span class="op">=</span><span class="kw">lambda</span> col: col.<span class="bu">apply</span>(<span class="kw">lambda</span> x: {<span class="st">'GRCh38'</span>:<span class="dv">0</span>, <span class="st">'CHM13v2.0'</span>:<span class="dv">1</span>}.get(x,x)))</span>
<span id="cb3-80"><a href="#cb3-80" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-81"><a href="#cb3-81" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> binned_maf_data[<span class="st">'switch_error_rate'</span>].<span class="bu">max</span>() <span class="op">&lt;</span> <span class="dv">1</span>:</span>
<span id="cb3-82"><a href="#cb3-82" aria-hidden="true" tabindex="-1"></a>    binned_maf_data[<span class="st">'switch_error_rate'</span>] <span class="op">*=</span> <span class="dv">100</span></span>
<span id="cb3-83"><a href="#cb3-83" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> binned_maf_data[<span class="st">'gt_error_rate'</span>].<span class="bu">max</span>() <span class="op">&lt;</span> <span class="dv">1</span>:</span>
<span id="cb3-84"><a href="#cb3-84" aria-hidden="true" tabindex="-1"></a>    binned_maf_data[<span class="st">'gt_error_rate'</span>] <span class="op">*=</span> <span class="dv">100</span></span>
<span id="cb3-85"><a href="#cb3-85" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> binned_maf_data[<span class="st">'MAF'</span>].<span class="bu">max</span>() <span class="op">&lt;</span> <span class="dv">1</span>:</span>
<span id="cb3-86"><a href="#cb3-86" aria-hidden="true" tabindex="-1"></a>    binned_maf_data[<span class="st">'MAF'</span>] <span class="op">*=</span> <span class="dv">100</span></span>
<span id="cb3-87"><a href="#cb3-87" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-88"><a href="#cb3-88" aria-hidden="true" tabindex="-1"></a><span class="co"># a handful of variants are miscategorized as singletons. There are no true GRCh38 singletons. We can fix this.</span></span>
<span id="cb3-89"><a href="#cb3-89" aria-hidden="true" tabindex="-1"></a>binned_maf_data <span class="op">=</span> binned_maf_data.loc[<span class="op">~</span>((binned_maf_data.rounded_MAF<span class="op">==</span><span class="st">'singleton'</span>)<span class="op">&amp;</span>(binned_maf_data.genome<span class="op">==</span><span class="st">'GRCh38'</span>)<span class="op">&amp;</span>(binned_maf_data.method_of_phasing<span class="op">==</span><span class="st">'phased_with_parents_and_pedigree'</span>)<span class="op">&amp;</span>(binned_maf_data.ground_truth_data_source<span class="op">!=</span><span class="st">'trios'</span>))]</span>
<span id="cb3-90"><a href="#cb3-90" aria-hidden="true" tabindex="-1"></a></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-4" class="cell" data-execution_count="4">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>oneKGP_meta <span class="op">=</span> pd.read_csv(<span class="st">'resources/1000_genomes_meta.tsv'</span>, sep<span class="op">=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>overlaps <span class="op">=</span> oneKGP_meta.loc[oneKGP_meta[<span class="st">'Sample name'</span>].isin(all_samples.sample_id.unique()) <span class="op">&amp;</span> oneKGP_meta[<span class="st">'Data collections'</span>].<span class="bu">str</span>.contains(<span class="st">'Simons Genome'</span>), <span class="st">'Sample name'</span>]</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>overlaps.to_csv(<span class="st">'resources/sample_subsets/SGDP_samples_in_1KGP.txt'</span>, index<span class="op">=</span><span class="va">False</span>, header<span class="op">=</span><span class="va">False</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-5" class="cell" data-execution_count="5">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>pedigree <span class="op">=</span> pd.read_csv(<span class="st">'resources/pedigrees/1kGP.3202_samples.pedigree_info.txt'</span>, sep<span class="op">=</span><span class="st">' '</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Stats in paragraph after “Full length, telomere to telomere maps of human genetic recombination rates are largely consistent with deCODE recombination rates”</p>
<div id="cell-7" class="cell" data-execution_count="6">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>sum_stats_var_filters <span class="op">=</span> pd.read_parquet(<span class="ss">f"</span><span class="sc">{</span>summary_statistics_folder<span class="sc">}</span><span class="ss">/filter_summary_stats.parquet"</span>)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>grch38_filter_columns<span class="op">=</span>[<span class="st">'MERR_filter'</span>, <span class="st">'HWE_pop_filter'</span>,<span class="st">'AC_filter'</span>,</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>       <span class="st">'f_missing_filter'</span>, <span class="st">'var_len_filter'</span>, <span class="st">'alt_star_filter'</span>, <span class="st">'pass_filter'</span>]</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>chm13_filter_columns<span class="op">=</span>[<span class="co">#'neg_train_site_filter',</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>       <span class="st">'VQSLOD_filter'</span>, <span class="st">'MERR_filter'</span>, <span class="st">'HWE_pop_filter'</span>, <span class="st">'MAC_filter'</span>,</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>       <span class="st">'f_missing_filter'</span>, <span class="st">'var_len_filter'</span>, <span class="st">'alt_star_filter'</span>, <span class="st">'pass_filter'</span>]</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>sum_stats_var_filters[<span class="st">'gatk_filters'</span>] <span class="op">=</span> sum_stats_var_filters.VQSLOD_filter <span class="co">#| sum_stats_var_filters.neg_train_site_filter</span></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>sum_stats_var_filters[<span class="st">'GRCh38_criteria_pass_if_singletons_passed'</span>] <span class="op">=</span> <span class="op">~</span>sum_stats_var_filters[[x <span class="cf">for</span> x <span class="kw">in</span> grch38_filter_columns <span class="cf">if</span> x <span class="op">!=</span> <span class="st">'AC_filter'</span>] <span class="op">+</span> [<span class="st">'MAC_filter'</span>]].<span class="bu">any</span>(axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>num_singletons_gained <span class="op">=</span> sum_stats_var_filters.loc[ (sum_stats_var_filters.genome<span class="op">==</span><span class="st">'CHM13v2.0'</span>) <span class="op">&amp;</span></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>                                                   <span class="op">~</span>sum_stats_var_filters.CHM13_criteria_fail <span class="op">&amp;</span></span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>                                                    sum_stats_var_filters.singleton <span class="op">&amp;</span></span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>                                                    sum_stats_var_filters.GRCh38_criteria_fail <span class="op">&amp;</span></span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a>                                                    sum_stats_var_filters.GRCh38_criteria_pass_if_singletons_passed].<span class="bu">len</span>.<span class="bu">sum</span>()</span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a>num_variants_lost_to_gatk <span class="op">=</span> sum_stats_var_filters.loc[ (sum_stats_var_filters.genome<span class="op">==</span><span class="st">'CHM13v2.0'</span>) <span class="op">&amp;</span></span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a>                                                        sum_stats_var_filters.CHM13_criteria_fail <span class="op">&amp;</span></span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a>                                                       <span class="op">~</span>sum_stats_var_filters.GRCh38_criteria_fail <span class="op">&amp;</span></span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a>                                                        sum_stats_var_filters.gatk_filters <span class="op">&amp;</span></span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a>                                                        sum_stats_var_filters.GRCh38_criteria_pass_if_singletons_passed].<span class="bu">len</span>.<span class="bu">sum</span>()</span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-23"><a href="#cb6-23" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"CHM13 variants lost to gatk-based filters:</span><span class="ch">\n</span><span class="sc">{</span>num_variants_lost_to_gatk<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb6-24"><a href="#cb6-24" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"CHM13 variants gained due to ability to include singletons:</span><span class="ch">\n</span><span class="sc">{</span>num_singletons_gained<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>CHM13 variants lost to gatk-based filters:
14491949
CHM13 variants gained due to ability to include singletons:
40226623</code></pre>
</div>
</div>
<p>The resulting panel contains 6404 haplotypes and XXX unique variants, including XXX phased variants in novel regions of the CHM13v2.0 genome.</p>
<div id="cell-9" class="cell" data-execution_count="7">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>number_unique_CHM13_variants <span class="op">=</span> sum_stats_var_filters.loc[<span class="op">~</span>sum_stats_var_filters.CHM13_filtered <span class="op">&amp;</span> (sum_stats_var_filters.genome<span class="op">==</span><span class="st">'CHM13v2.0'</span>)].<span class="bu">len</span>.<span class="bu">sum</span>()</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>number_unique_syntenic_CHM13_variants <span class="op">=</span> sum_stats_var_filters.loc[<span class="op">~</span>sum_stats_var_filters.Syntenic <span class="op">&amp;</span> <span class="op">~</span>sum_stats_var_filters.CHM13_filtered <span class="op">&amp;</span> (sum_stats_var_filters.genome<span class="op">==</span><span class="st">'CHM13v2.0'</span>)].<span class="bu">len</span>.<span class="bu">sum</span>()</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Unique variants: </span><span class="sc">{</span>number_unique_CHM13_variants<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Unique novel variants: </span><span class="sc">{</span>number_unique_syntenic_CHM13_variants<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Unique variants: 75135193
Unique novel variants: 244718</code></pre>
</div>
</div>
<p>We observed that samples that underwent Mendelian pre-phasing (Figure 2a, c) had lower SERs than samples that underwent statistical phasing (Figure 2b, d; XXX% vs XXX%).</p>
<div id="cell-11" class="cell" data-execution_count="8">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>trio_vs_nontrio <span class="op">=</span> all_samples.loc[all_samples.ground_truth_data_source.isin([<span class="st">'HPRC_samples'</span>, <span class="st">'HGSVC_samples_nontrios_only'</span>]) <span class="op">&amp;</span> </span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>                                    (all_samples.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)<span class="op">&amp;</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>                                    (all_samples.genome<span class="op">==</span><span class="st">'GRCh38'</span>)</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>                                    ].groupby(<span class="st">'ground_truth_data_source'</span>, observed<span class="op">=</span><span class="va">True</span>).<span class="bu">sum</span>(numeric_only<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>trio_vs_nontrio[<span class="st">'switch_error_rate'</span>] <span class="op">=</span> trio_vs_nontrio.n_switch_errors<span class="op">/</span>trio_vs_nontrio.n_checked <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>trio_vs_nontrio.switch_error_rate</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="8">
<pre><code>ground_truth_data_source
HPRC_samples                   0.407948
HGSVC_samples_nontrios_only    1.280732
Name: switch_error_rate, dtype: float64</code></pre>
</div>
</div>
<section id="number-of-reference-samples-that-are-probandsparentsunrelateds" class="level3" data-number="0.1">
<h3 data-number="0.1" class="anchored" data-anchor-id="number-of-reference-samples-that-are-probandsparentsunrelateds"><span class="header-section-number">0.1</span> Number of reference samples that are probands/parents/unrelateds</h3>
<div id="cell-13" class="cell" data-execution_count="9">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>pedigree <span class="op">=</span> pedigree.replace(<span class="st">'0'</span>, np.nan)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>num_that_are_children_and_parents <span class="op">=</span> pedigree.loc[pedigree.sampleID.isin(pedigree.fatherID) <span class="op">|</span> pedigree.sampleID.isin(pedigree.motherID) <span class="op">&amp;</span> <span class="op">~</span>(pedigree.fatherID.isna() <span class="op">|</span> pedigree.motherID.isna())]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-14" class="cell" data-execution_count="10">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>children_that_are_parents <span class="op">=</span> pedigree.loc[(pedigree.sampleID.isin(pedigree.fatherID) <span class="op">|</span> pedigree.sampleID.isin(pedigree.motherID)) <span class="op">&amp;</span> (<span class="op">~</span>pedigree.fatherID.isna() <span class="op">|</span> <span class="op">~</span>pedigree.motherID.isna())].sampleID</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>pedigree[<span class="st">'trio_status'</span>] <span class="op">=</span> np.nan</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>complex_pedigrees <span class="op">=</span> pedigree.loc[pedigree.sampleID.isin(children_that_are_parents) <span class="op">|</span> pedigree.motherID.isin(children_that_are_parents) <span class="op">|</span> pedigree.fatherID.isin(children_that_are_parents)]</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>simple_trios <span class="op">=</span> pedigree.loc[<span class="op">~</span>pedigree.index.isin(complex_pedigrees.index) <span class="op">&amp;</span> (<span class="op">~</span>pedigree.fatherID.isna() <span class="op">&amp;</span> <span class="op">~</span>pedigree.motherID.isna())]</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>simple_duos <span class="op">=</span> pedigree.loc[<span class="op">~</span>pedigree.index.isin(complex_pedigrees.index) <span class="op">&amp;</span> (<span class="op">~</span>pedigree.fatherID.isna() <span class="op">^</span> <span class="op">~</span>pedigree.motherID.isna())]</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>relatives <span class="op">=</span> pd.concat((simple_duos, simple_trios, complex_pedigrees))</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>relatives <span class="op">=</span> relatives.fatherID.dropna().tolist() <span class="op">+</span> relatives.motherID.dropna().tolist() <span class="op">+</span> relatives.sampleID.tolist()</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>relatives <span class="op">=</span> pedigree.loc[pedigree.sampleID.isin(relatives)]</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>no_relatives <span class="op">=</span> pedigree.loc[<span class="op">~</span>pedigree.sampleID.isin(relatives.sampleID)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-15" class="cell" data-execution_count="11">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>proband_or_not_trio <span class="op">=</span> pedigree.loc[<span class="op">~</span>(pedigree.sampleID.isin(pedigree.fatherID) <span class="op">|</span> pedigree.sampleID.isin(pedigree.motherID))]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-16" class="cell" data-execution_count="12">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_parents(df):</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> df.fatherID.dropna().to_list() <span class="op">+</span> df.motherID.dropna().to_list()</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>complex_people <span class="op">=</span> [x <span class="cf">for</span> x <span class="kw">in</span> get_parents(complex_pedigrees) <span class="cf">if</span> x <span class="kw">in</span> complex_pedigrees.sampleID.tolist()]</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>complex_trio_parents <span class="op">=</span> [x <span class="cf">for</span> x <span class="kw">in</span> get_parents(complex_pedigrees.loc[<span class="op">~</span>(complex_pedigrees.fatherID.isna() <span class="op">|</span> complex_pedigrees.motherID.isna())]) <span class="cf">if</span> x <span class="kw">not</span> <span class="kw">in</span> complex_pedigrees.sampleID.tolist()]</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>complex_duo_parents <span class="op">=</span> [x <span class="cf">for</span> x <span class="kw">in</span> get_parents(complex_pedigrees.loc[(complex_pedigrees.fatherID.isna() <span class="op">^</span> complex_pedigrees.motherID.isna())]) <span class="cf">if</span> x <span class="kw">not</span> <span class="kw">in</span> complex_pedigrees.sampleID.tolist()]</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>complex_trio_children <span class="op">=</span> [x <span class="cf">for</span> x <span class="kw">in</span> complex_pedigrees.loc[<span class="op">~</span>(complex_pedigrees.fatherID.isna() <span class="op">|</span> complex_pedigrees.motherID.isna())].sampleID.to_list() <span class="cf">if</span> x <span class="kw">not</span> <span class="kw">in</span> get_parents(complex_pedigrees)]</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>complex_duo_children <span class="op">=</span> [x <span class="cf">for</span> x <span class="kw">in</span> complex_pedigrees.loc[(complex_pedigrees.fatherID.isna() <span class="op">^</span> complex_pedigrees.motherID.isna())].sampleID.to_list() <span class="cf">if</span> x <span class="kw">not</span> <span class="kw">in</span> get_parents(complex_pedigrees)]</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>simple_pedigrees <span class="op">=</span> pedigree.loc[<span class="op">~</span>pedigree.index.isin(complex_pedigrees.index)]</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>simple_trios <span class="op">=</span> simple_pedigrees.loc[<span class="op">~</span>(simple_pedigrees.fatherID.isna() <span class="op">|</span> simple_pedigrees.motherID.isna())]</span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>trio_parents <span class="op">=</span> get_parents(simple_trios) <span class="op">+</span> complex_trio_parents</span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>trio_children <span class="op">=</span> simple_trios.sampleID.to_list() <span class="op">+</span> complex_trio_children</span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a>simple_duos <span class="op">=</span> simple_pedigrees.loc[(simple_pedigrees.fatherID.isna() <span class="op">^</span> simple_pedigrees.motherID.isna())]</span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a>duo_parents <span class="op">=</span> get_parents(simple_duos) <span class="op">+</span> complex_duo_parents</span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a>duo_children <span class="op">=</span> simple_duos.sampleID.to_list() <span class="op">+</span> complex_duo_children</span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a>trio_dads <span class="op">=</span> simple_trios.fatherID.dropna().to_list() <span class="op">+</span> [x <span class="cf">for</span> x <span class="kw">in</span> complex_trio_parents <span class="cf">if</span> x <span class="kw">in</span> complex_pedigrees.fatherID.tolist()]</span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a>trio_moms <span class="op">=</span> simple_trios.motherID.dropna().to_list() <span class="op">+</span> [x <span class="cf">for</span> x <span class="kw">in</span> complex_trio_parents <span class="cf">if</span> x <span class="kw">in</span> complex_pedigrees.motherID.tolist()]</span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true" tabindex="-1"></a><span class="bu">len</span>(<span class="bu">set</span>(trio_moms)), <span class="bu">len</span>(<span class="bu">set</span>(trio_dads)), <span class="bu">len</span>(<span class="bu">set</span>(trio_parents)), <span class="bu">len</span>(<span class="bu">set</span>(duo_parents)), <span class="bu">len</span>(<span class="bu">set</span>(trio_children)), <span class="bu">len</span>(<span class="bu">set</span>(duo_children)), <span class="bu">len</span>(complex_people)</span>
<span id="cb15-22"><a href="#cb15-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-23"><a href="#cb15-23" aria-hidden="true" tabindex="-1"></a><span class="co">## Describes the parent who is in two duo/trios</span></span>
<span id="cb15-24"><a href="#cb15-24" aria-hidden="true" tabindex="-1"></a>pedigree.loc[(pedigree.sampleID<span class="op">==</span><span class="st">'NA19238'</span>)<span class="op">|</span>(pedigree.fatherID<span class="op">==</span><span class="st">'NA19238'</span>)<span class="op">|</span>(pedigree.motherID<span class="op">==</span><span class="st">'NA19238'</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="12">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">sampleID</th>
<th data-quarto-table-cell-role="th">fatherID</th>
<th data-quarto-table-cell-role="th">motherID</th>
<th data-quarto-table-cell-role="th">sex</th>
<th data-quarto-table-cell-role="th">trio_status</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">2504</td>
<td>NA18913</td>
<td>NaN</td>
<td>NA19238</td>
<td>1</td>
<td>NaN</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">2731</td>
<td>NA19238</td>
<td>NaN</td>
<td>NaN</td>
<td>2</td>
<td>NaN</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2733</td>
<td>NA19240</td>
<td>NA19239</td>
<td>NA19238</td>
<td>2</td>
<td>NaN</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
</section>
<section id="export-per-sample-statistics-for-supplemental-table-1" class="level3" data-number="0.2">
<h3 data-number="0.2" class="anchored" data-anchor-id="export-per-sample-statistics-for-supplemental-table-1"><span class="header-section-number">0.2</span> Export per-sample statistics for supplemental table 1</h3>
<div id="cell-18" class="cell" data-execution_count="13">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>samples_for_export <span class="op">=</span> all_samples.loc[(all_samples.method_of_phasing<span class="op">==</span><span class="st">'phased_with_parents_and_pedigree'</span>)<span class="op">&amp;</span>(all_samples.ground_truth_data_source.isin((<span class="st">'HPRC_samples'</span>,<span class="st">'HGSVC_samples'</span>)))]</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>sample_export_columns <span class="op">=</span> [<span class="st">'sample_id'</span>, <span class="st">'panel_genome'</span>, <span class="st">'ground_truth_data_source'</span>, <span class="st">'family_status'</span>,<span class="st">'population'</span>, <span class="st">'superpopulation'</span>, <span class="st">'sex'</span>, <span class="st">'n_switch_errors'</span>, <span class="st">'n_checked'</span>, <span class="st">'switch_error_rate'</span>,</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>       <span class="st">'n_total_switch_errors'</span>, <span class="st">'n_flips'</span>, <span class="st">'n_consecutive_flips'</span>,</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>       <span class="st">'n_true_switch_errors'</span>, <span class="st">'num_correct_phased_hets'</span>, <span class="st">'n_total_hets'</span>,</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>       <span class="st">'flip_and_switches_SER'</span>, <span class="st">'flip_error_rate'</span>, <span class="st">'true_switch_error_rate'</span>,</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>       <span class="st">'accurately_phased_rate'</span>, <span class="st">'n_gt_errors'</span>, <span class="st">'n_gt_checked'</span>,</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>       <span class="st">'gt_error_rate'</span>, <span class="st">'N50'</span>]</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>source_accuracy_columns <span class="op">=</span> [<span class="st">'panel_genome'</span>, <span class="st">'ground_truth_data_source'</span>, <span class="st">'family_status'</span>, <span class="st">'n_switch_errors'</span>, <span class="st">'n_checked'</span>, <span class="st">'switch_error_rate'</span>,</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>       <span class="st">'n_total_switch_errors'</span>, <span class="st">'n_flips'</span>, <span class="st">'flip_error_rate'</span>,</span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>       <span class="st">'n_true_switch_errors'</span>, <span class="st">'n_total_hets'</span>,<span class="st">'true_switch_error_rate'</span>,</span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>       <span class="st">'n_gt_errors'</span>, <span class="st">'n_gt_checked'</span>,</span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>       <span class="st">'gt_error_rate'</span>, <span class="st">'N50'</span>]</span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a>samples_for_export <span class="op">=</span> samples_for_export.sort_values([<span class="st">'genome'</span>,<span class="st">'ground_truth_data_source'</span>, <span class="st">'sample_id'</span>]).rename(columns<span class="op">=</span>{<span class="st">'genome'</span>:<span class="st">'panel_genome'</span>})</span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a>samples_for_export[<span class="st">'family_status'</span>] <span class="op">=</span> <span class="st">'non_trio'</span></span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a>samples_for_export.loc[samples_for_export.trio_phased,<span class="st">'family_status'</span>] <span class="op">=</span> <span class="st">'trio_proband'</span></span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a>samples_for_export.loc[samples_for_export.sample_id.isin(pedigree.fatherID.to_list()<span class="op">+</span>pedigree.motherID.to_list()),<span class="st">'family_status'</span>] <span class="op">=</span> <span class="st">'trio_parent'</span></span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true" tabindex="-1"></a>source_accuracy <span class="op">=</span> samples_for_export[sample_export_columns].groupby([<span class="st">'panel_genome'</span>,<span class="st">'ground_truth_data_source'</span>,<span class="st">'family_status'</span>], observed<span class="op">=</span><span class="va">True</span>).<span class="bu">sum</span>(numeric_only<span class="op">=</span><span class="va">True</span>).reset_index()</span>
<span id="cb16-20"><a href="#cb16-20" aria-hidden="true" tabindex="-1"></a>source_accuracy[<span class="st">'switch_error_rate'</span>] <span class="op">=</span> source_accuracy.n_switch_errors<span class="op">/</span>source_accuracy[<span class="st">'n_checked'</span>] <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb16-21"><a href="#cb16-21" aria-hidden="true" tabindex="-1"></a>source_accuracy[<span class="st">'gt_error_rate'</span>] <span class="op">=</span> source_accuracy[<span class="st">'n_gt_errors'</span>]<span class="op">/</span>source_accuracy[<span class="st">'n_gt_checked'</span>] <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb16-22"><a href="#cb16-22" aria-hidden="true" tabindex="-1"></a>source_accuracy[<span class="st">'true_switch_error_rate'</span>] <span class="op">=</span> source_accuracy.n_true_switch_errors<span class="op">/</span>source_accuracy.n_total_hets <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb16-23"><a href="#cb16-23" aria-hidden="true" tabindex="-1"></a>source_accuracy[<span class="st">'flip_error_rate'</span>] <span class="op">=</span> source_accuracy.n_flips<span class="op">/</span>source_accuracy.n_total_hets <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb16-24"><a href="#cb16-24" aria-hidden="true" tabindex="-1"></a>os.makedirs(<span class="st">'figures/supplemental/tables'</span>, exist_ok<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb16-25"><a href="#cb16-25" aria-hidden="true" tabindex="-1"></a>source_accuracy[source_accuracy_columns].to_csv(<span class="st">'figures/supplemental/tables/supplemental_ground_truth_source_accuracy_data_unformatted.tsv'</span>, sep<span class="op">=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>, index<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb16-26"><a href="#cb16-26" aria-hidden="true" tabindex="-1"></a>samples_for_export[sample_export_columns].to_csv(<span class="st">'figures/supplemental/tables/supplemental_per_sample_data_unformatted.tsv'</span>, sep<span class="op">=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>, index<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb16-27"><a href="#cb16-27" aria-hidden="true" tabindex="-1"></a>source_accuracy.T</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="13">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">0</th>
<th data-quarto-table-cell-role="th">1</th>
<th data-quarto-table-cell-role="th">2</th>
<th data-quarto-table-cell-role="th">3</th>
<th data-quarto-table-cell-role="th">4</th>
<th data-quarto-table-cell-role="th">5</th>
<th data-quarto-table-cell-role="th">6</th>
<th data-quarto-table-cell-role="th">7</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">panel_genome</td>
<td>GRCh38</td>
<td>GRCh38</td>
<td>GRCh38</td>
<td>GRCh38</td>
<td>CHM13v2.0</td>
<td>CHM13v2.0</td>
<td>CHM13v2.0</td>
<td>CHM13v2.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">ground_truth_data_source</td>
<td>HPRC_samples</td>
<td>HGSVC_samples</td>
<td>HGSVC_samples</td>
<td>HGSVC_samples</td>
<td>HPRC_samples</td>
<td>HGSVC_samples</td>
<td>HGSVC_samples</td>
<td>HGSVC_samples</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">family_status</td>
<td>trio_proband</td>
<td>non_trio</td>
<td>trio_parent</td>
<td>trio_proband</td>
<td>trio_proband</td>
<td>non_trio</td>
<td>trio_parent</td>
<td>trio_proband</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">n_switch_errors</td>
<td>474899</td>
<td>721620</td>
<td>282761</td>
<td>396029</td>
<td>361950</td>
<td>511618</td>
<td>174592</td>
<td>333248</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">n_checked</td>
<td>116411762</td>
<td>55578917</td>
<td>40080199</td>
<td>85364549</td>
<td>101999287</td>
<td>46079440</td>
<td>35237358</td>
<td>77833807</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">switch_error_rate</td>
<td>0.407948</td>
<td>1.29837</td>
<td>0.705488</td>
<td>0.463927</td>
<td>0.354855</td>
<td>1.110296</td>
<td>0.495474</td>
<td>0.428153</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">n_total_switch_errors</td>
<td>474899</td>
<td>721620</td>
<td>282761</td>
<td>396029</td>
<td>361950</td>
<td>511618</td>
<td>174592</td>
<td>333248</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">n_flips</td>
<td>216913</td>
<td>277200</td>
<td>121977</td>
<td>184339</td>
<td>169732</td>
<td>219366</td>
<td>83125</td>
<td>157725</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">n_consecutive_flips</td>
<td>7184</td>
<td>27455</td>
<td>6508</td>
<td>5243</td>
<td>3390</td>
<td>12846</td>
<td>1439</td>
<td>3281</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">n_true_switch_errors</td>
<td>33889</td>
<td>139765</td>
<td>32299</td>
<td>22108</td>
<td>19096</td>
<td>60040</td>
<td>6903</td>
<td>14517</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">num_correct_phased_hets</td>
<td>115936863</td>
<td>54857297</td>
<td>39797438</td>
<td>84968520</td>
<td>101637337</td>
<td>45567822</td>
<td>35062766</td>
<td>77500559</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">n_total_hets</td>
<td>116635859</td>
<td>55883572</td>
<td>40208684</td>
<td>85554131</td>
<td>102172409</td>
<td>46311652</td>
<td>35321922</td>
<td>77994813</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">flip_and_switches_SER</td>
<td>15.848738</td>
<td>25.362946</td>
<td>9.033922</td>
<td>13.873697</td>
<td>13.827223</td>
<td>21.464301</td>
<td>6.393531</td>
<td>13.165651</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">flip_error_rate</td>
<td>0.185975</td>
<td>0.496031</td>
<td>0.30336</td>
<td>0.215465</td>
<td>0.166123</td>
<td>0.473673</td>
<td>0.235335</td>
<td>0.202225</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">true_switch_error_rate</td>
<td>0.029055</td>
<td>0.2501</td>
<td>0.080328</td>
<td>0.025841</td>
<td>0.01869</td>
<td>0.129643</td>
<td>0.019543</td>
<td>0.018613</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">accurately_phased_rate</td>
<td>3876.686593</td>
<td>1964.020542</td>
<td>1286.860834</td>
<td>2979.492545</td>
<td>3879.585276</td>
<td>1868.825082</td>
<td>1290.513027</td>
<td>3080.486457</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">n_gt_errors</td>
<td>9552682.0</td>
<td>6916313.0</td>
<td>4587144.0</td>
<td>9665325.0</td>
<td>6867529.0</td>
<td>5090289.0</td>
<td>3571836.0</td>
<td>7676802.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">n_gt_checked</td>
<td>857820709</td>
<td>481426123</td>
<td>314802029</td>
<td>724957722</td>
<td>747787272</td>
<td>406219522</td>
<td>279764783</td>
<td>665772735</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">gt_error_rate</td>
<td>1.113599</td>
<td>1.43663</td>
<td>1.457152</td>
<td>1.333226</td>
<td>0.91838</td>
<td>1.253088</td>
<td>1.276728</td>
<td>1.153066</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">N50</td>
<td>5972868442</td>
<td>3108474708</td>
<td>1990430925</td>
<td>4593002664</td>
<td>5922189400</td>
<td>2905799316</td>
<td>1972578898</td>
<td>4705087212</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="cell-19" class="cell" data-execution_count="14">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> group_phasing_statistics(df, columns_to_groupby):</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>    df_cols <span class="op">=</span> df.columns</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>    <span class="cf">assert</span> <span class="bu">all</span>([c <span class="kw">in</span> df_cols <span class="cf">for</span> c <span class="kw">in</span> columns_to_groupby])</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>    df <span class="op">=</span> df.groupby(columns_to_groupby, observed<span class="op">=</span><span class="va">True</span>).<span class="bu">sum</span>(numeric_only<span class="op">=</span><span class="va">True</span>).reset_index()</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="st">'n_gt_errors'</span> <span class="kw">in</span> df_cols:</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>        df[<span class="st">'gt_error_rate'</span>] <span class="op">=</span> df.n_gt_errors<span class="op">/</span>df.n_gt_checked <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="st">'n_true_switch_errors'</span> <span class="kw">in</span> df_cols:</span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>        <span class="co"># See definitions below.</span></span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>        df[<span class="st">'flip_error_rate'</span>] <span class="op">=</span> df.n_flips<span class="op">/</span>df.n_checked<span class="op">*</span><span class="dv">100</span></span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a>        df[<span class="st">'true_switch_error_rate'</span>] <span class="op">=</span> df.n_true_switch_errors<span class="op">/</span>df.n_checked<span class="op">*</span><span class="dv">100</span></span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="st">'n_switch_errors'</span> <span class="kw">in</span> df_cols:</span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a>        df[<span class="st">'switch_error_rate'</span>] <span class="op">=</span> df.n_switch_errors<span class="op">/</span>df.n_checked<span class="op">*</span><span class="dv">100</span></span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> df</span>
<span id="cb17-17"><a href="#cb17-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-18"><a href="#cb17-18" aria-hidden="true" tabindex="-1"></a>chrX <span class="op">=</span> all_chroms.loc[(all_chroms.sex<span class="op">==</span><span class="st">'Female'</span>)<span class="op">&amp;</span>(all_chroms.chrom<span class="op">==</span><span class="st">'chrX'</span>) <span class="op">&amp;</span> (all_chroms.method_of_phasing<span class="op">==</span><span class="st">'phased_with_parents_and_pedigree'</span>) <span class="op">&amp;</span> all_chroms.ground_truth_data_source.isin([<span class="st">'HPRC_samples'</span>,<span class="st">'HGSVC_samples_nontrios_only'</span>,])]</span>
<span id="cb17-19"><a href="#cb17-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-20"><a href="#cb17-20" aria-hidden="true" tabindex="-1"></a>group_phasing_statistics(chrX, [<span class="st">'genome'</span>,<span class="st">'ground_truth_data_source'</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="14">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">genome</th>
<th data-quarto-table-cell-role="th">ground_truth_data_source</th>
<th data-quarto-table-cell-role="th">n_switch_errors</th>
<th data-quarto-table-cell-role="th">n_checked</th>
<th data-quarto-table-cell-role="th">switch_error_rate</th>
<th data-quarto-table-cell-role="th">n_total_switch_errors</th>
<th data-quarto-table-cell-role="th">n_flips</th>
<th data-quarto-table-cell-role="th">n_consecutive_flips</th>
<th data-quarto-table-cell-role="th">n_true_switch_errors</th>
<th data-quarto-table-cell-role="th">num_correct_phased_hets</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">flip_and_switches_SER</th>
<th data-quarto-table-cell-role="th">flip_error_rate</th>
<th data-quarto-table-cell-role="th">true_switch_error_rate</th>
<th data-quarto-table-cell-role="th">accurately_phased_rate</th>
<th data-quarto-table-cell-role="th">n_gt_errors</th>
<th data-quarto-table-cell-role="th">n_gt_checked</th>
<th data-quarto-table-cell-role="th">gt_error_rate</th>
<th data-quarto-table-cell-role="th">trio_phased</th>
<th data-quarto-table-cell-role="th">N50</th>
<th data-quarto-table-cell-role="th">percent_new</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>GRCh38</td>
<td>HPRC_samples</td>
<td>22059</td>
<td>2414208</td>
<td>0.913716</td>
<td>22059</td>
<td>7768</td>
<td>330</td>
<td>6193</td>
<td>2392149</td>
<td>...</td>
<td>21.045021</td>
<td>0.321762</td>
<td>0.256523</td>
<td>2271.275904</td>
<td>268160.0</td>
<td>18126553</td>
<td>1.479377</td>
<td>23</td>
<td>11296600.0</td>
<td>76.052737</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>GRCh38</td>
<td>HGSVC_samples_nontrios_only</td>
<td>9007</td>
<td>597692</td>
<td>1.506963</td>
<td>9007</td>
<td>2851</td>
<td>264</td>
<td>3041</td>
<td>588685</td>
<td>...</td>
<td>10.400772</td>
<td>0.477002</td>
<td>0.508790</td>
<td>686.032609</td>
<td>82156.0</td>
<td>2585422</td>
<td>3.177663</td>
<td>0</td>
<td>2833560.0</td>
<td>23.146485</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>CHM13v2.0</td>
<td>HPRC_samples</td>
<td>9068</td>
<td>2173783</td>
<td>0.417153</td>
<td>9068</td>
<td>4334</td>
<td>58</td>
<td>342</td>
<td>2164715</td>
<td>...</td>
<td>9.536867</td>
<td>0.199376</td>
<td>0.015733</td>
<td>2285.865213</td>
<td>238853.0</td>
<td>16257872</td>
<td>1.469153</td>
<td>23</td>
<td>34129171.0</td>
<td>76.052737</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>CHM13v2.0</td>
<td>HGSVC_samples_nontrios_only</td>
<td>8618</td>
<td>548108</td>
<td>1.572318</td>
<td>8618</td>
<td>3585</td>
<td>335</td>
<td>1113</td>
<td>539490</td>
<td>...</td>
<td>11.224752</td>
<td>0.654068</td>
<td>0.203062</td>
<td>683.683798</td>
<td>74578.0</td>
<td>2387889</td>
<td>3.123177</td>
<td>0</td>
<td>3651775.0</td>
<td>23.146485</td>
</tr>
</tbody>
</table>

<p>4 rows × 21 columns</p>
</div>
</div>
</div>
<div id="cell-20" class="cell" data-execution_count="15">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co">### Data generated in make_figures.ipynb</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>error_rate_corr <span class="op">=</span> pd.read_csv(<span class="st">'figures/correlation_error_rates_and_gt_error_rate.csv'</span>)</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>error_rate_corr</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-error">
<div class="ansi-escaped-output">
<pre><span class="ansi-red-fg">---------------------------------------------------------------------------</span>
<span class="ansi-red-fg">FileNotFoundError</span>                         Traceback (most recent call last)
<span class="ansi-cyan-fg">Cell</span><span class="ansi-cyan-fg"> </span><span class="ansi-green-fg">In[15]</span><span class="ansi-green-fg">, line 3</span>
<span class="ansi-green-fg">      1</span> <span style="font-style:italic;color:rgb(95,135,135)">### Data generated in make_figures.ipynb</span>
<span class="ansi-green-fg">----&gt; </span><span class="ansi-green-fg">3</span> error_rate_corr = <span class="ansi-yellow-bg">pd</span><span class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">read_csv</span><span class="ansi-yellow-bg">(</span><span class="ansi-yellow-fg ansi-yellow-bg">'</span><span class="ansi-yellow-fg ansi-yellow-bg">figures/correlation_error_rates_and_gt_error_rate.csv</span><span class="ansi-yellow-fg ansi-yellow-bg">'</span><span class="ansi-yellow-bg">)</span>
<span class="ansi-green-fg">      4</span> error_rate_corr

<span class="ansi-cyan-fg">File </span><span class="ansi-green-fg">/home/lalli/.linuxbrew/opt/python@3.11/lib/python3.11/site-packages/pandas/io/parsers/readers.py:1026</span>, in <span class="ansi-cyan-fg">read_csv</span><span class="ansi-blue-fg">(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)</span>
<span class="ansi-green-fg">   1013</span> kwds_defaults = _refine_defaults_read(
<span class="ansi-green-fg">   1014</span>     dialect,
<span class="ansi-green-fg">   1015</span>     delimiter,
<span class="ansi-green-fg">   (...)</span><span class="ansi-green-fg">   1022</span>     dtype_backend=dtype_backend,
<span class="ansi-green-fg">   1023</span> )
<span class="ansi-green-fg">   1024</span> kwds.update(kwds_defaults)
<span class="ansi-green-fg">-&gt; </span><span class="ansi-green-fg">1026</span> <span style="font-weight:bold;color:rgb(0,135,0)">return</span> <span class="ansi-yellow-bg">_read</span><span class="ansi-yellow-bg">(</span><span class="ansi-yellow-bg">filepath_or_buffer</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-bg">kwds</span><span class="ansi-yellow-bg">)</span>

<span class="ansi-cyan-fg">File </span><span class="ansi-green-fg">/home/lalli/.linuxbrew/opt/python@3.11/lib/python3.11/site-packages/pandas/io/parsers/readers.py:620</span>, in <span class="ansi-cyan-fg">_read</span><span class="ansi-blue-fg">(filepath_or_buffer, kwds)</span>
<span class="ansi-green-fg">    617</span> _validate_names(kwds.get(<span class="ansi-yellow-fg">"</span><span class="ansi-yellow-fg">names</span><span class="ansi-yellow-fg">"</span>, <span style="font-weight:bold;color:rgb(0,135,0)">None</span>))
<span class="ansi-green-fg">    619</span> <span style="font-style:italic;color:rgb(95,135,135)"># Create the parser.</span>
<span class="ansi-green-fg">--&gt; </span><span class="ansi-green-fg">620</span> parser = <span class="ansi-yellow-bg">TextFileReader</span><span class="ansi-yellow-bg">(</span><span class="ansi-yellow-bg">filepath_or_buffer</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-bg">*</span><span class="ansi-yellow-bg">*</span><span class="ansi-yellow-bg">kwds</span><span class="ansi-yellow-bg">)</span>
<span class="ansi-green-fg">    622</span> <span style="font-weight:bold;color:rgb(0,135,0)">if</span> chunksize <span style="font-weight:bold;color:rgb(175,0,255)">or</span> iterator:
<span class="ansi-green-fg">    623</span>     <span style="font-weight:bold;color:rgb(0,135,0)">return</span> parser

<span class="ansi-cyan-fg">File </span><span class="ansi-green-fg">/home/lalli/.linuxbrew/opt/python@3.11/lib/python3.11/site-packages/pandas/io/parsers/readers.py:1620</span>, in <span class="ansi-cyan-fg">TextFileReader.__init__</span><span class="ansi-blue-fg">(self, f, engine, **kwds)</span>
<span class="ansi-green-fg">   1617</span>     <span style="color:rgb(0,135,0)">self</span>.options[<span class="ansi-yellow-fg">"</span><span class="ansi-yellow-fg">has_index_names</span><span class="ansi-yellow-fg">"</span>] = kwds[<span class="ansi-yellow-fg">"</span><span class="ansi-yellow-fg">has_index_names</span><span class="ansi-yellow-fg">"</span>]
<span class="ansi-green-fg">   1619</span> <span style="color:rgb(0,135,0)">self</span>.handles: IOHandles | <span style="font-weight:bold;color:rgb(0,135,0)">None</span> = <span style="font-weight:bold;color:rgb(0,135,0)">None</span>
<span class="ansi-green-fg">-&gt; </span><span class="ansi-green-fg">1620</span> <span style="color:rgb(0,135,0)">self</span>._engine = <span style="color:rgb(0,135,0)" class="ansi-yellow-bg">self</span><span class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">_make_engine</span><span class="ansi-yellow-bg">(</span><span class="ansi-yellow-bg">f</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span style="color:rgb(0,135,0)" class="ansi-yellow-bg">self</span><span class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">engine</span><span class="ansi-yellow-bg">)</span>

<span class="ansi-cyan-fg">File </span><span class="ansi-green-fg">/home/lalli/.linuxbrew/opt/python@3.11/lib/python3.11/site-packages/pandas/io/parsers/readers.py:1880</span>, in <span class="ansi-cyan-fg">TextFileReader._make_engine</span><span class="ansi-blue-fg">(self, f, engine)</span>
<span class="ansi-green-fg">   1878</span>     <span style="font-weight:bold;color:rgb(0,135,0)">if</span> <span class="ansi-yellow-fg">"</span><span class="ansi-yellow-fg">b</span><span class="ansi-yellow-fg">"</span> <span style="font-weight:bold;color:rgb(175,0,255)">not</span> <span style="font-weight:bold;color:rgb(175,0,255)">in</span> mode:
<span class="ansi-green-fg">   1879</span>         mode += <span class="ansi-yellow-fg">"</span><span class="ansi-yellow-fg">b</span><span class="ansi-yellow-fg">"</span>
<span class="ansi-green-fg">-&gt; </span><span class="ansi-green-fg">1880</span> <span style="color:rgb(0,135,0)">self</span>.handles = <span class="ansi-yellow-bg">get_handle</span><span class="ansi-yellow-bg">(</span>
<span class="ansi-green-fg">   1881</span> <span class="ansi-yellow-bg">    </span><span class="ansi-yellow-bg">f</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">   1882</span> <span class="ansi-yellow-bg">    </span><span class="ansi-yellow-bg">mode</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">   1883</span> <span class="ansi-yellow-bg">    </span><span class="ansi-yellow-bg">encoding</span><span class="ansi-yellow-bg">=</span><span style="color:rgb(0,135,0)" class="ansi-yellow-bg">self</span><span class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">options</span><span class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">get</span><span class="ansi-yellow-bg">(</span><span class="ansi-yellow-fg ansi-yellow-bg">"</span><span class="ansi-yellow-fg ansi-yellow-bg">encoding</span><span class="ansi-yellow-fg ansi-yellow-bg">"</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span style="font-weight:bold;color:rgb(0,135,0)" class="ansi-yellow-bg">None</span><span class="ansi-yellow-bg">)</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">   1884</span> <span class="ansi-yellow-bg">    </span><span class="ansi-yellow-bg">compression</span><span class="ansi-yellow-bg">=</span><span style="color:rgb(0,135,0)" class="ansi-yellow-bg">self</span><span class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">options</span><span class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">get</span><span class="ansi-yellow-bg">(</span><span class="ansi-yellow-fg ansi-yellow-bg">"</span><span class="ansi-yellow-fg ansi-yellow-bg">compression</span><span class="ansi-yellow-fg ansi-yellow-bg">"</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span style="font-weight:bold;color:rgb(0,135,0)" class="ansi-yellow-bg">None</span><span class="ansi-yellow-bg">)</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">   1885</span> <span class="ansi-yellow-bg">    </span><span class="ansi-yellow-bg">memory_map</span><span class="ansi-yellow-bg">=</span><span style="color:rgb(0,135,0)" class="ansi-yellow-bg">self</span><span class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">options</span><span class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">get</span><span class="ansi-yellow-bg">(</span><span class="ansi-yellow-fg ansi-yellow-bg">"</span><span class="ansi-yellow-fg ansi-yellow-bg">memory_map</span><span class="ansi-yellow-fg ansi-yellow-bg">"</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span style="font-weight:bold;color:rgb(0,135,0)" class="ansi-yellow-bg">False</span><span class="ansi-yellow-bg">)</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">   1886</span> <span class="ansi-yellow-bg">    </span><span class="ansi-yellow-bg">is_text</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">is_text</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">   1887</span> <span class="ansi-yellow-bg">    </span><span class="ansi-yellow-bg">errors</span><span class="ansi-yellow-bg">=</span><span style="color:rgb(0,135,0)" class="ansi-yellow-bg">self</span><span class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">options</span><span class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">get</span><span class="ansi-yellow-bg">(</span><span class="ansi-yellow-fg ansi-yellow-bg">"</span><span class="ansi-yellow-fg ansi-yellow-bg">encoding_errors</span><span class="ansi-yellow-fg ansi-yellow-bg">"</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-fg ansi-yellow-bg">"</span><span class="ansi-yellow-fg ansi-yellow-bg">strict</span><span class="ansi-yellow-fg ansi-yellow-bg">"</span><span class="ansi-yellow-bg">)</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">   1888</span> <span class="ansi-yellow-bg">    </span><span class="ansi-yellow-bg">storage_options</span><span class="ansi-yellow-bg">=</span><span style="color:rgb(0,135,0)" class="ansi-yellow-bg">self</span><span class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">options</span><span class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">get</span><span class="ansi-yellow-bg">(</span><span class="ansi-yellow-fg ansi-yellow-bg">"</span><span class="ansi-yellow-fg ansi-yellow-bg">storage_options</span><span class="ansi-yellow-fg ansi-yellow-bg">"</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span style="font-weight:bold;color:rgb(0,135,0)" class="ansi-yellow-bg">None</span><span class="ansi-yellow-bg">)</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">   1889</span> <span class="ansi-yellow-bg">)</span>
<span class="ansi-green-fg">   1890</span> <span style="font-weight:bold;color:rgb(0,135,0)">assert</span> <span style="color:rgb(0,135,0)">self</span>.handles <span style="font-weight:bold;color:rgb(175,0,255)">is</span> <span style="font-weight:bold;color:rgb(175,0,255)">not</span> <span style="font-weight:bold;color:rgb(0,135,0)">None</span>
<span class="ansi-green-fg">   1891</span> f = <span style="color:rgb(0,135,0)">self</span>.handles.handle

<span class="ansi-cyan-fg">File </span><span class="ansi-green-fg">/home/lalli/.linuxbrew/opt/python@3.11/lib/python3.11/site-packages/pandas/io/common.py:873</span>, in <span class="ansi-cyan-fg">get_handle</span><span class="ansi-blue-fg">(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)</span>
<span class="ansi-green-fg">    868</span> <span style="font-weight:bold;color:rgb(0,135,0)">elif</span> <span style="color:rgb(0,135,0)">isinstance</span>(handle, <span style="color:rgb(0,135,0)">str</span>):
<span class="ansi-green-fg">    869</span>     <span style="font-style:italic;color:rgb(95,135,135)"># Check whether the filename is to be opened in binary mode.</span>
<span class="ansi-green-fg">    870</span>     <span style="font-style:italic;color:rgb(95,135,135)"># Binary mode does not support 'encoding' and 'newline'.</span>
<span class="ansi-green-fg">    871</span>     <span style="font-weight:bold;color:rgb(0,135,0)">if</span> ioargs.encoding <span style="font-weight:bold;color:rgb(175,0,255)">and</span> <span class="ansi-yellow-fg">"</span><span class="ansi-yellow-fg">b</span><span class="ansi-yellow-fg">"</span> <span style="font-weight:bold;color:rgb(175,0,255)">not</span> <span style="font-weight:bold;color:rgb(175,0,255)">in</span> ioargs.mode:
<span class="ansi-green-fg">    872</span>         <span style="font-style:italic;color:rgb(95,135,135)"># Encoding</span>
<span class="ansi-green-fg">--&gt; </span><span class="ansi-green-fg">873</span>         handle = <span style="color:rgb(0,135,0)">open</span>(
<span class="ansi-green-fg">    874</span>             handle,
<span class="ansi-green-fg">    875</span>             ioargs.mode,
<span class="ansi-green-fg">    876</span>             encoding=ioargs.encoding,
<span class="ansi-green-fg">    877</span>             errors=errors,
<span class="ansi-green-fg">    878</span>             newline=<span class="ansi-yellow-fg">"</span><span class="ansi-yellow-fg">"</span>,
<span class="ansi-green-fg">    879</span>         )
<span class="ansi-green-fg">    880</span>     <span style="font-weight:bold;color:rgb(0,135,0)">else</span>:
<span class="ansi-green-fg">    881</span>         <span style="font-style:italic;color:rgb(95,135,135)"># Binary mode</span>
<span class="ansi-green-fg">    882</span>         handle = <span style="color:rgb(0,135,0)">open</span>(handle, ioargs.mode)

<span class="ansi-red-fg">FileNotFoundError</span>: [Errno 2] No such file or directory: 'figures/correlation_error_rates_and_gt_error_rate.csv'</pre>
</div>
</div>
</div>
<div id="cell-21" class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>HPRC_samples_phased_with_panel <span class="op">=</span> binned_maf_data.loc[(binned_maf_data.ground_truth_data_source<span class="op">==</span><span class="st">'HPRC_samples'</span>)<span class="op">&amp;</span>(binned_maf_data.method_of_phasing <span class="op">==</span> <span class="st">'1kgp_variation_phased_with_reference_panel'</span>)]</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>HPRC_samples_phased_with_panel <span class="op">=</span> group_phasing_statistics(HPRC_samples_phased_with_panel, [<span class="st">'type'</span>,<span class="st">'genome'</span>,<span class="st">'Syntenic'</span>]).drop(columns<span class="op">=</span>[<span class="st">'MAC'</span>,<span class="st">'AN'</span>,<span class="st">'MAF'</span>])</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>error_rate_corr.to_csv(<span class="st">'figures/panel_phased_HPRC_samples_summary_data.csv'</span>, index<span class="op">=</span><span class="va">False</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="get-number-of-probandsparents-in-1kgp" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="get-number-of-probandsparents-in-1kgp"><span class="header-section-number">1</span> Get number of probands/parents in 1KGP</h2>
<div id="cell-23" class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>parents <span class="op">=</span> <span class="bu">len</span>(trio_parents) <span class="op">+</span> <span class="bu">len</span>(duo_parents)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>probands <span class="op">=</span> <span class="bu">len</span>(trio_children) <span class="op">+</span> <span class="bu">len</span>(duo_children)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>statistical <span class="op">=</span> <span class="dv">3202</span><span class="op">-</span>parents<span class="op">-</span>probands</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>parents, probands, statistical</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="19">
<pre><code>(1206, 604, 1392)</code></pre>
</div>
</div>
<div id="cell-24" class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>number_of_datasets_per_sample <span class="op">=</span> all_samples.loc[(all_samples.genome<span class="op">==</span><span class="st">'GRCh38'</span>) <span class="op">&amp;</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>                                (all_samples[<span class="st">'method_of_phasing'</span>]<span class="op">==</span><span class="st">'phased_with_parents_and_pedigree'</span>) <span class="op">&amp;</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>                                (all_samples.ground_truth_data_source.isin((<span class="st">'HPRC_samples'</span>,<span class="st">'HGSVC_samples'</span>)))</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>                                ].groupby([<span class="st">'sample_id'</span>], observed<span class="op">=</span><span class="va">True</span>).size().sort_values()</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="st">"These samples are in both the HPRC and HGSVC datasets. We will use the HPRC genomes."</span>)</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (number_of_datasets_per_sample.loc[number_of_datasets_per_sample<span class="op">&gt;</span><span class="dv">1</span>].index.tolist())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>These samples are in both the HPRC and HGSVC datasets. We will use the HPRC genomes.
['HG02818', 'HG00733']</code></pre>
</div>
</div>
<div id="cell-25" class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>all_methods.loc[(all_methods.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)<span class="op">&amp;</span> (all_methods.ground_truth_data_source<span class="op">==</span><span class="st">'HPRC_samples'</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="21">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">genome</th>
<th data-quarto-table-cell-role="th">method_of_phasing</th>
<th data-quarto-table-cell-role="th">ground_truth_data_source</th>
<th data-quarto-table-cell-role="th">trio_phased</th>
<th data-quarto-table-cell-role="th">n_switch_errors</th>
<th data-quarto-table-cell-role="th">n_checked</th>
<th data-quarto-table-cell-role="th">switch_error_rate</th>
<th data-quarto-table-cell-role="th">n_total_switch_errors</th>
<th data-quarto-table-cell-role="th">n_flips</th>
<th data-quarto-table-cell-role="th">n_consecutive_flips</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">n_total_hets</th>
<th data-quarto-table-cell-role="th">flip_and_switches_SER</th>
<th data-quarto-table-cell-role="th">flip_error_rate</th>
<th data-quarto-table-cell-role="th">true_switch_error_rate</th>
<th data-quarto-table-cell-role="th">accurately_phased_rate</th>
<th data-quarto-table-cell-role="th">n_gt_errors</th>
<th data-quarto-table-cell-role="th">n_gt_checked</th>
<th data-quarto-table-cell-role="th">gt_error_rate</th>
<th data-quarto-table-cell-role="th">N50</th>
<th data-quarto-table-cell-role="th">L50</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>CHM13v2.0</td>
<td>phased_with_parents_and_pedigree</td>
<td>HPRC_samples</td>
<td>39</td>
<td>361950</td>
<td>101999287</td>
<td>0.354855</td>
<td>361950</td>
<td>169732</td>
<td>3390</td>
<td>...</td>
<td>102172409</td>
<td>0.354254</td>
<td>0.166123</td>
<td>0.018690</td>
<td>99.476305</td>
<td>6867529.0</td>
<td>747787272</td>
<td>0.918380</td>
<td>1.518510e+08</td>
<td>None</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">8</td>
<td>GRCh38</td>
<td>phased_with_parents_and_pedigree</td>
<td>HPRC_samples</td>
<td>39</td>
<td>474899</td>
<td>116411762</td>
<td>0.407948</td>
<td>474899</td>
<td>216913</td>
<td>7184</td>
<td>...</td>
<td>116635859</td>
<td>0.407164</td>
<td>0.185975</td>
<td>0.029055</td>
<td>99.400702</td>
<td>9552682.0</td>
<td>857820709</td>
<td>1.113599</td>
<td>1.531505e+08</td>
<td>None</td>
</tr>
</tbody>
</table>

<p>2 rows × 22 columns</p>
</div>
</div>
</div>
<div id="cell-26" class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>calc_ancestries <span class="op">=</span> all_ancestries.loc[(all_ancestries.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>) <span class="op">&amp;</span> </span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>                                     (all_ancestries.ground_truth_data_source.isin((<span class="st">'HPRC_samples'</span>,<span class="st">'HGSVC_samples'</span>)))]</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>calc_ancestries <span class="op">=</span> calc_ancestries.groupby([<span class="st">'superpopulation'</span>,<span class="st">'genome'</span>], observed<span class="op">=</span><span class="va">True</span>).<span class="bu">sum</span>(numeric_only<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>calc_ancestries[<span class="st">'switch_error_rate'</span>] <span class="op">=</span> calc_ancestries.n_switch_errors<span class="op">/</span>calc_ancestries.n_checked<span class="op">*</span><span class="dv">100</span></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>calc_ancestries[<span class="st">'flip_error_rate'</span>] <span class="op">=</span> calc_ancestries.n_flips<span class="op">/</span>calc_ancestries.n_checked<span class="op">*</span><span class="dv">100</span></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>calc_ancestries[<span class="st">'true_switch_error_rate'</span>] <span class="op">=</span> calc_ancestries.n_true_switch_errors<span class="op">/</span>calc_ancestries.n_checked<span class="op">*</span><span class="dv">100</span></span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>calc_ancestries <span class="op">=</span> calc_ancestries.reset_index()</span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>calc_ancestries.loc[calc_ancestries.genome<span class="op">==</span><span class="st">'GRCh38'</span>, <span class="st">'switch_error_ratio'</span>] <span class="op">=</span> calc_ancestries.loc[calc_ancestries.genome<span class="op">==</span><span class="st">'GRCh38'</span>, <span class="st">'switch_error_rate'</span>] <span class="op">/</span> calc_ancestries.loc[calc_ancestries.genome<span class="op">==</span><span class="st">'GRCh38'</span>, <span class="st">'switch_error_rate'</span>].<span class="bu">min</span>()</span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>calc_ancestries.loc[calc_ancestries.genome<span class="op">==</span><span class="st">'CHM13v2.0'</span>, <span class="st">'switch_error_ratio'</span>] <span class="op">=</span> calc_ancestries.loc[calc_ancestries.genome<span class="op">==</span><span class="st">'CHM13v2.0'</span>, <span class="st">'switch_error_rate'</span>] <span class="op">/</span> calc_ancestries.loc[calc_ancestries.genome<span class="op">==</span><span class="st">'CHM13v2.0'</span>, <span class="st">'switch_error_rate'</span>].<span class="bu">min</span>()</span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> sns.barplot(data<span class="op">=</span>calc_ancestries.reset_index(), x<span class="op">=</span><span class="st">'superpopulation'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, y<span class="op">=</span><span class="st">'switch_error_ratio'</span>)</span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a>calc_ancestries.sort_values([<span class="st">'genome'</span>,<span class="st">'superpopulation'</span>])[[<span class="st">'genome'</span>,<span class="st">'superpopulation'</span>, <span class="st">'switch_error_ratio'</span>]]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="22">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">genome</th>
<th data-quarto-table-cell-role="th">superpopulation</th>
<th data-quarto-table-cell-role="th">switch_error_ratio</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>GRCh38</td>
<td>AFR</td>
<td>1.606182</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">2</td>
<td>GRCh38</td>
<td>AMR</td>
<td>1.000000</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>GRCh38</td>
<td>EAS</td>
<td>1.934932</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">6</td>
<td>GRCh38</td>
<td>EUR</td>
<td>1.790890</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">8</td>
<td>GRCh38</td>
<td>SAS</td>
<td>1.542042</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>CHM13v2.0</td>
<td>AFR</td>
<td>1.496699</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">3</td>
<td>CHM13v2.0</td>
<td>AMR</td>
<td>1.000000</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">5</td>
<td>CHM13v2.0</td>
<td>EAS</td>
<td>2.538928</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">7</td>
<td>CHM13v2.0</td>
<td>EUR</td>
<td>2.204572</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">9</td>
<td>CHM13v2.0</td>
<td>SAS</td>
<td>1.712655</td>
</tr>
</tbody>
</table>

</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="calc_figures_for_paper_files/figure-html/cell-21-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>