<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.34">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Nature figure requirements</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="make_plots_files/libs/clipboard/clipboard.min.js"></script>
<script src="make_plots_files/libs/quarto-html/quarto.js" type="module"></script>
<script src="make_plots_files/libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="make_plots_files/libs/quarto-html/popper.min.js"></script>
<script src="make_plots_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="make_plots_files/libs/quarto-html/anchor.min.js"></script>
<link href="make_plots_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="make_plots_files/libs/quarto-html/quarto-syntax-highlighting-c8ad9e5dbd60b7b70b38521ab19b7da4.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="make_plots_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="make_plots_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="make_plots_files/libs/bootstrap/bootstrap-996b3f8b7236a2c8b17e0060dfcdce54.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


</head>

<body class="quarto-light">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#define-line-graph-styling" id="toc-define-line-graph-styling" class="nav-link active" data-scroll-target="#define-line-graph-styling"><span class="header-section-number">0.0.1</span> Define line graph styling</a></li>
  <li><a href="#figures" id="toc-figures" class="nav-link" data-scroll-target="#figures"><span class="header-section-number">1</span> Figures</a>
  <ul class="collapse">
  <li><a href="#figure-1-recombination-maps" id="toc-figure-1-recombination-maps" class="nav-link" data-scroll-target="#figure-1-recombination-maps"><span class="header-section-number">1.0.1</span> Figure 1: Recombination maps</a></li>
  <li><a href="#fig-2" id="toc-fig-2" class="nav-link" data-scroll-target="#fig-2"><span class="header-section-number">1.1</span> Fig 2</a>
  <ul class="collapse">
  <li><a href="#introduction-to-the-phased-panels-basic-performance-statistics" id="toc-introduction-to-the-phased-panels-basic-performance-statistics" class="nav-link" data-scroll-target="#introduction-to-the-phased-panels-basic-performance-statistics"><span class="header-section-number">1.1.1</span> Introduction to the phased panels, basic performance statistics</a></li>
  </ul></li>
  <li><a href="#figure-3" id="toc-figure-3" class="nav-link" data-scroll-target="#figure-3"><span class="header-section-number">1.2</span> Figure 3</a>
  <ul class="collapse">
  <li><a href="#per-chromosome-error-rate-performance-in-chm13-vs-grch38-including-chrx" id="toc-per-chromosome-error-rate-performance-in-chm13-vs-grch38-including-chrx" class="nav-link" data-scroll-target="#per-chromosome-error-rate-performance-in-chm13-vs-grch38-including-chrx"><span class="header-section-number">1.2.1</span> Per-chromosome error rate performance in CHM13 vs GRCh38, including chrX</a></li>
  </ul></li>
  <li><a href="#figure-4-switch-error-rate-vs-true-switch-error-rate" id="toc-figure-4-switch-error-rate-vs-true-switch-error-rate" class="nav-link" data-scroll-target="#figure-4-switch-error-rate-vs-true-switch-error-rate"><span class="header-section-number">1.3</span> Figure 4: Switch error rate vs true switch error rate</a></li>
  <li><a href="#figure-5" id="toc-figure-5" class="nav-link" data-scroll-target="#figure-5"><span class="header-section-number">1.4</span> Figure 5</a>
  <ul class="collapse">
  <li><a href="#phasing-performance-as-reference-dataset" id="toc-phasing-performance-as-reference-dataset" class="nav-link" data-scroll-target="#phasing-performance-as-reference-dataset"><span class="header-section-number">1.4.1</span> Phasing performance as reference dataset</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#figure-6" id="toc-figure-6" class="nav-link" data-scroll-target="#figure-6"><span class="header-section-number">2</span> Figure 6</a>
  <ul class="collapse">
  <li><a href="#figure-7" id="toc-figure-7" class="nav-link" data-scroll-target="#figure-7"><span class="header-section-number">2.1</span> Figure 7</a>
  <ul class="collapse">
  <li><a href="#imputation-performance-as-reference-dataset" id="toc-imputation-performance-as-reference-dataset" class="nav-link" data-scroll-target="#imputation-performance-as-reference-dataset"><span class="header-section-number">2.1.1</span> Imputation performance as reference dataset</a></li>
  </ul></li>
  <li><a href="#table-2" id="toc-table-2" class="nav-link" data-scroll-target="#table-2"><span class="header-section-number">2.2</span> Table 2</a></li>
  </ul></li>
  <li><a href="#additional-regions" id="toc-additional-regions" class="nav-link" data-scroll-target="#additional-regions"><span class="header-section-number">3</span> Additional regions</a>
  <ul class="collapse">
  <li><a href="#figure-2-style-plots-of-switch-error-rate-and-genotyping-error-rate-by-region-of-interest" id="toc-figure-2-style-plots-of-switch-error-rate-and-genotyping-error-rate-by-region-of-interest" class="nav-link" data-scroll-target="#figure-2-style-plots-of-switch-error-rate-and-genotyping-error-rate-by-region-of-interest"><span class="header-section-number">3.0.1</span> Figure 2-style plots of switch error rate and genotyping error rate by region of interest</a></li>
  </ul></li>
  </ul>
</nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Nature figure requirements</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<div id="cell-1" class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> scipy <span class="im">import</span> stats</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib <span class="im">as</span> mpl</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> matplotlib.ticker <span class="im">import</span> PercentFormatter, MultipleLocator</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> matplotlib.transforms <span class="im">import</span> ScaledTranslation</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> statannotations.Annotator <span class="im">import</span> Annotator</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scipy</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> matplotlib.ticker <span class="im">import</span> FuncFormatter</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> copy <span class="im">import</span> deepcopy</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> string <span class="im">import</span> ascii_lowercase <span class="im">as</span> lowercase<span class="co">#, ascii_uppercase as uppercase</span></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> upsetplot <span class="im">import</span> UpSet</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> itertools</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a>intermediate_data_folder<span class="op">=</span><span class="st">'intermediate_data_fullgenome'</span></span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> os.getcwd()<span class="op">!=</span><span class="st">'phasing_T2T_project'</span>:</span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a>    os.chdir(<span class="st">'..'</span>) <span class="co"># Change to the parent directory</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>1 col fig = 3.46” wide 2 col fig = 7.08” wide</p>
<p>300dpi</p>
<p>font: Sans serif (Helvetica, Arial) font size: 5-7 pt</p>
<p>Save as AI, EPS, or PDF.</p>
<p>Max 50MB size per fig.</p>
<p>All text needs to be in editable vector format</p>
<p>Legend: brief figure title, short description for each panel</p>
<div id="cell-3" class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co">#### Display config settings</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>plt.rcParams[<span class="st">'font.family'</span>] <span class="op">=</span> <span class="st">'sans-serif'</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>plt.rcParams[<span class="st">'font.sans-serif'</span>] <span class="op">=</span> <span class="st">'Arial'</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>plt.rcParams[<span class="st">'font.size'</span>] <span class="op">=</span> <span class="dv">6</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>one_col<span class="op">=</span><span class="fl">3.34</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>oneandahalf_col<span class="op">=</span><span class="fl">4.49</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>two_col<span class="op">=</span><span class="fl">7.08</span></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a><span class="co">## Padding (in inches) around axes; defaults to 3/72 inches, i.e. 3 points.</span></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a><span class="co">#figure.constrained_layout.h_pad:  0.04167</span></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a><span class="co">#figure.constrained_layout.w_pad:  0.04167</span></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>plt.rcParams[<span class="st">'axes.linewidth'</span>] <span class="op">=</span> <span class="fl">0.75</span></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>plt.rcParams[<span class="st">'axes.axisbelow'</span>] <span class="op">=</span> <span class="va">True</span></span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>plt.rcParams[<span class="st">'grid.alpha'</span>] <span class="op">=</span> <span class="fl">0.5</span></span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>plt.rcParams[<span class="st">'legend.frameon'</span>] <span class="op">=</span> <span class="va">True</span></span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>plt.rcParams[<span class="st">'legend.edgecolor'</span>] <span class="op">=</span> <span class="st">'white'</span></span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a>plt.rcParams[<span class="st">'figure.figsize'</span>] <span class="op">=</span> (two_col,two_col)</span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a>plt.rcParams[<span class="st">"figure.constrained_layout.use"</span>] <span class="op">=</span> <span class="va">True</span></span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a>plt.rcParams[<span class="st">"errorbar.capsize"</span>] <span class="op">=</span> <span class="fl">0.3</span></span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a>plt.rcParams[<span class="st">"savefig.dpi"</span>] <span class="op">=</span> <span class="dv">300</span></span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> add_letter_to_ax(ax: mpl.axes, label: <span class="bu">str</span> <span class="op">|</span> <span class="bu">int</span>, points_offset<span class="op">=</span>(<span class="op">-</span><span class="dv">25</span>, <span class="dv">7</span>),</span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true" tabindex="-1"></a>                         va<span class="op">=</span><span class="st">'bottom'</span>, fontsize<span class="op">=</span><span class="dv">10</span>, weight<span class="op">=</span><span class="st">'bold'</span>, fontfamily<span class="op">=</span><span class="st">'sans'</span>, alphabet<span class="op">=</span>lowercase, <span class="op">**</span>kwargs) <span class="op">-&gt;</span> mpl.axes:</span>
<span id="cb2-30"><a href="#cb2-30" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Use ScaledTranslation to put the label</span></span>
<span id="cb2-31"><a href="#cb2-31" aria-hidden="true" tabindex="-1"></a>    <span class="co"># - at the top left corner (axes fraction (0, 1)),</span></span>
<span id="cb2-32"><a href="#cb2-32" aria-hidden="true" tabindex="-1"></a>    <span class="co"># - offset 20 pixels left and 7 pixels up (offset points (-20, +7)),</span></span>
<span id="cb2-33"><a href="#cb2-33" aria-hidden="true" tabindex="-1"></a>    <span class="co"># i.e. just outside the axes.</span></span>
<span id="cb2-34"><a href="#cb2-34" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="bu">type</span>(label) <span class="op">==</span> <span class="bu">int</span>:</span>
<span id="cb2-35"><a href="#cb2-35" aria-hidden="true" tabindex="-1"></a>        label<span class="op">=</span>alphabet[label]</span>
<span id="cb2-36"><a href="#cb2-36" aria-hidden="true" tabindex="-1"></a>    ax.text(</span>
<span id="cb2-37"><a href="#cb2-37" aria-hidden="true" tabindex="-1"></a>        <span class="fl">0.0</span>, <span class="fl">1.0</span>, label, transform<span class="op">=</span>(</span>
<span id="cb2-38"><a href="#cb2-38" aria-hidden="true" tabindex="-1"></a>            ax.transAxes <span class="op">+</span> ScaledTranslation(points_offset[<span class="dv">0</span>]<span class="op">/</span><span class="dv">72</span>, points_offset[<span class="dv">1</span>]<span class="op">/</span><span class="dv">72</span>, ax.get_figure().dpi_scale_trans)),</span>
<span id="cb2-39"><a href="#cb2-39" aria-hidden="true" tabindex="-1"></a>            va<span class="op">=</span>va, fontsize<span class="op">=</span>fontsize, weight<span class="op">=</span>weight, fontfamily<span class="op">=</span>fontfamily, <span class="op">**</span>kwargs)</span>
<span id="cb2-40"><a href="#cb2-40" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> ax</span>
<span id="cb2-41"><a href="#cb2-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-42"><a href="#cb2-42" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> add_letter_to_subfig(subfig, label, points_offset<span class="op">=</span>(<span class="op">-</span><span class="dv">25</span>, <span class="dv">7</span>),</span>
<span id="cb2-43"><a href="#cb2-43" aria-hidden="true" tabindex="-1"></a>                         va<span class="op">=</span><span class="st">'bottom'</span>, fontsize<span class="op">=</span><span class="dv">10</span>, weight<span class="op">=</span><span class="st">'bold'</span>, fontfamily<span class="op">=</span><span class="st">'sans'</span>, alphabet<span class="op">=</span>lowercase, <span class="op">**</span>kwargs):</span>
<span id="cb2-44"><a href="#cb2-44" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Use ScaledTranslation to put the label</span></span>
<span id="cb2-45"><a href="#cb2-45" aria-hidden="true" tabindex="-1"></a>    <span class="co"># - at the top left corner (axes fraction (0, 1)),</span></span>
<span id="cb2-46"><a href="#cb2-46" aria-hidden="true" tabindex="-1"></a>    <span class="co"># - offset 20 pixels left and 7 pixels up (offset points (-20, +7)),</span></span>
<span id="cb2-47"><a href="#cb2-47" aria-hidden="true" tabindex="-1"></a>    <span class="co"># i.e. just outside the axes.</span></span>
<span id="cb2-48"><a href="#cb2-48" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="bu">type</span>(label) <span class="op">==</span> <span class="bu">int</span>:</span>
<span id="cb2-49"><a href="#cb2-49" aria-hidden="true" tabindex="-1"></a>        label<span class="op">=</span>alphabet[label]</span>
<span id="cb2-50"><a href="#cb2-50" aria-hidden="true" tabindex="-1"></a>    subfig.text(</span>
<span id="cb2-51"><a href="#cb2-51" aria-hidden="true" tabindex="-1"></a>        <span class="fl">0.0</span>, <span class="fl">1.0</span>, label, transform<span class="op">=</span>(</span>
<span id="cb2-52"><a href="#cb2-52" aria-hidden="true" tabindex="-1"></a>            subfig.transSubfigure <span class="op">+</span> ScaledTranslation(points_offset[<span class="dv">0</span>]<span class="op">/</span><span class="dv">72</span>, points_offset[<span class="dv">1</span>]<span class="op">/</span><span class="dv">72</span>, subfig.get_figure().dpi_scale_trans)),</span>
<span id="cb2-53"><a href="#cb2-53" aria-hidden="true" tabindex="-1"></a>            va<span class="op">=</span>va, fontsize<span class="op">=</span>fontsize, weight<span class="op">=</span>weight, fontfamily<span class="op">=</span>fontfamily, <span class="op">**</span>kwargs)</span>
<span id="cb2-54"><a href="#cb2-54" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> subfig</span>
<span id="cb2-55"><a href="#cb2-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-56"><a href="#cb2-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-57"><a href="#cb2-57" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> percent_formatter(x, pos, max_decimals<span class="op">=</span><span class="dv">4</span>):</span>
<span id="cb2-58"><a href="#cb2-58" aria-hidden="true" tabindex="-1"></a>    x <span class="op">=</span> <span class="bu">str</span>(x)</span>
<span id="cb2-59"><a href="#cb2-59" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="st">'.'</span> <span class="kw">in</span> x:</span>
<span id="cb2-60"><a href="#cb2-60" aria-hidden="true" tabindex="-1"></a>        x <span class="op">=</span> <span class="st">'.'</span>.join([x.split(<span class="st">'.'</span>)[<span class="dv">0</span>], x.split(<span class="st">'.'</span>)[<span class="dv">1</span>][:max_decimals]])</span>
<span id="cb2-61"><a href="#cb2-61" aria-hidden="true" tabindex="-1"></a>    <span class="cf">while</span> x[<span class="op">-</span><span class="dv">1</span>] <span class="op">==</span> <span class="st">'0'</span>:</span>
<span id="cb2-62"><a href="#cb2-62" aria-hidden="true" tabindex="-1"></a>        x <span class="op">=</span> x[:<span class="op">-</span><span class="dv">1</span>]</span>
<span id="cb2-63"><a href="#cb2-63" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> x[<span class="op">-</span><span class="dv">1</span>] <span class="op">==</span> <span class="st">'.'</span>:</span>
<span id="cb2-64"><a href="#cb2-64" aria-hidden="true" tabindex="-1"></a>        x <span class="op">=</span> x[:<span class="op">-</span><span class="dv">1</span>]</span>
<span id="cb2-65"><a href="#cb2-65" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> x<span class="op">+</span><span class="st">'%'</span></span>
<span id="cb2-66"><a href="#cb2-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-67"><a href="#cb2-67" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> axis_has_legend(ax) <span class="op">-&gt;</span> <span class="bu">bool</span>:</span>
<span id="cb2-68"><a href="#cb2-68" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> ax.get_legend() <span class="kw">is</span> <span class="kw">not</span> <span class="va">None</span>  <span class="co"># or: getattr(ax, "legend_", None) is not None</span></span>
<span id="cb2-69"><a href="#cb2-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-70"><a href="#cb2-70" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> figure_has_legend(ax) <span class="op">-&gt;</span> <span class="bu">bool</span>:</span>
<span id="cb2-71"><a href="#cb2-71" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> axis_has_legend(ax) <span class="kw">or</span> <span class="bu">bool</span>(ax.figure.legends)</span>
<span id="cb2-72"><a href="#cb2-72" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-73"><a href="#cb2-73" aria-hidden="true" tabindex="-1"></a>perc_formatter <span class="op">=</span> FuncFormatter(percent_formatter)</span>
<span id="cb2-74"><a href="#cb2-74" aria-hidden="true" tabindex="-1"></a>MAF_xticks <span class="op">=</span> [<span class="fl">0.01</span>, <span class="fl">0.1</span>, <span class="dv">1</span>, <span class="dv">10</span>, <span class="dv">50</span>]</span>
<span id="cb2-75"><a href="#cb2-75" aria-hidden="true" tabindex="-1"></a>SER_xticks <span class="op">=</span> [<span class="fl">0.1</span>, <span class="dv">1</span>, <span class="dv">10</span>, <span class="dv">50</span>]</span>
<span id="cb2-76"><a href="#cb2-76" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-77"><a href="#cb2-77" aria-hidden="true" tabindex="-1"></a>genome_palette<span class="op">=</span>[(<span class="fl">0.4</span>,<span class="fl">0.4</span>,<span class="fl">0.4</span>)] <span class="op">+</span> sns.color_palette()[<span class="dv">4</span>:<span class="dv">5</span>]</span>
<span id="cb2-78"><a href="#cb2-78" aria-hidden="true" tabindex="-1"></a>type_markers <span class="op">=</span> {<span class="st">'SNPs + Indels'</span>:<span class="st">','</span>, <span class="st">'SNP'</span>:<span class="st">'o'</span>, <span class="st">'Indel'</span>:<span class="st">'o'</span>}</span>
<span id="cb2-79"><a href="#cb2-79" aria-hidden="true" tabindex="-1"></a>type_dashes <span class="op">=</span> {<span class="st">'SNPs + Indels'</span>:<span class="st">''</span>, <span class="st">'SNP'</span>:(<span class="dv">1</span>,<span class="dv">1</span>), <span class="st">'Indel'</span>:(<span class="dv">3</span>,<span class="fl">0.5</span>)}</span>
<span id="cb2-80"><a href="#cb2-80" aria-hidden="true" tabindex="-1"></a>syntenic_markers <span class="op">=</span> {<span class="st">'Syntenic'</span>:<span class="st">'o'</span>, <span class="st">'Nonsyntenic'</span>:<span class="st">'o'</span>}</span>
<span id="cb2-81"><a href="#cb2-81" aria-hidden="true" tabindex="-1"></a>syntenic_dashes <span class="op">=</span> {<span class="st">'Syntenic'</span>:<span class="st">''</span>, <span class="st">'Nonsyntenic'</span>:(<span class="dv">1</span>,<span class="dv">1</span>)}</span>
<span id="cb2-82"><a href="#cb2-82" aria-hidden="true" tabindex="-1"></a>update_legend_values<span class="op">=</span>{<span class="st">'genome'</span>:<span class="st">'Genome'</span>,</span>
<span id="cb2-83"><a href="#cb2-83" aria-hidden="true" tabindex="-1"></a>                      <span class="st">'SNPs + Indels'</span>: <span class="st">'All Variants'</span>,</span>
<span id="cb2-84"><a href="#cb2-84" aria-hidden="true" tabindex="-1"></a>                      <span class="st">'type'</span>: <span class="st">'Variant Type'</span>,</span>
<span id="cb2-85"><a href="#cb2-85" aria-hidden="true" tabindex="-1"></a>                      <span class="st">'syntenic'</span>:<span class="st">'Genomic Region'</span>,</span>
<span id="cb2-86"><a href="#cb2-86" aria-hidden="true" tabindex="-1"></a>                      <span class="st">'Nonsyntenic'</span>:<span class="st">'CHM13v2.0</span><span class="ch">\n</span><span class="st">(Newly Resolved)'</span>,</span>
<span id="cb2-87"><a href="#cb2-87" aria-hidden="true" tabindex="-1"></a>                      <span class="st">'CHM13v2.0'</span>:<span class="st">'CHM13v2.0'</span>,</span>
<span id="cb2-88"><a href="#cb2-88" aria-hidden="true" tabindex="-1"></a>                      <span class="st">'GRCh38'</span>:<span class="st">'GRCh38'</span>,</span>
<span id="cb2-89"><a href="#cb2-89" aria-hidden="true" tabindex="-1"></a>                      <span class="st">'not_in_STRs'</span>:<span class="st">'Not in GIAB STR region'</span>,</span>
<span id="cb2-90"><a href="#cb2-90" aria-hidden="true" tabindex="-1"></a>                      <span class="st">'in_STRs'</span>:<span class="st">'In GIAB STR region'</span>,</span>
<span id="cb2-91"><a href="#cb2-91" aria-hidden="true" tabindex="-1"></a>                      <span class="st">'in_segdups'</span>:<span class="st">'In Segmental Duplication'</span>,</span>
<span id="cb2-92"><a href="#cb2-92" aria-hidden="true" tabindex="-1"></a>                      <span class="st">'not_in_segdups'</span>:<span class="st">'Not in Segmental Duplication'</span>,</span>
<span id="cb2-93"><a href="#cb2-93" aria-hidden="true" tabindex="-1"></a>                      <span class="st">'biallelic'</span>:<span class="st">'Isolated, biallelic variant'</span>,</span>
<span id="cb2-94"><a href="#cb2-94" aria-hidden="true" tabindex="-1"></a>                      <span class="st">'multiallelic'</span>:<span class="st">'Variant overlaps another variant'</span>,</span>
<span id="cb2-95"><a href="#cb2-95" aria-hidden="true" tabindex="-1"></a>                      <span class="st">'not_in_platinum_STRs'</span>:<span class="st">'Not in Platinum Genomes STR region'</span>,</span>
<span id="cb2-96"><a href="#cb2-96" aria-hidden="true" tabindex="-1"></a>                      <span class="st">'in_platinum_STRs'</span>:<span class="st">'In Platinum Genomes STR region'</span>}</span>
<span id="cb2-97"><a href="#cb2-97" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-98"><a href="#cb2-98" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-99"><a href="#cb2-99" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-100"><a href="#cb2-100" aria-hidden="true" tabindex="-1"></a>genome_order <span class="op">=</span> [<span class="st">'GRCh38'</span>,<span class="st">'CHM13v2.0'</span>]</span>
<span id="cb2-101"><a href="#cb2-101" aria-hidden="true" tabindex="-1"></a>type_order <span class="op">=</span> [<span class="st">'SNPs + Indels'</span>, <span class="st">'SNP'</span>, <span class="st">'Indel'</span>]</span>
<span id="cb2-102"><a href="#cb2-102" aria-hidden="true" tabindex="-1"></a>syntenic_order <span class="op">=</span> [<span class="st">'Syntenic'</span>, <span class="st">'Nonsyntenic'</span>]</span>
<span id="cb2-103"><a href="#cb2-103" aria-hidden="true" tabindex="-1"></a>chrom_order <span class="op">=</span> [<span class="st">'Whole</span><span class="ch">\n</span><span class="st">Genome'</span>] <span class="op">+</span> [<span class="st">'chr'</span><span class="op">+</span><span class="bu">str</span>(i) <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">1</span>,<span class="dv">23</span>)] <span class="op">+</span> [<span class="st">'PAR1'</span>,<span class="st">'PAR2'</span>,<span class="st">'chrX'</span>]</span>
<span id="cb2-104"><a href="#cb2-104" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-105"><a href="#cb2-105" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-106"><a href="#cb2-106" aria-hidden="true" tabindex="-1"></a>default_style <span class="op">=</span> {</span>
<span id="cb2-107"><a href="#cb2-107" aria-hidden="true" tabindex="-1"></a>    <span class="st">'errorbar'</span>: <span class="va">None</span>,</span>
<span id="cb2-108"><a href="#cb2-108" aria-hidden="true" tabindex="-1"></a>    <span class="st">'palette'</span>:genome_palette,</span>
<span id="cb2-109"><a href="#cb2-109" aria-hidden="true" tabindex="-1"></a>    <span class="st">'markers'</span>:type_markers,</span>
<span id="cb2-110"><a href="#cb2-110" aria-hidden="true" tabindex="-1"></a>    <span class="st">'dashes'</span>:type_dashes,</span>
<span id="cb2-111"><a href="#cb2-111" aria-hidden="true" tabindex="-1"></a>    <span class="st">'markersize'</span>:<span class="dv">3</span>,</span>
<span id="cb2-112"><a href="#cb2-112" aria-hidden="true" tabindex="-1"></a>    <span class="st">'markeredgewidth'</span>:<span class="fl">0.2</span>,</span>
<span id="cb2-113"><a href="#cb2-113" aria-hidden="true" tabindex="-1"></a>    <span class="st">'markeredgecolor'</span>: <span class="st">'white'</span>,</span>
<span id="cb2-114"><a href="#cb2-114" aria-hidden="true" tabindex="-1"></a>    <span class="st">'linewidth'</span>:<span class="dv">1</span></span>
<span id="cb2-115"><a href="#cb2-115" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb2-116"><a href="#cb2-116" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-117"><a href="#cb2-117" aria-hidden="true" tabindex="-1"></a>formatted_ancestry_names <span class="op">=</span> {<span class="st">'all'</span>:<span class="st">'All</span><span class="ch">\n</span><span class="st">populations'</span>,<span class="st">'WestEurasia'</span>:<span class="st">'West</span><span class="ch">\n</span><span class="st">Eurasia'</span>,<span class="st">'SouthAsia'</span>:<span class="st">"South</span><span class="ch">\n</span><span class="st">Asia"</span>,</span>
<span id="cb2-118"><a href="#cb2-118" aria-hidden="true" tabindex="-1"></a>                            <span class="st">'EastAsia'</span>:<span class="st">'East</span><span class="ch">\n</span><span class="st">Asia'</span>,<span class="st">'Africa'</span>:<span class="st">'Africa'</span>, <span class="st">'America'</span>:<span class="st">'Americas'</span>,<span class="st">'Oceania'</span>:<span class="st">'Oceania'</span>,</span>
<span id="cb2-119"><a href="#cb2-119" aria-hidden="true" tabindex="-1"></a>                            <span class="st">'CentralAsiaSiberia'</span>:<span class="st">'Central Asia</span><span class="ch">\n</span><span class="st">&amp; Siberia'</span>}</span>
<span id="cb2-120"><a href="#cb2-120" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-121"><a href="#cb2-121" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> add_grid(ax, <span class="op">**</span>kwargs):</span>
<span id="cb2-122"><a href="#cb2-122" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> ax.grid(alpha<span class="op">=</span><span class="fl">0.5</span>, <span class="op">**</span>kwargs)</span>
<span id="cb2-123"><a href="#cb2-123" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-124"><a href="#cb2-124" aria-hidden="true" tabindex="-1"></a>syntenic_style <span class="op">=</span> deepcopy(default_style)</span>
<span id="cb2-125"><a href="#cb2-125" aria-hidden="true" tabindex="-1"></a>syntenic_style[<span class="st">'markers'</span>] <span class="op">=</span> syntenic_markers</span>
<span id="cb2-126"><a href="#cb2-126" aria-hidden="true" tabindex="-1"></a>syntenic_style[<span class="st">'dashes'</span>] <span class="op">=</span> syntenic_dashes</span>
<span id="cb2-127"><a href="#cb2-127" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-128"><a href="#cb2-128" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-129"><a href="#cb2-129" aria-hidden="true" tabindex="-1"></a>figures_folder<span class="op">=</span><span class="st">'figures'</span></span>
<span id="cb2-130"><a href="#cb2-130" aria-hidden="true" tabindex="-1"></a>supplemental_figures_folder <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>figures_folder<span class="sc">}</span><span class="ss">/supplemental'</span></span>
<span id="cb2-131"><a href="#cb2-131" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-132"><a href="#cb2-132" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> os.path.exists(figures_folder):</span>
<span id="cb2-133"><a href="#cb2-133" aria-hidden="true" tabindex="-1"></a>    os.mkdir(figures_folder)</span>
<span id="cb2-134"><a href="#cb2-134" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> os.path.exists(supplemental_figures_folder):</span>
<span id="cb2-135"><a href="#cb2-135" aria-hidden="true" tabindex="-1"></a>    os.mkdir(supplemental_figures_folder)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-4" class="cell" data-execution_count="3">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>binned_maf_data <span class="op">=</span> pd.read_parquet(<span class="ss">f"</span><span class="sc">{</span>intermediate_data_folder<span class="sc">}</span><span class="ss">/binned_maf_data.parquet"</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>all_chroms <span class="op">=</span> pd.read_parquet(<span class="ss">f'</span><span class="sc">{</span>intermediate_data_folder<span class="sc">}</span><span class="ss">/chroms.parquet'</span>)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>all_samples <span class="op">=</span> pd.read_parquet(<span class="ss">f'</span><span class="sc">{</span>intermediate_data_folder<span class="sc">}</span><span class="ss">/samples.parquet'</span>)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>all_ancestries <span class="op">=</span> pd.read_parquet(<span class="ss">f'</span><span class="sc">{</span>intermediate_data_folder<span class="sc">}</span><span class="ss">/ancestries.parquet'</span>)</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>all_methods <span class="op">=</span> pd.read_parquet(<span class="ss">f'</span><span class="sc">{</span>intermediate_data_folder<span class="sc">}</span><span class="ss">/methods.parquet'</span>)</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>per_variant_category_imputation_performance <span class="op">=</span> pd.read_parquet(<span class="ss">f'</span><span class="sc">{</span>intermediate_data_folder<span class="sc">}</span><span class="ss">/per_variant_category_imputation_performance.parquet'</span>)</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>per_sample_imputation_performance <span class="op">=</span> pd.read_parquet(<span class="ss">f'</span><span class="sc">{</span>intermediate_data_folder<span class="sc">}</span><span class="ss">/per_sample_imputation_performance.parquet'</span>)</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>sum_stats_var_filters <span class="op">=</span> pd.read_parquet(<span class="ss">f"</span><span class="sc">{</span>intermediate_data_folder<span class="sc">}</span><span class="ss">/filter_summary_stats.parquet"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-5" class="cell" data-execution_count="4">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Post hoc fixes - one off events that are tricky to deal with when compiling stats</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="co"># there is one variant (I have not investigated, but likely a denovo SNP) that is a singleton variant in the HPRC dataset. </span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="co"># That variant passed all filters, but makes for a very odd figure (correctly phased: 0% error rate, incorrect: 100% error rate)</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="co"># We will trim it here</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>binned_maf_data.loc[(binned_maf_data.ground_truth_data_source <span class="op">==</span> <span class="st">'HPRC_samples'</span>)<span class="op">&amp;</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>                    (binned_maf_data.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)<span class="op">&amp;</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>                    (binned_maf_data.rounded_MAF <span class="op">==</span> <span class="st">'singleton'</span>), </span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>                    [<span class="st">'n_switch_errors'</span>,<span class="st">'n_checked'</span>,<span class="st">'switch_error_rate'</span>]] <span class="op">=</span> np.nan</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>binned_maf_data_regions <span class="op">=</span> binned_maf_data.loc[<span class="op">~</span>binned_maf_data.region.<span class="bu">str</span>.contains(<span class="st">'teni'</span>)]</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>binned_maf_data <span class="op">=</span> binned_maf_data.loc[binned_maf_data.region.<span class="bu">str</span>.contains(<span class="st">'teni'</span>)<span class="op">|</span>(binned_maf_data.region <span class="op">==</span> <span class="st">'All'</span>)].rename(columns<span class="op">=</span>{<span class="st">'region'</span>:<span class="st">'syntenic'</span>})</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a><span class="co"># I never converted syntenic representation from true/false to something named. Let's do that.</span></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>binned_maf_data <span class="op">=</span> binned_maf_data.assign(syntenic<span class="op">=</span>binned_maf_data.syntenic</span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>                                                                 .astype(<span class="bu">str</span>)</span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a>                                                                 .fillna(<span class="st">'All'</span>)</span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a>                                                                 .replace(<span class="va">True</span>,<span class="st">'Syntenic'</span>)</span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a>                                                                 .replace(<span class="va">False</span>,<span class="st">'Nonsyntenic'</span>)</span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a>                                                                 )</span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a><span class="co"># Nonsyntenic GRCh38 regions are mostly GRCh38 errors, and we are not interested in the performance in these regions</span></span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a>binned_maf_data <span class="op">=</span> binned_maf_data.loc[<span class="op">~</span>((binned_maf_data.genome<span class="op">==</span><span class="st">'GRCh38'</span>)<span class="op">&amp;</span>(binned_maf_data.syntenic<span class="op">==</span><span class="st">'Nonsyntenic'</span>))]</span>
<span id="cb4-26"><a href="#cb4-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-27"><a href="#cb4-27" aria-hidden="true" tabindex="-1"></a><span class="co"># NA20355 is a duo member, and as such was not part of duohmm trio correction for the GRCh38 2022 effort.</span></span>
<span id="cb4-28"><a href="#cb4-28" aria-hidden="true" tabindex="-1"></a>all_chroms.loc[(all_chroms.genome<span class="op">==</span><span class="st">'GRCh38'</span>)<span class="op">&amp;</span>(all_chroms[<span class="st">'sample_id'</span>]<span class="op">==</span><span class="st">'NA20355'</span>), <span class="st">'trio_phased'</span>] <span class="op">=</span> <span class="va">False</span></span>
<span id="cb4-29"><a href="#cb4-29" aria-hidden="true" tabindex="-1"></a>all_samples.loc[(all_samples.genome<span class="op">==</span><span class="st">'GRCh38'</span>)<span class="op">&amp;</span>(all_samples[<span class="st">'sample_id'</span>]<span class="op">==</span><span class="st">'NA20355'</span>), <span class="st">'trio_phased'</span>] <span class="op">=</span> <span class="va">False</span></span>
<span id="cb4-30"><a href="#cb4-30" aria-hidden="true" tabindex="-1"></a>all_chroms <span class="op">=</span> all_chroms.loc[<span class="op">~</span>((all_chroms.genome<span class="op">==</span><span class="st">'GRCh38'</span>)<span class="op">&amp;</span>(all_chroms[<span class="st">'sample_id'</span>]<span class="op">==</span><span class="st">'NA20355'</span>)<span class="op">&amp;</span>(all_chroms.ground_truth_data_source.isin([<span class="st">'HPRC_HGSVC_probands'</span>,<span class="st">'HGSVC_probands'</span>])))]</span>
<span id="cb4-31"><a href="#cb4-31" aria-hidden="true" tabindex="-1"></a>all_samples <span class="op">=</span> all_samples.loc[<span class="op">~</span>((all_samples.genome<span class="op">==</span><span class="st">'GRCh38'</span>)<span class="op">&amp;</span>(all_samples[<span class="st">'sample_id'</span>]<span class="op">==</span><span class="st">'NA20355'</span>)<span class="op">&amp;</span></span>
<span id="cb4-32"><a href="#cb4-32" aria-hidden="true" tabindex="-1"></a>                                (all_samples.ground_truth_data_source.isin([<span class="st">'HPRC_HGSVC_probands'</span>,<span class="st">'HGSVC_probands'</span>])))]</span>
<span id="cb4-33"><a href="#cb4-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-34"><a href="#cb4-34" aria-hidden="true" tabindex="-1"></a><span class="co"># very high variance stats when n_checked is less than 10. Replace switch error rate at these sites with np.nan</span></span>
<span id="cb4-35"><a href="#cb4-35" aria-hidden="true" tabindex="-1"></a>binned_maf_data.loc[(binned_maf_data.n_checked<span class="op">&gt;</span><span class="dv">10</span>), <span class="st">'switch_error_rate'</span>] <span class="op">=</span> binned_maf_data.loc[(binned_maf_data.n_checked<span class="op">&gt;</span><span class="dv">10</span>), <span class="st">'switch_error_rate'</span>].replace(<span class="dv">0</span>, np.nan)</span>
<span id="cb4-36"><a href="#cb4-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-37"><a href="#cb4-37" aria-hidden="true" tabindex="-1"></a><span class="co"># bins with a switch error rate of 0 make for insane figures when plotted on a log scale. Replace with np.nan, and maybe note</span></span>
<span id="cb4-38"><a href="#cb4-38" aria-hidden="true" tabindex="-1"></a>binned_maf_data[<span class="st">'switch_error_rate'</span>] <span class="op">=</span> binned_maf_data[<span class="st">'switch_error_rate'</span>].replace(<span class="dv">0</span>, np.nan)</span>
<span id="cb4-39"><a href="#cb4-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-40"><a href="#cb4-40" aria-hidden="true" tabindex="-1"></a><span class="co"># A reviewer suggests combining chrX, PAR1, and PAR2 statistics. I will do that, and set male chrX switch error rates to np.nan.</span></span>
<span id="cb4-41"><a href="#cb4-41" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate percent discordance in imputation performance</span></span>
<span id="cb4-42"><a href="#cb4-42" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="st">'perc_discordance'</span> <span class="kw">not</span> <span class="kw">in</span> per_variant_category_imputation_performance.columns:</span>
<span id="cb4-43"><a href="#cb4-43" aria-hidden="true" tabindex="-1"></a>    per_variant_category_imputation_performance <span class="op">=</span> per_variant_category_imputation_performance.loc[per_variant_category_imputation_performance.num_AA <span class="op">&gt;</span> <span class="dv">0</span>]</span>
<span id="cb4-44"><a href="#cb4-44" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="bu">max</span>(per_variant_category_imputation_performance.mean_AF) <span class="op">&lt;</span> <span class="dv">1</span>:</span>
<span id="cb4-45"><a href="#cb4-45" aria-hidden="true" tabindex="-1"></a>        per_variant_category_imputation_performance.mean_AF <span class="op">*=</span> <span class="dv">100</span></span>
<span id="cb4-46"><a href="#cb4-46" aria-hidden="true" tabindex="-1"></a>    per_variant_category_imputation_performance[<span class="st">'num_alt_mismatches'</span>] <span class="op">=</span> per_variant_category_imputation_performance.num_Aa_mismatches <span class="op">+</span> per_variant_category_imputation_performance.num_aa_mismatches</span>
<span id="cb4-47"><a href="#cb4-47" aria-hidden="true" tabindex="-1"></a>    per_variant_category_imputation_performance[<span class="st">'num_alt_variants'</span>] <span class="op">=</span> per_variant_category_imputation_performance.num_Aa <span class="op">+</span> per_variant_category_imputation_performance.num_aa</span>
<span id="cb4-48"><a href="#cb4-48" aria-hidden="true" tabindex="-1"></a>    per_variant_category_imputation_performance[<span class="st">'perc_discordance'</span>] <span class="op">=</span> per_variant_category_imputation_performance[<span class="st">'num_alt_mismatches'</span>]<span class="op">/</span>per_variant_category_imputation_performance[<span class="st">'num_alt_variants'</span>]<span class="op">*</span><span class="dv">100</span></span>
<span id="cb4-49"><a href="#cb4-49" aria-hidden="true" tabindex="-1"></a>    per_variant_category_imputation_performance[<span class="st">'perc_concordance'</span>] <span class="op">=</span> <span class="dv">100</span><span class="op">-</span>per_variant_category_imputation_performance[<span class="st">'perc_discordance'</span>]</span>
<span id="cb4-50"><a href="#cb4-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-51"><a href="#cb4-51" aria-hidden="true" tabindex="-1"></a><span class="co"># imputed df needs to have CHM13v2.0 as genome name instead of T2T</span></span>
<span id="cb4-52"><a href="#cb4-52" aria-hidden="true" tabindex="-1"></a>per_variant_category_imputation_performance <span class="op">=</span> per_variant_category_imputation_performance.replace(<span class="st">'T2T'</span>,<span class="st">'CHM13v2.0'</span>)</span>
<span id="cb4-53"><a href="#cb4-53" aria-hidden="true" tabindex="-1"></a>per_sample_imputation_performance <span class="op">=</span> per_sample_imputation_performance.replace(<span class="st">'T2T'</span>,<span class="st">'CHM13v2.0'</span>)</span>
<span id="cb4-54"><a href="#cb4-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-55"><a href="#cb4-55" aria-hidden="true" tabindex="-1"></a><span class="co"># A supplemental calculates SER by % of chromosome that is new. we calc that here.</span></span>
<span id="cb4-56"><a href="#cb4-56" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="st">'percent_new'</span> <span class="kw">not</span> <span class="kw">in</span> all_chroms.columns:</span>
<span id="cb4-57"><a href="#cb4-57" aria-hidden="true" tabindex="-1"></a>    syntenic_bed <span class="op">=</span> pd.read_csv(<span class="st">'resources/hg38.GCA_009914755.4.synNet.summary.bed.gz'</span>, sep<span class="op">=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>, header<span class="op">=</span><span class="va">None</span>)</span>
<span id="cb4-58"><a href="#cb4-58" aria-hidden="true" tabindex="-1"></a>    syntenic_bed.columns <span class="op">=</span> [<span class="st">'chrom'</span>,<span class="st">'start'</span>,<span class="st">'end'</span>,<span class="st">'status'</span>]</span>
<span id="cb4-59"><a href="#cb4-59" aria-hidden="true" tabindex="-1"></a>    starts <span class="op">=</span> syntenic_bed.groupby(<span class="st">'chrom'</span>).first().reset_index()</span>
<span id="cb4-60"><a href="#cb4-60" aria-hidden="true" tabindex="-1"></a>    syntenic_bed[<span class="st">'length'</span>] <span class="op">=</span> syntenic_bed.end<span class="op">-</span>syntenic_bed.start</span>
<span id="cb4-61"><a href="#cb4-61" aria-hidden="true" tabindex="-1"></a>    syntenic_bed[<span class="st">'dif'</span>] <span class="op">=</span> syntenic_bed.end.diff()</span>
<span id="cb4-62"><a href="#cb4-62" aria-hidden="true" tabindex="-1"></a>    syntenic_bed[<span class="st">'nonsyn_len'</span>] <span class="op">=</span> syntenic_bed.dif <span class="op">-</span> syntenic_bed.length</span>
<span id="cb4-63"><a href="#cb4-63" aria-hidden="true" tabindex="-1"></a>    syntenic_bed.loc[syntenic_bed.nonsyn_len <span class="op">&lt;</span> <span class="dv">0</span>, <span class="st">'nonsyn_len'</span>] <span class="op">=</span> syntenic_bed.loc[syntenic_bed.nonsyn_len <span class="op">&lt;</span> <span class="dv">0</span>, <span class="st">'start'</span>]</span>
<span id="cb4-64"><a href="#cb4-64" aria-hidden="true" tabindex="-1"></a>    syntenic_bed.loc[syntenic_bed.nonsyn_len.isna(), <span class="st">'nonsyn_len'</span>] <span class="op">=</span> syntenic_bed.loc[syntenic_bed.nonsyn_len.isna(), <span class="st">'start'</span>]</span>
<span id="cb4-65"><a href="#cb4-65" aria-hidden="true" tabindex="-1"></a>    rough_length <span class="op">=</span> syntenic_bed.groupby(<span class="st">'chrom'</span>).end.<span class="bu">max</span>().reset_index()</span>
<span id="cb4-66"><a href="#cb4-66" aria-hidden="true" tabindex="-1"></a>    chr_nonsyn_lengths <span class="op">=</span> syntenic_bed.groupby(<span class="st">'chrom'</span>).<span class="bu">sum</span>(numeric_only<span class="op">=</span><span class="va">True</span>).reset_index()[[<span class="st">'chrom'</span>,<span class="st">'nonsyn_len'</span>]]</span>
<span id="cb4-67"><a href="#cb4-67" aria-hidden="true" tabindex="-1"></a>    chr_nonsyn_lengths <span class="op">=</span> chr_nonsyn_lengths.merge(rough_length, on<span class="op">=</span><span class="st">'chrom'</span>)</span>
<span id="cb4-68"><a href="#cb4-68" aria-hidden="true" tabindex="-1"></a>    chr_nonsyn_lengths[<span class="st">'percent_new'</span>] <span class="op">=</span> chr_nonsyn_lengths.nonsyn_len<span class="op">/</span>chr_nonsyn_lengths.end <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb4-69"><a href="#cb4-69" aria-hidden="true" tabindex="-1"></a>    all_chroms <span class="op">=</span> all_chroms.merge(chr_nonsyn_lengths[[<span class="st">'chrom'</span>, <span class="st">'percent_new'</span>]], on<span class="op">=</span>[<span class="st">'chrom'</span>],how<span class="op">=</span><span class="st">'left'</span>)</span>
<span id="cb4-70"><a href="#cb4-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-71"><a href="#cb4-71" aria-hidden="true" tabindex="-1"></a>all_chroms_chrX <span class="op">=</span> all_chroms.loc[all_chroms.chrom.isin((<span class="st">'chrX'</span>,<span class="st">'PAR1'</span>,<span class="st">'PAR2'</span>))].assign(chrom<span class="op">=</span><span class="st">'chrX'</span>)</span>
<span id="cb4-72"><a href="#cb4-72" aria-hidden="true" tabindex="-1"></a>all_chroms_chrX <span class="op">=</span> all_chroms_chrX.groupby([<span class="st">'chrom'</span>,<span class="st">'genome'</span>,<span class="st">'sample_id'</span>,<span class="st">'method_of_phasing'</span>,<span class="st">'ground_truth_data_source'</span>,<span class="st">'population'</span>, <span class="st">'superpopulation'</span>, <span class="st">'sex'</span>, <span class="st">'trio_phased'</span>],observed<span class="op">=</span><span class="va">True</span>).<span class="bu">sum</span>(numeric_only<span class="op">=</span><span class="va">True</span>).reset_index()</span>
<span id="cb4-73"><a href="#cb4-73" aria-hidden="true" tabindex="-1"></a>all_chroms_chrX.loc[all_chroms_chrX.sex<span class="op">==</span><span class="st">'Male'</span>, [<span class="st">'n_switch_errors'</span>,<span class="st">'n_total_switch_errors'</span>,<span class="st">'n_flips'</span>,<span class="st">'n_consecutive_flips'</span>,<span class="st">'n_true_switch_errors'</span>,<span class="st">'num_correct_phased_hets'</span>,</span>
<span id="cb4-74"><a href="#cb4-74" aria-hidden="true" tabindex="-1"></a>                                                  <span class="st">'n_total_hets'</span>,<span class="st">'flip_and_switches_SER'</span>, <span class="st">'flip_error_rate'</span>, <span class="st">'true_switch_error_rate'</span>,<span class="st">'accurately_phased_rate'</span>]] <span class="op">=</span> np.nan</span>
<span id="cb4-75"><a href="#cb4-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-76"><a href="#cb4-76" aria-hidden="true" tabindex="-1"></a>all_chroms_chrX[<span class="st">'switch_error_rate'</span>] <span class="op">=</span> all_chroms_chrX.n_switch_errors<span class="op">/</span>all_chroms_chrX.n_checked <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb4-77"><a href="#cb4-77" aria-hidden="true" tabindex="-1"></a>all_chroms_chrX[<span class="st">'gt_error_rate'</span>] <span class="op">=</span> all_chroms_chrX.n_gt_errors<span class="op">/</span>all_chroms_chrX.n_gt_checked <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb4-78"><a href="#cb4-78" aria-hidden="true" tabindex="-1"></a>all_chroms_chrX[<span class="st">'flip_and_switches_SER'</span>] <span class="op">=</span> all_chroms_chrX.n_total_switch_errors<span class="op">/</span>all_chroms_chrX.n_total_hets <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb4-79"><a href="#cb4-79" aria-hidden="true" tabindex="-1"></a>all_chroms_chrX[<span class="st">'flip_error_rate'</span>] <span class="op">=</span> all_chroms_chrX.n_flips<span class="op">/</span>all_chroms_chrX.n_total_hets <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb4-80"><a href="#cb4-80" aria-hidden="true" tabindex="-1"></a>all_chroms_chrX[<span class="st">'true_switch_error_rate'</span>] <span class="op">=</span> all_chroms_chrX.n_true_switch_errors<span class="op">/</span>all_chroms_chrX.n_total_hets <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb4-81"><a href="#cb4-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-82"><a href="#cb4-82" aria-hidden="true" tabindex="-1"></a>all_chroms_chrX <span class="op">=</span> pd.concat((all_chroms_chrX, all_chroms.loc[<span class="op">~</span>all_chroms.chrom.isin((<span class="st">'chrX'</span>,<span class="st">'PAR1'</span>,<span class="st">'PAR2'</span>))])).drop(columns<span class="op">=</span>[<span class="st">'L50'</span>,<span class="st">'N50'</span>,<span class="st">'percent_new'</span>])</span>
<span id="cb4-83"><a href="#cb4-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-84"><a href="#cb4-84" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-85"><a href="#cb4-85" aria-hidden="true" tabindex="-1"></a>binned_maf_data <span class="op">=</span> binned_maf_data.sort_values(<span class="st">'type'</span>, key<span class="op">=</span><span class="kw">lambda</span> col: col.<span class="bu">apply</span>(<span class="kw">lambda</span> x: {<span class="st">'Indel'</span>:<span class="dv">0</span>, <span class="st">'SNP'</span>:<span class="dv">1</span>, <span class="st">'SNPs + Indels'</span>:<span class="dv">2</span>}.get(x,x)))</span>
<span id="cb4-86"><a href="#cb4-86" aria-hidden="true" tabindex="-1"></a>binned_maf_data <span class="op">=</span> binned_maf_data.sort_values(<span class="st">'genome'</span>, key<span class="op">=</span><span class="kw">lambda</span> col: col.<span class="bu">apply</span>(<span class="kw">lambda</span> x: {<span class="st">'GRCh38'</span>:<span class="dv">0</span>, <span class="st">'CHM13v2.0'</span>:<span class="dv">1</span>}.get(x,x)))</span>
<span id="cb4-87"><a href="#cb4-87" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-88"><a href="#cb4-88" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> binned_maf_data[<span class="st">'switch_error_rate'</span>].<span class="bu">max</span>() <span class="op">&lt;</span> <span class="dv">1</span>:</span>
<span id="cb4-89"><a href="#cb4-89" aria-hidden="true" tabindex="-1"></a>    binned_maf_data[<span class="st">'switch_error_rate'</span>] <span class="op">*=</span> <span class="dv">100</span></span>
<span id="cb4-90"><a href="#cb4-90" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> binned_maf_data[<span class="st">'gt_error_rate'</span>].<span class="bu">max</span>() <span class="op">&lt;</span> <span class="dv">1</span>:</span>
<span id="cb4-91"><a href="#cb4-91" aria-hidden="true" tabindex="-1"></a>    binned_maf_data[<span class="st">'gt_error_rate'</span>] <span class="op">*=</span> <span class="dv">100</span></span>
<span id="cb4-92"><a href="#cb4-92" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> binned_maf_data[<span class="st">'MAF'</span>].<span class="bu">max</span>() <span class="op">&lt;</span> <span class="dv">1</span>:</span>
<span id="cb4-93"><a href="#cb4-93" aria-hidden="true" tabindex="-1"></a>    binned_maf_data[<span class="st">'MAF'</span>] <span class="op">*=</span> <span class="dv">100</span></span>
<span id="cb4-94"><a href="#cb4-94" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-95"><a href="#cb4-95" aria-hidden="true" tabindex="-1"></a><span class="co"># a handful of variants are miscategorized as singletons. There are no true GRCh38 singletons. We can fix this.</span></span>
<span id="cb4-96"><a href="#cb4-96" aria-hidden="true" tabindex="-1"></a>binned_maf_data <span class="op">=</span> binned_maf_data.loc[<span class="op">~</span>((binned_maf_data.rounded_MAF<span class="op">==</span><span class="st">'singleton'</span>)<span class="op">&amp;</span>(binned_maf_data.genome<span class="op">==</span><span class="st">'GRCh38'</span>)<span class="op">&amp;</span>(binned_maf_data.method_of_phasing<span class="op">==</span><span class="st">'phased_with_parents_and_pedigree'</span>)<span class="op">&amp;</span>(binned_maf_data.ground_truth_data_source<span class="op">!=</span><span class="st">'trios'</span>))]</span>
<span id="cb4-97"><a href="#cb4-97" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-98"><a href="#cb4-98" aria-hidden="true" tabindex="-1"></a>binned_maf_data.to_parquet(<span class="ss">f'</span><span class="sc">{</span>intermediate_data_folder<span class="sc">}</span><span class="ss">/per_MAF_bin.parquet'</span>)</span>
<span id="cb4-99"><a href="#cb4-99" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-100"><a href="#cb4-100" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate weighted "overall" error rates</span></span>
<span id="cb4-101"><a href="#cb4-101" aria-hidden="true" tabindex="-1"></a>cat_data <span class="op">=</span> binned_maf_data.loc[binned_maf_data.ground_truth_data_source.isin((<span class="st">'HPRC_samples'</span>, <span class="st">'HGSVC_parents'</span>, <span class="st">'HGSVC_parents'</span>, <span class="st">'HGSVC_samples_nontrios_only'</span>)) <span class="op">&amp;</span></span>
<span id="cb4-102"><a href="#cb4-102" aria-hidden="true" tabindex="-1"></a>                               (binned_maf_data.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)].copy()</span>
<span id="cb4-103"><a href="#cb4-103" aria-hidden="true" tabindex="-1"></a>cat_data.loc[cat_data.n_switch_errors <span class="op">==</span> <span class="dv">0</span>, <span class="st">'switch_error_rate'</span>] <span class="op">=</span> np.nan</span>
<span id="cb4-104"><a href="#cb4-104" aria-hidden="true" tabindex="-1"></a>cat_data[<span class="st">'support'</span>] <span class="op">=</span> cat_data.ground_truth_data_source.<span class="bu">map</span>({<span class="st">'HPRC_samples'</span>:<span class="st">'proband'</span>, <span class="st">'HGSVC_probands'</span>:<span class="st">'proband'</span>,<span class="st">'HGSVC_parents'</span>:<span class="st">'parent'</span>, <span class="st">'HGSVC_samples_nontrios_only'</span>:<span class="st">'nontrio'</span>})</span>
<span id="cb4-105"><a href="#cb4-105" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-106"><a href="#cb4-106" aria-hidden="true" tabindex="-1"></a>num_probands_in_dataset <span class="op">=</span> <span class="dv">68</span></span>
<span id="cb4-107"><a href="#cb4-107" aria-hidden="true" tabindex="-1"></a>num_parents_in_dataset <span class="op">=</span>  <span class="dv">13</span></span>
<span id="cb4-108"><a href="#cb4-108" aria-hidden="true" tabindex="-1"></a>num_nontrios_in_dataset <span class="op">=</span> <span class="dv">19</span></span>
<span id="cb4-109"><a href="#cb4-109" aria-hidden="true" tabindex="-1"></a>num_probands_in_1kgp <span class="op">=</span> <span class="dv">608</span></span>
<span id="cb4-110"><a href="#cb4-110" aria-hidden="true" tabindex="-1"></a>num_parents_in_1kgp  <span class="op">=</span> <span class="dv">1199</span></span>
<span id="cb4-111"><a href="#cb4-111" aria-hidden="true" tabindex="-1"></a>num_nontrios_in_1kgp <span class="op">=</span> <span class="dv">1399</span></span>
<span id="cb4-112"><a href="#cb4-112" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-113"><a href="#cb4-113" aria-hidden="true" tabindex="-1"></a>cat_data.loc[(cat_data.support<span class="op">==</span><span class="st">'proband'</span>) <span class="op">&amp;</span> (cat_data.genome<span class="op">==</span><span class="st">'GRCh38'</span>), [<span class="st">'n_switch_errors'</span>,<span class="st">'n_checked'</span>, <span class="st">'n_gt_errors'</span>,<span class="st">'n_gt_checked'</span>,<span class="st">'MAC'</span>,<span class="st">'AN'</span>]] <span class="op">=</span> (((num_probands_in_1kgp<span class="op">-</span><span class="dv">1</span>)<span class="op">/</span>num_probands_in_dataset<span class="op">-</span><span class="dv">1</span>) <span class="op">*</span> cat_data.loc[(cat_data.support<span class="op">==</span><span class="st">'proband'</span>) <span class="op">&amp;</span> (cat_data.genome<span class="op">==</span><span class="st">'GRCh38'</span>), [<span class="st">'n_switch_errors'</span>,<span class="st">'n_checked'</span>, <span class="st">'n_gt_errors'</span>,<span class="st">'n_gt_checked'</span>,<span class="st">'MAC'</span>,<span class="st">'AN'</span>]]).fillna(<span class="dv">0</span>).astype(<span class="bu">int</span>)</span>
<span id="cb4-114"><a href="#cb4-114" aria-hidden="true" tabindex="-1"></a>cat_data.loc[(cat_data.support<span class="op">==</span><span class="st">'proband'</span>) <span class="op">&amp;</span> (cat_data.genome<span class="op">==</span><span class="st">'CHM13v2.0'</span>), [<span class="st">'n_switch_errors'</span>,<span class="st">'n_checked'</span>, <span class="st">'n_gt_errors'</span>,<span class="st">'n_gt_checked'</span>,<span class="st">'MAC'</span>,<span class="st">'AN'</span>]] <span class="op">=</span> ((num_probands_in_1kgp<span class="op">/</span>num_probands_in_dataset) <span class="op">*</span> cat_data.loc[(cat_data.support<span class="op">==</span><span class="st">'proband'</span>) <span class="op">&amp;</span> (cat_data.genome<span class="op">==</span><span class="st">'CHM13v2.0'</span>), [<span class="st">'n_switch_errors'</span>,<span class="st">'n_checked'</span>, <span class="st">'n_gt_errors'</span>,<span class="st">'n_gt_checked'</span>,<span class="st">'MAC'</span>,<span class="st">'AN'</span>]]).fillna(<span class="dv">0</span>).astype(<span class="bu">int</span>)</span>
<span id="cb4-115"><a href="#cb4-115" aria-hidden="true" tabindex="-1"></a>cat_data.loc[cat_data.support<span class="op">==</span><span class="st">'parent'</span>, [<span class="st">'n_switch_errors'</span>,<span class="st">'n_checked'</span>, <span class="st">'n_gt_errors'</span>,<span class="st">'n_gt_checked'</span>,<span class="st">'MAC'</span>,<span class="st">'AN'</span>]] <span class="op">=</span> ((num_parents_in_1kgp<span class="op">/</span>num_parents_in_dataset) <span class="op">*</span> cat_data.loc[cat_data.support<span class="op">==</span><span class="st">'parent'</span>, [<span class="st">'n_switch_errors'</span>,<span class="st">'n_checked'</span>, <span class="st">'n_gt_errors'</span>,<span class="st">'n_gt_checked'</span>,<span class="st">'MAC'</span>,<span class="st">'AN'</span>]]).fillna(<span class="dv">0</span>).astype(<span class="bu">int</span>)</span>
<span id="cb4-116"><a href="#cb4-116" aria-hidden="true" tabindex="-1"></a>cat_data.loc[cat_data.support<span class="op">==</span><span class="st">'nontrio'</span>, [<span class="st">'n_switch_errors'</span>,<span class="st">'n_checked'</span>, <span class="st">'n_gt_errors'</span>,<span class="st">'n_gt_checked'</span>,<span class="st">'MAC'</span>,<span class="st">'AN'</span>]] <span class="op">=</span> ((num_nontrios_in_1kgp<span class="op">/</span>num_nontrios_in_dataset) <span class="op">*</span> cat_data.loc[cat_data.support<span class="op">==</span><span class="st">'nontrio'</span>, [<span class="st">'n_switch_errors'</span>,<span class="st">'n_checked'</span>, <span class="st">'n_gt_errors'</span>,<span class="st">'n_gt_checked'</span>,<span class="st">'MAC'</span>,<span class="st">'AN'</span>]]).fillna(<span class="dv">0</span>).astype(<span class="bu">int</span>)</span>
<span id="cb4-117"><a href="#cb4-117" aria-hidden="true" tabindex="-1"></a>cat_data.loc[(cat_data.n_switch_errors <span class="op">==</span> <span class="dv">0</span>), <span class="st">'switch_error_rate'</span>] <span class="op">=</span> np.nan</span>
<span id="cb4-118"><a href="#cb4-118" aria-hidden="true" tabindex="-1"></a>cat_data <span class="op">=</span> cat_data.groupby([<span class="st">'genome'</span>,<span class="st">'type'</span>,<span class="st">'rounded_MAF'</span>,<span class="st">'syntenic'</span>], observed<span class="op">=</span><span class="va">True</span>).<span class="bu">sum</span>(numeric_only<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb4-119"><a href="#cb4-119" aria-hidden="true" tabindex="-1"></a>cat_data[<span class="st">'gt_error_rate'</span>] <span class="op">=</span> cat_data.n_switch_errors<span class="op">/</span>cat_data.n_checked <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb4-120"><a href="#cb4-120" aria-hidden="true" tabindex="-1"></a>cat_data[<span class="st">'gt_error_rate'</span>] <span class="op">=</span> cat_data.n_gt_errors<span class="op">/</span>cat_data.n_gt_checked <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb4-121"><a href="#cb4-121" aria-hidden="true" tabindex="-1"></a>cat_data[<span class="st">'MAF'</span>] <span class="op">=</span> cat_data.MAC<span class="op">/</span>cat_data.AN <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb4-122"><a href="#cb4-122" aria-hidden="true" tabindex="-1"></a>cat_data <span class="op">=</span> cat_data.reset_index()</span>
<span id="cb4-123"><a href="#cb4-123" aria-hidden="true" tabindex="-1"></a>cat_data[<span class="st">'switch_error_rate'</span>] <span class="op">=</span> cat_data[<span class="st">'switch_error_rate'</span>].replace(<span class="dv">0</span>, np.nan)</span>
<span id="cb4-124"><a href="#cb4-124" aria-hidden="true" tabindex="-1"></a>binned_maf_data <span class="op">=</span> pd.concat([binned_maf_data, cat_data.assign(ground_truth_data_source<span class="op">=</span><span class="st">'estimated'</span>, method_of_phasing<span class="op">=</span><span class="st">'phased_with_parents_and_pedigree'</span>)])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<section id="define-line-graph-styling" class="level3" data-number="0.0.1">
<h3 data-number="0.0.1" class="anchored" data-anchor-id="define-line-graph-styling"><span class="header-section-number">0.0.1</span> Define line graph styling</h3>
<div id="cell-7" class="cell" data-execution_count="5">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> highlight_value(df, var, val):</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>    df[<span class="st">'highlight'</span>] <span class="op">=</span> <span class="va">False</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>    df.loc[df[var] <span class="op">==</span> val, <span class="st">'highlight'</span>] <span class="op">=</span> <span class="va">True</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> df</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>highlight_style <span class="op">=</span> {<span class="st">'size'</span>:<span class="st">'highlight'</span>, <span class="st">'size_order'</span>:[<span class="va">True</span>, <span class="va">False</span>], <span class="st">'sizes'</span>:(<span class="fl">1.5</span>,<span class="dv">3</span>)}</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> add_highlights_legend(ax, var, val, sizes<span class="op">=</span>highlight_style[<span class="st">'sizes'</span>]):</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> is_variable_title(handle):</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> handle.get_color()<span class="op">==</span><span class="st">'w'</span></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>    new_handles <span class="op">=</span> <span class="bu">list</span>()</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>    new_labels <span class="op">=</span> <span class="bu">list</span>()</span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>    handles, labels <span class="op">=</span> ax.get_legend_handles_labels()</span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>    in_higlight_var_section<span class="op">=</span><span class="va">False</span></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> handle, label <span class="kw">in</span> <span class="bu">zip</span>(handles, labels):</span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> is_variable_title(handle):</span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a>            in_higlight_var_section <span class="op">=</span> label <span class="op">==</span> var</span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> label <span class="kw">not</span> <span class="kw">in</span> [<span class="st">'highlight'</span>,<span class="st">'True'</span>,<span class="st">'False'</span>]:</span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> in_higlight_var_section:</span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a>                <span class="cf">if</span> label <span class="op">==</span> val:</span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a>                    handle.set_linewidth(sizes[<span class="dv">1</span>])</span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a>                <span class="cf">else</span>:</span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true" tabindex="-1"></a>                    handle.set_linewidth(sizes[<span class="dv">0</span>])</span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true" tabindex="-1"></a>            new_labels.append(update_legend_values.get(label,label))</span>
<span id="cb5-28"><a href="#cb5-28" aria-hidden="true" tabindex="-1"></a>            new_handles.append(handle)</span>
<span id="cb5-29"><a href="#cb5-29" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span>:</span>
<span id="cb5-30"><a href="#cb5-30" aria-hidden="true" tabindex="-1"></a>            <span class="bu">print</span>(label)</span>
<span id="cb5-31"><a href="#cb5-31" aria-hidden="true" tabindex="-1"></a>    ax.legend(handles<span class="op">=</span>new_handles, labels<span class="op">=</span>new_labels)</span>
<span id="cb5-32"><a href="#cb5-32" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> ax</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="figures" class="level1" data-number="1">
<h1 data-number="1"><span class="header-section-number">1</span> Figures</h1>
<section id="figure-1-recombination-maps" class="level3" data-number="1.0.1">
<h3 data-number="1.0.1" class="anchored" data-anchor-id="figure-1-recombination-maps"><span class="header-section-number">1.0.1</span> Figure 1: Recombination maps</h3>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="make_plots_files/figure-html/cell-9-1-image.png" class="img-fluid figure-img"></p>
<figcaption>image.png</figcaption>
</figure>
</div>
</section>
<section id="fig-2" class="level2" data-number="1.1">
<h2 data-number="1.1" class="anchored" data-anchor-id="fig-2"><span class="header-section-number">1.1</span> Fig 2</h2>
<section id="introduction-to-the-phased-panels-basic-performance-statistics" class="level3" data-number="1.1.1">
<h3 data-number="1.1.1" class="anchored" data-anchor-id="introduction-to-the-phased-panels-basic-performance-statistics"><span class="header-section-number">1.1.1</span> Introduction to the phased panels, basic performance statistics</h3>
<ul>
<li><p>2x2 figure: Panel accuracy, split by SNPs, indels, mendelian- and statistically phased samples (4 figure panels)</p></li>
<li><p>Supplemental: Swarm plots of error rates by probands, parents, stat-phased can go in supplement</p>
<ul>
<li>Genotyping error rates, same 2x2 as figure 2</li>
</ul></li>
</ul>
<div id="cell-11" class="cell" data-execution_count="6">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Figure 2</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>fig, ((axA, axB),(axC, axD)) <span class="op">=</span> plt.subplots(<span class="dv">2</span>,<span class="dv">2</span>, figsize<span class="op">=</span>(two_col, <span class="dv">5</span>))</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>fig.set_layout_engine(layout<span class="op">=</span><span class="st">'constrained'</span>, h_pad<span class="op">=</span><span class="fl">0.15</span>, w_pad<span class="op">=</span><span class="fl">0.15</span>)</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'switch_error_rate'</span>, style<span class="op">=</span><span class="st">'syntenic'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, <span class="co">#hue_order = genome_order, style_order = syntenic_order,</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data.loc[(binned_maf_data.ground_truth_data_source <span class="op">==</span> <span class="st">'HPRC_samples'</span>)</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.<span class="bu">type</span> <span class="op">==</span> <span class="st">'SNP'</span>)</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.syntenic <span class="op">!=</span> <span class="st">'All'</span>)],</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axA, <span class="op">**</span>syntenic_style)</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'switch_error_rate'</span>, style<span class="op">=</span><span class="st">'syntenic'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> syntenic_order,</span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data.loc[(binned_maf_data.ground_truth_data_source <span class="op">==</span> <span class="st">'HGSVC_samples_nontrios_only'</span>) <span class="co">#trios only</span></span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.<span class="bu">type</span> <span class="op">==</span> <span class="st">'SNP'</span>)</span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.syntenic <span class="op">!=</span> <span class="st">'All'</span>)],</span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axB, <span class="op">**</span>syntenic_style)</span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'switch_error_rate'</span>, style<span class="op">=</span><span class="st">'syntenic'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> syntenic_order,</span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data.loc[(binned_maf_data.ground_truth_data_source <span class="op">==</span> <span class="st">'HPRC_samples'</span>)</span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.<span class="bu">type</span> <span class="op">==</span> <span class="st">'Indel'</span>)</span>
<span id="cb6-23"><a href="#cb6-23" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.syntenic <span class="op">!=</span> <span class="st">'All'</span>)],</span>
<span id="cb6-24"><a href="#cb6-24" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axC, <span class="op">**</span>syntenic_style)</span>
<span id="cb6-25"><a href="#cb6-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-26"><a href="#cb6-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-27"><a href="#cb6-27" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'switch_error_rate'</span>, style<span class="op">=</span><span class="st">'syntenic'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> syntenic_order,</span>
<span id="cb6-28"><a href="#cb6-28" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data.loc[(binned_maf_data.ground_truth_data_source <span class="op">==</span> <span class="st">'HGSVC_samples_nontrios_only'</span>) <span class="co">#trios only</span></span>
<span id="cb6-29"><a href="#cb6-29" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb6-30"><a href="#cb6-30" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.<span class="bu">type</span> <span class="op">==</span> <span class="st">'Indel'</span>)</span>
<span id="cb6-31"><a href="#cb6-31" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.syntenic <span class="op">!=</span> <span class="st">'All'</span>)],</span>
<span id="cb6-32"><a href="#cb6-32" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axD, <span class="op">**</span>syntenic_style)</span>
<span id="cb6-33"><a href="#cb6-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-34"><a href="#cb6-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-35"><a href="#cb6-35" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, ax <span class="kw">in</span> <span class="bu">enumerate</span>([axA, axB, axC, axD]):</span>
<span id="cb6-36"><a href="#cb6-36" aria-hidden="true" tabindex="-1"></a>    ax.set_yscale(<span class="st">'log'</span>)</span>
<span id="cb6-37"><a href="#cb6-37" aria-hidden="true" tabindex="-1"></a>    ax.set_xscale(<span class="st">'log'</span>)</span>
<span id="cb6-38"><a href="#cb6-38" aria-hidden="true" tabindex="-1"></a>    ax.set_ylim(<span class="fl">1e-1</span>, <span class="dv">50</span>)</span>
<span id="cb6-39"><a href="#cb6-39" aria-hidden="true" tabindex="-1"></a>    ax.set_xlim(<span class="fl">1e-2</span>, <span class="dv">55</span>)</span>
<span id="cb6-40"><a href="#cb6-40" aria-hidden="true" tabindex="-1"></a>    <span class="co"># ax.set_box_aspect(1)</span></span>
<span id="cb6-41"><a href="#cb6-41" aria-hidden="true" tabindex="-1"></a>    ax.grid()</span>
<span id="cb6-42"><a href="#cb6-42" aria-hidden="true" tabindex="-1"></a>    ax.set_axisbelow(<span class="va">True</span>)</span>
<span id="cb6-43"><a href="#cb6-43" aria-hidden="true" tabindex="-1"></a>    ax.set_xticks(MAF_xticks)</span>
<span id="cb6-44"><a href="#cb6-44" aria-hidden="true" tabindex="-1"></a>    ax.set_yticks(SER_xticks)</span>
<span id="cb6-45"><a href="#cb6-45" aria-hidden="true" tabindex="-1"></a>    ax.xaxis.set_major_formatter(perc_formatter)</span>
<span id="cb6-46"><a href="#cb6-46" aria-hidden="true" tabindex="-1"></a>    ax.yaxis.set_major_formatter(perc_formatter)</span>
<span id="cb6-47"><a href="#cb6-47" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="kw">in</span> (<span class="dv">2</span>,<span class="dv">3</span>):</span>
<span id="cb6-48"><a href="#cb6-48" aria-hidden="true" tabindex="-1"></a>        ax.set_xlabel(<span class="st">'Minor Allele Frequency (%)'</span>, labelpad<span class="op">=</span><span class="dv">3</span>)</span>
<span id="cb6-49"><a href="#cb6-49" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb6-50"><a href="#cb6-50" aria-hidden="true" tabindex="-1"></a>        ax.set_xlabel(<span class="va">None</span>)</span>
<span id="cb6-51"><a href="#cb6-51" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="kw">in</span> (<span class="dv">0</span>,<span class="dv">2</span>):</span>
<span id="cb6-52"><a href="#cb6-52" aria-hidden="true" tabindex="-1"></a>        ax.set_ylabel(<span class="st">'Switch Error Rate (%)'</span>, labelpad<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb6-53"><a href="#cb6-53" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb6-54"><a href="#cb6-54" aria-hidden="true" tabindex="-1"></a>        ax.set_ylabel(<span class="va">None</span>)</span>
<span id="cb6-55"><a href="#cb6-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-56"><a href="#cb6-56" aria-hidden="true" tabindex="-1"></a>axA.text(<span class="op">-</span><span class="fl">.2</span>,<span class="fl">0.5</span>, <span class="st">'SNPs'</span>, fontsize<span class="op">=</span><span class="dv">10</span>, transform<span class="op">=</span>axA.transAxes, weight<span class="op">=</span><span class="st">'bold'</span>, rotation<span class="op">=</span><span class="st">'vertical'</span>, horizontalalignment<span class="op">=</span><span class="st">'right'</span>, verticalalignment<span class="op">=</span><span class="st">'center'</span>)</span>
<span id="cb6-57"><a href="#cb6-57" aria-hidden="true" tabindex="-1"></a>axC.text(<span class="op">-</span><span class="fl">.2</span>,<span class="fl">0.5</span>, <span class="st">'Indels'</span>, fontsize<span class="op">=</span><span class="dv">10</span>, transform<span class="op">=</span>axC.transAxes, weight<span class="op">=</span><span class="st">'bold'</span>, rotation<span class="op">=</span><span class="st">'vertical'</span>, horizontalalignment<span class="op">=</span><span class="st">'right'</span>, verticalalignment<span class="op">=</span><span class="st">'center'</span>)</span>
<span id="cb6-58"><a href="#cb6-58" aria-hidden="true" tabindex="-1"></a>axA.set_title(<span class="st">'Mendelian phased HPRC samples'</span>, fontsize<span class="op">=</span><span class="dv">9</span>, weight<span class="op">=</span><span class="st">'bold'</span>, pad<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb6-59"><a href="#cb6-59" aria-hidden="true" tabindex="-1"></a>axB.set_title(<span class="st">'Statistically phased HGSVC samples'</span>, fontsize<span class="op">=</span><span class="dv">9</span>, weight<span class="op">=</span><span class="st">'bold'</span>, pad<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb6-60"><a href="#cb6-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-61"><a href="#cb6-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-62"><a href="#cb6-62" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, ax <span class="kw">in</span> <span class="bu">enumerate</span>([axA, axB, axC, axD]): <span class="co"># legend axes</span></span>
<span id="cb6-63"><a href="#cb6-63" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="op">==</span> <span class="dv">0</span>:</span>
<span id="cb6-64"><a href="#cb6-64" aria-hidden="true" tabindex="-1"></a>        loc<span class="op">=</span><span class="st">'upper right'</span></span>
<span id="cb6-65"><a href="#cb6-65" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb6-66"><a href="#cb6-66" aria-hidden="true" tabindex="-1"></a>        loc<span class="op">=</span><span class="st">'upper right'</span></span>
<span id="cb6-67"><a href="#cb6-67" aria-hidden="true" tabindex="-1"></a>    handles, labels <span class="op">=</span> ax.get_legend_handles_labels()</span>
<span id="cb6-68"><a href="#cb6-68" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> handle <span class="kw">in</span> handles:</span>
<span id="cb6-69"><a href="#cb6-69" aria-hidden="true" tabindex="-1"></a>        handle.set_marker(<span class="st">'o'</span>)</span>
<span id="cb6-70"><a href="#cb6-70" aria-hidden="true" tabindex="-1"></a>    handles <span class="op">=</span> handles[<span class="dv">1</span>:<span class="dv">3</span>] <span class="op">+</span> handles[<span class="dv">5</span>:<span class="dv">6</span>]</span>
<span id="cb6-71"><a href="#cb6-71" aria-hidden="true" tabindex="-1"></a>    labels <span class="op">=</span> labels[<span class="dv">1</span>:<span class="dv">3</span>] <span class="op">+</span> labels[<span class="dv">5</span>:<span class="dv">6</span>]</span>
<span id="cb6-72"><a href="#cb6-72" aria-hidden="true" tabindex="-1"></a>    handles[<span class="op">-</span><span class="dv">1</span>].set_color(handles[<span class="dv">1</span>].get_color())</span>
<span id="cb6-73"><a href="#cb6-73" aria-hidden="true" tabindex="-1"></a>    legend <span class="op">=</span> ax.legend(ncols<span class="op">=</span><span class="dv">1</span>, loc<span class="op">=</span>loc, handles<span class="op">=</span>handles, labels<span class="op">=</span>labels)</span>
<span id="cb6-74"><a href="#cb6-74" aria-hidden="true" tabindex="-1"></a>    legend.set_title(<span class="va">None</span>)</span>
<span id="cb6-75"><a href="#cb6-75" aria-hidden="true" tabindex="-1"></a>    legend.get_frame().<span class="bu">set</span>(</span>
<span id="cb6-76"><a href="#cb6-76" aria-hidden="true" tabindex="-1"></a>        alpha<span class="op">=</span><span class="fl">0.75</span>, boxstyle<span class="op">=</span><span class="st">'square'</span>, edgecolor<span class="op">=</span><span class="st">'grey'</span>, linewidth<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb6-77"><a href="#cb6-77" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> text <span class="kw">in</span> legend.get_texts():</span>
<span id="cb6-78"><a href="#cb6-78" aria-hidden="true" tabindex="-1"></a>        t <span class="op">=</span> text.get_text()</span>
<span id="cb6-79"><a href="#cb6-79" aria-hidden="true" tabindex="-1"></a>        new_t <span class="op">=</span> update_legend_values.get(t,t)</span>
<span id="cb6-80"><a href="#cb6-80" aria-hidden="true" tabindex="-1"></a>        text.set_text(new_t)</span>
<span id="cb6-81"><a href="#cb6-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-82"><a href="#cb6-82" aria-hidden="true" tabindex="-1"></a>x_pos <span class="op">=</span> <span class="fl">0.5275</span></span>
<span id="cb6-83"><a href="#cb6-83" aria-hidden="true" tabindex="-1"></a>y_pos <span class="op">=</span> <span class="fl">0.5</span></span>
<span id="cb6-84"><a href="#cb6-84" aria-hidden="true" tabindex="-1"></a>perc_of_figure<span class="op">=</span><span class="fl">0.95</span></span>
<span id="cb6-85"><a href="#cb6-85" aria-hidden="true" tabindex="-1"></a>fig.add_artist(mpl.lines.Line2D([(<span class="dv">1</span><span class="op">-</span>perc_of_figure)<span class="op">/</span><span class="dv">2</span>, <span class="dv">1</span> <span class="op">-</span> (<span class="dv">1</span><span class="op">-</span>perc_of_figure)<span class="op">/</span><span class="dv">2</span>], [y_pos, y_pos], color<span class="op">=</span><span class="st">'grey'</span>, alpha<span class="op">=</span><span class="dv">1</span>, linewidth<span class="op">=</span><span class="dv">1</span>))</span>
<span id="cb6-86"><a href="#cb6-86" aria-hidden="true" tabindex="-1"></a>fig.add_artist(mpl.lines.Line2D([x_pos, x_pos], [(<span class="dv">1</span><span class="op">-</span>perc_of_figure)<span class="op">/</span><span class="dv">2</span>, <span class="dv">1</span> <span class="op">-</span> (<span class="dv">1</span><span class="op">-</span>perc_of_figure)<span class="op">/</span><span class="dv">2</span>], color<span class="op">=</span><span class="st">'grey'</span>, alpha<span class="op">=</span><span class="dv">1</span>, linewidth<span class="op">=</span><span class="dv">1</span>))</span>
<span id="cb6-87"><a href="#cb6-87" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-88"><a href="#cb6-88" aria-hidden="true" tabindex="-1"></a>add_letter_to_ax(axA, <span class="st">'a'</span>)</span>
<span id="cb6-89"><a href="#cb6-89" aria-hidden="true" tabindex="-1"></a>add_letter_to_ax(axB, <span class="st">'b'</span>)</span>
<span id="cb6-90"><a href="#cb6-90" aria-hidden="true" tabindex="-1"></a>add_letter_to_ax(axC, <span class="st">'c'</span>)</span>
<span id="cb6-91"><a href="#cb6-91" aria-hidden="true" tabindex="-1"></a>add_letter_to_ax(axD, <span class="st">'d'</span>)</span>
<span id="cb6-92"><a href="#cb6-92" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-93"><a href="#cb6-93" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/Figure 2.png'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span>
<span id="cb6-94"><a href="#cb6-94" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/Figure 2.svg'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="make_plots_files/figure-html/cell-7-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Fig 2. Phased panel switch error rates binned by minor allele frequency. Of the 3202 participants in the 1KGP, 39 had their genomes assembled by the HPRC. An additional 61 had their genomes assembled by the HGSVC. a) The switch error rate of haplotype panel SNPs from 39 indiviudals, using their HPRC-assembled genomes as ground truth. Variants located in regions of CHM13v2.0 that are non-syntenic with GRCh38 are included in overall variant bin error rates, but are also plotted separately as CHM13v2.0 (Novel) variants. b) SER of haplotype panel SNPs from 19 individuals, using their HGSVC-assembled genomes as ground truth. Unlike the 39 HPRC samples, these 19 individuals are not part of a 1KGP trio, and therefore were phased without any mendelian-based error correction. c) Indel SER from the mendelian-phased HPRC samples. d) Indel SER from the 19 individuals phased without mendelian error correction. Switch error rates and average minor allele frequency per bin are displayed on a log scale.</p>
<div id="cell-13" class="cell" data-execution_count="7">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Figure 2 supplementals</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Supplemental 1: effects of variant filtering</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> importlib</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> upsetplot</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>importlib.<span class="bu">reload</span>(upsetplot)</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> upsetplot <span class="im">import</span> UpSet</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="ss">f'Using upsetplot version </span><span class="sc">{</span>upsetplot<span class="sc">.</span><span class="va">__file__</span><span class="sc">}</span><span class="ss">'</span>)</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> move_all_passing_to_front(df):</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>    passing<span class="op">=</span>(<span class="va">False</span>,)<span class="op">*</span><span class="bu">len</span>(df.index[<span class="dv">0</span>])</span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>    new_index <span class="op">=</span> df.groupby(df.index).<span class="bu">sum</span>().sort_values(<span class="st">'num_variants'</span>, ascending<span class="op">=</span><span class="va">False</span>).index.drop(passing)</span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>    new_index <span class="op">=</span> new_index.insert(<span class="dv">0</span>, passing)</span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> df.loc[new_index,:]</span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> reshape_data_for_upsetplot(data, genome_name, filter_columns, color_scheme):</span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""Process genome-specific data for UpSet plots"""</span></span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a>    genome_data <span class="op">=</span> data.loc[data.genome<span class="op">==</span>genome_name].rename(columns<span class="op">=</span>{<span class="st">'len'</span>:<span class="st">'num_variants'</span>,<span class="st">'CHM13_criteria_fail'</span>:<span class="st">'other_criteria_fail'</span>} <span class="cf">if</span> genome_name <span class="op">==</span> <span class="st">'GRCh38'</span> <span class="cf">else</span> {<span class="st">'len'</span>:<span class="st">'num_variants'</span>,<span class="st">'GRCh38_criteria_fail'</span>:<span class="st">'other_criteria_fail'</span>})</span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> genome_name <span class="op">==</span> <span class="st">'CHM13v2.0'</span>:</span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a>        genome_data[<span class="st">'grch38_fail_besides_singleton'</span>] <span class="op">=</span> genome_data[data[[x <span class="cf">for</span> x <span class="kw">in</span> [<span class="st">'MERR_filter'</span>, <span class="st">'HWE_pop_filter'</span>,<span class="st">'AC_filter'</span>, <span class="st">'f_missing_filter'</span>, <span class="st">'var_len_filter'</span>, <span class="st">'alt_star_filter'</span>, <span class="st">'pass_filter'</span>] <span class="cf">if</span> x <span class="op">!=</span> <span class="st">'AC_filter'</span>] <span class="op">+</span> [<span class="st">'MAC_filter'</span>]].columns].<span class="bu">any</span>(axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb7-22"><a href="#cb7-22" aria-hidden="true" tabindex="-1"></a>    genome_data[<span class="st">'other_criteria_fail'</span>] <span class="op">=</span> genome_data[<span class="st">'other_criteria_fail'</span>].astype(<span class="bu">str</span>)</span>
<span id="cb7-23"><a href="#cb7-23" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb7-24"><a href="#cb7-24" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> genome_name <span class="op">==</span> <span class="st">'GRCh38'</span>:</span>
<span id="cb7-25"><a href="#cb7-25" aria-hidden="true" tabindex="-1"></a>        genome_data.loc[genome_data.other_criteria_fail<span class="op">==</span><span class="st">'False'</span>, <span class="st">'other_criteria_fail'</span>] <span class="op">=</span> <span class="st">'No'</span></span>
<span id="cb7-26"><a href="#cb7-26" aria-hidden="true" tabindex="-1"></a>        genome_data.loc[genome_data.other_criteria_fail<span class="op">==</span><span class="st">'True'</span>, <span class="st">'other_criteria_fail'</span>] <span class="op">=</span> <span class="st">'Yes'</span></span>
<span id="cb7-27"><a href="#cb7-27" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:  <span class="co"># CHM13v2.0</span></span>
<span id="cb7-28"><a href="#cb7-28" aria-hidden="true" tabindex="-1"></a>        genome_data.loc[genome_data.singleton <span class="op">&amp;</span> (genome_data.grch38_fail_besides_singleton <span class="op">!=</span> <span class="st">'True'</span>), <span class="st">'other_criteria_fail'</span>] <span class="op">=</span> <span class="st">'Yes: singleton variant'</span></span>
<span id="cb7-29"><a href="#cb7-29" aria-hidden="true" tabindex="-1"></a>        genome_data.loc[genome_data.other_criteria_fail<span class="op">==</span><span class="st">'False'</span>, <span class="st">'other_criteria_fail'</span>] <span class="op">=</span> <span class="st">'No'</span></span>
<span id="cb7-30"><a href="#cb7-30" aria-hidden="true" tabindex="-1"></a>        genome_data.loc[genome_data.other_criteria_fail<span class="op">==</span><span class="st">'True'</span>, <span class="st">'other_criteria_fail'</span>] <span class="op">=</span> <span class="st">'Yes: Other reason'</span></span>
<span id="cb7-31"><a href="#cb7-31" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb7-32"><a href="#cb7-32" aria-hidden="true" tabindex="-1"></a>    filters <span class="op">=</span> genome_data.groupby([<span class="st">'other_criteria_fail'</span>]<span class="op">+</span>filter_columns).<span class="bu">sum</span>(numeric_only<span class="op">=</span><span class="va">True</span>).reset_index().set_index(filter_columns)</span>
<span id="cb7-33"><a href="#cb7-33" aria-hidden="true" tabindex="-1"></a>    syntenic_annotated <span class="op">=</span> genome_data.groupby([<span class="st">'other_criteria_fail'</span>,<span class="st">'Syntenic'</span>]<span class="op">+</span>filter_columns).<span class="bu">sum</span>(numeric_only<span class="op">=</span><span class="va">True</span>).reset_index().set_index(filter_columns)</span>
<span id="cb7-34"><a href="#cb7-34" aria-hidden="true" tabindex="-1"></a>    syntenic_only_fail <span class="op">=</span> genome_data.loc[genome_data[filter_columns].<span class="bu">sum</span>(axis<span class="op">=</span><span class="dv">1</span>)<span class="op">&gt;</span><span class="dv">0</span>].groupby([<span class="st">'other_criteria_fail'</span>,<span class="st">'Syntenic'</span>]<span class="op">+</span>filter_columns).<span class="bu">sum</span>(numeric_only<span class="op">=</span><span class="va">True</span>).reset_index().set_index(filter_columns)</span>
<span id="cb7-35"><a href="#cb7-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-36"><a href="#cb7-36" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> {</span>
<span id="cb7-37"><a href="#cb7-37" aria-hidden="true" tabindex="-1"></a>        <span class="st">'filters'</span>: filters,</span>
<span id="cb7-38"><a href="#cb7-38" aria-hidden="true" tabindex="-1"></a>        <span class="st">'syntenic_annotated'</span>: syntenic_annotated,</span>
<span id="cb7-39"><a href="#cb7-39" aria-hidden="true" tabindex="-1"></a>        <span class="st">'syntenic_only_fail'</span>: syntenic_only_fail,</span>
<span id="cb7-40"><a href="#cb7-40" aria-hidden="true" tabindex="-1"></a>        <span class="st">'color_scheme'</span>: color_scheme</span>
<span id="cb7-41"><a href="#cb7-41" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb7-42"><a href="#cb7-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-43"><a href="#cb7-43" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> create_upset_plot(data, is_syntenic, genome_name, cutoff_perc, element_size, top_x_categories, colors, filter_column_rename):</span>
<span id="cb7-44"><a href="#cb7-44" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""Create an UpSet plot for the given data"""</span></span>
<span id="cb7-45"><a href="#cb7-45" aria-hidden="true" tabindex="-1"></a>    subset_data <span class="op">=</span> move_all_passing_to_front(data.loc[(data.Syntenic<span class="op">==</span>is_syntenic)].rename_axis(index<span class="op">=</span>filter_column_rename))</span>
<span id="cb7-46"><a href="#cb7-46" aria-hidden="true" tabindex="-1"></a>    cutoff <span class="op">=</span> subset_data.num_variants.<span class="bu">sum</span>().item() <span class="op">*</span> cutoff_perc</span>
<span id="cb7-47"><a href="#cb7-47" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb7-48"><a href="#cb7-48" aria-hidden="true" tabindex="-1"></a>    upset <span class="op">=</span> UpSet(</span>
<span id="cb7-49"><a href="#cb7-49" aria-hidden="true" tabindex="-1"></a>        subset_data[[<span class="st">'other_criteria_fail'</span>,<span class="st">'num_variants'</span>]], </span>
<span id="cb7-50"><a href="#cb7-50" aria-hidden="true" tabindex="-1"></a>        show_counts<span class="op">=</span><span class="va">True</span>, </span>
<span id="cb7-51"><a href="#cb7-51" aria-hidden="true" tabindex="-1"></a>        sum_over<span class="op">=</span><span class="st">'num_variants'</span>, </span>
<span id="cb7-52"><a href="#cb7-52" aria-hidden="true" tabindex="-1"></a>        sort_by<span class="op">=</span><span class="st">'input'</span>,</span>
<span id="cb7-53"><a href="#cb7-53" aria-hidden="true" tabindex="-1"></a>        orientation<span class="op">=</span><span class="st">"horizontal"</span>, </span>
<span id="cb7-54"><a href="#cb7-54" aria-hidden="true" tabindex="-1"></a>        min_subset_size<span class="op">=</span>cutoff,</span>
<span id="cb7-55"><a href="#cb7-55" aria-hidden="true" tabindex="-1"></a>        intersection_plot_elements<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb7-56"><a href="#cb7-56" aria-hidden="true" tabindex="-1"></a>        element_size<span class="op">=</span>element_size</span>
<span id="cb7-57"><a href="#cb7-57" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb7-58"><a href="#cb7-58" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(subset_data.index.names)</span>
<span id="cb7-59"><a href="#cb7-59" aria-hidden="true" tabindex="-1"></a>    upset.add_stacked_bars(</span>
<span id="cb7-60"><a href="#cb7-60" aria-hidden="true" tabindex="-1"></a>        by<span class="op">=</span><span class="st">"other_criteria_fail"</span>, </span>
<span id="cb7-61"><a href="#cb7-61" aria-hidden="true" tabindex="-1"></a>        title<span class="op">=</span><span class="st">"Number of variants filtered"</span>, </span>
<span id="cb7-62"><a href="#cb7-62" aria-hidden="true" tabindex="-1"></a>        sum_over<span class="op">=</span><span class="st">'num_variants'</span>, </span>
<span id="cb7-63"><a href="#cb7-63" aria-hidden="true" tabindex="-1"></a>        elements<span class="op">=</span><span class="bu">len</span>(subset_data.index.names),</span>
<span id="cb7-64"><a href="#cb7-64" aria-hidden="true" tabindex="-1"></a>        colors<span class="op">=</span>colors</span>
<span id="cb7-65"><a href="#cb7-65" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb7-66"><a href="#cb7-66" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb7-67"><a href="#cb7-67" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> upset</span>
<span id="cb7-68"><a href="#cb7-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-69"><a href="#cb7-69" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> create_upset_figure_layout(upset_plots_config):</span>
<span id="cb7-70"><a href="#cb7-70" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""Create the main figure with 2x2 subplot layout and labels"""</span></span>
<span id="cb7-71"><a href="#cb7-71" aria-hidden="true" tabindex="-1"></a>    fig <span class="op">=</span> plt.figure(figsize<span class="op">=</span>(<span class="dv">13</span>,<span class="dv">9</span>), layout<span class="op">=</span><span class="st">'constrained'</span>)</span>
<span id="cb7-72"><a href="#cb7-72" aria-hidden="true" tabindex="-1"></a>    ((subA, subB), (subC, subD)) <span class="op">=</span> fig.subfigures(<span class="dv">2</span>, <span class="dv">2</span>, wspace<span class="op">=-</span><span class="fl">0.5</span>, hspace<span class="op">=-</span><span class="fl">0.5</span>)</span>
<span id="cb7-73"><a href="#cb7-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-74"><a href="#cb7-74" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Plot each UpSet plot in its designated subfigure</span></span>
<span id="cb7-75"><a href="#cb7-75" aria-hidden="true" tabindex="-1"></a>    subplots <span class="op">=</span> [subA, subB, subC, subD]</span>
<span id="cb7-76"><a href="#cb7-76" aria-hidden="true" tabindex="-1"></a>    configs <span class="op">=</span> upset_plots_config</span>
<span id="cb7-77"><a href="#cb7-77" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb7-78"><a href="#cb7-78" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i, (subplot, config) <span class="kw">in</span> <span class="bu">enumerate</span>(<span class="bu">zip</span>(subplots, configs)):</span>
<span id="cb7-79"><a href="#cb7-79" aria-hidden="true" tabindex="-1"></a>        axes <span class="op">=</span> config[<span class="st">'upset'</span>].plot(fig<span class="op">=</span>subplot)</span>
<span id="cb7-80"><a href="#cb7-80" aria-hidden="true" tabindex="-1"></a>        axes[<span class="st">'extra0'</span>].legend().set_title(config[<span class="st">'legend_title'</span>])</span>
<span id="cb7-81"><a href="#cb7-81" aria-hidden="true" tabindex="-1"></a>        add_letter_to_subfig(subplot, <span class="bu">chr</span>(<span class="bu">ord</span>(<span class="st">'a'</span>) <span class="op">+</span> i), fontsize<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb7-82"><a href="#cb7-82" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb7-83"><a href="#cb7-83" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Add main labels</span></span>
<span id="cb7-84"><a href="#cb7-84" aria-hidden="true" tabindex="-1"></a>    fig.text(<span class="fl">.01</span>,<span class="fl">0.25</span>, <span class="st">'1KGP CHM13v2.0 panel'</span>, fontsize<span class="op">=</span><span class="dv">9</span>, weight<span class="op">=</span><span class="st">'bold'</span>, rotation<span class="op">=</span><span class="st">'vertical'</span>, horizontalalignment<span class="op">=</span><span class="st">'right'</span>, verticalalignment<span class="op">=</span><span class="st">'center'</span>)</span>
<span id="cb7-85"><a href="#cb7-85" aria-hidden="true" tabindex="-1"></a>    fig.text(<span class="fl">.01</span>,<span class="fl">0.75</span>, <span class="st">'1KGP NYGC GRCh38 panel'</span>, fontsize<span class="op">=</span><span class="dv">9</span>, weight<span class="op">=</span><span class="st">'bold'</span>, rotation<span class="op">=</span><span class="st">'vertical'</span>, horizontalalignment<span class="op">=</span><span class="st">'right'</span>, verticalalignment<span class="op">=</span><span class="st">'center'</span>)</span>
<span id="cb7-86"><a href="#cb7-86" aria-hidden="true" tabindex="-1"></a>    fig.text(<span class="fl">.25</span>,<span class="dv">1</span>, <span class="st">'Syntenic variants'</span>, fontsize<span class="op">=</span><span class="dv">9</span>, weight<span class="op">=</span><span class="st">'bold'</span>, rotation<span class="op">=</span><span class="st">'horizontal'</span>, horizontalalignment<span class="op">=</span><span class="st">'center'</span>, verticalalignment<span class="op">=</span><span class="st">'center'</span>)</span>
<span id="cb7-87"><a href="#cb7-87" aria-hidden="true" tabindex="-1"></a>    fig.text(<span class="fl">.75</span>,<span class="dv">1</span>, <span class="st">'Nonsyntenic variants'</span>, fontsize<span class="op">=</span><span class="dv">9</span>, weight<span class="op">=</span><span class="st">'bold'</span>, rotation<span class="op">=</span><span class="st">'horizontal'</span>, horizontalalignment<span class="op">=</span><span class="st">'center'</span>, verticalalignment<span class="op">=</span><span class="st">'center'</span>)</span>
<span id="cb7-88"><a href="#cb7-88" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb7-89"><a href="#cb7-89" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> fig</span>
<span id="cb7-90"><a href="#cb7-90" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-91"><a href="#cb7-91" aria-hidden="true" tabindex="-1"></a><span class="co"># Load and prepare data</span></span>
<span id="cb7-92"><a href="#cb7-92" aria-hidden="true" tabindex="-1"></a>grch38_filter_columns<span class="op">=</span>[<span class="st">'AC_filter'</span>,</span>
<span id="cb7-93"><a href="#cb7-93" aria-hidden="true" tabindex="-1"></a>                       <span class="st">'MERR_filter'</span>, <span class="st">'HWE_pop_filter'</span>, <span class="st">'f_missing_filter'</span>, </span>
<span id="cb7-94"><a href="#cb7-94" aria-hidden="true" tabindex="-1"></a>                       <span class="st">'var_len_filter'</span>, <span class="st">'alt_star_filter'</span>, <span class="st">'pass_filter'</span>]</span>
<span id="cb7-95"><a href="#cb7-95" aria-hidden="true" tabindex="-1"></a>chm13_filter_columns<span class="op">=</span>[<span class="st">'VQSLOD_filter'</span>, <span class="st">'MAC_filter'</span>,</span>
<span id="cb7-96"><a href="#cb7-96" aria-hidden="true" tabindex="-1"></a>                      <span class="st">'MERR_filter'</span>, <span class="st">'HWE_pop_filter'</span>, <span class="st">'f_missing_filter'</span>, </span>
<span id="cb7-97"><a href="#cb7-97" aria-hidden="true" tabindex="-1"></a>                      <span class="st">'var_len_filter'</span>, <span class="st">'alt_star_filter'</span>, <span class="st">'pass_filter'</span>]</span>
<span id="cb7-98"><a href="#cb7-98" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-99"><a href="#cb7-99" aria-hidden="true" tabindex="-1"></a>filter_column_rename <span class="op">=</span> {<span class="st">'HWE_pop_filter'</span>:<span class="st">'Low population HWE'</span>,</span>
<span id="cb7-100"><a href="#cb7-100" aria-hidden="true" tabindex="-1"></a>                        <span class="st">'AC_filter'</span>: <span class="st">'AC &lt;= 1'</span>,</span>
<span id="cb7-101"><a href="#cb7-101" aria-hidden="true" tabindex="-1"></a>                        <span class="st">'MAC_filter'</span>: <span class="st">'MAC == 0'</span>,</span>
<span id="cb7-102"><a href="#cb7-102" aria-hidden="true" tabindex="-1"></a>                        <span class="st">'f_missing_filter'</span>:<span class="st">"&gt;5% missing"</span>,</span>
<span id="cb7-103"><a href="#cb7-103" aria-hidden="true" tabindex="-1"></a>                        <span class="st">'var_len_filter'</span>:<span class="st">'Variant over 50bp'</span>,</span>
<span id="cb7-104"><a href="#cb7-104" aria-hidden="true" tabindex="-1"></a>                        <span class="st">'alt_star_filter'</span>:<span class="st">'ALT allele is star (</span><span class="ch">\'</span><span class="er">\</span><span class="st">*</span><span class="ch">\'</span><span class="st">)'</span>,</span>
<span id="cb7-105"><a href="#cb7-105" aria-hidden="true" tabindex="-1"></a>                        <span class="st">'pass_filter'</span>: <span class="st">'PASS Filter'</span>,</span>
<span id="cb7-106"><a href="#cb7-106" aria-hidden="true" tabindex="-1"></a>                        <span class="st">'MERR_filter'</span>: <span class="st">'Mend. Error Rate &gt; 5%'</span>,</span>
<span id="cb7-107"><a href="#cb7-107" aria-hidden="true" tabindex="-1"></a>                        <span class="st">'VQSLOD_filter'</span>:<span class="st">'VQSLOD &lt; 0'</span>}</span>
<span id="cb7-108"><a href="#cb7-108" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-109"><a href="#cb7-109" aria-hidden="true" tabindex="-1"></a>sum_stats_var_filters[<span class="st">'only_filtered_from_grch38_because_singleton'</span>] <span class="op">=</span> sum_stats_var_filters[[x <span class="cf">for</span> x <span class="kw">in</span> grch38_filter_columns <span class="cf">if</span> x <span class="op">!=</span> <span class="st">'AC_filter'</span>] <span class="op">+</span> [<span class="st">'MAC_filter'</span>]].<span class="bu">any</span>(axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb7-110"><a href="#cb7-110" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-111"><a href="#cb7-111" aria-hidden="true" tabindex="-1"></a><span class="co"># Define color schemes</span></span>
<span id="cb7-112"><a href="#cb7-112" aria-hidden="true" tabindex="-1"></a>grch38_color_scheme<span class="op">=</span>{<span class="st">'No: Singleton variant'</span>: sns.color_palette()[<span class="dv">6</span>],</span>
<span id="cb7-113"><a href="#cb7-113" aria-hidden="true" tabindex="-1"></a>                     <span class="st">'Yes'</span>:sns.color_palette()[<span class="dv">1</span>],</span>
<span id="cb7-114"><a href="#cb7-114" aria-hidden="true" tabindex="-1"></a>                     <span class="st">'No'</span>: sns.color_palette()[<span class="dv">0</span>]}</span>
<span id="cb7-115"><a href="#cb7-115" aria-hidden="true" tabindex="-1"></a>chm13_color_scheme<span class="op">=</span>{<span class="st">'Yes: singleton variant'</span>: sns.color_palette()[<span class="dv">6</span>],</span>
<span id="cb7-116"><a href="#cb7-116" aria-hidden="true" tabindex="-1"></a>                    <span class="st">'Yes: Other reason'</span>:sns.color_palette()[<span class="dv">1</span>],</span>
<span id="cb7-117"><a href="#cb7-117" aria-hidden="true" tabindex="-1"></a>                    <span class="st">'No'</span>: sns.color_palette()[<span class="dv">0</span>]}</span>
<span id="cb7-118"><a href="#cb7-118" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-119"><a href="#cb7-119" aria-hidden="true" tabindex="-1"></a><span class="co"># Process data for both genomes</span></span>
<span id="cb7-120"><a href="#cb7-120" aria-hidden="true" tabindex="-1"></a>grch38_processed <span class="op">=</span> reshape_data_for_upsetplot(sum_stats_var_filters, <span class="st">'GRCh38'</span>, grch38_filter_columns, grch38_color_scheme)</span>
<span id="cb7-121"><a href="#cb7-121" aria-hidden="true" tabindex="-1"></a>chm13_processed <span class="op">=</span> reshape_data_for_upsetplot(sum_stats_var_filters, <span class="st">'CHM13v2.0'</span>, chm13_filter_columns, chm13_color_scheme)</span>
<span id="cb7-122"><a href="#cb7-122" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-123"><a href="#cb7-123" aria-hidden="true" tabindex="-1"></a><span class="co"># Configuration</span></span>
<span id="cb7-124"><a href="#cb7-124" aria-hidden="true" tabindex="-1"></a>element_size <span class="op">=</span> <span class="dv">30</span></span>
<span id="cb7-125"><a href="#cb7-125" aria-hidden="true" tabindex="-1"></a>top_x_categories <span class="op">=</span> <span class="dv">12</span></span>
<span id="cb7-126"><a href="#cb7-126" aria-hidden="true" tabindex="-1"></a>cutoff_perc <span class="op">=</span> <span class="fl">0.005</span></span>
<span id="cb7-127"><a href="#cb7-127" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-128"><a href="#cb7-128" aria-hidden="true" tabindex="-1"></a><span class="co"># Create UpSet plots</span></span>
<span id="cb7-129"><a href="#cb7-129" aria-hidden="true" tabindex="-1"></a>upset_syn_grch38 <span class="op">=</span> create_upset_plot(grch38_processed[<span class="st">'syntenic_annotated'</span>], <span class="va">True</span>, <span class="st">'GRCh38'</span>, cutoff_perc, element_size, <span class="dv">16</span>, grch38_color_scheme, filter_column_rename)</span>
<span id="cb7-130"><a href="#cb7-130" aria-hidden="true" tabindex="-1"></a>upset_nonsyn_grch38 <span class="op">=</span> create_upset_plot(grch38_processed[<span class="st">'syntenic_annotated'</span>], <span class="va">False</span>, <span class="st">'GRCh38'</span>, cutoff_perc, element_size, top_x_categories, grch38_color_scheme, filter_column_rename)</span>
<span id="cb7-131"><a href="#cb7-131" aria-hidden="true" tabindex="-1"></a>upset_syn_chm13 <span class="op">=</span> create_upset_plot(chm13_processed[<span class="st">'syntenic_annotated'</span>], <span class="va">True</span>, <span class="st">'CHM13v2.0'</span>, cutoff_perc, element_size, top_x_categories, chm13_color_scheme, filter_column_rename)</span>
<span id="cb7-132"><a href="#cb7-132" aria-hidden="true" tabindex="-1"></a>upset_nonsyn_chm13 <span class="op">=</span> create_upset_plot(chm13_processed[<span class="st">'syntenic_annotated'</span>], <span class="va">False</span>, <span class="st">'CHM13v2.0'</span>, cutoff_perc, element_size, top_x_categories, chm13_color_scheme, filter_column_rename)</span>
<span id="cb7-133"><a href="#cb7-133" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-134"><a href="#cb7-134" aria-hidden="true" tabindex="-1"></a><span class="co"># Configure plots for figure layout</span></span>
<span id="cb7-135"><a href="#cb7-135" aria-hidden="true" tabindex="-1"></a>upset_plots_config <span class="op">=</span> [</span>
<span id="cb7-136"><a href="#cb7-136" aria-hidden="true" tabindex="-1"></a>    {<span class="st">'upset'</span>: upset_syn_grch38, <span class="st">'legend_title'</span>: <span class="st">'Would be filtered</span><span class="ch">\n</span><span class="st">in CHM13v2.0 panel?'</span>},</span>
<span id="cb7-137"><a href="#cb7-137" aria-hidden="true" tabindex="-1"></a>    {<span class="st">'upset'</span>: upset_nonsyn_grch38, <span class="st">'legend_title'</span>: <span class="st">'Would be filtered</span><span class="ch">\n</span><span class="st">in CHM13v2.0 panel?'</span>},</span>
<span id="cb7-138"><a href="#cb7-138" aria-hidden="true" tabindex="-1"></a>    {<span class="st">'upset'</span>: upset_syn_chm13, <span class="st">'legend_title'</span>: <span class="st">'Would be filtered</span><span class="ch">\n</span><span class="st">in GRCh38 panel?'</span>},</span>
<span id="cb7-139"><a href="#cb7-139" aria-hidden="true" tabindex="-1"></a>    {<span class="st">'upset'</span>: upset_nonsyn_chm13, <span class="st">'legend_title'</span>: <span class="st">'Would be filtered</span><span class="ch">\n</span><span class="st">in GRCh38 panel?'</span>}</span>
<span id="cb7-140"><a href="#cb7-140" aria-hidden="true" tabindex="-1"></a>]</span>
<span id="cb7-141"><a href="#cb7-141" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-142"><a href="#cb7-142" aria-hidden="true" tabindex="-1"></a><span class="co"># Create and save figure</span></span>
<span id="cb7-143"><a href="#cb7-143" aria-hidden="true" tabindex="-1"></a>fig <span class="op">=</span> create_upset_figure_layout(upset_plots_config)</span>
<span id="cb7-144"><a href="#cb7-144" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-145"><a href="#cb7-145" aria-hidden="true" tabindex="-1"></a>fig.savefig(<span class="st">'figures/supplemental/Supplemental 1.png'</span>,pad_inches<span class="op">=</span><span class="fl">.5</span>)</span>
<span id="cb7-146"><a href="#cb7-146" aria-hidden="true" tabindex="-1"></a>fig.savefig(<span class="st">'figures/supplemental/Supplemental 1.svg'</span>,bbox_inches<span class="op">=</span><span class="st">'tight'</span>)</span>
<span id="cb7-147"><a href="#cb7-147" aria-hidden="true" tabindex="-1"></a>fig.savefig(<span class="st">'figures/supplemental/Supplemental 1.eps'</span>,pad_inches<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="cb7-148"><a href="#cb7-148" aria-hidden="true" tabindex="-1"></a>fig.savefig(<span class="st">'figures/supplemental/Supplemental 1.pdf'</span>,pad_inches<span class="op">=</span><span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Using upsetplot version /home/lalli/.linuxbrew/opt/python@3.11/lib/python3.11/site-packages/upsetplot/__init__.py
['AC &lt;= 1', 'Mend. Error Rate &gt; 5%', 'Low population HWE', '&gt;5% missing', 'Variant over 50bp', 'ALT allele is star ('\*')', 'PASS Filter']
['AC &lt;= 1', 'Mend. Error Rate &gt; 5%', 'Low population HWE', '&gt;5% missing', 'Variant over 50bp', 'ALT allele is star ('\*')', 'PASS Filter']
['VQSLOD &lt; 0', 'MAC == 0', 'Mend. Error Rate &gt; 5%', 'Low population HWE', '&gt;5% missing', 'Variant over 50bp', 'ALT allele is star ('\*')', 'PASS Filter']</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>/tmp/ipykernel_2892998/4020693012.py:48: DeprecationWarning: element_size parameter is deprecated and will be removed in a future version. Use element_size_fraction instead for fractional sizing relative to figure dimensions.
  upset = UpSet(
/tmp/ipykernel_2892998/4020693012.py:48: DeprecationWarning: element_size parameter is deprecated and will be removed in a future version. Use element_size_fraction instead for fractional sizing relative to figure dimensions.
  upset = UpSet(
/tmp/ipykernel_2892998/4020693012.py:48: DeprecationWarning: element_size parameter is deprecated and will be removed in a future version. Use element_size_fraction instead for fractional sizing relative to figure dimensions.
  upset = UpSet(
/tmp/ipykernel_2892998/4020693012.py:48: DeprecationWarning: element_size parameter is deprecated and will be removed in a future version. Use element_size_fraction instead for fractional sizing relative to figure dimensions.
  upset = UpSet(</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>['VQSLOD &lt; 0', 'MAC == 0', 'Mend. Error Rate &gt; 5%', 'Low population HWE', '&gt;5% missing', 'Variant over 50bp', 'ALT allele is star ('\*')', 'PASS Filter']
14 = 7 + (7 or 0)
Max grid height: 225.0 pixels
Figure height: 450.0 pixels
16.071428571428573 = 225.0 / 14
14 = 7 + (7 or 0)
Max grid height: 225.0 pixels
Figure height: 450.0 pixels
16.071428571428573 = 225.0 / 14
16 = 8 + (8 or 0)
Max grid height: 225.0 pixels
Figure height: 450.0 pixels
14.0625 = 225.0 / 16
16 = 8 + (8 or 0)
Max grid height: 225.0 pixels
Figure height: 450.0 pixels
14.0625 = 225.0 / 16</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The PostScript backend does not support transparency; partially transparent artists will be rendered opaque.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="make_plots_files/figure-html/cell-8-output-5.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="make_plots_files/figure-html/cell-14-1-image.png" class="img-fluid figure-img"></p>
<figcaption>image.png</figcaption>
</figure>
</div>
<p>Supplemental Figure 2. Illustration describing different methods of evaluating phasing quality. Phasing is the process of assigning heterozygous variant calls to one of two haplotypes (in the case of diploid organisms.) One of the challenges of assessing the accuracy of statistically phased haplotypes is identifying a ground truth to compare to. In the case of trio-based phasing evaluation, the ground truth is parental genomes. While not every heterozygous site can be assigned to a haplotype of paternal or maternal origin, most can confidently be assigned to one or the other. Alternately, one can use an empirically phased reference assembly as a source of ground truth. When a callset’s phasing describes a variant that is from a different parent of origin than the preceeding variant, that is called a switch error. Genotyping errors, as illustrated in the fourth panel, can force statistical phasing software to assign a variant to a haplotype that is different than nearby heterozygous sites. When two switch errors occur back-to-back, that is described as a flip error. Importantly, flip errors do not affect the phasing of surrounding heterozygous sites. Discerning flip errors from true switch errors is important when evaluating phasing methods, as the majority of switch errors can frequently be found in flip errors. Counting these switches as “true” switch errors inflates switch error rates by including sites that do not affect regional phasing.</p>
<div id="cell-15" class="cell" data-execution_count="8">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Figure 2 supplementals</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Figure 2</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="co">## Figure 2 with parents and estimated overall</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>fig, ((axA, axB), (axC, axD), (axE, axF), (axG, axH), (axI, axJ)) <span class="op">=</span> plt.subplots(<span class="dv">5</span>,<span class="dv">2</span>, figsize<span class="op">=</span>(oneandahalf_col,<span class="dv">8</span>))</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a><span class="co"># fig.set_layout_engine(layout='constrained', h_pad=0.15, w_pad=0.15)</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'switch_error_rate'</span>, style<span class="op">=</span><span class="st">'syntenic'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> syntenic_order,</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data.loc[(binned_maf_data.ground_truth_data_source <span class="op">==</span> <span class="st">'HPRC_samples'</span>)</span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.<span class="bu">type</span> <span class="op">==</span> <span class="st">'SNP'</span>)</span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.syntenic <span class="op">!=</span> <span class="st">'All'</span>)],</span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axA, <span class="op">**</span>syntenic_style)</span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'switch_error_rate'</span>, style<span class="op">=</span><span class="st">'syntenic'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> syntenic_order,</span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data.loc[(binned_maf_data.ground_truth_data_source <span class="op">==</span> <span class="st">'HGSVC_probands'</span>) <span class="co">#trios only</span></span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.<span class="bu">type</span> <span class="op">==</span> <span class="st">'SNP'</span>)</span>
<span id="cb12-20"><a href="#cb12-20" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.syntenic <span class="op">!=</span> <span class="st">'All'</span>)],</span>
<span id="cb12-21"><a href="#cb12-21" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axC, <span class="op">**</span>syntenic_style)</span>
<span id="cb12-22"><a href="#cb12-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-23"><a href="#cb12-23" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'switch_error_rate'</span>, style<span class="op">=</span><span class="st">'syntenic'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> syntenic_order,</span>
<span id="cb12-24"><a href="#cb12-24" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data.loc[(binned_maf_data.ground_truth_data_source <span class="op">==</span> <span class="st">'HGSVC_parents'</span>) <span class="co">#trios only</span></span>
<span id="cb12-25"><a href="#cb12-25" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb12-26"><a href="#cb12-26" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.<span class="bu">type</span> <span class="op">==</span> <span class="st">'SNP'</span>)</span>
<span id="cb12-27"><a href="#cb12-27" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.syntenic <span class="op">!=</span> <span class="st">'All'</span>)],</span>
<span id="cb12-28"><a href="#cb12-28" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axE, <span class="op">**</span>syntenic_style)</span>
<span id="cb12-29"><a href="#cb12-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-30"><a href="#cb12-30" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'switch_error_rate'</span>, style<span class="op">=</span><span class="st">'syntenic'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> syntenic_order,</span>
<span id="cb12-31"><a href="#cb12-31" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data.loc[(binned_maf_data.ground_truth_data_source <span class="op">==</span> <span class="st">'HGSVC_samples_nontrios_only'</span>) <span class="co">#trios only</span></span>
<span id="cb12-32"><a href="#cb12-32" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb12-33"><a href="#cb12-33" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.<span class="bu">type</span> <span class="op">==</span> <span class="st">'SNP'</span>)</span>
<span id="cb12-34"><a href="#cb12-34" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.syntenic <span class="op">!=</span> <span class="st">'All'</span>)],</span>
<span id="cb12-35"><a href="#cb12-35" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axG, <span class="op">**</span>syntenic_style)</span>
<span id="cb12-36"><a href="#cb12-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-37"><a href="#cb12-37" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'switch_error_rate'</span>, style<span class="op">=</span><span class="st">'syntenic'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> syntenic_order,</span>
<span id="cb12-38"><a href="#cb12-38" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>cat_data.loc[(cat_data.<span class="bu">type</span> <span class="op">==</span> <span class="st">'SNP'</span>)</span>
<span id="cb12-39"><a href="#cb12-39" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (cat_data.syntenic <span class="op">!=</span> <span class="st">'All'</span>)],</span>
<span id="cb12-40"><a href="#cb12-40" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axI, <span class="op">**</span>syntenic_style)</span>
<span id="cb12-41"><a href="#cb12-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-42"><a href="#cb12-42" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'switch_error_rate'</span>, style<span class="op">=</span><span class="st">'syntenic'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> syntenic_order,</span>
<span id="cb12-43"><a href="#cb12-43" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data.loc[(binned_maf_data.ground_truth_data_source <span class="op">==</span> <span class="st">'HPRC_samples'</span>)</span>
<span id="cb12-44"><a href="#cb12-44" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb12-45"><a href="#cb12-45" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.<span class="bu">type</span> <span class="op">==</span> <span class="st">'Indel'</span>)</span>
<span id="cb12-46"><a href="#cb12-46" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.syntenic <span class="op">!=</span> <span class="st">'All'</span>)],</span>
<span id="cb12-47"><a href="#cb12-47" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axB, <span class="op">**</span>syntenic_style)</span>
<span id="cb12-48"><a href="#cb12-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-49"><a href="#cb12-49" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'switch_error_rate'</span>, style<span class="op">=</span><span class="st">'syntenic'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> syntenic_order,</span>
<span id="cb12-50"><a href="#cb12-50" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data.loc[(binned_maf_data.ground_truth_data_source <span class="op">==</span> <span class="st">'HGSVC_probands'</span>)</span>
<span id="cb12-51"><a href="#cb12-51" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb12-52"><a href="#cb12-52" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.<span class="bu">type</span> <span class="op">==</span> <span class="st">'Indel'</span>)</span>
<span id="cb12-53"><a href="#cb12-53" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.syntenic <span class="op">!=</span> <span class="st">'All'</span>)],</span>
<span id="cb12-54"><a href="#cb12-54" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axD, <span class="op">**</span>syntenic_style)</span>
<span id="cb12-55"><a href="#cb12-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-56"><a href="#cb12-56" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'switch_error_rate'</span>, style<span class="op">=</span><span class="st">'syntenic'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> syntenic_order,</span>
<span id="cb12-57"><a href="#cb12-57" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data.loc[(binned_maf_data.ground_truth_data_source <span class="op">==</span> <span class="st">'HGSVC_parents'</span>)</span>
<span id="cb12-58"><a href="#cb12-58" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb12-59"><a href="#cb12-59" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.<span class="bu">type</span> <span class="op">==</span> <span class="st">'Indel'</span>)</span>
<span id="cb12-60"><a href="#cb12-60" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.syntenic <span class="op">!=</span> <span class="st">'All'</span>)],</span>
<span id="cb12-61"><a href="#cb12-61" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axF, <span class="op">**</span>syntenic_style)</span>
<span id="cb12-62"><a href="#cb12-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-63"><a href="#cb12-63" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'switch_error_rate'</span>, style<span class="op">=</span><span class="st">'syntenic'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> syntenic_order,</span>
<span id="cb12-64"><a href="#cb12-64" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data.loc[(binned_maf_data.ground_truth_data_source <span class="op">==</span> <span class="st">'HGSVC_samples_nontrios_only'</span>) <span class="co">#trios only</span></span>
<span id="cb12-65"><a href="#cb12-65" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb12-66"><a href="#cb12-66" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.<span class="bu">type</span> <span class="op">==</span> <span class="st">'Indel'</span>)</span>
<span id="cb12-67"><a href="#cb12-67" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.syntenic <span class="op">!=</span> <span class="st">'All'</span>)],</span>
<span id="cb12-68"><a href="#cb12-68" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axH, <span class="op">**</span>syntenic_style)</span>
<span id="cb12-69"><a href="#cb12-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-70"><a href="#cb12-70" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'switch_error_rate'</span>, style<span class="op">=</span><span class="st">'syntenic'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> syntenic_order,</span>
<span id="cb12-71"><a href="#cb12-71" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>cat_data.loc[(cat_data.<span class="bu">type</span> <span class="op">==</span> <span class="st">'Indel'</span>)</span>
<span id="cb12-72"><a href="#cb12-72" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (cat_data.syntenic <span class="op">!=</span> <span class="st">'All'</span>)],</span>
<span id="cb12-73"><a href="#cb12-73" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axJ, <span class="op">**</span>syntenic_style)</span>
<span id="cb12-74"><a href="#cb12-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-75"><a href="#cb12-75" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, ax <span class="kw">in</span> <span class="bu">enumerate</span>([axA, axB, axC, axD, axE, axF, axG, axH, axI, axJ]):</span>
<span id="cb12-76"><a href="#cb12-76" aria-hidden="true" tabindex="-1"></a>    ax.set_yscale(<span class="st">'log'</span>)</span>
<span id="cb12-77"><a href="#cb12-77" aria-hidden="true" tabindex="-1"></a>    ax.set_xscale(<span class="st">'log'</span>)</span>
<span id="cb12-78"><a href="#cb12-78" aria-hidden="true" tabindex="-1"></a>    ax.set_ylim(<span class="fl">1e-1</span>, <span class="dv">50</span>)</span>
<span id="cb12-79"><a href="#cb12-79" aria-hidden="true" tabindex="-1"></a>    ax.set_xlim(<span class="fl">1e-2</span>, <span class="dv">55</span>)</span>
<span id="cb12-80"><a href="#cb12-80" aria-hidden="true" tabindex="-1"></a>    <span class="co"># ax.set_box_aspect(1)</span></span>
<span id="cb12-81"><a href="#cb12-81" aria-hidden="true" tabindex="-1"></a>    ax.grid()</span>
<span id="cb12-82"><a href="#cb12-82" aria-hidden="true" tabindex="-1"></a>    ax.set_axisbelow(<span class="va">True</span>)</span>
<span id="cb12-83"><a href="#cb12-83" aria-hidden="true" tabindex="-1"></a>    ax.set_xticks(MAF_xticks)</span>
<span id="cb12-84"><a href="#cb12-84" aria-hidden="true" tabindex="-1"></a>    ax.set_yticks(SER_xticks)</span>
<span id="cb12-85"><a href="#cb12-85" aria-hidden="true" tabindex="-1"></a>    ax.xaxis.set_major_formatter(perc_formatter)</span>
<span id="cb12-86"><a href="#cb12-86" aria-hidden="true" tabindex="-1"></a>    ax.yaxis.set_major_formatter(perc_formatter)</span>
<span id="cb12-87"><a href="#cb12-87" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="kw">in</span> (<span class="dv">8</span>,<span class="dv">9</span>):</span>
<span id="cb12-88"><a href="#cb12-88" aria-hidden="true" tabindex="-1"></a>        ax.set_xlabel(<span class="st">'Minor Allele Frequency (%)'</span>, labelpad<span class="op">=</span><span class="dv">3</span>)</span>
<span id="cb12-89"><a href="#cb12-89" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb12-90"><a href="#cb12-90" aria-hidden="true" tabindex="-1"></a>        ax.set_xlabel(<span class="va">None</span>)</span>
<span id="cb12-91"><a href="#cb12-91" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="kw">in</span> (<span class="dv">0</span>,<span class="dv">2</span>,<span class="dv">4</span>,<span class="dv">6</span>,<span class="dv">8</span>):</span>
<span id="cb12-92"><a href="#cb12-92" aria-hidden="true" tabindex="-1"></a>        ax.set_ylabel(<span class="st">'Switch Error Rate (%)'</span>, labelpad<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb12-93"><a href="#cb12-93" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb12-94"><a href="#cb12-94" aria-hidden="true" tabindex="-1"></a>        ax.set_ylabel(<span class="va">None</span>)</span>
<span id="cb12-95"><a href="#cb12-95" aria-hidden="true" tabindex="-1"></a>    add_letter_to_ax(ax, i)</span>
<span id="cb12-96"><a href="#cb12-96" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-97"><a href="#cb12-97" aria-hidden="true" tabindex="-1"></a>x_offset <span class="op">=</span> <span class="op">-</span><span class="fl">0.4</span></span>
<span id="cb12-98"><a href="#cb12-98" aria-hidden="true" tabindex="-1"></a>axA.text(x_offset, <span class="fl">0.5</span>, <span class="st">'Mendelian phased</span><span class="ch">\n</span><span class="st">HPRC samples'</span>, fontsize<span class="op">=</span><span class="dv">10</span>, transform<span class="op">=</span>axA.transAxes, weight<span class="op">=</span><span class="st">'bold'</span>, rotation<span class="op">=</span><span class="st">'vertical'</span>, horizontalalignment<span class="op">=</span><span class="st">'center'</span>, verticalalignment<span class="op">=</span><span class="st">'center'</span>)</span>
<span id="cb12-99"><a href="#cb12-99" aria-hidden="true" tabindex="-1"></a>axC.text(x_offset, <span class="fl">0.5</span>, <span class="st">'Mendelian phased</span><span class="ch">\n</span><span class="st">HGSVC probands'</span>, fontsize<span class="op">=</span><span class="dv">10</span>, transform<span class="op">=</span>axC.transAxes, weight<span class="op">=</span><span class="st">'bold'</span>, rotation<span class="op">=</span><span class="st">'vertical'</span>, horizontalalignment<span class="op">=</span><span class="st">'center'</span>, verticalalignment<span class="op">=</span><span class="st">'center'</span>)</span>
<span id="cb12-100"><a href="#cb12-100" aria-hidden="true" tabindex="-1"></a>axE.text(x_offset, <span class="fl">0.5</span>, <span class="st">'Mendelian phased</span><span class="ch">\n</span><span class="st">HGSVC parents'</span>, fontsize<span class="op">=</span><span class="dv">10</span>, transform<span class="op">=</span>axE.transAxes, weight<span class="op">=</span><span class="st">'bold'</span>, rotation<span class="op">=</span><span class="st">'vertical'</span>, horizontalalignment<span class="op">=</span><span class="st">'center'</span>, verticalalignment<span class="op">=</span><span class="st">'center'</span>)</span>
<span id="cb12-101"><a href="#cb12-101" aria-hidden="true" tabindex="-1"></a>axG.text(x_offset, <span class="fl">0.5</span>, <span class="st">'Statistically phased</span><span class="ch">\n</span><span class="st">HGSVC samples'</span>, fontsize<span class="op">=</span><span class="dv">10</span>, transform<span class="op">=</span>axG.transAxes, weight<span class="op">=</span><span class="st">'bold'</span>, rotation<span class="op">=</span><span class="st">'vertical'</span>, horizontalalignment<span class="op">=</span><span class="st">'center'</span>, verticalalignment<span class="op">=</span><span class="st">'center'</span>)</span>
<span id="cb12-102"><a href="#cb12-102" aria-hidden="true" tabindex="-1"></a>axI.text(x_offset, <span class="fl">0.5</span>, <span class="st">'Estimated</span><span class="ch">\n</span><span class="st">panel-wide accuracy'</span>, fontsize<span class="op">=</span><span class="dv">10</span>, transform<span class="op">=</span>axI.transAxes, weight<span class="op">=</span><span class="st">'bold'</span>, rotation<span class="op">=</span><span class="st">'vertical'</span>, horizontalalignment<span class="op">=</span><span class="st">'center'</span>, verticalalignment<span class="op">=</span><span class="st">'center'</span>)</span>
<span id="cb12-103"><a href="#cb12-103" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-104"><a href="#cb12-104" aria-hidden="true" tabindex="-1"></a>axA.set_title(<span class="st">'SNPs'</span>, fontsize<span class="op">=</span><span class="dv">9</span>, weight<span class="op">=</span><span class="st">'bold'</span>, pad<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb12-105"><a href="#cb12-105" aria-hidden="true" tabindex="-1"></a>axB.set_title(<span class="st">'Indels'</span>, fontsize<span class="op">=</span><span class="dv">9</span>, weight<span class="op">=</span><span class="st">'bold'</span>, pad<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb12-106"><a href="#cb12-106" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-107"><a href="#cb12-107" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, ax <span class="kw">in</span> <span class="bu">enumerate</span>([axA, axB, axC, axD, axE, axF, axG, axH, axI, axJ]): <span class="co"># legend axes</span></span>
<span id="cb12-108"><a href="#cb12-108" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="kw">in</span> (<span class="dv">5</span>,<span class="dv">7</span>,<span class="dv">9</span>):</span>
<span id="cb12-109"><a href="#cb12-109" aria-hidden="true" tabindex="-1"></a>        loc<span class="op">=</span><span class="st">'lower right'</span></span>
<span id="cb12-110"><a href="#cb12-110" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb12-111"><a href="#cb12-111" aria-hidden="true" tabindex="-1"></a>        loc<span class="op">=</span><span class="st">'upper right'</span></span>
<span id="cb12-112"><a href="#cb12-112" aria-hidden="true" tabindex="-1"></a>    handles, labels <span class="op">=</span> ax.get_legend_handles_labels()</span>
<span id="cb12-113"><a href="#cb12-113" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> handle <span class="kw">in</span> handles:</span>
<span id="cb12-114"><a href="#cb12-114" aria-hidden="true" tabindex="-1"></a>        handle.set_marker(<span class="st">'o'</span>)</span>
<span id="cb12-115"><a href="#cb12-115" aria-hidden="true" tabindex="-1"></a>    handles <span class="op">=</span> handles[<span class="dv">1</span>:<span class="dv">3</span>] <span class="op">+</span> handles[<span class="dv">5</span>:<span class="dv">6</span>]</span>
<span id="cb12-116"><a href="#cb12-116" aria-hidden="true" tabindex="-1"></a>    labels <span class="op">=</span> labels[<span class="dv">1</span>:<span class="dv">3</span>] <span class="op">+</span> labels[<span class="dv">5</span>:<span class="dv">6</span>]</span>
<span id="cb12-117"><a href="#cb12-117" aria-hidden="true" tabindex="-1"></a>    handles[<span class="op">-</span><span class="dv">1</span>].set_color(handles[<span class="dv">1</span>].get_color())</span>
<span id="cb12-118"><a href="#cb12-118" aria-hidden="true" tabindex="-1"></a>    legend <span class="op">=</span> ax.legend(ncols<span class="op">=</span><span class="dv">1</span>, loc<span class="op">=</span>loc, handles<span class="op">=</span>handles, labels<span class="op">=</span>labels)</span>
<span id="cb12-119"><a href="#cb12-119" aria-hidden="true" tabindex="-1"></a>    legend.set_title(<span class="va">None</span>)</span>
<span id="cb12-120"><a href="#cb12-120" aria-hidden="true" tabindex="-1"></a>    legend.get_frame().<span class="bu">set</span>(</span>
<span id="cb12-121"><a href="#cb12-121" aria-hidden="true" tabindex="-1"></a>        alpha<span class="op">=</span><span class="dv">0</span>, boxstyle<span class="op">=</span><span class="st">'square'</span>, edgecolor<span class="op">=</span><span class="st">'grey'</span>, linewidth<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb12-122"><a href="#cb12-122" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> text <span class="kw">in</span> legend.get_texts():</span>
<span id="cb12-123"><a href="#cb12-123" aria-hidden="true" tabindex="-1"></a>        t <span class="op">=</span> text.get_text()</span>
<span id="cb12-124"><a href="#cb12-124" aria-hidden="true" tabindex="-1"></a>        new_t <span class="op">=</span> update_legend_values.get(t,t)</span>
<span id="cb12-125"><a href="#cb12-125" aria-hidden="true" tabindex="-1"></a>        text.set_text(new_t)</span>
<span id="cb12-126"><a href="#cb12-126" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-127"><a href="#cb12-127" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-128"><a href="#cb12-128" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/supplemental/Supplemental 4.png'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span>
<span id="cb12-129"><a href="#cb12-129" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/supplemental/Supplemental 4.svg'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span>
<span id="cb12-130"><a href="#cb12-130" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/supplemental/Supplemental 4.eps'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span>
<span id="cb12-131"><a href="#cb12-131" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/supplemental/Supplemental 4.pdf'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>The PostScript backend does not support transparency; partially transparent artists will be rendered opaque.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="make_plots_files/figure-html/cell-9-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Supplemental Figure 3. SNP (L column) and Indel (R column) switch error rates for 1KGP phased haplotype panel variation, binned by minor allele frequency. Switch error rates were determined by comparing estimated phasing to ground-truth assemblies for the following subsets of samples: a) Mendelian-corrected trio probands with ground truth assemblies present in the HPRC human pangenome; b) Mendelian-corrected trio probands with ground truth assemblies produced by the HGSVC; c) Mendelian-corrected trio parents with ground truth assemblies produced by the HGSVC; d) Uncorrected samples which are not part of a 1KGP trio with ground truth assemblies produced by the HGSVC. e) The switch error rates of trio probands, trio parents, and non-trio samples were weighted by each category’s prevelance in the 1000 Genomes Project to produce estimated panel-wide average switch error rates. Switch error rates and average minor allele frequency per bin are displayed on a log scale.</p>
<div id="cell-17" class="cell" data-execution_count="9">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Figure 2 supplementals</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Figure 2</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="co">## Figure 2 with parents and estimated overall</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>fig, ((axA, axB), (axC, axD), (axE, axF), (axG, axH), (axI, axJ)) <span class="op">=</span> plt.subplots(<span class="dv">5</span>,<span class="dv">2</span>, figsize<span class="op">=</span>(oneandahalf_col,<span class="dv">8</span>))</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a><span class="co"># fig.set_layout_engine(layout='constrained', h_pad=0.15, w_pad=0.15)</span></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'gt_error_rate'</span>, style<span class="op">=</span><span class="st">'syntenic'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> syntenic_order,</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data.loc[(binned_maf_data.ground_truth_data_source <span class="op">==</span> <span class="st">'HPRC_samples'</span>)</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.<span class="bu">type</span> <span class="op">==</span> <span class="st">'SNP'</span>)</span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.syntenic <span class="op">!=</span> <span class="st">'All'</span>)],</span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axA, <span class="op">**</span>syntenic_style)</span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'gt_error_rate'</span>, style<span class="op">=</span><span class="st">'syntenic'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> syntenic_order,</span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data.loc[(binned_maf_data.ground_truth_data_source <span class="op">==</span> <span class="st">'HGSVC_probands'</span>) <span class="co">#trios only</span></span>
<span id="cb14-18"><a href="#cb14-18" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb14-19"><a href="#cb14-19" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.<span class="bu">type</span> <span class="op">==</span> <span class="st">'SNP'</span>)</span>
<span id="cb14-20"><a href="#cb14-20" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.syntenic <span class="op">!=</span> <span class="st">'All'</span>)],</span>
<span id="cb14-21"><a href="#cb14-21" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axC, <span class="op">**</span>syntenic_style)</span>
<span id="cb14-22"><a href="#cb14-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-23"><a href="#cb14-23" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'gt_error_rate'</span>, style<span class="op">=</span><span class="st">'syntenic'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> syntenic_order,</span>
<span id="cb14-24"><a href="#cb14-24" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data.loc[(binned_maf_data.ground_truth_data_source <span class="op">==</span> <span class="st">'HGSVC_parents'</span>) <span class="co">#trios only</span></span>
<span id="cb14-25"><a href="#cb14-25" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb14-26"><a href="#cb14-26" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.<span class="bu">type</span> <span class="op">==</span> <span class="st">'SNP'</span>)</span>
<span id="cb14-27"><a href="#cb14-27" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.syntenic <span class="op">!=</span> <span class="st">'All'</span>)],</span>
<span id="cb14-28"><a href="#cb14-28" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axE, <span class="op">**</span>syntenic_style)</span>
<span id="cb14-29"><a href="#cb14-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-30"><a href="#cb14-30" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'gt_error_rate'</span>, style<span class="op">=</span><span class="st">'syntenic'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> syntenic_order,</span>
<span id="cb14-31"><a href="#cb14-31" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data.loc[(binned_maf_data.ground_truth_data_source <span class="op">==</span> <span class="st">'HGSVC_samples_nontrios_only'</span>) <span class="co">#trios only</span></span>
<span id="cb14-32"><a href="#cb14-32" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb14-33"><a href="#cb14-33" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.<span class="bu">type</span> <span class="op">==</span> <span class="st">'SNP'</span>)</span>
<span id="cb14-34"><a href="#cb14-34" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.syntenic <span class="op">!=</span> <span class="st">'All'</span>)],</span>
<span id="cb14-35"><a href="#cb14-35" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axG, <span class="op">**</span>syntenic_style)</span>
<span id="cb14-36"><a href="#cb14-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-37"><a href="#cb14-37" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'gt_error_rate'</span>, style<span class="op">=</span><span class="st">'syntenic'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> syntenic_order,</span>
<span id="cb14-38"><a href="#cb14-38" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>cat_data.loc[(cat_data.<span class="bu">type</span> <span class="op">==</span> <span class="st">'SNP'</span>)</span>
<span id="cb14-39"><a href="#cb14-39" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (cat_data.syntenic <span class="op">!=</span> <span class="st">'All'</span>)],</span>
<span id="cb14-40"><a href="#cb14-40" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axI, <span class="op">**</span>syntenic_style)</span>
<span id="cb14-41"><a href="#cb14-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-42"><a href="#cb14-42" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'gt_error_rate'</span>, style<span class="op">=</span><span class="st">'syntenic'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> syntenic_order,</span>
<span id="cb14-43"><a href="#cb14-43" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data.loc[(binned_maf_data.ground_truth_data_source <span class="op">==</span> <span class="st">'HPRC_samples'</span>)</span>
<span id="cb14-44"><a href="#cb14-44" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb14-45"><a href="#cb14-45" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.<span class="bu">type</span> <span class="op">==</span> <span class="st">'Indel'</span>)</span>
<span id="cb14-46"><a href="#cb14-46" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.syntenic <span class="op">!=</span> <span class="st">'All'</span>)],</span>
<span id="cb14-47"><a href="#cb14-47" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axB, <span class="op">**</span>syntenic_style)</span>
<span id="cb14-48"><a href="#cb14-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-49"><a href="#cb14-49" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'gt_error_rate'</span>, style<span class="op">=</span><span class="st">'syntenic'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> syntenic_order,</span>
<span id="cb14-50"><a href="#cb14-50" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data.loc[(binned_maf_data.ground_truth_data_source <span class="op">==</span> <span class="st">'HGSVC_probands'</span>)</span>
<span id="cb14-51"><a href="#cb14-51" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb14-52"><a href="#cb14-52" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.<span class="bu">type</span> <span class="op">==</span> <span class="st">'Indel'</span>)</span>
<span id="cb14-53"><a href="#cb14-53" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.syntenic <span class="op">!=</span> <span class="st">'All'</span>)],</span>
<span id="cb14-54"><a href="#cb14-54" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axD, <span class="op">**</span>syntenic_style)</span>
<span id="cb14-55"><a href="#cb14-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-56"><a href="#cb14-56" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'gt_error_rate'</span>, style<span class="op">=</span><span class="st">'syntenic'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> syntenic_order,</span>
<span id="cb14-57"><a href="#cb14-57" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data.loc[(binned_maf_data.ground_truth_data_source <span class="op">==</span> <span class="st">'HGSVC_parents'</span>)</span>
<span id="cb14-58"><a href="#cb14-58" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb14-59"><a href="#cb14-59" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.<span class="bu">type</span> <span class="op">==</span> <span class="st">'Indel'</span>)</span>
<span id="cb14-60"><a href="#cb14-60" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.syntenic <span class="op">!=</span> <span class="st">'All'</span>)],</span>
<span id="cb14-61"><a href="#cb14-61" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axF, <span class="op">**</span>syntenic_style)</span>
<span id="cb14-62"><a href="#cb14-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-63"><a href="#cb14-63" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'gt_error_rate'</span>, style<span class="op">=</span><span class="st">'syntenic'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> syntenic_order,</span>
<span id="cb14-64"><a href="#cb14-64" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data.loc[(binned_maf_data.ground_truth_data_source <span class="op">==</span> <span class="st">'HGSVC_samples_nontrios_only'</span>) <span class="co">#trios only</span></span>
<span id="cb14-65"><a href="#cb14-65" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb14-66"><a href="#cb14-66" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.<span class="bu">type</span> <span class="op">==</span> <span class="st">'Indel'</span>)</span>
<span id="cb14-67"><a href="#cb14-67" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data.syntenic <span class="op">!=</span> <span class="st">'All'</span>)],</span>
<span id="cb14-68"><a href="#cb14-68" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axH, <span class="op">**</span>syntenic_style)</span>
<span id="cb14-69"><a href="#cb14-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-70"><a href="#cb14-70" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'gt_error_rate'</span>, style<span class="op">=</span><span class="st">'syntenic'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> syntenic_order,</span>
<span id="cb14-71"><a href="#cb14-71" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>cat_data.loc[(cat_data.<span class="bu">type</span> <span class="op">==</span> <span class="st">'Indel'</span>)</span>
<span id="cb14-72"><a href="#cb14-72" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (cat_data.syntenic <span class="op">!=</span> <span class="st">'All'</span>)],</span>
<span id="cb14-73"><a href="#cb14-73" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axJ, <span class="op">**</span>syntenic_style)</span>
<span id="cb14-74"><a href="#cb14-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-75"><a href="#cb14-75" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, ax <span class="kw">in</span> <span class="bu">enumerate</span>([axA, axB, axC, axD, axE, axF, axG, axH, axI, axJ]):</span>
<span id="cb14-76"><a href="#cb14-76" aria-hidden="true" tabindex="-1"></a>    ax.set_yscale(<span class="st">'log'</span>)</span>
<span id="cb14-77"><a href="#cb14-77" aria-hidden="true" tabindex="-1"></a>    ax.set_xscale(<span class="st">'log'</span>)</span>
<span id="cb14-78"><a href="#cb14-78" aria-hidden="true" tabindex="-1"></a>    ax.set_ylim(<span class="fl">1e-1</span>, <span class="dv">50</span>)</span>
<span id="cb14-79"><a href="#cb14-79" aria-hidden="true" tabindex="-1"></a>    ax.set_xlim(<span class="fl">1e-2</span>, <span class="dv">55</span>)</span>
<span id="cb14-80"><a href="#cb14-80" aria-hidden="true" tabindex="-1"></a>    <span class="co"># ax.set_box_aspect(1)</span></span>
<span id="cb14-81"><a href="#cb14-81" aria-hidden="true" tabindex="-1"></a>    ax.grid()</span>
<span id="cb14-82"><a href="#cb14-82" aria-hidden="true" tabindex="-1"></a>    ax.set_axisbelow(<span class="va">True</span>)</span>
<span id="cb14-83"><a href="#cb14-83" aria-hidden="true" tabindex="-1"></a>    ax.set_xticks(MAF_xticks)</span>
<span id="cb14-84"><a href="#cb14-84" aria-hidden="true" tabindex="-1"></a>    ax.set_yticks(SER_xticks)</span>
<span id="cb14-85"><a href="#cb14-85" aria-hidden="true" tabindex="-1"></a>    ax.xaxis.set_major_formatter(perc_formatter)</span>
<span id="cb14-86"><a href="#cb14-86" aria-hidden="true" tabindex="-1"></a>    ax.yaxis.set_major_formatter(perc_formatter)</span>
<span id="cb14-87"><a href="#cb14-87" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="kw">in</span> (<span class="dv">8</span>,<span class="dv">9</span>):</span>
<span id="cb14-88"><a href="#cb14-88" aria-hidden="true" tabindex="-1"></a>        ax.set_xlabel(<span class="st">'Minor Allele Frequency (%)'</span>, labelpad<span class="op">=</span><span class="dv">3</span>)</span>
<span id="cb14-89"><a href="#cb14-89" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb14-90"><a href="#cb14-90" aria-hidden="true" tabindex="-1"></a>        ax.set_xlabel(<span class="va">None</span>)</span>
<span id="cb14-91"><a href="#cb14-91" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="kw">in</span> (<span class="dv">0</span>,<span class="dv">2</span>,<span class="dv">4</span>,<span class="dv">6</span>,<span class="dv">8</span>):</span>
<span id="cb14-92"><a href="#cb14-92" aria-hidden="true" tabindex="-1"></a>        ax.set_ylabel(<span class="st">'Genotyping Error Rate (%)'</span>, labelpad<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb14-93"><a href="#cb14-93" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb14-94"><a href="#cb14-94" aria-hidden="true" tabindex="-1"></a>        ax.set_ylabel(<span class="va">None</span>)</span>
<span id="cb14-95"><a href="#cb14-95" aria-hidden="true" tabindex="-1"></a>    add_letter_to_ax(ax, i)</span>
<span id="cb14-96"><a href="#cb14-96" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-97"><a href="#cb14-97" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-98"><a href="#cb14-98" aria-hidden="true" tabindex="-1"></a>x_offset <span class="op">=</span> <span class="op">-</span><span class="fl">0.4</span></span>
<span id="cb14-99"><a href="#cb14-99" aria-hidden="true" tabindex="-1"></a>axA.text(x_offset, <span class="fl">0.5</span>, <span class="st">'Mendelian phased</span><span class="ch">\n</span><span class="st">HPRC samples'</span>, fontsize<span class="op">=</span><span class="dv">10</span>, transform<span class="op">=</span>axA.transAxes, weight<span class="op">=</span><span class="st">'bold'</span>, rotation<span class="op">=</span><span class="st">'vertical'</span>, horizontalalignment<span class="op">=</span><span class="st">'center'</span>, verticalalignment<span class="op">=</span><span class="st">'center'</span>)</span>
<span id="cb14-100"><a href="#cb14-100" aria-hidden="true" tabindex="-1"></a>axC.text(x_offset, <span class="fl">0.5</span>, <span class="st">'Mendelian phased</span><span class="ch">\n</span><span class="st">HGSVC probands'</span>, fontsize<span class="op">=</span><span class="dv">10</span>, transform<span class="op">=</span>axC.transAxes, weight<span class="op">=</span><span class="st">'bold'</span>, rotation<span class="op">=</span><span class="st">'vertical'</span>, horizontalalignment<span class="op">=</span><span class="st">'center'</span>, verticalalignment<span class="op">=</span><span class="st">'center'</span>)</span>
<span id="cb14-101"><a href="#cb14-101" aria-hidden="true" tabindex="-1"></a>axE.text(x_offset, <span class="fl">0.5</span>, <span class="st">'Mendelian phased</span><span class="ch">\n</span><span class="st">HGSVC parents'</span>, fontsize<span class="op">=</span><span class="dv">10</span>, transform<span class="op">=</span>axE.transAxes, weight<span class="op">=</span><span class="st">'bold'</span>, rotation<span class="op">=</span><span class="st">'vertical'</span>, horizontalalignment<span class="op">=</span><span class="st">'center'</span>, verticalalignment<span class="op">=</span><span class="st">'center'</span>)</span>
<span id="cb14-102"><a href="#cb14-102" aria-hidden="true" tabindex="-1"></a>axG.text(x_offset, <span class="fl">0.5</span>, <span class="st">'Statistically phased</span><span class="ch">\n</span><span class="st">HGSVC samples'</span>, fontsize<span class="op">=</span><span class="dv">10</span>, transform<span class="op">=</span>axG.transAxes, weight<span class="op">=</span><span class="st">'bold'</span>, rotation<span class="op">=</span><span class="st">'vertical'</span>, horizontalalignment<span class="op">=</span><span class="st">'center'</span>, verticalalignment<span class="op">=</span><span class="st">'center'</span>)</span>
<span id="cb14-103"><a href="#cb14-103" aria-hidden="true" tabindex="-1"></a>axI.text(x_offset, <span class="fl">0.5</span>, <span class="st">'Estimated</span><span class="ch">\n</span><span class="st">panel-wide accuracy'</span>, fontsize<span class="op">=</span><span class="dv">10</span>, transform<span class="op">=</span>axI.transAxes, weight<span class="op">=</span><span class="st">'bold'</span>, rotation<span class="op">=</span><span class="st">'vertical'</span>, horizontalalignment<span class="op">=</span><span class="st">'center'</span>, verticalalignment<span class="op">=</span><span class="st">'center'</span>)</span>
<span id="cb14-104"><a href="#cb14-104" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-105"><a href="#cb14-105" aria-hidden="true" tabindex="-1"></a>axA.set_title(<span class="st">'SNPs'</span>, fontsize<span class="op">=</span><span class="dv">9</span>, weight<span class="op">=</span><span class="st">'bold'</span>, pad<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb14-106"><a href="#cb14-106" aria-hidden="true" tabindex="-1"></a>axB.set_title(<span class="st">'Indels'</span>, fontsize<span class="op">=</span><span class="dv">9</span>, weight<span class="op">=</span><span class="st">'bold'</span>, pad<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb14-107"><a href="#cb14-107" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-108"><a href="#cb14-108" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, ax <span class="kw">in</span> <span class="bu">enumerate</span>([axA, axB, axC, axD, axE, axF, axG, axH, axI, axJ]): <span class="co"># legend axes</span></span>
<span id="cb14-109"><a href="#cb14-109" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="kw">in</span> (<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">5</span>,<span class="dv">7</span>,<span class="dv">9</span>):</span>
<span id="cb14-110"><a href="#cb14-110" aria-hidden="true" tabindex="-1"></a>        loc<span class="op">=</span><span class="st">'lower right'</span></span>
<span id="cb14-111"><a href="#cb14-111" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb14-112"><a href="#cb14-112" aria-hidden="true" tabindex="-1"></a>        loc<span class="op">=</span><span class="st">'upper right'</span></span>
<span id="cb14-113"><a href="#cb14-113" aria-hidden="true" tabindex="-1"></a>    handles, labels <span class="op">=</span> ax.get_legend_handles_labels()</span>
<span id="cb14-114"><a href="#cb14-114" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> handle <span class="kw">in</span> handles:</span>
<span id="cb14-115"><a href="#cb14-115" aria-hidden="true" tabindex="-1"></a>        handle.set_marker(<span class="st">'o'</span>)</span>
<span id="cb14-116"><a href="#cb14-116" aria-hidden="true" tabindex="-1"></a>    handles <span class="op">=</span> handles[<span class="dv">1</span>:<span class="dv">3</span>] <span class="op">+</span> handles[<span class="dv">5</span>:<span class="dv">6</span>]</span>
<span id="cb14-117"><a href="#cb14-117" aria-hidden="true" tabindex="-1"></a>    labels <span class="op">=</span> labels[<span class="dv">1</span>:<span class="dv">3</span>] <span class="op">+</span> labels[<span class="dv">5</span>:<span class="dv">6</span>]</span>
<span id="cb14-118"><a href="#cb14-118" aria-hidden="true" tabindex="-1"></a>    handles[<span class="op">-</span><span class="dv">1</span>].set_color(handles[<span class="dv">1</span>].get_color())</span>
<span id="cb14-119"><a href="#cb14-119" aria-hidden="true" tabindex="-1"></a>    legend <span class="op">=</span> ax.legend(ncols<span class="op">=</span><span class="dv">1</span>, loc<span class="op">=</span>loc, handles<span class="op">=</span>handles, labels<span class="op">=</span>labels)</span>
<span id="cb14-120"><a href="#cb14-120" aria-hidden="true" tabindex="-1"></a>    legend.set_title(<span class="va">None</span>)</span>
<span id="cb14-121"><a href="#cb14-121" aria-hidden="true" tabindex="-1"></a>    legend.get_frame().<span class="bu">set</span>(</span>
<span id="cb14-122"><a href="#cb14-122" aria-hidden="true" tabindex="-1"></a>        alpha<span class="op">=</span><span class="dv">0</span>, boxstyle<span class="op">=</span><span class="st">'square'</span>, edgecolor<span class="op">=</span><span class="st">'grey'</span>, linewidth<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb14-123"><a href="#cb14-123" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> text <span class="kw">in</span> legend.get_texts():</span>
<span id="cb14-124"><a href="#cb14-124" aria-hidden="true" tabindex="-1"></a>        t <span class="op">=</span> text.get_text()</span>
<span id="cb14-125"><a href="#cb14-125" aria-hidden="true" tabindex="-1"></a>        new_t <span class="op">=</span> update_legend_values.get(t,t)</span>
<span id="cb14-126"><a href="#cb14-126" aria-hidden="true" tabindex="-1"></a>        text.set_text(new_t)</span>
<span id="cb14-127"><a href="#cb14-127" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-128"><a href="#cb14-128" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-129"><a href="#cb14-129" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/supplemental/Supplemental 3.png'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span>
<span id="cb14-130"><a href="#cb14-130" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/supplemental/Supplemental 3.svg'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span>
<span id="cb14-131"><a href="#cb14-131" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/supplemental/Supplemental 3.eps'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span>
<span id="cb14-132"><a href="#cb14-132" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/supplemental/Supplemental 3.pdf'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span>
<span id="cb14-133"><a href="#cb14-133" aria-hidden="true" tabindex="-1"></a></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>The PostScript backend does not support transparency; partially transparent artists will be rendered opaque.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="make_plots_files/figure-html/cell-10-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Supplemental figure 3: SNP (L column) and Indel (R column) genotyping error rates for 1KGP phased haplotype panel variation, binned by minor allele frequency. Genotyping error rates were defined as the number of inaccurate heterozygous or homozygous alternative genotypes (both alleles must be called correctly; there was no partial credit for correctly calling one allele) divided by the number of heterozygous or homozygous alternative genotypes. Correct and incorrect variant calls were determined via comparison to ground-truth assemblies for the following subsets of samples: a) Mendelian-corrected trio probands with ground truth assemblies present in the HPRC human pangenome; b) Mendelian-corrected trio probands with ground truth assemblies produced by the HGSVC; c) Mendelian-corrected trio parents with ground truth assemblies produced by the HGSVC; d) Uncorrected samples which are not part of a 1KGP trio with ground truth assemblies produced by the HGSVC. e) The genotype error rates of trio probands, trio parents, and non-trio samples were weighted by each category’s prevelance in the 1000 Genomes Project to produce estimated panel-wide average genotype error rates. Genotype error rates and average minor allele frequency per bin are displayed on a log scale.</p>
</section>
</section>
<section id="figure-3" class="level2" data-number="1.2">
<h2 data-number="1.2" class="anchored" data-anchor-id="figure-3"><span class="header-section-number">1.2</span> Figure 3</h2>
<section id="per-chromosome-error-rate-performance-in-chm13-vs-grch38-including-chrx" class="level3" data-number="1.2.1">
<h3 data-number="1.2.1" class="anchored" data-anchor-id="per-chromosome-error-rate-performance-in-chm13-vs-grch38-including-chrx"><span class="header-section-number">1.2.1</span> Per-chromosome error rate performance in CHM13 vs GRCh38, including chrX</h3>
<ul>
<li>supplemental separating out PAR regions</li>
</ul>
<div id="cell-20" class="cell" data-execution_count="10">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Figure 3: Per-chromosome error rate performance in CHM13 vs GRCh38, including chrX</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>all_chroms_plus_chrX <span class="op">=</span> pd.concat((all_chroms_chrX, all_samples.assign(chrom <span class="op">=</span> <span class="st">'chr1-22+X'</span>))).reset_index(drop<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>full_panel_HPRC_compared <span class="op">=</span> all_chroms_plus_chrX.loc[ (all_chroms_plus_chrX.method_of_phasing<span class="op">==</span><span class="st">'phased_with_parents_and_pedigree'</span>) <span class="op">&amp;</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>                                                (all_chroms_plus_chrX.ground_truth_data_source.isin((<span class="st">'HPRC_samples'</span>,<span class="st">'HGSVC_probands'</span>, <span class="st">'HGSVC_parents'</span>, <span class="st">'HGSVC_samples_nontrios_only'</span>)))].replace(<span class="st">'chr1-22+X'</span>, <span class="st">'Whole</span><span class="ch">\n</span><span class="st">Genome'</span>)</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(<span class="dv">1</span>,<span class="dv">1</span>,figsize<span class="op">=</span>(two_col,<span class="fl">1.5</span>), layout<span class="op">=</span><span class="st">'tight'</span>)</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>sns.barplot(x<span class="op">=</span><span class="st">'chrom'</span>, y<span class="op">=</span><span class="st">'switch_error_rate'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, ax<span class="op">=</span>ax,</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>            order<span class="op">=</span>[x <span class="cf">for</span> x <span class="kw">in</span> chrom_order <span class="cf">if</span> <span class="st">'PAR'</span> <span class="kw">not</span> <span class="kw">in</span> x],</span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>            hue_order<span class="op">=</span>genome_order, palette<span class="op">=</span>genome_palette,</span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>            err_kws<span class="op">=</span>{<span class="st">'linewidth'</span>:<span class="fl">0.5</span>}, capsize<span class="op">=</span><span class="fl">0.3</span>,</span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>            <span class="co"># split=True, inner=None, linewidth=0.5, cut=0, density_norm='area',</span></span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a>            data<span class="op">=</span>full_panel_HPRC_compared.loc[(full_panel_HPRC_compared.ground_truth_data_source<span class="op">==</span><span class="st">'HPRC_samples'</span>)])</span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a>ax.set_ylabel(<span class="st">'Switch Error Rate'</span>)</span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a>ax.set_xlabel(<span class="st">'Chromosome'</span>)</span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a>ax.set_ylim(<span class="dv">0</span>,full_panel_HPRC_compared.loc[(full_panel_HPRC_compared.ground_truth_data_source<span class="op">==</span><span class="st">'HPRC_samples'</span>)].switch_error_rate.<span class="bu">max</span>()<span class="op">*</span><span class="fl">1.08</span>)</span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true" tabindex="-1"></a>legend <span class="op">=</span> ax.legend()</span>
<span id="cb16-20"><a href="#cb16-20" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> text <span class="kw">in</span> legend.get_texts():</span>
<span id="cb16-21"><a href="#cb16-21" aria-hidden="true" tabindex="-1"></a>    t <span class="op">=</span> text.get_text()</span>
<span id="cb16-22"><a href="#cb16-22" aria-hidden="true" tabindex="-1"></a>    new_t <span class="op">=</span> update_legend_values.get(t,t)</span>
<span id="cb16-23"><a href="#cb16-23" aria-hidden="true" tabindex="-1"></a>    text.set_text(new_t)</span>
<span id="cb16-24"><a href="#cb16-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-25"><a href="#cb16-25" aria-hidden="true" tabindex="-1"></a>sns.move_legend(ax, <span class="st">"upper left"</span>)</span>
<span id="cb16-26"><a href="#cb16-26" aria-hidden="true" tabindex="-1"></a>ax.yaxis.set_major_formatter(PercentFormatter(decimals<span class="op">=</span><span class="dv">2</span>))</span>
<span id="cb16-27"><a href="#cb16-27" aria-hidden="true" tabindex="-1"></a>ax.grid(axis<span class="op">=</span><span class="st">'y'</span>, alpha<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="cb16-28"><a href="#cb16-28" aria-hidden="true" tabindex="-1"></a>ax.set_axisbelow(<span class="va">True</span>)</span>
<span id="cb16-29"><a href="#cb16-29" aria-hidden="true" tabindex="-1"></a>legend.set_title(<span class="st">'Genome'</span>)</span>
<span id="cb16-30"><a href="#cb16-30" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/Figure 3.png'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span>
<span id="cb16-31"><a href="#cb16-31" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/Figure 3.svg'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span>
<span id="cb16-32"><a href="#cb16-32" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/Figure 3.eps'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span>
<span id="cb16-33"><a href="#cb16-33" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/Figure 3.pdf'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>The PostScript backend does not support transparency; partially transparent artists will be rendered opaque.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="make_plots_files/figure-html/cell-11-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Figure 3: Whole genome and per-chromosome switch error rates from all 100 samples with HPRC or HGSVC assemblies. Error bars show +/- 95% ci.</p>
<div id="cell-22" class="cell" data-execution_count="11">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># fig, ax = plt.subplots(1,1,figsize=(one_col,one_col), layout='tight')</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(<span class="dv">1</span>,<span class="dv">1</span>,figsize<span class="op">=</span>(one_col,one_col), layout<span class="op">=</span><span class="st">'tight'</span>)</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="co">## Supplemental: PAR1, PAR2, chrX</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>all_chroms_plus <span class="op">=</span> pd.concat((all_chroms, all_samples.assign(chrom <span class="op">=</span> <span class="st">'chr1-22+X'</span>))).reset_index(drop<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>all_chroms_plus[<span class="st">'chrom'</span>] <span class="op">=</span> all_chroms_plus.chrom.<span class="bu">map</span>(<span class="kw">lambda</span> x: {<span class="st">'chrX'</span>:<span class="st">'Non-PAR ChrX'</span>}.get(x,x))</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a><span class="co">### PAR is full of outlier samples (think HG01175, with 1 switch error and 2 checked)</span></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>all_chroms_plus <span class="op">=</span> all_chroms_plus.loc[all_chroms_plus.n_checked<span class="op">&gt;</span><span class="dv">5</span>]</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>full_panel_HPRC_compared_sep_chr_X <span class="op">=</span> all_chroms_plus.loc[ (all_chroms_plus.method_of_phasing<span class="op">==</span><span class="st">'phased_with_parents_and_pedigree'</span>) <span class="op">&amp;</span></span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>                                                (all_chroms_plus.ground_truth_data_source.isin((<span class="st">'HPRC_samples'</span>,<span class="st">'HGSVC_probands'</span>)))].replace(<span class="st">'chr1-22+X'</span>, <span class="st">'Whole</span><span class="ch">\n</span><span class="st">Genome'</span>).rename(columns<span class="op">=</span>{<span class="st">'genome'</span>:<span class="st">"Genome"</span>})</span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> sns.violinplot(x<span class="op">=</span><span class="st">'chrom'</span>, y<span class="op">=</span><span class="st">'switch_error_rate'</span>, hue<span class="op">=</span><span class="st">'Genome'</span>, ax<span class="op">=</span>ax,</span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a>            hue_order<span class="op">=</span>genome_order, palette<span class="op">=</span>genome_palette, order<span class="op">=</span>[<span class="st">'Whole</span><span class="ch">\n</span><span class="st">Genome'</span>,<span class="st">'PAR1'</span>,<span class="st">'PAR2'</span>,<span class="st">'Non-PAR ChrX'</span>],</span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a>            <span class="co"># err_kws={'linewidth':1}, capsize=0.3,</span></span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a>            split<span class="op">=</span><span class="va">True</span>, inner<span class="op">=</span><span class="va">None</span>, linewidth<span class="op">=</span><span class="fl">0.5</span>, cut<span class="op">=</span><span class="dv">0</span>, density_norm<span class="op">=</span><span class="st">'count'</span>,</span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a>            data<span class="op">=</span>full_panel_HPRC_compared_sep_chr_X.loc[full_panel_HPRC_compared_sep_chr_X.chrom.isin((<span class="st">'PAR1'</span>,<span class="st">'PAR2'</span>,<span class="st">'Non-PAR ChrX'</span>, <span class="st">'Whole</span><span class="ch">\n</span><span class="st">Genome'</span>))],</span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a>            legend<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true" tabindex="-1"></a><span class="co"># ax.set_ylim(0,2)</span></span>
<span id="cb18-20"><a href="#cb18-20" aria-hidden="true" tabindex="-1"></a>ax.set_ylabel(<span class="st">'Switch Error Rate'</span>)</span>
<span id="cb18-21"><a href="#cb18-21" aria-hidden="true" tabindex="-1"></a>ax.set_xlabel(<span class="st">'Chromosome'</span>)</span>
<span id="cb18-22"><a href="#cb18-22" aria-hidden="true" tabindex="-1"></a>ax.yaxis.set_major_formatter(perc_formatter)</span>
<span id="cb18-23"><a href="#cb18-23" aria-hidden="true" tabindex="-1"></a>ax.grid(axis<span class="op">=</span><span class="st">'y'</span>, alpha<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="cb18-24"><a href="#cb18-24" aria-hidden="true" tabindex="-1"></a>ax.set_axisbelow(<span class="va">True</span>)</span>
<span id="cb18-25"><a href="#cb18-25" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/supplemental/Supplemental 5.png'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span>
<span id="cb18-26"><a href="#cb18-26" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/supplemental/Supplemental 5.eps'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span>
<span id="cb18-27"><a href="#cb18-27" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/supplemental/Supplemental 5.svg'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span>
<span id="cb18-28"><a href="#cb18-28" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/supplemental/Supplemental 5.pdf'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>The PostScript backend does not support transparency; partially transparent artists will be rendered opaque.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="make_plots_files/figure-html/cell-12-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Supplemental Figure 5. Whole genome, pseudoautosomal region 1 (PAR1), PAR2, and non-PAR chromosome X switch error rates. Switch error rates for trio probands with HPRC or HGSVC assemblies from either the GRCh38 or CHM13v2.0 1KGP reference haplotype panels, with each sample’s assembly used as ground truth. PAR switch error rates are calculated seperately from the non-PAR region of chromosome X. Non-PAR chromosome X switch error rates are only calculated from female samples.</p>
<div id="cell-24" class="cell" data-execution_count="12">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Supplemental 6: </span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>superpop_order <span class="op">=</span> [<span class="st">'AFR'</span>,<span class="st">'EUR'</span>,<span class="st">'AMR'</span>,<span class="st">'SAS'</span>,<span class="st">'EAS'</span>]</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(<span class="dv">1</span>,<span class="dv">1</span>,figsize<span class="op">=</span>(oneandahalf_col, <span class="dv">3</span>))</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>superpop_data <span class="op">=</span> all_samples.loc[(all_samples.method_of_phasing<span class="op">==</span><span class="st">'phased_with_parents_and_pedigree'</span>) <span class="op">&amp;</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>                                (all_samples.ground_truth_data_source<span class="op">==</span><span class="st">'trios'</span>)].copy()</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>superpop_data[<span class="st">'switch_error_rate'</span>] <span class="op">=</span> superpop_data.n_switch_errors<span class="op">/</span>superpop_data.n_checked <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>ax.set_yscale(<span class="st">'log'</span>)</span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> sns.swarmplot(y<span class="op">=</span><span class="st">'switch_error_rate'</span>, x<span class="op">=</span><span class="st">'genome'</span>, hue<span class="op">=</span><span class="st">'superpopulation'</span>, ax<span class="op">=</span>ax,</span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a>            data<span class="op">=</span>superpop_data,</span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>            order<span class="op">=</span>genome_order, hue_order<span class="op">=</span>superpop_order, s<span class="op">=</span><span class="dv">1</span>)<span class="co">#             err_kws={'linewidth':0.75}, capsize=0.3)</span></span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a>ax.set_ylabel(<span class="st">'Switch Error Rate'</span>)</span>
<span id="cb20-16"><a href="#cb20-16" aria-hidden="true" tabindex="-1"></a>ax.set_xlabel(<span class="st">'1KGP Reference Panel'</span>)</span>
<span id="cb20-17"><a href="#cb20-17" aria-hidden="true" tabindex="-1"></a>ax.legend().set_title(<span class="st">'Superpopulation'</span>)</span>
<span id="cb20-18"><a href="#cb20-18" aria-hidden="true" tabindex="-1"></a>sns.move_legend(ax, <span class="st">"upper left"</span>, bbox_to_anchor<span class="op">=</span>(<span class="dv">1</span>, <span class="dv">1</span>))</span>
<span id="cb20-19"><a href="#cb20-19" aria-hidden="true" tabindex="-1"></a>ax.grid(axis<span class="op">=</span><span class="st">'y'</span>, alpha<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="cb20-20"><a href="#cb20-20" aria-hidden="true" tabindex="-1"></a>ax.set_axisbelow(<span class="va">True</span>)</span>
<span id="cb20-21"><a href="#cb20-21" aria-hidden="true" tabindex="-1"></a>ax.yaxis.set_major_formatter(PercentFormatter(decimals<span class="op">=</span><span class="dv">2</span>))</span>
<span id="cb20-22"><a href="#cb20-22" aria-hidden="true" tabindex="-1"></a>ax.set_ylim(bottom<span class="op">=</span><span class="fl">0.06</span>) <span class="co"># drop three samples in CHM13v2.0 this way.</span></span>
<span id="cb20-23"><a href="#cb20-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-24"><a href="#cb20-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-25"><a href="#cb20-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-26"><a href="#cb20-26" aria-hidden="true" tabindex="-1"></a>handles, labels <span class="op">=</span>ax.get_legend_handles_labels()</span>
<span id="cb20-27"><a href="#cb20-27" aria-hidden="true" tabindex="-1"></a>new_handles<span class="op">=</span><span class="bu">list</span>()</span>
<span id="cb20-28"><a href="#cb20-28" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, (handle, label) <span class="kw">in</span> <span class="bu">enumerate</span>(<span class="bu">zip</span>(handles, labels)):</span>
<span id="cb20-29"><a href="#cb20-29" aria-hidden="true" tabindex="-1"></a>    handle.set_alpha(<span class="dv">1</span>)</span>
<span id="cb20-30"><a href="#cb20-30" aria-hidden="true" tabindex="-1"></a>    handle.set_markersize(<span class="dv">5</span>)</span>
<span id="cb20-31"><a href="#cb20-31" aria-hidden="true" tabindex="-1"></a>    new_handles.append(handle)</span>
<span id="cb20-32"><a href="#cb20-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-33"><a href="#cb20-33" aria-hidden="true" tabindex="-1"></a>ax.legend(new_handles, labels, title<span class="op">=</span><span class="st">"1KGP Superpopulation"</span>)</span>
<span id="cb20-34"><a href="#cb20-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-35"><a href="#cb20-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-36"><a href="#cb20-36" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/supplemental/Supplemental 6.png'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span>
<span id="cb20-37"><a href="#cb20-37" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/supplemental/Supplemental 6.svg'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span>
<span id="cb20-38"><a href="#cb20-38" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/supplemental/Supplemental 6.eps'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span>
<span id="cb20-39"><a href="#cb20-39" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/supplemental/Supplemental 6.pdf'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span>
<span id="cb20-40"><a href="#cb20-40" aria-hidden="true" tabindex="-1"></a><span class="co"># 22% and 27.3% of points cannot be plotted, always, since they are off the log axis at 0</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>/tmp/ipykernel_2892998/991315246.py:17: UserWarning: No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.
  ax.legend().set_title('Superpopulation')
The PostScript backend does not support transparency; partially transparent artists will be rendered opaque.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="make_plots_files/figure-html/cell-13-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Supplemental Figure 6. Per sample switch error rates of 1698 samples which are members of family trios. Unlike most other measures of switch error rate used in this paper, switch error rate here was measured by mendelian consistency. Each point is the genome-wide switch error rate of one sample from either the GRCh38 or CHM13v2.0 panels. Points are colored by their 1KGP-defined superpopulation membership. The bimodal distribution of the GRCh38 error rates is due to the difference in phasing error between trio parents and probands – the GRCh38 1KGP panel underwent post-HMM error correction in probands only, while the CHM13v2.0 panel underwent pre-HMM trio-based phasing of both parents and probands.</p>
</section>
</section>
<section id="figure-4-switch-error-rate-vs-true-switch-error-rate" class="level2" data-number="1.3">
<h2 data-number="1.3" class="anchored" data-anchor-id="figure-4-switch-error-rate-vs-true-switch-error-rate"><span class="header-section-number">1.3</span> Figure 4: Switch error rate vs true switch error rate</h2>
<ul>
<li>graphic a la slide 12 from ASHG describing the difference between switch errors and flip errors, and why we only really care about true switch errors</li>
<li>That graphic will fill the space between the first and second parts of this figure</li>
<li>figure outlining per-sample switch error rate, flip rate, and true switch error rate</li>
<li>Per-contig true switch error rate</li>
<li>Supplementals:
<ul>
<li>Figures describing the relationship between true switch error rate and % of chromosome that is novel</li>
<li>Figures describing the relationship between true switch error rate and genotyping error rate (to show that flip errors are more closely tied to switch errors)</li>
</ul></li>
</ul>
<div id="cell-27" class="cell" data-execution_count="14">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co">### Figure 4a illustrator image</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> IPython.display <span class="im">import</span> Image</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>Image(filename<span class="op">=</span><span class="st">'figures/Fig 4a.png'</span>, width<span class="op">=</span><span class="dv">500</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="14">
<div>
<figure class="figure">
<p><img src="make_plots_files/figure-html/cell-14-output-1.png" class="img-fluid figure-img" width="500"></p>
</figure>
</div>
</div>
</div>
<div id="cell-28" class="cell" data-execution_count="15">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co">## Figure 4</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>figure_layout<span class="op">=</span><span class="st">"""</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="st">AAABB</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a><span class="st">AAABB</span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a><span class="st">AAABB</span></span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a><span class="st">CCCCC</span></span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a><span class="st">DDDDD</span></span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a><span class="st">"""</span></span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>fig, axes <span class="op">=</span> plt.subplot_mosaic(figure_layout, figsize<span class="op">=</span>(two_col, <span class="dv">5</span>))<span class="co">#layout='tight')</span></span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>axA <span class="op">=</span> axes[<span class="st">'A'</span>]</span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a>axB <span class="op">=</span> axes[<span class="st">'B'</span>]</span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a>axC <span class="op">=</span> axes[<span class="st">'C'</span>]</span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a>axD <span class="op">=</span> axes[<span class="st">'D'</span>]</span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a><span class="co">## Fig 4b</span></span>
<span id="cb23-17"><a href="#cb23-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-18"><a href="#cb23-18" aria-hidden="true" tabindex="-1"></a>error_order <span class="op">=</span> [<span class="st">'Total SER'</span>, <span class="st">'Flip rate'</span>,<span class="st">'True SER'</span>]</span>
<span id="cb23-19"><a href="#cb23-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-20"><a href="#cb23-20" aria-hidden="true" tabindex="-1"></a>compare_stats <span class="op">=</span> all_samples.loc[all_samples.ground_truth_data_source.isin((<span class="st">'HPRC_samples'</span>,<span class="st">'HGSVC_probands'</span>,<span class="st">'HGSVC_parents'</span>,<span class="st">'HGSVC_samples_nontrios_only'</span>)) <span class="op">&amp;</span></span>
<span id="cb23-21"><a href="#cb23-21" aria-hidden="true" tabindex="-1"></a>                                (all_samples.method_of_phasing<span class="op">==</span><span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb23-22"><a href="#cb23-22" aria-hidden="true" tabindex="-1"></a>                               ].rename(columns<span class="op">=</span>{<span class="st">'true_switch_error_rate'</span>:<span class="st">'True SER'</span>, <span class="st">'flip_error_rate'</span>:<span class="st">'Flip rate'</span>, <span class="st">'flip_and_switches_SER'</span>:<span class="st">'Total SER'</span>, <span class="st">'genome'</span>:<span class="st">'Genome'</span>}</span>
<span id="cb23-23"><a href="#cb23-23" aria-hidden="true" tabindex="-1"></a>                               ).melt(value_vars<span class="op">=</span>[<span class="st">'True SER'</span>,<span class="st">'Flip rate'</span>,<span class="st">'Total SER'</span>],</span>
<span id="cb23-24"><a href="#cb23-24" aria-hidden="true" tabindex="-1"></a>                                       id_vars<span class="op">=</span>[<span class="st">'sample_id'</span>, <span class="st">'Genome'</span>, <span class="st">'population'</span>, <span class="st">'superpopulation'</span>, <span class="st">'sex'</span>,<span class="st">'trio_phased'</span>, <span class="st">'method_of_phasing'</span>, <span class="st">'ground_truth_data_source'</span>],</span>
<span id="cb23-25"><a href="#cb23-25" aria-hidden="true" tabindex="-1"></a>                                      var_name<span class="op">=</span><span class="st">'Measure of Phasing Error'</span>,</span>
<span id="cb23-26"><a href="#cb23-26" aria-hidden="true" tabindex="-1"></a>                                      value_name<span class="op">=</span><span class="st">'Error Rate'</span>)</span>
<span id="cb23-27"><a href="#cb23-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-28"><a href="#cb23-28" aria-hidden="true" tabindex="-1"></a>dot_size<span class="op">=</span><span class="dv">3</span></span>
<span id="cb23-29"><a href="#cb23-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-30"><a href="#cb23-30" aria-hidden="true" tabindex="-1"></a>sns.violinplot(x<span class="op">=</span><span class="st">'Measure of Phasing Error'</span>, y<span class="op">=</span><span class="st">'Error Rate'</span>, hue<span class="op">=</span><span class="st">'Genome'</span>,</span>
<span id="cb23-31"><a href="#cb23-31" aria-hidden="true" tabindex="-1"></a>               order <span class="op">=</span> error_order, hue_order<span class="op">=</span>genome_order, palette<span class="op">=</span>genome_palette, linewidth<span class="op">=</span><span class="fl">0.5</span>,</span>
<span id="cb23-32"><a href="#cb23-32" aria-hidden="true" tabindex="-1"></a>               dodge<span class="op">=</span><span class="va">True</span>, split<span class="op">=</span><span class="va">True</span>, cut<span class="op">=</span><span class="fl">0.5</span>, inner<span class="op">=</span><span class="va">None</span>, density_norm<span class="op">=</span><span class="st">'count'</span>, fill<span class="op">=</span><span class="va">True</span>, alpha<span class="op">=</span><span class="fl">0.2</span>, bw_adjust<span class="op">=</span><span class="fl">.5</span>,</span>
<span id="cb23-33"><a href="#cb23-33" aria-hidden="true" tabindex="-1"></a>               data <span class="op">=</span> compare_stats, ax<span class="op">=</span>axB, legend<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb23-34"><a href="#cb23-34" aria-hidden="true" tabindex="-1"></a>sns.violinplot(x<span class="op">=</span><span class="st">'Measure of Phasing Error'</span>, y<span class="op">=</span><span class="st">'Error Rate'</span>, hue<span class="op">=</span><span class="st">'Genome'</span>,</span>
<span id="cb23-35"><a href="#cb23-35" aria-hidden="true" tabindex="-1"></a>               order <span class="op">=</span> error_order, hue_order<span class="op">=</span>genome_order, palette<span class="op">=</span>genome_palette, linewidth<span class="op">=</span><span class="fl">0.5</span>,</span>
<span id="cb23-36"><a href="#cb23-36" aria-hidden="true" tabindex="-1"></a>               dodge<span class="op">=</span><span class="va">True</span>, split<span class="op">=</span><span class="va">True</span>, cut<span class="op">=</span><span class="fl">0.5</span>, inner<span class="op">=</span><span class="va">None</span>, density_norm<span class="op">=</span><span class="st">'count'</span>, fill<span class="op">=</span><span class="va">False</span>, bw_adjust<span class="op">=</span><span class="fl">.5</span>,</span>
<span id="cb23-37"><a href="#cb23-37" aria-hidden="true" tabindex="-1"></a>               data <span class="op">=</span> compare_stats, ax<span class="op">=</span>axB, legend<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb23-38"><a href="#cb23-38" aria-hidden="true" tabindex="-1"></a>sns.stripplot(x<span class="op">=</span><span class="st">'Measure of Phasing Error'</span>, y<span class="op">=</span><span class="st">'Error Rate'</span>, hue<span class="op">=</span><span class="st">'Genome'</span>,</span>
<span id="cb23-39"><a href="#cb23-39" aria-hidden="true" tabindex="-1"></a>               order <span class="op">=</span> error_order, jitter<span class="op">=</span><span class="fl">0.25</span>, s<span class="op">=</span>dot_size,</span>
<span id="cb23-40"><a href="#cb23-40" aria-hidden="true" tabindex="-1"></a>               dodge<span class="op">=</span><span class="va">True</span>, alpha<span class="op">=</span><span class="fl">0.6</span>, linewidth<span class="op">=</span><span class="fl">0.1</span>, palette<span class="op">=</span>{<span class="st">'GRCh38'</span>:<span class="st">'dimgrey'</span>, <span class="st">'CHM13v2.0'</span>:<span class="st">'rebeccapurple'</span>},</span>
<span id="cb23-41"><a href="#cb23-41" aria-hidden="true" tabindex="-1"></a>               data <span class="op">=</span> compare_stats.loc[compare_stats.trio_phased<span class="op">==</span><span class="va">True</span>], ax<span class="op">=</span>axB, legend<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb23-42"><a href="#cb23-42" aria-hidden="true" tabindex="-1"></a>sns.stripplot(x<span class="op">=</span><span class="st">'Measure of Phasing Error'</span>, y<span class="op">=</span><span class="st">'Error Rate'</span>, hue<span class="op">=</span><span class="st">'Genome'</span>,</span>
<span id="cb23-43"><a href="#cb23-43" aria-hidden="true" tabindex="-1"></a>               order <span class="op">=</span> error_order, jitter<span class="op">=</span><span class="fl">0.25</span>, s<span class="op">=</span>dot_size,</span>
<span id="cb23-44"><a href="#cb23-44" aria-hidden="true" tabindex="-1"></a>               dodge<span class="op">=</span><span class="va">True</span>, alpha<span class="op">=</span><span class="fl">0.6</span>, linewidth<span class="op">=</span><span class="fl">0.5</span>, palette<span class="op">=</span>{<span class="st">'GRCh38'</span>:<span class="st">'slategrey'</span>, <span class="st">'CHM13v2.0'</span>:<span class="st">'plum'</span>},</span>
<span id="cb23-45"><a href="#cb23-45" aria-hidden="true" tabindex="-1"></a>               data <span class="op">=</span> compare_stats.loc[compare_stats.trio_phased<span class="op">==</span><span class="va">False</span>], ax<span class="op">=</span>axB, legend<span class="op">=</span><span class="va">True</span>, edgecolor<span class="op">=</span><span class="st">'black'</span>)</span>
<span id="cb23-46"><a href="#cb23-46" aria-hidden="true" tabindex="-1"></a>axB.yaxis.set_major_formatter(perc_formatter)</span>
<span id="cb23-47"><a href="#cb23-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-48"><a href="#cb23-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-49"><a href="#cb23-49" aria-hidden="true" tabindex="-1"></a><span class="co">#Fig 4c</span></span>
<span id="cb23-50"><a href="#cb23-50" aria-hidden="true" tabindex="-1"></a>sns.barplot(x<span class="op">=</span><span class="st">'chrom'</span>, y<span class="op">=</span><span class="st">'true_switch_error_rate'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, ax<span class="op">=</span>axC,</span>
<span id="cb23-51"><a href="#cb23-51" aria-hidden="true" tabindex="-1"></a>            order<span class="op">=</span>[x <span class="cf">for</span> x <span class="kw">in</span> chrom_order <span class="cf">if</span> <span class="st">'PAR'</span> <span class="kw">not</span> <span class="kw">in</span> x],</span>
<span id="cb23-52"><a href="#cb23-52" aria-hidden="true" tabindex="-1"></a>            hue_order<span class="op">=</span>genome_order, palette<span class="op">=</span>genome_palette,</span>
<span id="cb23-53"><a href="#cb23-53" aria-hidden="true" tabindex="-1"></a>            err_kws<span class="op">=</span>{<span class="st">'linewidth'</span>:<span class="fl">0.5</span>}, capsize<span class="op">=</span><span class="fl">0.3</span>,</span>
<span id="cb23-54"><a href="#cb23-54" aria-hidden="true" tabindex="-1"></a>            data<span class="op">=</span>full_panel_HPRC_compared.loc[(full_panel_HPRC_compared.ground_truth_data_source<span class="op">==</span><span class="st">'HPRC_samples'</span>)])</span>
<span id="cb23-55"><a href="#cb23-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-56"><a href="#cb23-56" aria-hidden="true" tabindex="-1"></a>sns.barplot(x<span class="op">=</span><span class="st">'chrom'</span>, y<span class="op">=</span><span class="st">'true_switch_error_rate'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, ax<span class="op">=</span>axD,</span>
<span id="cb23-57"><a href="#cb23-57" aria-hidden="true" tabindex="-1"></a>            order<span class="op">=</span>[x <span class="cf">for</span> x <span class="kw">in</span> chrom_order <span class="cf">if</span> <span class="st">'PAR'</span> <span class="kw">not</span> <span class="kw">in</span> x],</span>
<span id="cb23-58"><a href="#cb23-58" aria-hidden="true" tabindex="-1"></a>            hue_order<span class="op">=</span>genome_order, palette<span class="op">=</span>genome_palette,</span>
<span id="cb23-59"><a href="#cb23-59" aria-hidden="true" tabindex="-1"></a>            err_kws<span class="op">=</span>{<span class="st">'linewidth'</span>:<span class="fl">0.5</span>}, capsize<span class="op">=</span><span class="fl">0.3</span>,</span>
<span id="cb23-60"><a href="#cb23-60" aria-hidden="true" tabindex="-1"></a>            data<span class="op">=</span>full_panel_HPRC_compared.loc[(full_panel_HPRC_compared.ground_truth_data_source<span class="op">==</span><span class="st">'HGSVC_samples_nontrios_only'</span>)])</span>
<span id="cb23-61"><a href="#cb23-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-62"><a href="#cb23-62" aria-hidden="true" tabindex="-1"></a>axC.yaxis.set_major_locator(MultipleLocator(<span class="fl">0.15</span>))</span>
<span id="cb23-63"><a href="#cb23-63" aria-hidden="true" tabindex="-1"></a>axD.yaxis.set_major_locator(MultipleLocator(<span class="fl">0.15</span>))</span>
<span id="cb23-64"><a href="#cb23-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-65"><a href="#cb23-65" aria-hidden="true" tabindex="-1"></a>axC.set_xlabel(<span class="va">None</span>)</span>
<span id="cb23-66"><a href="#cb23-66" aria-hidden="true" tabindex="-1"></a>axD.set_xlabel(<span class="st">'Chromosome'</span>)</span>
<span id="cb23-67"><a href="#cb23-67" aria-hidden="true" tabindex="-1"></a>axC.set_title(<span class="st">'HPRC Mendelian Phased'</span>, pad<span class="op">=</span><span class="dv">6</span>)</span>
<span id="cb23-68"><a href="#cb23-68" aria-hidden="true" tabindex="-1"></a>axD.set_title(<span class="st">'HGSVC Statistically Phased'</span>, pad<span class="op">=</span><span class="dv">6</span>)</span>
<span id="cb23-69"><a href="#cb23-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-70"><a href="#cb23-70" aria-hidden="true" tabindex="-1"></a>handles, labels <span class="op">=</span>axB.get_legend_handles_labels()</span>
<span id="cb23-71"><a href="#cb23-71" aria-hidden="true" tabindex="-1"></a>new_handles<span class="op">=</span><span class="bu">list</span>()</span>
<span id="cb23-72"><a href="#cb23-72" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, (handle, label) <span class="kw">in</span> <span class="bu">enumerate</span>(<span class="bu">zip</span>(handles, labels)):</span>
<span id="cb23-73"><a href="#cb23-73" aria-hidden="true" tabindex="-1"></a>    handle.set_alpha(<span class="fl">.8</span>)</span>
<span id="cb23-74"><a href="#cb23-74" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="op">==</span> <span class="dv">0</span>:</span>
<span id="cb23-75"><a href="#cb23-75" aria-hidden="true" tabindex="-1"></a>        handle.set_facecolor(<span class="st">'#E0E0E0'</span>)</span>
<span id="cb23-76"><a href="#cb23-76" aria-hidden="true" tabindex="-1"></a>        handle.set_edgecolor(<span class="st">'#666666'</span>)</span>
<span id="cb23-77"><a href="#cb23-77" aria-hidden="true" tabindex="-1"></a>        handle.set_alpha(<span class="dv">1</span>)</span>
<span id="cb23-78"><a href="#cb23-78" aria-hidden="true" tabindex="-1"></a>    <span class="cf">elif</span> i <span class="op">==</span> <span class="dv">1</span>:</span>
<span id="cb23-79"><a href="#cb23-79" aria-hidden="true" tabindex="-1"></a>        handle.set_facecolor(<span class="st">'#E9E3F0'</span>)</span>
<span id="cb23-80"><a href="#cb23-80" aria-hidden="true" tabindex="-1"></a>        handle.set_edgecolor(<span class="st">'#9467BD'</span>)</span>
<span id="cb23-81"><a href="#cb23-81" aria-hidden="true" tabindex="-1"></a>        handle.set_alpha(<span class="dv">1</span>)</span>
<span id="cb23-82"><a href="#cb23-82" aria-hidden="true" tabindex="-1"></a>    new_handles.append(handle)</span>
<span id="cb23-83"><a href="#cb23-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-84"><a href="#cb23-84" aria-hidden="true" tabindex="-1"></a>labels <span class="op">=</span> labels[:<span class="dv">2</span>] <span class="op">+</span> [l <span class="op">+</span> <span class="st">' (Mendelian phased)'</span> <span class="cf">for</span> l <span class="kw">in</span> labels[<span class="dv">2</span>:<span class="dv">4</span>]] <span class="op">+</span> [l<span class="op">+</span> <span class="st">' (Statistically phased)'</span> <span class="cf">for</span> l <span class="kw">in</span> labels[<span class="dv">4</span>:<span class="dv">6</span>]]</span>
<span id="cb23-85"><a href="#cb23-85" aria-hidden="true" tabindex="-1"></a>axB.legend(new_handles,labels)</span>
<span id="cb23-86"><a href="#cb23-86" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-87"><a href="#cb23-87" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> ax <span class="kw">in</span> [axC, axD]:</span>
<span id="cb23-88"><a href="#cb23-88" aria-hidden="true" tabindex="-1"></a>    ax.set_ylabel(<span class="st">'True Switch Error Rate'</span>)</span>
<span id="cb23-89"><a href="#cb23-89" aria-hidden="true" tabindex="-1"></a>    legend <span class="op">=</span> ax.legend()</span>
<span id="cb23-90"><a href="#cb23-90" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> text <span class="kw">in</span> legend.get_texts():</span>
<span id="cb23-91"><a href="#cb23-91" aria-hidden="true" tabindex="-1"></a>        t <span class="op">=</span> text.get_text()</span>
<span id="cb23-92"><a href="#cb23-92" aria-hidden="true" tabindex="-1"></a>        new_t <span class="op">=</span> update_legend_values.get(t,t)</span>
<span id="cb23-93"><a href="#cb23-93" aria-hidden="true" tabindex="-1"></a>        text.set_text(new_t)</span>
<span id="cb23-94"><a href="#cb23-94" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-95"><a href="#cb23-95" aria-hidden="true" tabindex="-1"></a>    sns.move_legend(ax, <span class="st">"upper left"</span>)</span>
<span id="cb23-96"><a href="#cb23-96" aria-hidden="true" tabindex="-1"></a>    ax.yaxis.set_major_formatter(PercentFormatter(decimals<span class="op">=</span><span class="dv">2</span>))</span>
<span id="cb23-97"><a href="#cb23-97" aria-hidden="true" tabindex="-1"></a>    ax.grid(axis<span class="op">=</span><span class="st">'y'</span>, alpha<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="cb23-98"><a href="#cb23-98" aria-hidden="true" tabindex="-1"></a>    ax.set_axisbelow(<span class="va">True</span>)</span>
<span id="cb23-99"><a href="#cb23-99" aria-hidden="true" tabindex="-1"></a>    legend.set_title(<span class="st">'Genome'</span>)</span>
<span id="cb23-100"><a href="#cb23-100" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-101"><a href="#cb23-101" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, ax <span class="kw">in</span> <span class="bu">enumerate</span>(axes.values()):</span>
<span id="cb23-102"><a href="#cb23-102" aria-hidden="true" tabindex="-1"></a>    add_letter_to_ax(ax, i)</span>
<span id="cb23-103"><a href="#cb23-103" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-104"><a href="#cb23-104" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-105"><a href="#cb23-105" aria-hidden="true" tabindex="-1"></a><span class="co">## Fig 4a - graphic</span></span>
<span id="cb23-106"><a href="#cb23-106" aria-hidden="true" tabindex="-1"></a>sns.despine(ax<span class="op">=</span>axA, top<span class="op">=</span><span class="va">True</span>, right<span class="op">=</span><span class="va">True</span>, left<span class="op">=</span><span class="va">True</span>, bottom<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb23-107"><a href="#cb23-107" aria-hidden="true" tabindex="-1"></a>axA.set_yticks([])</span>
<span id="cb23-108"><a href="#cb23-108" aria-hidden="true" tabindex="-1"></a>axA.set_xticks([])</span>
<span id="cb23-109"><a href="#cb23-109" aria-hidden="true" tabindex="-1"></a>fig</span>
<span id="cb23-110"><a href="#cb23-110" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/Figure 4.png'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span>
<span id="cb23-111"><a href="#cb23-111" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/Figure 4.svg'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span>
<span id="cb23-112"><a href="#cb23-112" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/Figure 4.eps'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span>
<span id="cb23-113"><a href="#cb23-113" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/Figure 4.pdf'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>The PostScript backend does not support transparency; partially transparent artists will be rendered opaque.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="make_plots_files/figure-html/cell-15-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Figure 4: Flip error rates and true switch error rates. a) Throughout this study, we have compared our panel’s phasing with ground truth reference assemblies to identify switch errors, defined as sites at which the maternal and paternal haplotypes switch strands. However, regions which are difficult to sequence can often have genotyping errors that result in flip errors, defined as two consecutive switch errors. Unlike true switch errors, flip errors do not impact the phasing accuracy of nearby variants. Eight switch errors are in the illustrated haplotypes, but only two of these are true switch errors - the other six switch errors are part of three flip errors. b) The switch error rate, flip error rate, and true switch error rate of all 100 samples with assembly-based ground truths. Data from the 19 HGSVC individuals that were phased without mendelian error correction are highlighted. Data points are plotted on top of a violin plot showing the overall distribution of error rates. c) Whole genome and per-chromosome true switch error rates from 39 HPRC-assembled samples. d) Whole genome and per-chromosome true switch error rates from 19 HGSVC-assembled samples phased without mendelian error correction.</p>
<div id="cell-30" class="cell" data-execution_count="16">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co">## Supplementals:</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="co">## Supplemental 6 - association btwn genotype accuracy and various switch error rates</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>hprc_evaluated_no_PAR <span class="op">=</span> full_panel_HPRC_compared.loc[full_panel_HPRC_compared.chrom.<span class="bu">str</span>.contains(<span class="st">'chr'</span>) <span class="op">&amp;</span> <span class="op">~</span>full_panel_HPRC_compared.chrom.<span class="bu">str</span>.contains(<span class="st">'Whole'</span>)].dropna(subset<span class="op">=</span>[<span class="st">'true_switch_error_rate'</span>])</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>hprc_evaluated_no_PAR[[<span class="st">'log_switch_error_rate'</span>,<span class="st">'log_flip_error_rate'</span>,<span class="st">'log_true_switch_error_rate'</span>,<span class="st">'log_gt_error_rate'</span>]] <span class="op">=</span> np.log10(hprc_evaluated_no_PAR[[<span class="st">'switch_error_rate'</span>,<span class="st">'flip_error_rate'</span>,<span class="st">'true_switch_error_rate'</span>,<span class="st">'gt_error_rate'</span>]])</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>hprc_evaluated_no_PAR.loc[hprc_evaluated_no_PAR.log_true_switch_error_rate <span class="op">==</span> <span class="op">-</span>np.inf,<span class="st">'log_true_switch_error_rate'</span>] <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> normalize_df(df):</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> ((df<span class="op">-</span>df.<span class="bu">min</span>())<span class="op">/</span>(df.<span class="bu">max</span>()<span class="op">-</span>df.<span class="bu">min</span>())).values</span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>tmp<span class="op">=</span>hprc_evaluated_no_PAR.melt(id_vars<span class="op">=</span>[<span class="st">'sample_id'</span>,<span class="st">'log_gt_error_rate'</span>,<span class="st">'trio_phased'</span>, <span class="st">'genome'</span>], value_vars<span class="op">=</span>[<span class="st">'log_switch_error_rate'</span>,<span class="st">'log_flip_error_rate'</span>,<span class="st">'log_true_switch_error_rate'</span>],var_name<span class="op">=</span><span class="st">'error_kind'</span>,value_name<span class="op">=</span><span class="st">'log_error_rate'</span>)</span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>tmp <span class="op">=</span> tmp.rename(columns<span class="op">=</span>{<span class="st">'genome'</span>:<span class="st">'Panel Genome'</span>, <span class="st">'trio_phased'</span>:<span class="st">'Mendelian Phased'</span>,<span class="st">'log_gt_error_rate'</span>:<span class="st">'log(Genotype Error Rate)'</span>, <span class="st">'error_kind'</span>:<span class="st">'Measurement of panel error'</span>, <span class="st">'log_error_rate'</span>:<span class="st">'log(Error Rate)'</span>})</span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a>tmp <span class="op">=</span> tmp.replace(<span class="st">'log_switch_error_rate'</span>, <span class="st">'log(Switch Error Rate)'</span>).replace(<span class="st">'log_flip_error_rate'</span>, <span class="st">'log(Flip Error Rate)'</span>).replace(<span class="st">'log_true_switch_error_rate'</span>, <span class="st">'log(True Switch Error Rate)'</span>)</span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a>height_width_ratio <span class="op">=</span> <span class="fl">0.5</span><span class="op">;</span> aspect_ratio <span class="op">=</span> <span class="fl">1.15</span> <span class="co">#(produces a two_col by 5 inch figure, found through trial and error)</span></span>
<span id="cb25-14"><a href="#cb25-14" aria-hidden="true" tabindex="-1"></a>g <span class="op">=</span> sns.lmplot(data<span class="op">=</span>tmp, col<span class="op">=</span><span class="st">'Mendelian Phased'</span>, row<span class="op">=</span><span class="st">'Panel Genome'</span>, col_order<span class="op">=</span>[<span class="va">True</span>, <span class="va">False</span>],</span>
<span id="cb25-15"><a href="#cb25-15" aria-hidden="true" tabindex="-1"></a>           x<span class="op">=</span><span class="st">'log(Genotype Error Rate)'</span>,y<span class="op">=</span><span class="st">'log(Error Rate)'</span>,hue<span class="op">=</span><span class="st">'Measurement of panel error'</span>, scatter_kws<span class="op">=</span>{<span class="st">'alpha'</span>:<span class="fl">0.1</span>}, units<span class="op">=</span><span class="st">'sample_id'</span>, height<span class="op">=</span><span class="dv">5</span><span class="op">*</span>height_width_ratio, aspect<span class="op">=</span>aspect_ratio)</span>
<span id="cb25-16"><a href="#cb25-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-17"><a href="#cb25-17" aria-hidden="true" tabindex="-1"></a>sns.move_legend(g, loc<span class="op">=</span><span class="st">'upper center'</span>, ncols<span class="op">=</span><span class="dv">3</span>, bbox_to_anchor<span class="op">=</span>(<span class="fl">.46</span>, <span class="dv">0</span>))</span>
<span id="cb25-18"><a href="#cb25-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-19"><a href="#cb25-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-20"><a href="#cb25-20" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, ax <span class="kw">in</span> <span class="bu">enumerate</span>(<span class="bu">list</span>(itertools.chain.from_iterable(g.axes))):</span>
<span id="cb25-21"><a href="#cb25-21" aria-hidden="true" tabindex="-1"></a>    add_letter_to_ax(ax, i, points_offset<span class="op">=</span>(<span class="op">-</span><span class="dv">25</span>, <span class="dv">0</span>))</span>
<span id="cb25-22"><a href="#cb25-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-23"><a href="#cb25-23" aria-hidden="true" tabindex="-1"></a>g.tight_layout()</span>
<span id="cb25-24"><a href="#cb25-24" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> lh <span class="kw">in</span> g._legend.legend_handles:</span>
<span id="cb25-25"><a href="#cb25-25" aria-hidden="true" tabindex="-1"></a>    lh.set_alpha(<span class="dv">1</span>)</span>
<span id="cb25-26"><a href="#cb25-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-27"><a href="#cb25-27" aria-hidden="true" tabindex="-1"></a>g.savefig(<span class="st">'figures/supplemental/Supplemental 7.svg'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span>
<span id="cb25-28"><a href="#cb25-28" aria-hidden="true" tabindex="-1"></a>g.savefig(<span class="st">'figures/supplemental/Supplemental 7.png'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span>
<span id="cb25-29"><a href="#cb25-29" aria-hidden="true" tabindex="-1"></a>g.savefig(<span class="st">'figures/supplemental/Supplemental 7.svg'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span>
<span id="cb25-30"><a href="#cb25-30" aria-hidden="true" tabindex="-1"></a>g.savefig(<span class="st">'figures/supplemental/Supplemental 7.png'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>/home/lalli/.linuxbrew/opt/python@3.11/lib/python3.11/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: divide by zero encountered in log10
  result = func(self.values, **kwargs)
/home/lalli/.linuxbrew/opt/python@3.11/lib/python3.11/site-packages/seaborn/axisgrid.py:123: UserWarning: The figure layout has changed to tight
  self._figure.tight_layout(*args, **kwargs)</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="make_plots_files/figure-html/cell-16-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Supplemental Figure 7. Association between flip error rate, switch error rate, true switch error rate, and genotyping error rate. The flip error rate, switch error rate, and true switch error rate was calculated for all chromosomes and all samples for which assembly-based ground truths existed. Genotype error rates were calculated for each chromosome in each sample. Log10 values of each error rate were taken to transform each rate to a gaussian distribution. Linear regression was sequentially performed with the formula log10(Error_Rate) ~ log10(genotype_error_rate), and r2 correlation values were calculated. a) Per chromosome per sample error rates in the GRCh38 haplotype panel for Mendelian error-corrected samples assemblies generated by the HPRC or the HGSVC. b) Per chromosome per sample error rates in the GRCh38 haplotype panel for non-trio samples with HGSVC assemblies that did not undergo mendelian error correction. c) Per chromosome per sample error rates in the CHM13v2.0 haplotype panel for Mendelian error-corrected samples assemblies generated by the HPRC or the HGSVC. d) Per chromosome per sample error rates in the CHM13v2.0 haplotype panel for non-trio samples with HGSVC assemblies that did not undergo mendelian error correction. All regression lines are plotted with a +/- 95% confidence interval. </p>
<div id="cell-32" class="cell" data-execution_count="17">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>grch38_trio_phased_corr <span class="op">=</span> (hprc_evaluated_no_PAR.loc[(hprc_evaluated_no_PAR.trio_phased)<span class="op">&amp;</span>(hprc_evaluated_no_PAR.genome<span class="op">==</span><span class="st">'GRCh38'</span>),[<span class="st">'log_switch_error_rate'</span>,<span class="st">'log_flip_error_rate'</span>,<span class="st">'log_true_switch_error_rate'</span>,<span class="st">'log_gt_error_rate'</span>]].corr(method<span class="op">=</span><span class="st">'pearson'</span>)<span class="op">**</span><span class="dv">2</span>)[[<span class="st">'log_gt_error_rate'</span>]].assign(genome<span class="op">=</span><span class="st">'GRCh38'</span>, mendelian_phased<span class="op">=</span><span class="va">True</span>).iloc[:<span class="op">-</span><span class="dv">1</span>]</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>grch38_stat_phased_corr <span class="op">=</span> (hprc_evaluated_no_PAR.loc[(<span class="op">~</span>hprc_evaluated_no_PAR.trio_phased)<span class="op">&amp;</span>(hprc_evaluated_no_PAR.genome<span class="op">==</span><span class="st">'GRCh38'</span>),[<span class="st">'log_switch_error_rate'</span>,<span class="st">'log_flip_error_rate'</span>,<span class="st">'log_true_switch_error_rate'</span>,<span class="st">'log_gt_error_rate'</span>]].corr(method<span class="op">=</span><span class="st">'pearson'</span>)<span class="op">**</span><span class="dv">2</span>)[[<span class="st">'log_gt_error_rate'</span>]].assign(genome<span class="op">=</span><span class="st">'GRCh38'</span>, mendelian_phased<span class="op">=</span><span class="va">False</span>).iloc[:<span class="op">-</span><span class="dv">1</span>]</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>chm13_trio_phased_corr <span class="op">=</span> (hprc_evaluated_no_PAR.loc[(hprc_evaluated_no_PAR.trio_phased)<span class="op">&amp;</span>(hprc_evaluated_no_PAR.genome<span class="op">==</span><span class="st">'CHM13v2.0'</span>),[<span class="st">'log_switch_error_rate'</span>,<span class="st">'log_flip_error_rate'</span>,<span class="st">'log_true_switch_error_rate'</span>,<span class="st">'log_gt_error_rate'</span>]].corr(method<span class="op">=</span><span class="st">'pearson'</span>)<span class="op">**</span><span class="dv">2</span>)[[<span class="st">'log_gt_error_rate'</span>]].assign(genome<span class="op">=</span><span class="st">'CHM13v2.0'</span>, mendelian_phased<span class="op">=</span><span class="va">True</span>).iloc[:<span class="op">-</span><span class="dv">1</span>]</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>chm13_stat_phased_corr <span class="op">=</span> (hprc_evaluated_no_PAR.loc[(<span class="op">~</span>hprc_evaluated_no_PAR.trio_phased)<span class="op">&amp;</span>(hprc_evaluated_no_PAR.genome<span class="op">==</span><span class="st">'CHM13v2.0'</span>),[<span class="st">'log_switch_error_rate'</span>,<span class="st">'log_flip_error_rate'</span>,<span class="st">'log_true_switch_error_rate'</span>,<span class="st">'log_gt_error_rate'</span>]].corr(method<span class="op">=</span><span class="st">'pearson'</span>)<span class="op">**</span><span class="dv">2</span>)[[<span class="st">'log_gt_error_rate'</span>]].assign(genome<span class="op">=</span><span class="st">'CHM13v2.0'</span>, mendelian_phased<span class="op">=</span><span class="va">False</span>).iloc[:<span class="op">-</span><span class="dv">1</span>]</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>chm13_phased_corr <span class="op">=</span> pd.concat((chm13_stat_phased_corr, chm13_trio_phased_corr)).drop(columns<span class="op">=</span><span class="st">'genome'</span>).reset_index().rename(columns<span class="op">=</span>{<span class="st">'index'</span>:<span class="st">'error_rate'</span>}).set_index([<span class="st">'error_rate'</span>,<span class="st">'mendelian_phased'</span>])</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>grch38_phased_corr <span class="op">=</span> pd.concat((grch38_stat_phased_corr, grch38_trio_phased_corr)).drop(columns<span class="op">=</span><span class="st">'genome'</span>).reset_index().rename(columns<span class="op">=</span>{<span class="st">'index'</span>:<span class="st">'error_rate'</span>}).set_index([<span class="st">'error_rate'</span>,<span class="st">'mendelian_phased'</span>])</span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>error_rate_corr<span class="op">=</span>chm13_phased_corr.join(grch38_phased_corr, lsuffix<span class="op">=</span><span class="st">"_chm13"</span>, rsuffix<span class="op">=</span><span class="st">'_grch38'</span>).reset_index().rename(columns<span class="op">=</span>{<span class="st">'error_rate'</span>:<span class="st">'Log error rate'</span>,<span class="st">'mendelian_phased'</span>:<span class="st">'Mendelian-based error correction'</span>,<span class="st">'log_gt_error_rate_chm13'</span>:<span class="st">'CHM13v2.0'</span>,<span class="st">'log_gt_error_rate_grch38'</span>:<span class="st">'GRCh38'</span>})</span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a>error_rate_corr <span class="op">=</span> error_rate_corr.replace(<span class="st">'log_switch_error_rate'</span>,<span class="st">'Log(Switch Error Rate)'</span>).replace(<span class="st">'log_flip_error_rate'</span>,<span class="st">'Log(Flip Error Rate)'</span>).replace(<span class="st">'log_true_switch_error_rate'</span>,<span class="st">'Log(True Switch Error Rate)'</span>).set_index([<span class="st">"Log error rate"</span>,<span class="st">'Mendelian-based error correction'</span>])</span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a>error_rate_corr.reset_index().to_csv(<span class="st">'figures/correlation_error_rates_and_gt_error_rate.csv'</span>, index<span class="op">=</span><span class="va">False</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-33" class="cell" data-execution_count="18">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co">## Supplemental 7: Association with true switch error rate and % of chromosome that's novel</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> plot_regression(x, y, hue, data, <span class="op">**</span>kws):</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>    fig <span class="op">=</span> sns.lmplot(x<span class="op">=</span>x,y<span class="op">=</span>y, hue<span class="op">=</span><span class="st">'genome'</span>, data <span class="op">=</span> data, robust<span class="op">=</span><span class="va">False</span>, palette<span class="op">=</span>default_style[<span class="st">'palette'</span>], aspect<span class="op">=</span><span class="fl">1.5</span>, height<span class="op">=</span><span class="dv">4</span>,</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>                     facet_kws<span class="op">=</span>{<span class="st">'legend_out'</span>:<span class="va">False</span>}, units<span class="op">=</span><span class="st">'sample_id'</span>, <span class="op">**</span>kws)<span class="co">#, x_jitter=0.1)</span></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>    ax <span class="op">=</span> fig.ax</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>    grch <span class="op">=</span> data.loc[data.genome<span class="op">==</span><span class="st">'GRCh38'</span>]</span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>    t2t <span class="op">=</span> data.loc[data.genome<span class="op">==</span><span class="st">'CHM13v2.0'</span>]</span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a>    s1, i1, r1, p1, err1 <span class="op">=</span> scipy.stats.linregress(grch[x], grch[y])</span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a>    s2, i2, r2, p2, err2 <span class="op">=</span> scipy.stats.linregress(t2t[x], t2t[y])</span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a>    s3, i3, r3, p3, err3 <span class="op">=</span> scipy.stats.linregress(grch[x], grch[y].values<span class="op">-</span>t2t[y].values)</span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a>    <span class="co"># r2, p2 = stats.pearsonr(t2t[x], t2t[y])</span></span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a>    <span class="co"># r3, p3 = stats.pearsonr(grch[x], grch[y].values-t2t[y].values)</span></span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span> (<span class="st">'GRCh38: '</span> <span class="op">+</span> <span class="st">"$</span><span class="er">\</span><span class="st">mathregular{r^2}$"</span> <span class="op">+</span> <span class="ss">f"=</span><span class="sc">{</span>r1<span class="sc">:.2f}</span><span class="ss">, p=</span><span class="sc">{</span>p1<span class="sc">:.2g}</span><span class="ss">, slope=</span><span class="sc">{</span>s1<span class="sc">:.1e}</span><span class="ss">"</span>)</span>
<span id="cb28-14"><a href="#cb28-14" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span> (<span class="st">'CHM13v2.0: '</span> <span class="op">+</span> <span class="st">"$</span><span class="er">\</span><span class="st">mathregular{r^2}$"</span> <span class="op">+</span> <span class="ss">f"=</span><span class="sc">{</span>r2<span class="sc">:.2f}</span><span class="ss">, p=</span><span class="sc">{</span>p2<span class="sc">:.2g}</span><span class="ss">, slope=</span><span class="sc">{</span>s2<span class="sc">:.1e}</span><span class="ss">"</span>)</span>
<span id="cb28-15"><a href="#cb28-15" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span> (<span class="st">'CHM13v2.0 - GRCh38: '</span> <span class="op">+</span> <span class="st">"$</span><span class="er">\</span><span class="st">mathregular{r^2}$"</span> <span class="op">+</span> <span class="ss">f"=</span><span class="sc">{</span>r3<span class="sc">:.2f}</span><span class="ss">, p=</span><span class="sc">{</span>p3<span class="sc">:.2g}</span><span class="ss">, slope=</span><span class="sc">{</span>s3<span class="sc">:.1e}</span><span class="ss">"</span>)</span>
<span id="cb28-16"><a href="#cb28-16" aria-hidden="true" tabindex="-1"></a>    ax.text(<span class="fl">.05</span>, <span class="fl">.96</span>, <span class="st">'GRCh38'</span>,</span>
<span id="cb28-17"><a href="#cb28-17" aria-hidden="true" tabindex="-1"></a>            transform<span class="op">=</span>ax.transAxes)</span>
<span id="cb28-18"><a href="#cb28-18" aria-hidden="true" tabindex="-1"></a>    ax.text(<span class="fl">.05</span>, <span class="fl">.92</span>, <span class="st">"$</span><span class="er">\</span><span class="st">mathregular{r^2}$"</span> <span class="op">+</span> <span class="ss">f"=</span><span class="sc">{</span>r1<span class="sc">:.2f}</span><span class="ss">, p=</span><span class="sc">{</span>p1<span class="sc">:.2g}</span><span class="ss">, slope=</span><span class="sc">{</span>s1<span class="op">*</span><span class="dv">100</span><span class="sc">:.3f}</span><span class="ss">%"</span>,</span>
<span id="cb28-19"><a href="#cb28-19" aria-hidden="true" tabindex="-1"></a>            transform<span class="op">=</span>ax.transAxes)</span>
<span id="cb28-20"><a href="#cb28-20" aria-hidden="true" tabindex="-1"></a>    ax.text(<span class="fl">.05</span>, <span class="fl">.85</span>, <span class="st">'CHM13v2.0'</span>,</span>
<span id="cb28-21"><a href="#cb28-21" aria-hidden="true" tabindex="-1"></a>            transform<span class="op">=</span>ax.transAxes)</span>
<span id="cb28-22"><a href="#cb28-22" aria-hidden="true" tabindex="-1"></a>    ax.text(<span class="fl">.05</span>, <span class="fl">.81</span>, <span class="st">"$</span><span class="er">\</span><span class="st">mathregular{r^2}$"</span> <span class="op">+</span> <span class="ss">f"=</span><span class="sc">{</span>r2<span class="sc">:.2f}</span><span class="ss">, p=</span><span class="sc">{</span>p2<span class="sc">:.2g}</span><span class="ss">, slope=</span><span class="sc">{</span>s2<span class="op">*</span><span class="dv">100</span><span class="sc">:.3f}</span><span class="ss">%"</span>,</span>
<span id="cb28-23"><a href="#cb28-23" aria-hidden="true" tabindex="-1"></a>            transform<span class="op">=</span>ax.transAxes)</span>
<span id="cb28-24"><a href="#cb28-24" aria-hidden="true" tabindex="-1"></a>    ax.text(<span class="fl">.05</span>, <span class="fl">.73</span>, <span class="st">'CHM13v2.0 - GRCh38'</span>,</span>
<span id="cb28-25"><a href="#cb28-25" aria-hidden="true" tabindex="-1"></a>            transform<span class="op">=</span>ax.transAxes)</span>
<span id="cb28-26"><a href="#cb28-26" aria-hidden="true" tabindex="-1"></a>    ax.text(<span class="fl">.05</span>, <span class="fl">.69</span>, <span class="st">"$</span><span class="er">\</span><span class="st">mathregular{r^2}$"</span> <span class="op">+</span> <span class="ss">f"=</span><span class="sc">{</span>r3<span class="sc">:.2f}</span><span class="ss">, p=</span><span class="sc">{</span>p3<span class="sc">:.2g}</span><span class="ss">, slope=</span><span class="sc">{</span>s3<span class="op">*</span><span class="dv">100</span><span class="sc">:.3f}</span><span class="ss">%"</span>,</span>
<span id="cb28-27"><a href="#cb28-27" aria-hidden="true" tabindex="-1"></a>            transform<span class="op">=</span>ax.transAxes)</span>
<span id="cb28-28"><a href="#cb28-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-29"><a href="#cb28-29" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> fig, ax</span>
<span id="cb28-30"><a href="#cb28-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-31"><a href="#cb28-31" aria-hidden="true" tabindex="-1"></a>fg, ax <span class="op">=</span> plot_regression(x<span class="op">=</span><span class="st">'percent_new'</span>,y<span class="op">=</span><span class="st">'true_switch_error_rate'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, data <span class="op">=</span> all_chroms.loc[(all_chroms.method_of_phasing<span class="op">==</span><span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb28-32"><a href="#cb28-32" aria-hidden="true" tabindex="-1"></a>                                                                                                    <span class="op">&amp;</span>(all_chroms.ground_truth_data_source <span class="op">==</span> <span class="st">'HPRC_samples'</span>) <span class="op">&amp;</span></span>
<span id="cb28-33"><a href="#cb28-33" aria-hidden="true" tabindex="-1"></a>                                                                                                    <span class="op">~</span>(all_chroms.chrom.isin((<span class="st">'chrX'</span>,<span class="st">'PAR1'</span>,<span class="st">'PAR2'</span>)))],</span>
<span id="cb28-34"><a href="#cb28-34" aria-hidden="true" tabindex="-1"></a>                          scatter_kws<span class="op">=</span>{<span class="st">'linewidths'</span>:<span class="fl">0.1</span>})</span>
<span id="cb28-35"><a href="#cb28-35" aria-hidden="true" tabindex="-1"></a><span class="co"># ax.set_ylim(0, .16)</span></span>
<span id="cb28-36"><a href="#cb28-36" aria-hidden="true" tabindex="-1"></a>ax.set_ylabel(<span class="st">'True Switch Error Rate'</span>)</span>
<span id="cb28-37"><a href="#cb28-37" aria-hidden="true" tabindex="-1"></a>ax.set_xlabel(<span class="st">'</span><span class="sc">% o</span><span class="st">f CHM13v2.0 chromosome</span><span class="ch">\n</span><span class="st">nonsyntenic with GRCh38'</span>)</span>
<span id="cb28-38"><a href="#cb28-38" aria-hidden="true" tabindex="-1"></a>ax.xaxis.set_major_formatter(perc_formatter)</span>
<span id="cb28-39"><a href="#cb28-39" aria-hidden="true" tabindex="-1"></a>ax.yaxis.set_major_formatter(perc_formatter)</span>
<span id="cb28-40"><a href="#cb28-40" aria-hidden="true" tabindex="-1"></a>ax.grid(axis<span class="op">=</span><span class="st">'y'</span>, alpha<span class="op">=</span><span class="fl">.5</span>)</span>
<span id="cb28-41"><a href="#cb28-41" aria-hidden="true" tabindex="-1"></a>fg.legend.set_title(<span class="st">'Genome'</span>)</span>
<span id="cb28-42"><a href="#cb28-42" aria-hidden="true" tabindex="-1"></a>ax.set_axisbelow(<span class="va">True</span>)</span>
<span id="cb28-43"><a href="#cb28-43" aria-hidden="true" tabindex="-1"></a>sns.despine(ax<span class="op">=</span>ax, top<span class="op">=</span><span class="va">False</span>, right<span class="op">=</span><span class="va">False</span>, left<span class="op">=</span><span class="va">False</span>, bottom<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb28-44"><a href="#cb28-44" aria-hidden="true" tabindex="-1"></a>sns.move_legend(fg, <span class="st">"upper left"</span>, bbox_to_anchor<span class="op">=</span>(<span class="fl">.8</span>,<span class="fl">0.98</span>))</span>
<span id="cb28-45"><a href="#cb28-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-46"><a href="#cb28-46" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/supplemental/Supplemental 8.svg'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span>
<span id="cb28-47"><a href="#cb28-47" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/supplemental/Supplemental 8.png'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span>
<span id="cb28-48"><a href="#cb28-48" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/supplemental/Supplemental 8.eps'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span>
<span id="cb28-49"><a href="#cb28-49" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/supplemental/Supplemental 8.pdf'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>/home/lalli/.linuxbrew/opt/python@3.11/lib/python3.11/site-packages/seaborn/axisgrid.py:123: UserWarning: The figure layout has changed to tight
  self._figure.tight_layout(*args, **kwargs)</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>GRCh38: $\mathregular{r^2}$=0.32, p=1.9e-22, slope=7.0e-04
CHM13v2.0: $\mathregular{r^2}$=0.14, p=5e-05, slope=3.1e-04
CHM13v2.0 - GRCh38: $\mathregular{r^2}$=0.20, p=4.5e-09, slope=3.9e-04</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The PostScript backend does not support transparency; partially transparent artists will be rendered opaque.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="make_plots_files/figure-html/cell-18-output-4.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Supplemental Figure 8: True switch error rate as a function of GRCh38 chromosomal completeness. Scatterplot of all chromosome-wide switch error rate in the GRCh38 and CHM13v2.0 1KGP panels for all chromosomes from HPRC-assembled samples by the percentage of each chromosome that is newly resolved in CHM13v2.0. A linear regression (true_switch_error_rate ~ %_chrom_novel), run separately for each 1KGP panel, is also plotted. Shaded areas indicate 95% ci of the regression slope. A linear regression was also performed of ΔSER% ~ %_chrom_novel. Summary statistics of all three regressions are presented in the upper left corner of the plot.</p>
</section>
<section id="figure-5" class="level2" data-number="1.4">
<h2 data-number="1.4" class="anchored" data-anchor-id="figure-5"><span class="header-section-number">1.4</span> Figure 5</h2>
<section id="phasing-performance-as-reference-dataset" class="level3" data-number="1.4.1">
<h3 data-number="1.4.1" class="anchored" data-anchor-id="phasing-performance-as-reference-dataset"><span class="header-section-number">1.4.1</span> Phasing performance as reference dataset</h3>
<ul>
<li>Out of sample confirmations, show 2 figure panels:</li>
<li><ol type="A">
<li>most likely use case of Illumina parent data</li>
</ol></li>
<li><ol start="2" type="A">
<li>without genotyping error, i.e.&nbsp;pangenome data from parents as ground truth</li>
</ol></li>
</ul>
<div id="cell-36" class="cell" data-execution_count="19">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Figure 5</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>fig, (axA, axB) <span class="op">=</span> plt.subplots(<span class="dv">1</span>,<span class="dv">2</span>,figsize<span class="op">=</span>(oneandahalf_col,<span class="fl">2.25</span>), layout<span class="op">=</span><span class="st">'tight'</span>)</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'switch_error_rate'</span>, style<span class="op">=</span><span class="st">'syntenic'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> syntenic_order,</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data.loc[(binned_maf_data.ground_truth_data_source <span class="op">==</span> <span class="st">'HPRC_samples'</span>)</span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>                                    <span class="op">&amp;</span> (binned_maf_data.method_of_phasing <span class="op">==</span> <span class="st">'1kgp_variation_phased_with_reference_panel'</span>)</span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>                                    <span class="op">&amp;</span> (binned_maf_data.syntenic<span class="op">!=</span><span class="st">'All'</span>)</span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>                                    <span class="op">&amp;</span> (binned_maf_data.<span class="bu">type</span><span class="op">==</span><span class="st">'SNP'</span>)],</span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axA, <span class="op">**</span>syntenic_style)</span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'switch_error_rate'</span>, style<span class="op">=</span><span class="st">'syntenic'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> syntenic_order,</span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data.loc[(binned_maf_data.ground_truth_data_source <span class="op">==</span> <span class="st">'HPRC_samples'</span>)</span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a>                                    <span class="op">&amp;</span> (binned_maf_data.method_of_phasing <span class="op">==</span> <span class="st">'1kgp_variation_phased_with_reference_panel'</span>)</span>
<span id="cb32-13"><a href="#cb32-13" aria-hidden="true" tabindex="-1"></a>                                    <span class="op">&amp;</span> (binned_maf_data.syntenic<span class="op">!=</span><span class="st">'All'</span>)</span>
<span id="cb32-14"><a href="#cb32-14" aria-hidden="true" tabindex="-1"></a>                                    <span class="op">&amp;</span> (binned_maf_data.<span class="bu">type</span><span class="op">==</span><span class="st">'Indel'</span>)],</span>
<span id="cb32-15"><a href="#cb32-15" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axB, <span class="op">**</span>syntenic_style)</span>
<span id="cb32-16"><a href="#cb32-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-17"><a href="#cb32-17" aria-hidden="true" tabindex="-1"></a>axA.text(<span class="fl">0.6</span>,<span class="fl">1.1</span>, <span class="st">'SNPs'</span>, fontsize<span class="op">=</span><span class="dv">8</span>, transform<span class="op">=</span>axA.transAxes, weight<span class="op">=</span><span class="st">'bold'</span>, horizontalalignment<span class="op">=</span><span class="st">'right'</span>, verticalalignment<span class="op">=</span><span class="st">'center'</span>)</span>
<span id="cb32-18"><a href="#cb32-18" aria-hidden="true" tabindex="-1"></a>axB.text(<span class="fl">0.6</span>,<span class="fl">1.1</span>, <span class="st">'Indels'</span>, fontsize<span class="op">=</span><span class="dv">8</span>, transform<span class="op">=</span>axB.transAxes, weight<span class="op">=</span><span class="st">'bold'</span>, horizontalalignment<span class="op">=</span><span class="st">'right'</span>, verticalalignment<span class="op">=</span><span class="st">'center'</span>)</span>
<span id="cb32-19"><a href="#cb32-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-20"><a href="#cb32-20" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, ax <span class="kw">in</span> <span class="bu">enumerate</span>([axA, axB]):</span>
<span id="cb32-21"><a href="#cb32-21" aria-hidden="true" tabindex="-1"></a>    ax.set_yscale(<span class="st">'log'</span>)</span>
<span id="cb32-22"><a href="#cb32-22" aria-hidden="true" tabindex="-1"></a>    ax.set_xscale(<span class="st">'log'</span>)</span>
<span id="cb32-23"><a href="#cb32-23" aria-hidden="true" tabindex="-1"></a>    ax.set_ylim(<span class="fl">1e-1</span>, <span class="dv">50</span>)</span>
<span id="cb32-24"><a href="#cb32-24" aria-hidden="true" tabindex="-1"></a>    ax.set_xlim(<span class="fl">1e-2</span>, <span class="dv">55</span>)</span>
<span id="cb32-25"><a href="#cb32-25" aria-hidden="true" tabindex="-1"></a>    <span class="co"># ax.set_box_aspect(1)</span></span>
<span id="cb32-26"><a href="#cb32-26" aria-hidden="true" tabindex="-1"></a>    ax.grid()</span>
<span id="cb32-27"><a href="#cb32-27" aria-hidden="true" tabindex="-1"></a>    ax.set_axisbelow(<span class="va">True</span>)</span>
<span id="cb32-28"><a href="#cb32-28" aria-hidden="true" tabindex="-1"></a>    ax.set_xticks(MAF_xticks)</span>
<span id="cb32-29"><a href="#cb32-29" aria-hidden="true" tabindex="-1"></a>    ax.set_yticks(SER_xticks)</span>
<span id="cb32-30"><a href="#cb32-30" aria-hidden="true" tabindex="-1"></a>    ax.xaxis.set_major_formatter(perc_formatter)</span>
<span id="cb32-31"><a href="#cb32-31" aria-hidden="true" tabindex="-1"></a>    ax.yaxis.set_major_formatter(perc_formatter)</span>
<span id="cb32-32"><a href="#cb32-32" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="kw">in</span> (<span class="dv">2</span>,<span class="dv">3</span>):</span>
<span id="cb32-33"><a href="#cb32-33" aria-hidden="true" tabindex="-1"></a>        ax.set_xlabel(<span class="st">'Minor Allele Frequency (%)'</span>, labelpad<span class="op">=</span><span class="dv">3</span>)</span>
<span id="cb32-34"><a href="#cb32-34" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb32-35"><a href="#cb32-35" aria-hidden="true" tabindex="-1"></a>        ax.set_xlabel(<span class="va">None</span>)</span>
<span id="cb32-36"><a href="#cb32-36" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="kw">in</span> (<span class="dv">0</span>,<span class="dv">2</span>):</span>
<span id="cb32-37"><a href="#cb32-37" aria-hidden="true" tabindex="-1"></a>        ax.set_ylabel(<span class="st">'Switch Error Rate (%)'</span>, labelpad<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb32-38"><a href="#cb32-38" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb32-39"><a href="#cb32-39" aria-hidden="true" tabindex="-1"></a>        ax.set_ylabel(<span class="va">None</span>)</span>
<span id="cb32-40"><a href="#cb32-40" aria-hidden="true" tabindex="-1"></a>fig.supxlabel(<span class="st">'Minor Allele Frequency (%)'</span>)<span class="co">#, labelpad=3)</span></span>
<span id="cb32-41"><a href="#cb32-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-42"><a href="#cb32-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-43"><a href="#cb32-43" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, ax <span class="kw">in</span> <span class="bu">enumerate</span>([axA, axB]): <span class="co"># legend axes</span></span>
<span id="cb32-44"><a href="#cb32-44" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="op">==</span> <span class="dv">0</span>:</span>
<span id="cb32-45"><a href="#cb32-45" aria-hidden="true" tabindex="-1"></a>        loc<span class="op">=</span><span class="st">'upper right'</span></span>
<span id="cb32-46"><a href="#cb32-46" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb32-47"><a href="#cb32-47" aria-hidden="true" tabindex="-1"></a>        loc<span class="op">=</span><span class="st">'upper right'</span></span>
<span id="cb32-48"><a href="#cb32-48" aria-hidden="true" tabindex="-1"></a>    <span class="co"># handles, labels = ax.get_legend_handles_labels()</span></span>
<span id="cb32-49"><a href="#cb32-49" aria-hidden="true" tabindex="-1"></a>    <span class="co"># for handle in handles:</span></span>
<span id="cb32-50"><a href="#cb32-50" aria-hidden="true" tabindex="-1"></a>    <span class="co">#     handle.set_marker('o')</span></span>
<span id="cb32-51"><a href="#cb32-51" aria-hidden="true" tabindex="-1"></a>    <span class="co"># handles = handles[1:3] + handles[5:6]</span></span>
<span id="cb32-52"><a href="#cb32-52" aria-hidden="true" tabindex="-1"></a>    <span class="co"># labels = labels[1:3] + labels[5:6]</span></span>
<span id="cb32-53"><a href="#cb32-53" aria-hidden="true" tabindex="-1"></a>    <span class="co"># try:</span></span>
<span id="cb32-54"><a href="#cb32-54" aria-hidden="true" tabindex="-1"></a>    <span class="co">#     handles[-1].set_color(handles[1].get_color())</span></span>
<span id="cb32-55"><a href="#cb32-55" aria-hidden="true" tabindex="-1"></a>    <span class="co"># except:</span></span>
<span id="cb32-56"><a href="#cb32-56" aria-hidden="true" tabindex="-1"></a>    <span class="co">#     print (handles)</span></span>
<span id="cb32-57"><a href="#cb32-57" aria-hidden="true" tabindex="-1"></a>    <span class="co"># legend = ax.legend(ncols=1, loc=loc, handles=handles, labels=labels)</span></span>
<span id="cb32-58"><a href="#cb32-58" aria-hidden="true" tabindex="-1"></a>    <span class="co"># legend.set_title(None)</span></span>
<span id="cb32-59"><a href="#cb32-59" aria-hidden="true" tabindex="-1"></a>    <span class="co"># legend.get_frame().set(</span></span>
<span id="cb32-60"><a href="#cb32-60" aria-hidden="true" tabindex="-1"></a>    <span class="co">#     alpha=0.9, boxstyle='square', edgecolor='grey', linewidth=0)</span></span>
<span id="cb32-61"><a href="#cb32-61" aria-hidden="true" tabindex="-1"></a>    <span class="co"># for text in legend.get_texts():</span></span>
<span id="cb32-62"><a href="#cb32-62" aria-hidden="true" tabindex="-1"></a>    <span class="co">#     t = text.get_text()</span></span>
<span id="cb32-63"><a href="#cb32-63" aria-hidden="true" tabindex="-1"></a>    <span class="co">#     new_t = update_legend_values.get(t,t)</span></span>
<span id="cb32-64"><a href="#cb32-64" aria-hidden="true" tabindex="-1"></a>    <span class="co">#     text.set_text(new_t)</span></span>
<span id="cb32-65"><a href="#cb32-65" aria-hidden="true" tabindex="-1"></a>    add_letter_to_ax(ax, i, points_offset<span class="op">=</span>(<span class="op">-</span><span class="dv">25</span>, <span class="dv">7</span>))</span>
<span id="cb32-66"><a href="#cb32-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-67"><a href="#cb32-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-68"><a href="#cb32-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-69"><a href="#cb32-69" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/Figure 5.svg'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span>
<span id="cb32-70"><a href="#cb32-70" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/Figure 5.png'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span>
<span id="cb32-71"><a href="#cb32-71" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/Figure 5.pdf'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span>
<span id="cb32-72"><a href="#cb32-72" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/Figure 5.eps'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>The PostScript backend does not support transparency; partially transparent artists will be rendered opaque.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="make_plots_files/figure-html/cell-19-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Figure 5: Switch error rates of SNPs from samples statistically phased with the 1KGP haplotype panel used as a reference. All HPRC-assembled samples and their relatives were removed from the GRCh38 1KGP and the CHM13v2.0 1KGP panels of 2504 unrelated samples, yielding ‘non-HPRC’ panels containing data from 2426 individuals. a) Filtered short read derived SNPs from 39 HPRC-assembled individuals were phased, using either the GRCh38 or CHM13v2.0 non-HPRC reference haplotype panel. Switch error rates were measured by mendelian concordance with parental genomes derived from short read variant calls. b) SNPs present in the HPRC human pangenome were unphased, then statistically phased with either the GRCh38 or CHM13v2.0 non-HPRC reference haplotype panel. Switch error rates were measured using the pangenome haplotypes as ground truth. Switch error rates and average minor allele frequency per bin are displayed on a log scale.</p>
</section>
</section>
</section>
<section id="figure-6" class="level1" data-number="2">
<h1 data-number="2"><span class="header-section-number">2</span> Figure 6</h1>
<p>Ideograms see R script</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="make_plots_files/figure-html/cell-39-1-image.png" class="img-fluid figure-img"></p>
<figcaption>image.png</figcaption>
</figure>
</div>
<p>Supplemental Figure 9. UCSC Genome Browser information on mappability of a region near the Prader-Willi breakpoint 2 (roughly chr15:24100000-24600000) that is enriched for genetic variants in the GRCh38 phased genomic panel. a) Mappability, blacklist status, location of segmental duplications, and genic regions in GRCh38. b) Mappability, location of segmental duplications, “difficult region” status from GIAB, and genic regions in CHM13v2.0.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="make_plots_files/figure-html/cell-40-1-image.png" class="img-fluid figure-img"></p>
<figcaption>image.png</figcaption>
</figure>
</div>
<p>Supplemental Figure 10. Improved liftover algorithm results in better imputation of common indel variants. a) Scatter plot of the r2 value of imputed indel variants when using a GRCh38 1KGP reference panel (x-axis) or the CHM13v2.0 panel lifted to GRCh38 coordinates using either a) GATK LiftoverVCF or b) the modified liftover algorithm implemented in this paper. Manual inspection of the high MAF variants that were poorly called by the lifted-over CHM13v2.0 panel showed that the majority of these variants are short tandem repeats that are lifted to a site that is several bases away from the same variant in the 1KGP GRCh38-native variant callset. Each dot represents one biallelic indel. The panel minor allele frequency (defined as 1-AF) is colored by 1KGP GRCh38 MAF.</p>
<section id="figure-7" class="level2" data-number="2.1">
<h2 data-number="2.1" class="anchored" data-anchor-id="figure-7"><span class="header-section-number">2.1</span> Figure 7</h2>
<section id="imputation-performance-as-reference-dataset" class="level3" data-number="2.1.1">
<h3 data-number="2.1.1" class="anchored" data-anchor-id="imputation-performance-as-reference-dataset"><span class="header-section-number">2.1.1</span> Imputation performance as reference dataset</h3>
<ul>
<li>plots as I currently have them</li>
<li>Supplement: Breakdown of performance by population</li>
</ul>
<div id="cell-42" class="cell" data-execution_count="20">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Figure 7</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>genome_order <span class="op">=</span> [<span class="st">'GRCh38'</span>,<span class="st">'CHM13v2.0'</span>]</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>panel_order<span class="op">=</span>[<span class="st">'native_panel.common_variants'</span>,<span class="st">'lifted_panel.common_variants'</span>]</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>separate_panel_and_subset<span class="op">=</span>pd.concat(((per_variant_category_imputation_performance.loc[per_variant_category_imputation_performance.panel.<span class="bu">str</span>.contains(<span class="st">'common'</span>)]</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>                                                                                .assign(var_subset<span class="op">=</span><span class="st">'Common variants'</span>)</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>                                                                                .replace(<span class="st">'native_panel.common_variants'</span>, <span class="st">'Native'</span>)</span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>                                                                                .replace(<span class="st">'lifted_panel.common_variants'</span>, <span class="st">'Lifted'</span>)),</span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>                                    (per_variant_category_imputation_performance.loc[<span class="op">~</span>per_variant_category_imputation_performance.panel.<span class="bu">str</span>.contains(<span class="st">'common'</span>)]</span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>                                                                                .assign(var_subset<span class="op">=</span><span class="st">'All panel variants'</span>)</span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a>                                                                                .replace(<span class="st">'native_panel'</span>,<span class="st">'Native'</span>)</span>
<span id="cb34-11"><a href="#cb34-11" aria-hidden="true" tabindex="-1"></a>                                                                                .replace(<span class="st">'lifted_panel'</span>,<span class="st">'Lifted'</span>))))</span>
<span id="cb34-12"><a href="#cb34-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-13"><a href="#cb34-13" aria-hidden="true" tabindex="-1"></a>separate_panel_and_subset <span class="op">=</span> separate_panel_and_subset.rename(columns<span class="op">=</span>{<span class="st">'var_category_id'</span>:<span class="st">'Variant Type'</span>}).replace(<span class="st">'CHM13'</span>,<span class="st">'CHM13v2.0'</span>)</span>
<span id="cb34-14"><a href="#cb34-14" aria-hidden="true" tabindex="-1"></a>separate_panel_and_subset[<span class="st">'Reference Panel'</span>] <span class="op">=</span> <span class="st">'GRCh38'</span></span>
<span id="cb34-15"><a href="#cb34-15" aria-hidden="true" tabindex="-1"></a>separate_panel_and_subset.loc[((separate_panel_and_subset.genome<span class="op">==</span><span class="st">'GRCh38'</span>) <span class="op">&amp;</span> (separate_panel_and_subset.panel<span class="op">==</span><span class="st">'Lifted'</span>)) <span class="op">|</span></span>
<span id="cb34-16"><a href="#cb34-16" aria-hidden="true" tabindex="-1"></a>                              ((separate_panel_and_subset.genome<span class="op">==</span><span class="st">'CHM13v2.0'</span>) <span class="op">&amp;</span> (separate_panel_and_subset.panel<span class="op">==</span><span class="st">'Native'</span>)), <span class="st">'Reference Panel'</span>] <span class="op">=</span> <span class="st">'CHM13v2.0'</span></span>
<span id="cb34-17"><a href="#cb34-17" aria-hidden="true" tabindex="-1"></a><span class="co"># fig=sns.relplot(kind='line', data = separate_panel_and_subset.loc[(separate_panel_and_subset.Synteny == 'All')#|(per_variant_category_imputation_performance.genome=='CHM13v2.0'))</span></span>
<span id="cb34-18"><a href="#cb34-18" aria-hidden="true" tabindex="-1"></a><span class="co">#                                                                 #  &amp; (separate_panel_and_subset['var_subset']=='Common variants')</span></span>
<span id="cb34-19"><a href="#cb34-19" aria-hidden="true" tabindex="-1"></a><span class="co">#                                                                   &amp; (separate_panel_and_subset.ancestry=='all')</span></span>
<span id="cb34-20"><a href="#cb34-20" aria-hidden="true" tabindex="-1"></a><span class="co">#                                                                   &amp;  (separate_panel_and_subset.dataset=='SGDP')],</span></span>
<span id="cb34-21"><a href="#cb34-21" aria-hidden="true" tabindex="-1"></a><span class="co">#                 x='mean_AF',y='imputed_ds_rsquared', hue='Reference Panel', style='Variant Type', hue_order=['GRCh38','CHM13v2.0'], style_order=['SNPs','Indels'],#style_order=['All variants','Indels','SNPs'],</span></span>
<span id="cb34-22"><a href="#cb34-22" aria-hidden="true" tabindex="-1"></a><span class="co">#                 palette=default_style['palette'], col='var_subset', row='genome', height=4, aspect=1, row_order=['GRCh38','CHM13v2.0']), #, col_wrap=4)#, , height=3, aspect=1, palette=None, row_order=None, col_order=None, hue_order=None, hue_kws=None, dropna=False, legend_out=True, despine=True, margin_titles=False, xlim=None, ylim=None, subplot_kws=None, gridspec_kws=None)</span></span>
<span id="cb34-23"><a href="#cb34-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-24"><a href="#cb34-24" aria-hidden="true" tabindex="-1"></a>fig<span class="op">=</span>sns.relplot(kind<span class="op">=</span><span class="st">'line'</span>, data <span class="op">=</span> separate_panel_and_subset.loc[(separate_panel_and_subset.Synteny <span class="op">==</span> <span class="st">'All'</span>)<span class="co">#|(per_variant_category_imputation_performance.genome=='CHM13v2.0'))</span></span>
<span id="cb34-25"><a href="#cb34-25" aria-hidden="true" tabindex="-1"></a>                                                                 <span class="op">&amp;</span> (separate_panel_and_subset[<span class="st">'var_subset'</span>]<span class="op">==</span><span class="st">'Common variants'</span>)</span>
<span id="cb34-26"><a href="#cb34-26" aria-hidden="true" tabindex="-1"></a>                                                                  <span class="op">&amp;</span> (separate_panel_and_subset.ancestry<span class="op">==</span><span class="st">'all'</span>)</span>
<span id="cb34-27"><a href="#cb34-27" aria-hidden="true" tabindex="-1"></a>                                                                  <span class="op">&amp;</span>  (separate_panel_and_subset.dataset<span class="op">==</span><span class="st">'SGDP'</span>)],</span>
<span id="cb34-28"><a href="#cb34-28" aria-hidden="true" tabindex="-1"></a>                x<span class="op">=</span><span class="st">'mean_AF'</span>,y<span class="op">=</span><span class="st">'imputed_ds_rsquared'</span>, hue<span class="op">=</span><span class="st">'Reference Panel'</span>, style<span class="op">=</span><span class="st">'Variant Type'</span>, hue_order<span class="op">=</span>[<span class="st">'GRCh38'</span>,<span class="st">'CHM13v2.0'</span>], style_order<span class="op">=</span>[<span class="st">'SNPs'</span>,<span class="st">'Indels'</span>],<span class="co">#style_order=['All variants','Indels','SNPs'],</span></span>
<span id="cb34-29"><a href="#cb34-29" aria-hidden="true" tabindex="-1"></a>                palette<span class="op">=</span>default_style[<span class="st">'palette'</span>], col<span class="op">=</span><span class="st">'genome'</span>, height<span class="op">=</span><span class="dv">6</span>, aspect<span class="op">=</span><span class="dv">1</span>, col_order<span class="op">=</span>[<span class="st">'GRCh38'</span>,<span class="st">'CHM13v2.0'</span>]), <span class="co">#, col_wrap=4)#, , height=3, aspect=1, palette=None, row_order=None, col_order=None, hue_order=None, hue_kws=None, dropna=False, legend_out=True, despine=True, margin_titles=False, xlim=None, ylim=None, subplot_kws=None, gridspec_kws=None)</span></span>
<span id="cb34-30"><a href="#cb34-30" aria-hidden="true" tabindex="-1"></a>fig <span class="op">=</span> (fig[<span class="dv">0</span>].<span class="bu">set</span>(xscale<span class="op">=</span><span class="st">'log'</span>, ylim<span class="op">=</span>(<span class="dv">0</span>,<span class="dv">1</span>))</span>
<span id="cb34-31"><a href="#cb34-31" aria-hidden="true" tabindex="-1"></a>  .set_axis_labels(<span class="st">"Allele Frequency"</span>, <span class="st">"Variant Imputation $</span><span class="er">\</span><span class="st">mathregular{r^2}$"</span>)</span>
<span id="cb34-32"><a href="#cb34-32" aria-hidden="true" tabindex="-1"></a>  .set_titles(template<span class="op">=</span><span class="st">"</span><span class="sc">{col_name}</span><span class="st"> SGDP variation"</span>)</span>
<span id="cb34-33"><a href="#cb34-33" aria-hidden="true" tabindex="-1"></a>  .tight_layout(w_pad<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb34-34"><a href="#cb34-34" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb34-35"><a href="#cb34-35" aria-hidden="true" tabindex="-1"></a>fig.legend.set_frame_on(<span class="va">True</span>)</span>
<span id="cb34-36"><a href="#cb34-36" aria-hidden="true" tabindex="-1"></a>fig.figure.set_size_inches((oneandahalf_col,<span class="dv">2</span>))</span>
<span id="cb34-37"><a href="#cb34-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-38"><a href="#cb34-38" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> col_val, ax <span class="kw">in</span> fig.axes_dict.items():</span>
<span id="cb34-39"><a href="#cb34-39" aria-hidden="true" tabindex="-1"></a>    ax.xaxis.set_major_formatter(PercentFormatter(decimals<span class="op">=</span><span class="dv">1</span>))</span>
<span id="cb34-40"><a href="#cb34-40" aria-hidden="true" tabindex="-1"></a>    ax.grid(alpha<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="cb34-41"><a href="#cb34-41" aria-hidden="true" tabindex="-1"></a>sns.move_legend(fig.figure, <span class="st">"lower right"</span>, bbox_to_anchor<span class="op">=</span>(<span class="fl">.92</span>, <span class="fl">.1</span>))</span>
<span id="cb34-42"><a href="#cb34-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-43"><a href="#cb34-43" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, ax <span class="kw">in</span> <span class="bu">enumerate</span>(<span class="bu">list</span>(itertools.chain.from_iterable(fig.axes))):</span>
<span id="cb34-44"><a href="#cb34-44" aria-hidden="true" tabindex="-1"></a>    add_letter_to_ax(ax, i, points_offset<span class="op">=</span>(<span class="op">-</span><span class="dv">7</span>, <span class="dv">7</span>))</span>
<span id="cb34-45"><a href="#cb34-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-46"><a href="#cb34-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-47"><a href="#cb34-47" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/Figure 7.png'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>, bbox_inches<span class="op">=</span><span class="st">'tight'</span>)</span>
<span id="cb34-48"><a href="#cb34-48" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/Figure 7.svg'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>, bbox_inches<span class="op">=</span><span class="st">'tight'</span>)</span>
<span id="cb34-49"><a href="#cb34-49" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/Figure 7.eps'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>, bbox_inches<span class="op">=</span><span class="st">'tight'</span>)</span>
<span id="cb34-50"><a href="#cb34-50" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/Figure 7.pdf'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>, bbox_inches<span class="op">=</span><span class="st">'tight'</span>)</span>
<span id="cb34-51"><a href="#cb34-51" aria-hidden="true" tabindex="-1"></a></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>/home/lalli/.linuxbrew/opt/python@3.11/lib/python3.11/site-packages/seaborn/axisgrid.py:123: UserWarning: The figure layout has changed to tight
  self._figure.tight_layout(*args, **kwargs)
The PostScript backend does not support transparency; partially transparent artists will be rendered opaque.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="make_plots_files/figure-html/cell-20-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Figure 6: Imputation of genomic variation in 256 non-1KGP Human Genome Diversity Project (HGDP) samples, using 1KGP haplotype panels as references. a) Genotyping array data was simulated by downsampling variation derived from short reads aligned to GRCh38 were downsampled to those variants present in the Infinium Omni2.5 genotyping array. We then phased this ‘array’ variation and imputed missing variants. At both steps, we provided a set of reference haplotypes to guide phasing and imputation - either the GRCh38 haplotype panel, or the CHM13v2.0 panel lifted to GRCh38 coordinates. R2 values were calculated using variants that were imputed by both the GRCh38 and the CHM13 panels. b) We downsampled variation by intersecting the Infinium Omni2.5 genotyping array variation (this time in CHM13v2.0 coordinates) with HGDP variation called form short reads aligned to CHM13v2.0. We then phased and imputed these samples using either the 1KGP CHM13v2.0 reference haplotype panel, or the 1KGP GRCh38 reference panel lifted to GRCh38 coordinates. All imputed variants were binned by the minor allele frequency of the variant in the reference panel used during imputation. R2 values were then calculated per bin. For both panels, R2 values were calculated using variants that were present in both the GRCh38-informed imputed variant data and the CHM13v2.0-informed imputed variant data. R2 statistics are stratified by SNPs and Indels. Average minor allele frequency per bin is displayed on a log scale.</p>
<div id="cell-44" class="cell" data-execution_count="21">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Figure 8 supplements</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(<span class="dv">1</span>,<span class="dv">1</span>,figsize<span class="op">=</span>(two_col,<span class="dv">3</span>), layout<span class="op">=</span><span class="st">'tight'</span>)</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>separate_sample_panel_and_subset<span class="op">=</span>pd.concat(((per_sample_imputation_performance.loc[per_sample_imputation_performance.panel.<span class="bu">str</span>.contains(<span class="st">'common'</span>)]</span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>                                                                                .assign(var_subset<span class="op">=</span><span class="st">'Common variants'</span>)</span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>                                                                                .replace(<span class="st">'native_panel.common_variants'</span>, <span class="st">'Native'</span>)</span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a>                                                                                .replace(<span class="st">'lifted_panel.common_variants'</span>, <span class="st">'Lifted'</span>)),</span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a>                                    (per_sample_imputation_performance.loc[<span class="op">~</span>per_sample_imputation_performance.panel.<span class="bu">str</span>.contains(<span class="st">'common'</span>)]</span>
<span id="cb36-10"><a href="#cb36-10" aria-hidden="true" tabindex="-1"></a>                                                                                .assign(var_subset<span class="op">=</span><span class="st">'All panel variants'</span>)</span>
<span id="cb36-11"><a href="#cb36-11" aria-hidden="true" tabindex="-1"></a>                                                                                .replace(<span class="st">'native_panel'</span>,<span class="st">'Native'</span>)</span>
<span id="cb36-12"><a href="#cb36-12" aria-hidden="true" tabindex="-1"></a>                                                                                .replace(<span class="st">'lifted_panel'</span>,<span class="st">'Lifted'</span>))))</span>
<span id="cb36-13"><a href="#cb36-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-14"><a href="#cb36-14" aria-hidden="true" tabindex="-1"></a>separate_sample_panel_and_subset <span class="op">=</span> separate_sample_panel_and_subset.rename(columns<span class="op">=</span>{<span class="st">'var_category_id'</span>:<span class="st">'Variant Type'</span>}).replace(<span class="st">'CHM13'</span>,<span class="st">'CHM13v2.0'</span>)</span>
<span id="cb36-15"><a href="#cb36-15" aria-hidden="true" tabindex="-1"></a>separate_sample_panel_and_subset[<span class="st">'Reference Panel'</span>] <span class="op">=</span> <span class="st">'GRCh38'</span></span>
<span id="cb36-16"><a href="#cb36-16" aria-hidden="true" tabindex="-1"></a>separate_sample_panel_and_subset.loc[((separate_sample_panel_and_subset.genome<span class="op">==</span><span class="st">'GRCh38'</span>) <span class="op">&amp;</span> (separate_sample_panel_and_subset.panel<span class="op">==</span><span class="st">'Lifted'</span>)) <span class="op">|</span></span>
<span id="cb36-17"><a href="#cb36-17" aria-hidden="true" tabindex="-1"></a>                              ((separate_sample_panel_and_subset.genome<span class="op">==</span><span class="st">'CHM13v2.0'</span>) <span class="op">&amp;</span> (separate_sample_panel_and_subset.panel<span class="op">==</span><span class="st">'Native'</span>)), <span class="st">'Reference Panel'</span>] <span class="op">=</span> <span class="st">'CHM13v2.0'</span></span>
<span id="cb36-18"><a href="#cb36-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-19"><a href="#cb36-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-20"><a href="#cb36-20" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> sns.violinplot(x<span class="op">=</span><span class="st">'ancestry'</span>, y<span class="op">=</span><span class="st">'imputed_ds_rsquared'</span>, hue<span class="op">=</span><span class="st">'Reference Panel'</span>, ax<span class="op">=</span>ax,</span>
<span id="cb36-21"><a href="#cb36-21" aria-hidden="true" tabindex="-1"></a>                 hue_order<span class="op">=</span>genome_order, palette<span class="op">=</span>genome_palette, <span class="co">#s=4,</span></span>
<span id="cb36-22"><a href="#cb36-22" aria-hidden="true" tabindex="-1"></a>                 data<span class="op">=</span>separate_sample_panel_and_subset.loc[(separate_sample_panel_and_subset.dataset<span class="op">==</span><span class="st">'SGDP'</span>)</span>
<span id="cb36-23"><a href="#cb36-23" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span>(separate_sample_panel_and_subset.var_subset<span class="op">==</span><span class="st">'Common variants'</span>)</span>
<span id="cb36-24"><a href="#cb36-24" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (separate_sample_panel_and_subset[<span class="st">'Variant Type'</span>]<span class="op">==</span><span class="st">'SNPs'</span>)</span>
<span id="cb36-25"><a href="#cb36-25" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (separate_sample_panel_and_subset.genome<span class="op">==</span><span class="st">'GRCh38'</span>)</span>
<span id="cb36-26"><a href="#cb36-26" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (separate_sample_panel_and_subset.ancestry <span class="op">!=</span> <span class="st">'all'</span>)],</span>
<span id="cb36-27"><a href="#cb36-27" aria-hidden="true" tabindex="-1"></a>                 legend<span class="op">=</span><span class="va">True</span>, alpha<span class="op">=</span><span class="fl">0.75</span>, split<span class="op">=</span><span class="va">True</span>, density_norm<span class="op">=</span><span class="st">'count'</span>, cut<span class="op">=</span><span class="dv">0</span>, gap<span class="op">=</span><span class="dv">0</span>, inner<span class="op">=</span><span class="st">'quart'</span>, saturation<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb36-28"><a href="#cb36-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-29"><a href="#cb36-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-30"><a href="#cb36-30" aria-hidden="true" tabindex="-1"></a>formatted_ancestry_names <span class="op">=</span> {<span class="st">'all'</span>:<span class="st">'All</span><span class="ch">\n</span><span class="st">populations'</span>,<span class="st">'WestEurasia'</span>:<span class="st">'West</span><span class="ch">\n</span><span class="st">Eurasia'</span>,<span class="st">'SouthAsia'</span>:<span class="st">"South</span><span class="ch">\n</span><span class="st">Asia"</span>,</span>
<span id="cb36-31"><a href="#cb36-31" aria-hidden="true" tabindex="-1"></a>                            <span class="st">'EastAsia'</span>:<span class="st">'East</span><span class="ch">\n</span><span class="st">Asia'</span>,<span class="st">'Africa'</span>:<span class="st">'Africa'</span>, <span class="st">'America'</span>:<span class="st">'Americas'</span>,<span class="st">'Oceania'</span>:<span class="st">'Oceania'</span>,</span>
<span id="cb36-32"><a href="#cb36-32" aria-hidden="true" tabindex="-1"></a>                            <span class="st">'CentralAsiaSiberia'</span>:<span class="st">'Central Asia</span><span class="ch">\n</span><span class="st">&amp; Siberia'</span>}</span>
<span id="cb36-33"><a href="#cb36-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-34"><a href="#cb36-34" aria-hidden="true" tabindex="-1"></a>xticks <span class="op">=</span> <span class="bu">list</span>(ax.get_xticklabels())</span>
<span id="cb36-35"><a href="#cb36-35" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> tick <span class="kw">in</span> xticks:</span>
<span id="cb36-36"><a href="#cb36-36" aria-hidden="true" tabindex="-1"></a>    t <span class="op">=</span> tick.get_text()</span>
<span id="cb36-37"><a href="#cb36-37" aria-hidden="true" tabindex="-1"></a>    tick.set_text(formatted_ancestry_names.get(t, t))</span>
<span id="cb36-38"><a href="#cb36-38" aria-hidden="true" tabindex="-1"></a>ax.set_xticklabels(xticks)</span>
<span id="cb36-39"><a href="#cb36-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-40"><a href="#cb36-40" aria-hidden="true" tabindex="-1"></a>ax.set_ylim(<span class="fl">0.85</span>,<span class="fl">1.01</span>)</span>
<span id="cb36-41"><a href="#cb36-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-42"><a href="#cb36-42" aria-hidden="true" tabindex="-1"></a>legend <span class="op">=</span> ax.legend()</span>
<span id="cb36-43"><a href="#cb36-43" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> text <span class="kw">in</span> legend.get_texts():</span>
<span id="cb36-44"><a href="#cb36-44" aria-hidden="true" tabindex="-1"></a>    t <span class="op">=</span> text.get_text()</span>
<span id="cb36-45"><a href="#cb36-45" aria-hidden="true" tabindex="-1"></a>    new_t <span class="op">=</span> update_legend_values.get(t,t)</span>
<span id="cb36-46"><a href="#cb36-46" aria-hidden="true" tabindex="-1"></a>    text.set_text(new_t)</span>
<span id="cb36-47"><a href="#cb36-47" aria-hidden="true" tabindex="-1"></a>legend.set_title(<span class="st">'Genome'</span>) <span class="co"># This shouldn't be necessary, but it is and I don't really care to dig into why</span></span>
<span id="cb36-48"><a href="#cb36-48" aria-hidden="true" tabindex="-1"></a>sns.move_legend(ax, <span class="st">"lower left"</span>)<span class="co">#, bbox_to_anchor=(1, 0))</span></span>
<span id="cb36-49"><a href="#cb36-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-50"><a href="#cb36-50" aria-hidden="true" tabindex="-1"></a>ax.set_ylabel(<span class="st">'Average overall sample $</span><span class="er">\</span><span class="st">mathregular{r^2}$'</span>)</span>
<span id="cb36-51"><a href="#cb36-51" aria-hidden="true" tabindex="-1"></a>ax.set_xlabel(<span class="st">'SGDP Ancestry Grouping'</span>)</span>
<span id="cb36-52"><a href="#cb36-52" aria-hidden="true" tabindex="-1"></a>ax.grid(axis<span class="op">=</span><span class="st">'y'</span>, alpha<span class="op">=</span><span class="fl">.5</span>)</span>
<span id="cb36-53"><a href="#cb36-53" aria-hidden="true" tabindex="-1"></a>ax.set_axisbelow(<span class="va">True</span>)</span>
<span id="cb36-54"><a href="#cb36-54" aria-hidden="true" tabindex="-1"></a><span class="co"># sns.despine()</span></span>
<span id="cb36-55"><a href="#cb36-55" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/supplemental/supplemental 11.png'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span>
<span id="cb36-56"><a href="#cb36-56" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/supplemental/supplemental 11.svg'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span>
<span id="cb36-57"><a href="#cb36-57" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/supplemental/supplemental 11.eps'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span>
<span id="cb36-58"><a href="#cb36-58" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">'figures/supplemental/supplemental 11.pdf'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>/tmp/ipykernel_2892998/3955385878.py:38: UserWarning: set_ticklabels() should only be used with a fixed number of ticks, i.e. after set_ticks() or using a FixedLocator.
  ax.set_xticklabels(xticks)
The PostScript backend does not support transparency; partially transparent artists will be rendered opaque.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="make_plots_files/figure-html/cell-21-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Supplemental Figure 10: A violin plot displaying the distribution of overall per-sample r2 values when imputing GRCh38 SGDP variation, stratified by SGDP-defined sample ancestry. Distributions are split by reference panel used during imputation: either the Byrska-Bishop 2022 GRCh38 reference haplotype panel, or the CHM13v2.0 panel lifted over to GRCh38 coordinates.</p>
<div id="cell-46" class="cell" data-execution_count="22">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>per_pop_data <span class="op">=</span> separate_panel_and_subset.loc[(separate_panel_and_subset.Synteny <span class="op">==</span> <span class="st">'All'</span>)<span class="co">#|(per_variant_category_imputation_performance.genome=='CHM13v2.0'))</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>                                                                 <span class="op">&amp;</span> (separate_panel_and_subset[<span class="st">'var_subset'</span>]<span class="op">==</span><span class="st">'Common variants'</span>)</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>                                                                  <span class="op">&amp;</span>  (separate_panel_and_subset.dataset<span class="op">==</span><span class="st">'SGDP'</span>)</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>                                                                  <span class="op">&amp;</span> (separate_panel_and_subset.genome<span class="op">==</span><span class="st">'GRCh38'</span>)]</span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>per_pop_data[<span class="st">'ancestry'</span>] <span class="op">=</span> per_pop_data.ancestry.<span class="bu">map</span>(<span class="kw">lambda</span> x: formatted_ancestry_names.get(x,x).replace(<span class="st">'</span><span class="ch">\n</span><span class="st">'</span>,<span class="st">' '</span>))</span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>panel_order<span class="op">=</span>[<span class="st">'native.common_variants'</span>,<span class="st">'lifted.common_variants'</span>]</span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>fig<span class="op">=</span>sns.relplot(kind<span class="op">=</span><span class="st">'line'</span>, data <span class="op">=</span> per_pop_data,</span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a>                x<span class="op">=</span><span class="st">'mean_AF'</span>,y<span class="op">=</span><span class="st">'imputed_ds_rsquared'</span>, hue<span class="op">=</span><span class="st">'Reference Panel'</span>, style<span class="op">=</span><span class="st">'Variant Type'</span>, hue_order<span class="op">=</span>[<span class="st">'GRCh38'</span>,<span class="st">'CHM13v2.0'</span>], style_order<span class="op">=</span>[<span class="st">'SNPs'</span>,<span class="st">'Indels'</span>],<span class="co">#style_order=['All variants','Indels','SNPs'],</span></span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a>                errorbar<span class="op">=</span><span class="va">None</span>, palette<span class="op">=</span>default_style[<span class="st">'palette'</span>], col<span class="op">=</span><span class="st">'ancestry'</span>, height<span class="op">=</span><span class="dv">4</span>, aspect<span class="op">=</span><span class="dv">1</span>, col_wrap<span class="op">=</span><span class="dv">4</span>,facet_kws<span class="op">=</span><span class="bu">dict</span>(legend_out<span class="op">=</span><span class="va">False</span>))</span>
<span id="cb38-10"><a href="#cb38-10" aria-hidden="true" tabindex="-1"></a><span class="co">#, sharex=True, sharey=True, height=3, aspect=1, palette=None, row_order=None, col_order=None, hue_order=None, hue_kws=None, dropna=False, legend_out=True, despine=True, margin_titles=False, xlim=None, ylim=None, subplot_kws=None, gridspec_kws=None)</span></span>
<span id="cb38-11"><a href="#cb38-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># .set_ylabel('Average overall sample $\mathregular{r^2}$')</span></span>
<span id="cb38-12"><a href="#cb38-12" aria-hidden="true" tabindex="-1"></a>fig.figure.set_tight_layout(<span class="va">True</span>)</span>
<span id="cb38-13"><a href="#cb38-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-14"><a href="#cb38-14" aria-hidden="true" tabindex="-1"></a>(fig.<span class="bu">set</span>(xscale<span class="op">=</span><span class="st">'log'</span>)</span>
<span id="cb38-15"><a href="#cb38-15" aria-hidden="true" tabindex="-1"></a>  .set_axis_labels(<span class="st">"Allele Frequency"</span>, <span class="st">"$</span><span class="er">\</span><span class="st">mathregular{r^2}$"</span>)</span>
<span id="cb38-16"><a href="#cb38-16" aria-hidden="true" tabindex="-1"></a>  .set_titles(<span class="st">"</span><span class="sc">{col_name}</span><span class="st">"</span>)</span>
<span id="cb38-17"><a href="#cb38-17" aria-hidden="true" tabindex="-1"></a>  .tight_layout(w_pad<span class="op">=</span><span class="dv">0</span>,h_pad<span class="op">=</span><span class="dv">5</span>))</span>
<span id="cb38-18"><a href="#cb38-18" aria-hidden="true" tabindex="-1"></a>fig.figure.set_size_inches((two_col,<span class="dv">4</span>))</span>
<span id="cb38-19"><a href="#cb38-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-20"><a href="#cb38-20" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> col_val, ax <span class="kw">in</span> fig.axes_dict.items():</span>
<span id="cb38-21"><a href="#cb38-21" aria-hidden="true" tabindex="-1"></a>    ax.xaxis.set_major_formatter(PercentFormatter(decimals<span class="op">=</span><span class="dv">1</span>))</span>
<span id="cb38-22"><a href="#cb38-22" aria-hidden="true" tabindex="-1"></a>    ax.grid(alpha<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="cb38-23"><a href="#cb38-23" aria-hidden="true" tabindex="-1"></a>    ax.set_ylim(<span class="dv">0</span>,<span class="dv">1</span>)</span>
<span id="cb38-24"><a href="#cb38-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-25"><a href="#cb38-25" aria-hidden="true" tabindex="-1"></a>sns.move_legend(</span>
<span id="cb38-26"><a href="#cb38-26" aria-hidden="true" tabindex="-1"></a>    fig, <span class="st">"upper center"</span>,</span>
<span id="cb38-27"><a href="#cb38-27" aria-hidden="true" tabindex="-1"></a>    bbox_to_anchor<span class="op">=</span>(<span class="fl">.48</span>, <span class="op">-</span><span class="fl">0.05</span>), ncol<span class="op">=</span><span class="dv">6</span>, title<span class="op">=</span><span class="va">None</span>, frameon<span class="op">=</span><span class="va">False</span></span>
<span id="cb38-28"><a href="#cb38-28" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb38-29"><a href="#cb38-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-30"><a href="#cb38-30" aria-hidden="true" tabindex="-1"></a>fig.savefig(<span class="st">'figures/supplemental/Supplemental 12.svg'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>,bbox_inches<span class="op">=</span><span class="st">'tight'</span>)</span>
<span id="cb38-31"><a href="#cb38-31" aria-hidden="true" tabindex="-1"></a>fig.savefig(<span class="st">'figures/supplemental/Supplemental 12.png'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>,bbox_inches<span class="op">=</span><span class="st">'tight'</span>)</span>
<span id="cb38-32"><a href="#cb38-32" aria-hidden="true" tabindex="-1"></a>fig.savefig(<span class="st">'figures/supplemental/Supplemental 12.eps'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>,bbox_inches<span class="op">=</span><span class="st">'tight'</span>)</span>
<span id="cb38-33"><a href="#cb38-33" aria-hidden="true" tabindex="-1"></a>fig.savefig(<span class="st">'figures/supplemental/Supplemental 12.pdf'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>,bbox_inches<span class="op">=</span><span class="st">'tight'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>/tmp/ipykernel_2892998/854782898.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  per_pop_data['ancestry'] = per_pop_data.ancestry.map(lambda x: formatted_ancestry_names.get(x,x).replace('\n',' '))
/home/lalli/.linuxbrew/opt/python@3.11/lib/python3.11/site-packages/seaborn/axisgrid.py:123: UserWarning: The figure layout has changed to tight
  self._figure.tight_layout(*args, **kwargs)
The PostScript backend does not support transparency; partially transparent artists will be rendered opaque.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="make_plots_files/figure-html/cell-22-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Supplemental Figure 9: Imputation of 256 non-1KGP HGDP genetic variation using 1KGP haplotype panels as references, stratified by SGDP population of origin. As previously presented, genotyping array data was simulated by downsampling variation derived from short reads aligned to GRCh38 were downsampled to those variants present in the Infinium Omni2.5 genotyping array. Variants present in our callset and in the genotyping array were then phased and imputed using either the GRCh38 1KGP haplotype panel, or the CHM13v2.0 1KGP panel lifted to GRCh38 coordinates. R2 values were calculated from the intersection of variants that were in both callsets. R2 statistics were calculated from imputed variants in samples assigned by the SGDP to the indicated ancestry group. R2 values were separately calculated in SNP variation and Indel variation. Average minor allele frequency per bin is displayed on a log scale.</p>
</section>
</section>
<section id="table-2" class="level2" data-number="2.2">
<h2 data-number="2.2" class="anchored" data-anchor-id="table-2"><span class="header-section-number">2.2</span> Table 2</h2>
<ul>
<li>top 15 most different cytobands from GRCh38 to CHM13</li>
<li>Per band stats to present:
<ul>
<li>GRCh38 switch errors, n_checked, and SER</li>
<li>T2T switch errors, n_checked, and SER</li>
<li>absolute reduction in SER</li>
</ul></li>
<li>Export as csv, copy/paste into work</li>
</ul>
<div id="cell-49" class="cell" data-execution_count="23">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>cyto <span class="op">=</span> pd.read_parquet(<span class="ss">f"</span><span class="sc">{</span>intermediate_data_folder<span class="sc">}</span><span class="ss">/per_cytoband_variant_data.parquet"</span>)</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>cnv <span class="op">=</span> pd.read_parquet(<span class="ss">f"</span><span class="sc">{</span>intermediate_data_folder<span class="sc">}</span><span class="ss">/per_genome_cnv_regions.parquet"</span>)</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>cnvslop <span class="op">=</span> pd.read_parquet(<span class="ss">f"</span><span class="sc">{</span>intermediate_data_folder<span class="sc">}</span><span class="ss">/per_genome_cnv_regions_with_slop.parquet"</span>)</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> concat_overlapping_sydromes(df):</span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>    df[<span class="st">'Syndrome'</span>] <span class="op">=</span> df[<span class="st">'Syndrome'</span>] <span class="op">+</span> <span class="st">'/'</span></span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a>    df <span class="op">=</span> df.groupby([<span class="st">'n_switch_errors'</span>, <span class="st">'n_checked'</span>, <span class="st">'n_gt_errors'</span>, <span class="st">'n_gt_checked'</span>, <span class="st">'method_of_phasing'</span>, <span class="st">'ground_truth_data_source'</span>, <span class="st">'genome'</span>, <span class="st">'MAC'</span>, <span class="st">'AN'</span>, <span class="st">'start'</span>, <span class="st">'end'</span>, <span class="st">'switch_error_rate'</span>]).<span class="bu">sum</span>()</span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a>    df[<span class="st">'Syndrome'</span>] <span class="op">=</span> df[<span class="st">'Syndrome'</span>].<span class="bu">str</span>[:<span class="op">-</span><span class="dv">1</span>]</span>
<span id="cb40-9"><a href="#cb40-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> df.reset_index()</span>
<span id="cb40-10"><a href="#cb40-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-11"><a href="#cb40-11" aria-hidden="true" tabindex="-1"></a>cnv <span class="op">=</span> concat_overlapping_sydromes(cnv)</span>
<span id="cb40-12"><a href="#cb40-12" aria-hidden="true" tabindex="-1"></a>cnvslop <span class="op">=</span> concat_overlapping_sydromes(cnvslop)</span>
<span id="cb40-13"><a href="#cb40-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-14"><a href="#cb40-14" aria-hidden="true" tabindex="-1"></a>decipher_cnvs <span class="op">=</span> pd.read_csv(<span class="st">'resources/decipher_syndromes.txt'</span>, sep<span class="op">=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>)</span>
<span id="cb40-15"><a href="#cb40-15" aria-hidden="true" tabindex="-1"></a>decipher_cnvs.loc[decipher_cnvs.end_grch38 <span class="op">&lt;</span> decipher_cnvs.start_grch38, [<span class="st">'start_grch38'</span>,<span class="st">'end_grch38'</span>]] <span class="op">=</span> decipher_cnvs.loc[decipher_cnvs.end_grch38 <span class="op">&lt;</span> decipher_cnvs.start_grch38, [<span class="st">'end_grch38'</span>,<span class="st">'start_grch38'</span>]].values</span>
<span id="cb40-16"><a href="#cb40-16" aria-hidden="true" tabindex="-1"></a>decipher_cnvs.loc[decipher_cnvs.end_chm13 <span class="op">&lt;</span> decipher_cnvs.start_chm13, [<span class="st">'start_chm13'</span>,<span class="st">'end_chm13'</span>]] <span class="op">=</span> decipher_cnvs.loc[decipher_cnvs.end_chm13 <span class="op">&lt;</span> decipher_cnvs.start_chm13, [<span class="st">'end_chm13'</span>,<span class="st">'start_chm13'</span>]].values</span>
<span id="cb40-17"><a href="#cb40-17" aria-hidden="true" tabindex="-1"></a>decipher_band_annotation <span class="op">=</span> decipher_cnvs.drop(columns<span class="op">=</span>[x <span class="cf">for</span> x <span class="kw">in</span> decipher_cnvs.columns <span class="cf">if</span> <span class="st">'start'</span> <span class="kw">in</span> x <span class="kw">or</span> <span class="st">'end'</span> <span class="kw">in</span> x])</span>
<span id="cb40-18"><a href="#cb40-18" aria-hidden="true" tabindex="-1"></a>decipher_band_annotation[<span class="st">'cytoband'</span>] <span class="op">=</span> decipher_band_annotation.cytoband.<span class="bu">str</span>.split(<span class="st">','</span>)</span>
<span id="cb40-19"><a href="#cb40-19" aria-hidden="true" tabindex="-1"></a>decipher_band_annotation <span class="op">=</span> decipher_band_annotation.explode(<span class="st">'cytoband'</span>).reset_index(drop<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb40-20"><a href="#cb40-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-21"><a href="#cb40-21" aria-hidden="true" tabindex="-1"></a>cyto[<span class="st">'interval'</span>] <span class="op">=</span> cyto.chrom.<span class="bu">str</span>.replace(<span class="st">'chr'</span>,<span class="st">''</span>) <span class="op">+</span> cyto[<span class="st">'interval'</span>].astype(<span class="bu">str</span>)</span>
<span id="cb40-22"><a href="#cb40-22" aria-hidden="true" tabindex="-1"></a>cyto[<span class="st">'decipher'</span>] <span class="op">=</span> cyto.interval.isin(decipher_band_annotation.cytoband)</span>
<span id="cb40-23"><a href="#cb40-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-24"><a href="#cb40-24" aria-hidden="true" tabindex="-1"></a>cyto[<span class="st">'switch_error_rate'</span>] <span class="op">=</span> cyto.n_switch_errors<span class="op">/</span>cyto.n_checked <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb40-25"><a href="#cb40-25" aria-hidden="true" tabindex="-1"></a>cyto[<span class="st">'genotype_error_rate'</span>] <span class="op">=</span> cyto.n_gt_errors<span class="op">/</span>cyto.n_gt_checked <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb40-26"><a href="#cb40-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-27"><a href="#cb40-27" aria-hidden="true" tabindex="-1"></a>cnv[<span class="st">'switch_error_rate'</span>] <span class="op">=</span> cnv.n_switch_errors<span class="op">/</span>cnv.n_checked <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb40-28"><a href="#cb40-28" aria-hidden="true" tabindex="-1"></a>cnv[<span class="st">'genotype_error_rate'</span>] <span class="op">=</span> cnv.n_gt_errors<span class="op">/</span>cnv.n_gt_checked<span class="op">*</span> <span class="dv">100</span></span>
<span id="cb40-29"><a href="#cb40-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-30"><a href="#cb40-30" aria-hidden="true" tabindex="-1"></a>cnvslop[<span class="st">'switch_error_rate'</span>] <span class="op">=</span> cnvslop.n_switch_errors<span class="op">/</span>cnvslop.n_checked <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb40-31"><a href="#cb40-31" aria-hidden="true" tabindex="-1"></a>cnvslop[<span class="st">'genotype_error_rate'</span>] <span class="op">=</span> cnvslop.n_gt_errors<span class="op">/</span>cnvslop.n_gt_checked <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb40-32"><a href="#cb40-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-33"><a href="#cb40-33" aria-hidden="true" tabindex="-1"></a>merge_columns<span class="op">=</span>[<span class="st">'method_of_phasing'</span>,<span class="st">'ground_truth_data_source'</span>,<span class="st">'chrom'</span>,<span class="st">'interval'</span>]</span>
<span id="cb40-34"><a href="#cb40-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-35"><a href="#cb40-35" aria-hidden="true" tabindex="-1"></a>cyto_grch38 <span class="op">=</span> cyto.loc[cyto.genome<span class="op">==</span><span class="st">'GRCh38'</span>]</span>
<span id="cb40-36"><a href="#cb40-36" aria-hidden="true" tabindex="-1"></a>cyto_chm13 <span class="op">=</span> cyto.loc[cyto.genome<span class="op">==</span><span class="st">'CHM13v2.0'</span>]</span>
<span id="cb40-37"><a href="#cb40-37" aria-hidden="true" tabindex="-1"></a>cnv_grch38 <span class="op">=</span> cnv.loc[cnv.genome<span class="op">==</span><span class="st">'GRCh38'</span>]</span>
<span id="cb40-38"><a href="#cb40-38" aria-hidden="true" tabindex="-1"></a>cnv_chm13 <span class="op">=</span> cnv.loc[cnv.genome<span class="op">==</span><span class="st">'CHM13v2.0'</span>]</span>
<span id="cb40-39"><a href="#cb40-39" aria-hidden="true" tabindex="-1"></a>cnvslop_grch38 <span class="op">=</span> cnvslop.loc[cnvslop.genome<span class="op">==</span><span class="st">'GRCh38'</span>]</span>
<span id="cb40-40"><a href="#cb40-40" aria-hidden="true" tabindex="-1"></a>cnvslop_chm13 <span class="op">=</span> cnvslop.loc[cnvslop.genome<span class="op">==</span><span class="st">'CHM13v2.0'</span>]</span>
<span id="cb40-41"><a href="#cb40-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-42"><a href="#cb40-42" aria-hidden="true" tabindex="-1"></a>cyto <span class="op">=</span> cyto_grch38.merge(cyto_chm13, on<span class="op">=</span>[<span class="st">'method_of_phasing'</span>,<span class="st">'ground_truth_data_source'</span>,<span class="st">'chrom'</span>,<span class="st">'interval'</span>,<span class="st">'decipher'</span>], suffixes<span class="op">=</span>[<span class="st">'_grch38'</span>,<span class="st">'_chm13'</span>])</span>
<span id="cb40-43"><a href="#cb40-43" aria-hidden="true" tabindex="-1"></a>cnv <span class="op">=</span> cnv_grch38.merge(cnv_chm13, on<span class="op">=</span>[<span class="st">'method_of_phasing'</span>,<span class="st">'ground_truth_data_source'</span>,<span class="st">'Syndrome'</span>], how<span class="op">=</span><span class="st">'outer'</span>, suffixes<span class="op">=</span>(<span class="st">'_grch38'</span>, <span class="st">'_chm13'</span>))</span>
<span id="cb40-44"><a href="#cb40-44" aria-hidden="true" tabindex="-1"></a>cnvslop <span class="op">=</span> cnvslop_grch38.merge(cnvslop_chm13, on<span class="op">=</span>[<span class="st">'method_of_phasing'</span>,<span class="st">'ground_truth_data_source'</span>,<span class="st">'Syndrome'</span>], how<span class="op">=</span><span class="st">'outer'</span>, suffixes<span class="op">=</span>(<span class="st">'_grch38'</span>, <span class="st">'_chm13'</span>))</span>
<span id="cb40-45"><a href="#cb40-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-46"><a href="#cb40-46" aria-hidden="true" tabindex="-1"></a>cyto[<span class="st">'abs_diff_SER'</span>] <span class="op">=</span> cyto.switch_error_rate_grch38.fillna(<span class="dv">0</span>) <span class="op">-</span> cyto.switch_error_rate_chm13.fillna(<span class="dv">0</span>)</span>
<span id="cb40-47"><a href="#cb40-47" aria-hidden="true" tabindex="-1"></a>cnv[<span class="st">'abs_diff_SER'</span>] <span class="op">=</span> cnv.switch_error_rate_grch38.fillna(<span class="dv">0</span>) <span class="op">-</span> cnv.switch_error_rate_chm13.fillna(<span class="dv">0</span>)</span>
<span id="cb40-48"><a href="#cb40-48" aria-hidden="true" tabindex="-1"></a>cnvslop[<span class="st">'abs_diff_SER'</span>] <span class="op">=</span> cnvslop.switch_error_rate_grch38.fillna(<span class="dv">0</span>) <span class="op">-</span> cnvslop.switch_error_rate_chm13.fillna(<span class="dv">0</span>)</span>
<span id="cb40-49"><a href="#cb40-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-50"><a href="#cb40-50" aria-hidden="true" tabindex="-1"></a>cyto[<span class="st">'rel_drop_SER'</span>] <span class="op">=</span> (<span class="dv">1</span><span class="op">-</span> (cyto.switch_error_rate_chm13.fillna(<span class="dv">0</span>) <span class="op">/</span> cyto.switch_error_rate_grch38.fillna(<span class="dv">0</span>))) <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb40-51"><a href="#cb40-51" aria-hidden="true" tabindex="-1"></a>cnv[<span class="st">'rel_drop_SER'</span>] <span class="op">=</span> (<span class="dv">1</span><span class="op">-</span> (cnv.switch_error_rate_chm13.fillna(<span class="dv">0</span>) <span class="op">/</span> cnv.switch_error_rate_grch38.fillna(<span class="dv">0</span>))) <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb40-52"><a href="#cb40-52" aria-hidden="true" tabindex="-1"></a>cnvslop[<span class="st">'rel_drop_SER'</span>] <span class="op">=</span> (<span class="dv">1</span><span class="op">-</span> (cnvslop.switch_error_rate_chm13.fillna(<span class="dv">0</span>) <span class="op">/</span> cnvslop.switch_error_rate_grch38.fillna(<span class="dv">0</span>))) <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb40-53"><a href="#cb40-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-54"><a href="#cb40-54" aria-hidden="true" tabindex="-1"></a>cyto[<span class="st">'abs_diff_GER'</span>] <span class="op">=</span> cyto.genotype_error_rate_grch38.fillna(<span class="dv">0</span>) <span class="op">-</span> cyto.genotype_error_rate_chm13.fillna(<span class="dv">0</span>)</span>
<span id="cb40-55"><a href="#cb40-55" aria-hidden="true" tabindex="-1"></a>cnv[<span class="st">'abs_diff_GER'</span>] <span class="op">=</span> cnv.genotype_error_rate_grch38.fillna(<span class="dv">0</span>) <span class="op">-</span> cnv.genotype_error_rate_chm13.fillna(<span class="dv">0</span>)</span>
<span id="cb40-56"><a href="#cb40-56" aria-hidden="true" tabindex="-1"></a>cnvslop[<span class="st">'abs_diff_GER'</span>] <span class="op">=</span> cnvslop.genotype_error_rate_grch38.fillna(<span class="dv">0</span>) <span class="op">-</span> cnvslop.genotype_error_rate_chm13.fillna(<span class="dv">0</span>)</span>
<span id="cb40-57"><a href="#cb40-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-58"><a href="#cb40-58" aria-hidden="true" tabindex="-1"></a>cyto[<span class="st">'rel_drop_GER'</span>] <span class="op">=</span> (<span class="dv">1</span><span class="op">-</span> (cyto.genotype_error_rate_chm13.fillna(<span class="dv">0</span>) <span class="op">/</span> cyto.genotype_error_rate_grch38.fillna(<span class="dv">0</span>))) <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb40-59"><a href="#cb40-59" aria-hidden="true" tabindex="-1"></a>cnv[<span class="st">'rel_drop_GER'</span>] <span class="op">=</span> (<span class="dv">1</span><span class="op">-</span> (cnv.genotype_error_rate_chm13.fillna(<span class="dv">0</span>) <span class="op">/</span> cnv.genotype_error_rate_grch38.fillna(<span class="dv">0</span>))) <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb40-60"><a href="#cb40-60" aria-hidden="true" tabindex="-1"></a>cnvslop[<span class="st">'rel_drop_GER'</span>] <span class="op">=</span> (<span class="dv">1</span><span class="op">-</span> (cnvslop.genotype_error_rate_chm13.fillna(<span class="dv">0</span>) <span class="op">/</span> cnvslop.genotype_error_rate_grch38.fillna(<span class="dv">0</span>))) <span class="op">*</span> <span class="dv">100</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-50" class="cell" data-execution_count="24">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>filtered_cyto <span class="op">=</span> cyto.loc[(cyto.method_of_phasing<span class="op">==</span><span class="st">'1kgp_variation_phased_with_reference_panel'</span>)<span class="op">&amp;</span>(cyto.ground_truth_data_source<span class="op">==</span><span class="st">'HPRC_samples'</span>)<span class="op">&amp;</span>(cyto.n_checked_grch38 <span class="op">&gt;</span> <span class="dv">50</span>)<span class="op">&amp;</span>(cyto.n_checked_chm13 <span class="op">&gt;</span> <span class="dv">50</span>),</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>        [<span class="st">'interval'</span>,<span class="st">'n_switch_errors_grch38'</span>,<span class="st">'n_switch_errors_chm13'</span>,<span class="st">'n_checked_grch38'</span>,<span class="st">'n_checked_chm13'</span>,<span class="st">'switch_error_rate_grch38'</span>,<span class="st">'switch_error_rate_chm13'</span>,</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>         <span class="st">'n_gt_checked_grch38'</span>,<span class="st">'n_gt_checked_chm13'</span>,<span class="st">'genotype_error_rate_grch38'</span>,<span class="st">'genotype_error_rate_chm13'</span>,</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>         <span class="st">'abs_diff_GER'</span>,<span class="st">'rel_drop_GER'</span>,<span class="st">'abs_diff_SER'</span>,<span class="st">'rel_drop_SER'</span>, <span class="st">'decipher'</span>]</span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>         ].sort_values(<span class="st">'rel_drop_SER'</span>,ascending<span class="op">=</span><span class="va">False</span>).reset_index(drop<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>filtered_cyto.to_csv(<span class="st">'tables/1kgp_variation_HPRC_ground_truth_summarized.tsv'</span>, sep<span class="op">=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>, index<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a>cyto.loc[(cyto.method_of_phasing<span class="op">==</span><span class="st">'1kgp_variation_phased_with_reference_panel'</span>)<span class="op">&amp;</span>(cyto.ground_truth_data_source<span class="op">==</span><span class="st">'HPRC_samples'</span>)</span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a>         ].sort_values(<span class="st">'rel_drop_SER'</span>,ascending<span class="op">=</span><span class="va">False</span>).reset_index(drop<span class="op">=</span><span class="va">True</span>).to_csv(<span class="st">'tables/1kgp_variation_HPRC_ground_truth.tsv'</span>, sep<span class="op">=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>, index<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a>cyto.loc[(cyto.method_of_phasing<span class="op">==</span><span class="st">'1kgp_variation_phased_with_reference_panel'</span>)<span class="op">&amp;</span>(cyto.ground_truth_data_source<span class="op">==</span><span class="st">'trios'</span>)</span>
<span id="cb41-12"><a href="#cb41-12" aria-hidden="true" tabindex="-1"></a>         ].sort_values(<span class="st">'rel_drop_SER'</span>,ascending<span class="op">=</span><span class="va">False</span>).reset_index(drop<span class="op">=</span><span class="va">True</span>).to_csv(<span class="st">'tables/1kgp_variation_parental_haplotypes_ground_truth.tsv'</span>, sep<span class="op">=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>, index<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb41-13"><a href="#cb41-13" aria-hidden="true" tabindex="-1"></a></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-51" class="cell" data-execution_count="25">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> tqdm <span class="im">import</span> trange</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> calculate_es(gene_list, ranked_genes, gene_set, weights):</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""</span></span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a><span class="co">    Calculates the enrichment score (ES) for a gene set within a ranked list of genes.</span></span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a><span class="co">    Args:</span></span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a><span class="co">        gene_list (list): A list of all genes considered in the analysis.</span></span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a><span class="co">        ranked_genes (list): A ranked list of genes, e.g., by differential expression.</span></span>
<span id="cb42-9"><a href="#cb42-9" aria-hidden="true" tabindex="-1"></a><span class="co">        gene_set (list): A list of genes belonging to the gene set of interest.</span></span>
<span id="cb42-10"><a href="#cb42-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-11"><a href="#cb42-11" aria-hidden="true" tabindex="-1"></a><span class="co">    Returns:</span></span>
<span id="cb42-12"><a href="#cb42-12" aria-hidden="true" tabindex="-1"></a><span class="co">        float: The enrichment score (ES).</span></span>
<span id="cb42-13"><a href="#cb42-13" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb42-14"><a href="#cb42-14" aria-hidden="true" tabindex="-1"></a>    N <span class="op">=</span> <span class="bu">len</span>(gene_list)</span>
<span id="cb42-15"><a href="#cb42-15" aria-hidden="true" tabindex="-1"></a>    Nh <span class="op">=</span> <span class="bu">len</span>(gene_set)</span>
<span id="cb42-16"><a href="#cb42-16" aria-hidden="true" tabindex="-1"></a>    ES <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb42-17"><a href="#cb42-17" aria-hidden="true" tabindex="-1"></a>    running_sum <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb42-18"><a href="#cb42-18" aria-hidden="true" tabindex="-1"></a>    tag_indicator <span class="op">=</span> np.isin(ranked_genes, gene_set)</span>
<span id="cb42-19"><a href="#cb42-19" aria-hidden="true" tabindex="-1"></a>    rank_list_indicator <span class="op">=</span> <span class="op">~</span>tag_indicator</span>
<span id="cb42-20"><a href="#cb42-20" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb42-21"><a href="#cb42-21" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i, tag <span class="kw">in</span> <span class="bu">enumerate</span>(tag_indicator):</span>
<span id="cb42-22"><a href="#cb42-22" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> tag:</span>
<span id="cb42-23"><a href="#cb42-23" aria-hidden="true" tabindex="-1"></a>        running_sum <span class="op">+=</span> np.sqrt((N <span class="op">-</span> Nh)<span class="op">/</span>Nh) <span class="op">*</span> weights[i]</span>
<span id="cb42-24"><a href="#cb42-24" aria-hidden="true" tabindex="-1"></a>      <span class="cf">else</span>:</span>
<span id="cb42-25"><a href="#cb42-25" aria-hidden="true" tabindex="-1"></a>        running_sum <span class="op">-=</span> np.sqrt(Nh<span class="op">/</span>(N <span class="op">-</span> Nh)) <span class="op">*</span> weights[i]</span>
<span id="cb42-26"><a href="#cb42-26" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb42-27"><a href="#cb42-27" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> running_sum <span class="op">&gt;</span> ES:</span>
<span id="cb42-28"><a href="#cb42-28" aria-hidden="true" tabindex="-1"></a>        ES <span class="op">=</span> running_sum</span>
<span id="cb42-29"><a href="#cb42-29" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb42-30"><a href="#cb42-30" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> ES</span>
<span id="cb42-31"><a href="#cb42-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-32"><a href="#cb42-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-33"><a href="#cb42-33" aria-hidden="true" tabindex="-1"></a>filtered_cyto <span class="op">=</span> filtered_cyto.sort_values(<span class="st">'rel_drop_SER'</span>, ascending<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb42-34"><a href="#cb42-34" aria-hidden="true" tabindex="-1"></a>gene_list <span class="op">=</span> filtered_cyto.interval.to_list()</span>
<span id="cb42-35"><a href="#cb42-35" aria-hidden="true" tabindex="-1"></a>weights <span class="op">=</span> filtered_cyto.rel_drop_SER.to_list()</span>
<span id="cb42-36"><a href="#cb42-36" aria-hidden="true" tabindex="-1"></a>true_gene_set <span class="op">=</span> filtered_cyto.loc[filtered_cyto.decipher].interval.to_list()</span>
<span id="cb42-37"><a href="#cb42-37" aria-hidden="true" tabindex="-1"></a>true_es <span class="op">=</span> calculate_es(gene_list, gene_list, true_gene_set, weights<span class="op">=</span>weights)</span>
<span id="cb42-38"><a href="#cb42-38" aria-hidden="true" tabindex="-1"></a>permuted_es <span class="op">=</span> <span class="bu">list</span>()</span>
<span id="cb42-39"><a href="#cb42-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-40"><a href="#cb42-40" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> trange(<span class="dv">100000</span>):</span>
<span id="cb42-41"><a href="#cb42-41" aria-hidden="true" tabindex="-1"></a>    gene_set <span class="op">=</span> filtered_cyto.interval.sample(<span class="bu">len</span>(true_gene_set), replace<span class="op">=</span><span class="va">False</span>, random_state<span class="op">=</span>np.random.seed(<span class="dv">42</span><span class="op">*</span>i))</span>
<span id="cb42-42"><a href="#cb42-42" aria-hidden="true" tabindex="-1"></a>    permuted_es.append(calculate_es(gene_list, gene_list, gene_set, weights<span class="op">=</span>weights))</span>
<span id="cb42-43"><a href="#cb42-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-44"><a href="#cb42-44" aria-hidden="true" tabindex="-1"></a>permuted_es <span class="op">=</span> np.array(permuted_es)</span>
<span id="cb42-45"><a href="#cb42-45" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Weighted enrichment score of </span><span class="sc">{</span>true_es<span class="sc">}</span><span class="ss"> (pval &lt; </span><span class="sc">{</span>np<span class="sc">.</span>mean(permuted_es<span class="op">&gt;</span>true_es)<span class="sc">}</span><span class="ss">)"</span>)</span>
<span id="cb42-46"><a href="#cb42-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-47"><a href="#cb42-47" aria-hidden="true" tabindex="-1"></a><span class="co"># weighted enrichment score w/ permutation</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>100%|██████████| 100000/100000 [08:15&lt;00:00, 201.93it/s]</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Weighted enrichment score of 858.9571296115195 (pval &lt; 0.00452)</code></pre>
</div>
</div>
<div id="cell-52" class="cell" data-execution_count="26">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> tqdm <span class="im">import</span> trange</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> calculate_es(gene_list, ranked_genes, gene_set):</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""</span></span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a><span class="co">    Calculates the enrichment score (ES) for a gene set within a ranked list of genes.</span></span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a><span class="co">    Args:</span></span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a><span class="co">        gene_list (list): A list of all genes considered in the analysis.</span></span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a><span class="co">        ranked_genes (list): A ranked list of genes, e.g., by differential expression.</span></span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a><span class="co">        gene_set (list): A list of genes belonging to the gene set of interest.</span></span>
<span id="cb45-10"><a href="#cb45-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-11"><a href="#cb45-11" aria-hidden="true" tabindex="-1"></a><span class="co">    Returns:</span></span>
<span id="cb45-12"><a href="#cb45-12" aria-hidden="true" tabindex="-1"></a><span class="co">        float: The enrichment score (ES).</span></span>
<span id="cb45-13"><a href="#cb45-13" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb45-14"><a href="#cb45-14" aria-hidden="true" tabindex="-1"></a>    N <span class="op">=</span> <span class="bu">len</span>(gene_list)</span>
<span id="cb45-15"><a href="#cb45-15" aria-hidden="true" tabindex="-1"></a>    Nh <span class="op">=</span> <span class="bu">len</span>(gene_set)</span>
<span id="cb45-16"><a href="#cb45-16" aria-hidden="true" tabindex="-1"></a>    ES <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb45-17"><a href="#cb45-17" aria-hidden="true" tabindex="-1"></a>    running_sum <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb45-18"><a href="#cb45-18" aria-hidden="true" tabindex="-1"></a>    tag_indicator <span class="op">=</span> np.isin(ranked_genes, gene_set)</span>
<span id="cb45-19"><a href="#cb45-19" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb45-20"><a href="#cb45-20" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> tag <span class="kw">in</span> tag_indicator:</span>
<span id="cb45-21"><a href="#cb45-21" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> tag:</span>
<span id="cb45-22"><a href="#cb45-22" aria-hidden="true" tabindex="-1"></a>        running_sum <span class="op">+=</span> np.sqrt((N <span class="op">-</span> Nh)<span class="op">/</span>Nh)</span>
<span id="cb45-23"><a href="#cb45-23" aria-hidden="true" tabindex="-1"></a>      <span class="cf">else</span>:</span>
<span id="cb45-24"><a href="#cb45-24" aria-hidden="true" tabindex="-1"></a>        running_sum <span class="op">-=</span> np.sqrt(Nh<span class="op">/</span>(N <span class="op">-</span> Nh))</span>
<span id="cb45-25"><a href="#cb45-25" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb45-26"><a href="#cb45-26" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> <span class="bu">abs</span>(running_sum) <span class="op">&gt;</span> <span class="bu">abs</span>(ES):</span>
<span id="cb45-27"><a href="#cb45-27" aria-hidden="true" tabindex="-1"></a>        ES <span class="op">=</span> running_sum</span>
<span id="cb45-28"><a href="#cb45-28" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb45-29"><a href="#cb45-29" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> ES</span>
<span id="cb45-30"><a href="#cb45-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-31"><a href="#cb45-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-32"><a href="#cb45-32" aria-hidden="true" tabindex="-1"></a>filtered_cyto <span class="op">=</span> filtered_cyto.sort_values(<span class="st">'rel_drop_SER'</span>, ascending<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb45-33"><a href="#cb45-33" aria-hidden="true" tabindex="-1"></a>gene_list <span class="op">=</span> filtered_cyto.interval.drop_duplicates().to_list()</span>
<span id="cb45-34"><a href="#cb45-34" aria-hidden="true" tabindex="-1"></a>true_gene_set <span class="op">=</span> filtered_cyto.loc[filtered_cyto.decipher].interval.drop_duplicates().to_list()</span>
<span id="cb45-35"><a href="#cb45-35" aria-hidden="true" tabindex="-1"></a>true_es <span class="op">=</span> calculate_es(gene_list, gene_list, true_gene_set)</span>
<span id="cb45-36"><a href="#cb45-36" aria-hidden="true" tabindex="-1"></a>permuted_es <span class="op">=</span> <span class="bu">list</span>()</span>
<span id="cb45-37"><a href="#cb45-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-38"><a href="#cb45-38" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> trange(<span class="dv">100000</span>):</span>
<span id="cb45-39"><a href="#cb45-39" aria-hidden="true" tabindex="-1"></a>    gene_set <span class="op">=</span> filtered_cyto.interval.sample(<span class="bu">len</span>(true_gene_set), replace<span class="op">=</span><span class="va">False</span>, random_state<span class="op">=</span><span class="dv">42</span>)</span>
<span id="cb45-40"><a href="#cb45-40" aria-hidden="true" tabindex="-1"></a>    permuted_es.append(calculate_es(gene_list, gene_list, gene_set))</span>
<span id="cb45-41"><a href="#cb45-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-42"><a href="#cb45-42" aria-hidden="true" tabindex="-1"></a>permuted_es <span class="op">=</span> np.array(permuted_es)</span>
<span id="cb45-43"><a href="#cb45-43" aria-hidden="true" tabindex="-1"></a>np.mean(permuted_es<span class="op">&gt;</span>true_es)</span>
<span id="cb45-44"><a href="#cb45-44" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Unweighted enrichment score of </span><span class="sc">{</span>true_es<span class="sc">}</span><span class="ss"> (pval &lt; </span><span class="sc">{</span>np<span class="sc">.</span>mean(permuted_es<span class="op">&gt;</span>true_es)<span class="sc">}</span><span class="ss">)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>100%|██████████| 100000/100000 [10:04&lt;00:00, 165.36it/s]</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Unweighted enrichment score of 31.754264805429546 (pval &lt; 0.0)</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code></code></pre>
</div>
</div>
<div id="cell-53" class="cell" data-execution_count="27">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(<span class="dv">1</span>,<span class="dv">1</span>,figsize<span class="op">=</span>(<span class="fl">3.33</span>,<span class="dv">3</span>))</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> sns.swarmplot(data<span class="op">=</span>filtered_cyto.loc[filtered_cyto.abs_diff_SER<span class="op">&lt;</span><span class="dv">1</span>], x<span class="op">=</span><span class="st">'decipher'</span>, y<span class="op">=</span><span class="st">'abs_diff_SER'</span>,dodge<span class="op">=</span><span class="va">True</span>, s<span class="op">=</span><span class="fl">1.5</span>, legend<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> sns.boxplot(data<span class="op">=</span>filtered_cyto.loc[filtered_cyto.abs_diff_SER<span class="op">&lt;</span><span class="dv">1</span>], x<span class="op">=</span><span class="st">'decipher'</span>, y<span class="op">=</span><span class="st">'abs_diff_SER'</span>, ax<span class="op">=</span>ax, fill<span class="op">=</span><span class="va">False</span>, flierprops<span class="op">=</span>{<span class="st">"marker"</span>: <span class="st">"None"</span>}, linewidth<span class="op">=</span><span class="fl">0.5</span>, palette<span class="op">=</span>[<span class="st">'black'</span>], legend<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> statannotations.stats.StatTest <span class="im">import</span> StatTest</span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> kstest(x, y):</span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> stats.kstest(x, y)</span>
<span id="cb49-8"><a href="#cb49-8" aria-hidden="true" tabindex="-1"></a>custom_test <span class="op">=</span> StatTest(kstest, <span class="st">'Kolmogorov-Smirnov (KS) non-parametric test'</span>, <span class="st">'kstest'</span>)</span>
<span id="cb49-9"><a href="#cb49-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-10"><a href="#cb49-10" aria-hidden="true" tabindex="-1"></a>annotator <span class="op">=</span> Annotator(ax, [(<span class="va">False</span>, <span class="va">True</span>)], data<span class="op">=</span>filtered_cyto, x<span class="op">=</span><span class="st">'decipher'</span>, order<span class="op">=</span>(<span class="va">False</span>,<span class="va">True</span>), y<span class="op">=</span><span class="st">'abs_diff_SER'</span>)</span>
<span id="cb49-11"><a href="#cb49-11" aria-hidden="true" tabindex="-1"></a>annotator.configure(test<span class="op">=</span><span class="st">'Mann-Whitney'</span>, text_format<span class="op">=</span><span class="st">'star'</span>, loc<span class="op">=</span><span class="st">'inside'</span>)</span>
<span id="cb49-12"><a href="#cb49-12" aria-hidden="true" tabindex="-1"></a>annotator.apply_and_annotate()</span>
<span id="cb49-13"><a href="#cb49-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-14"><a href="#cb49-14" aria-hidden="true" tabindex="-1"></a>ax.set_ylabel(<span class="st">'Reduction in Switch Error Rate</span><span class="ch">\n</span><span class="st">GRCh38 panel - CHM13v2.0 panel'</span>)</span>
<span id="cb49-15"><a href="#cb49-15" aria-hidden="true" tabindex="-1"></a>ax.set_xlabel(<span class="st">'Cytology band does/does not contain</span><span class="ch">\n</span><span class="st">a Decipher common CNV syndrome'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>/tmp/ipykernel_2892998/1003454266.py:3: FutureWarning: 

Passing `palette` without assigning `hue` is deprecated and will be removed in v0.14.0. Assign the `x` variable to `hue` and set `legend=False` for the same effect.

  ax = sns.boxplot(data=filtered_cyto.loc[filtered_cyto.abs_diff_SER&lt;1], x='decipher', y='abs_diff_SER', ax=ax, fill=False, flierprops={"marker": "None"}, linewidth=0.5, palette=['black'], legend=True)
/tmp/ipykernel_2892998/1003454266.py:3: UserWarning: 
The palette list has fewer values (1) than needed (2) and will cycle, which may produce an uninterpretable plot.
  ax = sns.boxplot(data=filtered_cyto.loc[filtered_cyto.abs_diff_SER&lt;1], x='decipher', y='abs_diff_SER', ax=ax, fill=False, flierprops={"marker": "None"}, linewidth=0.5, palette=['black'], legend=True)</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>p-value annotation legend:
      ns: 5.00e-02 &lt; p &lt;= 1.00e+00
       *: 1.00e-02 &lt; p &lt;= 5.00e-02
      **: 1.00e-03 &lt; p &lt;= 1.00e-02
     ***: 1.00e-04 &lt; p &lt;= 1.00e-03
    ****: p &lt;= 1.00e-04

False vs. True: Mann-Whitney-Wilcoxon test two-sided, P_val:1.445e-01 U_stat=1.425e+04</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="27">
<pre><code>Text(0.5, 10.166999999999996, 'Cytology band does/does not contain\na Decipher common CNV syndrome')</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="make_plots_files/figure-html/cell-27-output-4.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>^Bootstrap: how ofted do you see diff_SER in randomly selected 63 regions (w/ replacement) vs decipher regions</p>
<div id="cell-55" class="cell" data-execution_count="28">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>cnv_clean <span class="op">=</span> cnv.loc[(cnv.method_of_phasing<span class="op">==</span><span class="st">'1kgp_variation_phased_with_reference_panel'</span>)<span class="op">&amp;</span>(cnv.ground_truth_data_source<span class="op">==</span><span class="st">'HPRC_samples'</span>),</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>                            [<span class="st">'Syndrome'</span>,<span class="st">'n_switch_errors_grch38'</span>,<span class="st">'n_switch_errors_chm13'</span>,<span class="st">'n_checked_grch38'</span>,<span class="st">'n_checked_chm13'</span>,<span class="st">'switch_error_rate_grch38'</span>,<span class="st">'switch_error_rate_chm13'</span>,</span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>                            <span class="st">'n_gt_checked_grch38'</span>,<span class="st">'n_gt_checked_chm13'</span>,<span class="st">'genotype_error_rate_grch38'</span>,<span class="st">'genotype_error_rate_chm13'</span>,</span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a>                            <span class="st">'abs_diff_GER'</span>,<span class="st">'rel_drop_GER'</span>,<span class="st">'abs_diff_SER'</span>,<span class="st">'rel_drop_SER'</span>]</span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a>                    ].sort_values(<span class="st">'rel_drop_SER'</span>,ascending<span class="op">=</span><span class="va">False</span>).reset_index(drop<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb53-6"><a href="#cb53-6" aria-hidden="true" tabindex="-1"></a>cnv_clean.to_csv(<span class="st">'tables/decipher_cnv_region_performances_rephased_1KGP_variants_HPRC_ground_truth.tsv'</span>, sep<span class="op">=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>, index<span class="op">=</span><span class="va">False</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-56" class="cell" data-execution_count="29">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>cnv_slop_clean <span class="op">=</span> cnvslop.loc[(cnvslop.method_of_phasing<span class="op">==</span><span class="st">'1kgp_variation_phased_with_reference_panel'</span>)<span class="op">&amp;</span>(cnvslop.ground_truth_data_source<span class="op">==</span><span class="st">'HPRC_samples'</span>)<span class="op">&amp;</span>(cnvslop.n_checked_grch38 <span class="op">&gt;</span> <span class="dv">50</span>)<span class="op">&amp;</span>(cnvslop.n_checked_chm13 <span class="op">&gt;</span> <span class="dv">50</span>),</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>                                [<span class="st">'Syndrome'</span>,<span class="st">'n_switch_errors_grch38'</span>,<span class="st">'n_switch_errors_chm13'</span>,<span class="st">'n_checked_grch38'</span>,<span class="st">'n_checked_chm13'</span>,<span class="st">'switch_error_rate_grch38'</span>,<span class="st">'switch_error_rate_chm13'</span>,</span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a>                                            <span class="st">'n_gt_checked_grch38'</span>,<span class="st">'n_gt_checked_chm13'</span>,<span class="st">'genotype_error_rate_grch38'</span>,<span class="st">'genotype_error_rate_chm13'</span>,</span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a>                                            <span class="st">'abs_diff_GER'</span>,<span class="st">'rel_drop_GER'</span>, <span class="st">'abs_diff_SER'</span>,<span class="st">'rel_drop_SER'</span>]</span>
<span id="cb54-5"><a href="#cb54-5" aria-hidden="true" tabindex="-1"></a>                            ].sort_values(<span class="st">'rel_drop_SER'</span>,ascending<span class="op">=</span><span class="va">False</span>).reset_index(drop<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb54-6"><a href="#cb54-6" aria-hidden="true" tabindex="-1"></a>cnv_slop_clean.to_csv(<span class="st">'tables/decipher_cnv_region_performances_rephased_1KGP_variants_HPRC_ground_truth_plusminus_1mb.tsv'</span>, sep<span class="op">=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>, index<span class="op">=</span><span class="va">False</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-57" class="cell" data-execution_count="30">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="co">## Bonus easter egg - not sure what to do with this result, but I want to keep it</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a><span class="co">## Most cytobands show a very, very weak correlation between an improvement in genotyping and an improvement in switch error rate</span></span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a><span class="co">## A second cluster of cytobands show an improvement in switch error rate &gt; 0.1% that is uncorrleated with the genotyping improvement (which was basically nil)</span></span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a>fig,ax <span class="op">=</span> plt.subplots(<span class="dv">1</span>,<span class="dv">1</span>,figsize<span class="op">=</span>(<span class="dv">6</span>,<span class="dv">6</span>))</span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a>q <span class="op">=</span> cyto.loc[(cyto.method_of_phasing<span class="op">==</span><span class="st">'HPRC_variation_phased_with_reference_panel'</span>)<span class="op">&amp;</span>(cyto.ground_truth_data_source<span class="op">==</span><span class="st">'HPRC_samples'</span>)<span class="op">&amp;</span>(cyto.n_checked_grch38 <span class="op">&gt;</span> <span class="dv">50</span>)<span class="op">&amp;</span>(cyto.n_checked_chm13 <span class="op">&gt;</span> <span class="dv">50</span>)].sort_values([<span class="st">'method_of_phasing'</span>,<span class="st">'ground_truth_data_source'</span>])[</span>
<span id="cb55-7"><a href="#cb55-7" aria-hidden="true" tabindex="-1"></a>[<span class="st">'interval'</span>,<span class="st">'n_switch_errors_grch38'</span>,<span class="st">'n_switch_errors_chm13'</span>,<span class="st">'n_checked_grch38'</span>,<span class="st">'n_checked_chm13'</span>,<span class="st">'switch_error_rate_grch38'</span>,<span class="st">'switch_error_rate_chm13'</span>,<span class="st">'genotype_error_rate_grch38'</span>,<span class="st">'genotype_error_rate_chm13'</span>,<span class="st">'abs_diff_SER'</span>,<span class="st">'rel_drop_SER'</span>, <span class="st">'decipher'</span>]].sort_values(<span class="st">'rel_drop_SER'</span>,ascending<span class="op">=</span><span class="va">False</span>).reset_index(drop<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb55-8"><a href="#cb55-8" aria-hidden="true" tabindex="-1"></a>cyto[<span class="st">'abs_drop_GER'</span>] <span class="op">=</span> cyto.genotype_error_rate_grch38<span class="op">-</span>cyto.genotype_error_rate_chm13</span>
<span id="cb55-9"><a href="#cb55-9" aria-hidden="true" tabindex="-1"></a>cyto[<span class="st">'rel_drop_GER'</span>] <span class="op">=</span> <span class="dv">1</span><span class="op">-</span>(cyto.genotype_error_rate_chm13<span class="op">/</span>cyto.genotype_error_rate_grch38)</span>
<span id="cb55-10"><a href="#cb55-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-11"><a href="#cb55-11" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> sns.scatterplot(x<span class="op">=</span><span class="st">'abs_diff_SER'</span>, y<span class="op">=</span><span class="st">'abs_drop_GER'</span>, data<span class="op">=</span>cyto, ax<span class="op">=</span>ax,alpha<span class="op">=</span><span class="fl">0.4</span>)</span>
<span id="cb55-12"><a href="#cb55-12" aria-hidden="true" tabindex="-1"></a>ax.set_yscale(<span class="st">'log'</span>)</span>
<span id="cb55-13"><a href="#cb55-13" aria-hidden="true" tabindex="-1"></a>ax.set_xscale(<span class="st">'log'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="make_plots_files/figure-html/cell-30-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="additional-regions" class="level1" data-number="3">
<h1 data-number="3"><span class="header-section-number">3</span> Additional regions</h1>
<p>Reviewers brought up reasonable curiosity about the performance of our data in different regions of the genome. This was especially fair in light of our claims that indel performance was hampered by inadequate genotyping and phasing accuracy in complex STR loci.<br> <br> These cells plot out some of the same performance metrics seen above, but broken out by variant membership in either: - GIAB’s T2T/GRCh38 STR region annotation - GIAB’s T2T/GRCh38 Segmental Duplication annotations - Platinum Pedigree Project’s STR annotations (https://pubmed.ncbi.nlm.nih.gov/39149261/, https://github.com/dashnowlab/PlatinumTRs)</p>
<div id="cell-59" class="cell" data-execution_count="31">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>region_style <span class="op">=</span> {k:v <span class="cf">for</span> k,v <span class="kw">in</span> syntenic_style.items()}</span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>region_style[<span class="st">'markers'</span>] <span class="op">=</span> <span class="va">True</span></span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a><span class="kw">del</span> region_style[<span class="st">'dashes'</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-60" class="cell" data-execution_count="32">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>number_of_variants_by_region <span class="op">=</span> binned_maf_data_regions.groupby([<span class="st">'genome'</span>,<span class="st">'region'</span>], observed<span class="op">=</span><span class="va">True</span>).<span class="bu">sum</span>(numeric_only<span class="op">=</span><span class="va">True</span>)[<span class="st">'n_checked'</span>].reset_index().pivot(index<span class="op">=</span><span class="st">'region'</span>, columns<span class="op">=</span><span class="st">'genome'</span>, values<span class="op">=</span><span class="st">'n_checked'</span>).fillna(<span class="dv">0</span>)</span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>number_of_variants_by_region</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="32">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th">genome</th>
<th data-quarto-table-cell-role="th">GRCh38</th>
<th data-quarto-table-cell-role="th">CHM13v2.0</th>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">region</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">All</td>
<td>2.120897e+09</td>
<td>4.601699e+09</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">in_STRs</td>
<td>1.286402e+08</td>
<td>3.343351e+08</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">in_segdups</td>
<td>7.460767e+07</td>
<td>9.723856e+07</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">in_platinum_STRs</td>
<td>3.984057e+08</td>
<td>9.087063e+08</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">not_in_STRs</td>
<td>1.992256e+09</td>
<td>4.267364e+09</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">not_in_segdups</td>
<td>2.046289e+09</td>
<td>4.504460e+09</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">not_in_platinum_STRs</td>
<td>1.722491e+09</td>
<td>3.692993e+09</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">multiallelic</td>
<td>3.676842e+08</td>
<td>8.839478e+08</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">biallelic</td>
<td>1.753212e+09</td>
<td>3.717751e+09</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="cell-61" class="cell" data-execution_count="33">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>binned_maf_data_regions[<span class="st">'switch_error_rate'</span>] <span class="op">=</span> binned_maf_data_regions[<span class="st">'switch_error_rate'</span>].replace(<span class="dv">0</span>,np.nan)</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>binned_maf_data_regions[<span class="st">'gt_error_rate'</span>] <span class="op">=</span> binned_maf_data_regions[<span class="st">'gt_error_rate'</span>].replace(<span class="dv">0</span>,np.nan)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-62" class="cell" data-execution_count="34">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>plt.rcParams[<span class="st">"savefig.dpi"</span>] <span class="op">=</span> <span class="dv">300</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>plt.rcParams[<span class="st">"figure.dpi"</span>] <span class="op">=</span> <span class="dv">300</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<section id="figure-2-style-plots-of-switch-error-rate-and-genotyping-error-rate-by-region-of-interest" class="level3" data-number="3.0.1">
<h3 data-number="3.0.1" class="anchored" data-anchor-id="figure-2-style-plots-of-switch-error-rate-and-genotyping-error-rate-by-region-of-interest"><span class="header-section-number">3.0.1</span> Figure 2-style plots of switch error rate and genotyping error rate by region of interest</h3>
<div id="cell-64" class="cell" data-execution_count="35">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>region_order<span class="op">=</span>[<span class="st">'not_in_STRs'</span>,<span class="st">'in_STRs'</span>]</span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a>fig, ((axA, axB),(axC, axD)) <span class="op">=</span> plt.subplots(<span class="dv">2</span>,<span class="dv">2</span>, figsize<span class="op">=</span>(two_col, <span class="dv">5</span>))</span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a>fig.set_layout_engine(layout<span class="op">=</span><span class="st">'constrained'</span>, h_pad<span class="op">=</span><span class="fl">0.15</span>, w_pad<span class="op">=</span><span class="fl">0.15</span>)</span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'switch_error_rate'</span>, style<span class="op">=</span><span class="st">'region'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> region_order,</span>
<span id="cb60-6"><a href="#cb60-6" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data_regions.loc[(binned_maf_data_regions.ground_truth_data_source <span class="op">==</span> <span class="st">'HPRC_samples'</span>)</span>
<span id="cb60-7"><a href="#cb60-7" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb60-8"><a href="#cb60-8" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.<span class="bu">type</span> <span class="op">==</span> <span class="st">'SNP'</span>)],</span>
<span id="cb60-9"><a href="#cb60-9" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axA, <span class="op">**</span>region_style)</span>
<span id="cb60-10"><a href="#cb60-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-11"><a href="#cb60-11" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'switch_error_rate'</span>, style<span class="op">=</span><span class="st">'region'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> region_order,</span>
<span id="cb60-12"><a href="#cb60-12" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data_regions.loc[(binned_maf_data_regions.ground_truth_data_source <span class="op">==</span> <span class="st">'HGSVC_samples_nontrios_only'</span>) <span class="co">#trios only</span></span>
<span id="cb60-13"><a href="#cb60-13" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb60-14"><a href="#cb60-14" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.<span class="bu">type</span> <span class="op">==</span> <span class="st">'SNP'</span>)],</span>
<span id="cb60-15"><a href="#cb60-15" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axB, <span class="op">**</span>region_style)</span>
<span id="cb60-16"><a href="#cb60-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-17"><a href="#cb60-17" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'switch_error_rate'</span>, style<span class="op">=</span><span class="st">'region'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> region_order,</span>
<span id="cb60-18"><a href="#cb60-18" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data_regions.loc[(binned_maf_data_regions.ground_truth_data_source <span class="op">==</span> <span class="st">'HPRC_samples'</span>)</span>
<span id="cb60-19"><a href="#cb60-19" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb60-20"><a href="#cb60-20" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.<span class="bu">type</span> <span class="op">==</span> <span class="st">'Indel'</span>)],</span>
<span id="cb60-21"><a href="#cb60-21" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axC, <span class="op">**</span>region_style)</span>
<span id="cb60-22"><a href="#cb60-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-23"><a href="#cb60-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-24"><a href="#cb60-24" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'switch_error_rate'</span>, style<span class="op">=</span><span class="st">'region'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> region_order,</span>
<span id="cb60-25"><a href="#cb60-25" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data_regions.loc[(binned_maf_data_regions.ground_truth_data_source <span class="op">==</span> <span class="st">'HGSVC_samples_nontrios_only'</span>) <span class="co">#trios only</span></span>
<span id="cb60-26"><a href="#cb60-26" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb60-27"><a href="#cb60-27" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.<span class="bu">type</span> <span class="op">==</span> <span class="st">'Indel'</span>)],</span>
<span id="cb60-28"><a href="#cb60-28" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axD, <span class="op">**</span>region_style)</span>
<span id="cb60-29"><a href="#cb60-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-30"><a href="#cb60-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-31"><a href="#cb60-31" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, ax <span class="kw">in</span> <span class="bu">enumerate</span>([axA, axB, axC, axD]):</span>
<span id="cb60-32"><a href="#cb60-32" aria-hidden="true" tabindex="-1"></a>    ax.set_yscale(<span class="st">'log'</span>)</span>
<span id="cb60-33"><a href="#cb60-33" aria-hidden="true" tabindex="-1"></a>    ax.set_xscale(<span class="st">'log'</span>)</span>
<span id="cb60-34"><a href="#cb60-34" aria-hidden="true" tabindex="-1"></a>    ax.set_ylim(<span class="fl">1e-1</span>, <span class="dv">60</span>)<span class="co">#50)</span></span>
<span id="cb60-35"><a href="#cb60-35" aria-hidden="true" tabindex="-1"></a>    ax.set_xlim(<span class="fl">1e-2</span>, <span class="dv">55</span>)</span>
<span id="cb60-36"><a href="#cb60-36" aria-hidden="true" tabindex="-1"></a>    <span class="co"># ax.set_box_aspect(1)</span></span>
<span id="cb60-37"><a href="#cb60-37" aria-hidden="true" tabindex="-1"></a>    ax.grid()</span>
<span id="cb60-38"><a href="#cb60-38" aria-hidden="true" tabindex="-1"></a>    ax.set_axisbelow(<span class="va">True</span>)</span>
<span id="cb60-39"><a href="#cb60-39" aria-hidden="true" tabindex="-1"></a>    ax.set_xticks(MAF_xticks)</span>
<span id="cb60-40"><a href="#cb60-40" aria-hidden="true" tabindex="-1"></a>    ax.set_yticks(SER_xticks)</span>
<span id="cb60-41"><a href="#cb60-41" aria-hidden="true" tabindex="-1"></a>    ax.xaxis.set_major_formatter(perc_formatter)</span>
<span id="cb60-42"><a href="#cb60-42" aria-hidden="true" tabindex="-1"></a>    ax.yaxis.set_major_formatter(perc_formatter)</span>
<span id="cb60-43"><a href="#cb60-43" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="kw">in</span> (<span class="dv">2</span>,<span class="dv">3</span>):</span>
<span id="cb60-44"><a href="#cb60-44" aria-hidden="true" tabindex="-1"></a>        ax.set_xlabel(<span class="st">'Minor Allele Frequency (%)'</span>, labelpad<span class="op">=</span><span class="dv">3</span>)</span>
<span id="cb60-45"><a href="#cb60-45" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb60-46"><a href="#cb60-46" aria-hidden="true" tabindex="-1"></a>        ax.set_xlabel(<span class="va">None</span>)</span>
<span id="cb60-47"><a href="#cb60-47" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="kw">in</span> (<span class="dv">0</span>,<span class="dv">2</span>):</span>
<span id="cb60-48"><a href="#cb60-48" aria-hidden="true" tabindex="-1"></a>        ax.set_ylabel(<span class="st">'Switch Error Rate (%)'</span>, labelpad<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb60-49"><a href="#cb60-49" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb60-50"><a href="#cb60-50" aria-hidden="true" tabindex="-1"></a>        ax.set_ylabel(<span class="va">None</span>)</span>
<span id="cb60-51"><a href="#cb60-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-52"><a href="#cb60-52" aria-hidden="true" tabindex="-1"></a>axA.text(<span class="op">-</span><span class="fl">.2</span>,<span class="fl">0.5</span>, <span class="st">'SNPs'</span>, fontsize<span class="op">=</span><span class="dv">10</span>, transform<span class="op">=</span>axA.transAxes, weight<span class="op">=</span><span class="st">'bold'</span>, rotation<span class="op">=</span><span class="st">'vertical'</span>, horizontalalignment<span class="op">=</span><span class="st">'right'</span>, verticalalignment<span class="op">=</span><span class="st">'center'</span>)</span>
<span id="cb60-53"><a href="#cb60-53" aria-hidden="true" tabindex="-1"></a>axC.text(<span class="op">-</span><span class="fl">.2</span>,<span class="fl">0.5</span>, <span class="st">'Indels'</span>, fontsize<span class="op">=</span><span class="dv">10</span>, transform<span class="op">=</span>axC.transAxes, weight<span class="op">=</span><span class="st">'bold'</span>, rotation<span class="op">=</span><span class="st">'vertical'</span>, horizontalalignment<span class="op">=</span><span class="st">'right'</span>, verticalalignment<span class="op">=</span><span class="st">'center'</span>)</span>
<span id="cb60-54"><a href="#cb60-54" aria-hidden="true" tabindex="-1"></a>axA.set_title(<span class="st">'Mendelian phased HPRC samples'</span>, fontsize<span class="op">=</span><span class="dv">9</span>, weight<span class="op">=</span><span class="st">'bold'</span>, pad<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb60-55"><a href="#cb60-55" aria-hidden="true" tabindex="-1"></a>axB.set_title(<span class="st">'Statistically phased HGSVC samples'</span>, fontsize<span class="op">=</span><span class="dv">9</span>, weight<span class="op">=</span><span class="st">'bold'</span>, pad<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb60-56"><a href="#cb60-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-57"><a href="#cb60-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-58"><a href="#cb60-58" aria-hidden="true" tabindex="-1"></a>legend_locs<span class="op">=</span><span class="bu">dict</span>()</span>
<span id="cb60-59"><a href="#cb60-59" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, ax <span class="kw">in</span> <span class="bu">enumerate</span>([axA, axB, axC, axD]):</span>
<span id="cb60-60"><a href="#cb60-60" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="kw">not</span> axis_has_legend(ax):</span>
<span id="cb60-61"><a href="#cb60-61" aria-hidden="true" tabindex="-1"></a>        <span class="cf">continue</span></span>
<span id="cb60-62"><a href="#cb60-62" aria-hidden="true" tabindex="-1"></a>    loc <span class="op">=</span> legend_locs.get(i, <span class="st">'best'</span>)</span>
<span id="cb60-63"><a href="#cb60-63" aria-hidden="true" tabindex="-1"></a>    handles, labels <span class="op">=</span> ax.get_legend_handles_labels()</span>
<span id="cb60-64"><a href="#cb60-64" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> h_idx, handle <span class="kw">in</span> <span class="bu">enumerate</span>(handles):</span>
<span id="cb60-65"><a href="#cb60-65" aria-hidden="true" tabindex="-1"></a>        handle.set_marker(<span class="st">'o'</span>)</span>
<span id="cb60-66"><a href="#cb60-66" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> h_idx <span class="op">&gt;=</span><span class="dv">5</span>:</span>
<span id="cb60-67"><a href="#cb60-67" aria-hidden="true" tabindex="-1"></a>            handle.set_color(<span class="st">'black'</span>)</span>
<span id="cb60-68"><a href="#cb60-68" aria-hidden="true" tabindex="-1"></a>    handles <span class="op">=</span> handles[<span class="dv">1</span>:<span class="dv">3</span>] <span class="op">+</span> handles[<span class="dv">4</span>:<span class="dv">6</span>]</span>
<span id="cb60-69"><a href="#cb60-69" aria-hidden="true" tabindex="-1"></a>    labels <span class="op">=</span> labels[<span class="dv">1</span>:<span class="dv">3</span>] <span class="op">+</span> labels[<span class="dv">4</span>:<span class="dv">6</span>]</span>
<span id="cb60-70"><a href="#cb60-70" aria-hidden="true" tabindex="-1"></a>    legend <span class="op">=</span> ax.legend(ncols<span class="op">=</span><span class="dv">1</span>, loc<span class="op">=</span>loc, handles<span class="op">=</span>handles, labels<span class="op">=</span>labels)</span>
<span id="cb60-71"><a href="#cb60-71" aria-hidden="true" tabindex="-1"></a>    legend.set_title(<span class="va">None</span>)</span>
<span id="cb60-72"><a href="#cb60-72" aria-hidden="true" tabindex="-1"></a>    legend.get_frame().<span class="bu">set</span>(</span>
<span id="cb60-73"><a href="#cb60-73" aria-hidden="true" tabindex="-1"></a>        alpha<span class="op">=</span><span class="fl">0.75</span>, boxstyle<span class="op">=</span><span class="st">'square'</span>, edgecolor<span class="op">=</span><span class="st">'grey'</span>, linewidth<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb60-74"><a href="#cb60-74" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> text <span class="kw">in</span> legend.get_texts():</span>
<span id="cb60-75"><a href="#cb60-75" aria-hidden="true" tabindex="-1"></a>        t <span class="op">=</span> text.get_text()</span>
<span id="cb60-76"><a href="#cb60-76" aria-hidden="true" tabindex="-1"></a>        new_t <span class="op">=</span> update_legend_values.get(t,t)</span>
<span id="cb60-77"><a href="#cb60-77" aria-hidden="true" tabindex="-1"></a>        text.set_text(new_t)</span>
<span id="cb60-78"><a href="#cb60-78" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-79"><a href="#cb60-79" aria-hidden="true" tabindex="-1"></a>x_pos <span class="op">=</span> <span class="fl">0.5275</span></span>
<span id="cb60-80"><a href="#cb60-80" aria-hidden="true" tabindex="-1"></a>y_pos <span class="op">=</span> <span class="fl">0.5</span></span>
<span id="cb60-81"><a href="#cb60-81" aria-hidden="true" tabindex="-1"></a>perc_of_figure<span class="op">=</span><span class="fl">0.95</span></span>
<span id="cb60-82"><a href="#cb60-82" aria-hidden="true" tabindex="-1"></a>fig.add_artist(mpl.lines.Line2D([(<span class="dv">1</span><span class="op">-</span>perc_of_figure)<span class="op">/</span><span class="dv">2</span>, <span class="dv">1</span> <span class="op">-</span> (<span class="dv">1</span><span class="op">-</span>perc_of_figure)<span class="op">/</span><span class="dv">2</span>], [y_pos, y_pos], color<span class="op">=</span><span class="st">'grey'</span>, alpha<span class="op">=</span><span class="dv">1</span>, linewidth<span class="op">=</span><span class="dv">1</span>))</span>
<span id="cb60-83"><a href="#cb60-83" aria-hidden="true" tabindex="-1"></a>fig.add_artist(mpl.lines.Line2D([x_pos, x_pos], [(<span class="dv">1</span><span class="op">-</span>perc_of_figure)<span class="op">/</span><span class="dv">2</span>, <span class="dv">1</span> <span class="op">-</span> (<span class="dv">1</span><span class="op">-</span>perc_of_figure)<span class="op">/</span><span class="dv">2</span>], color<span class="op">=</span><span class="st">'grey'</span>, alpha<span class="op">=</span><span class="dv">1</span>, linewidth<span class="op">=</span><span class="dv">1</span>))</span>
<span id="cb60-84"><a href="#cb60-84" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-85"><a href="#cb60-85" aria-hidden="true" tabindex="-1"></a>add_letter_to_ax(axA, <span class="st">'a'</span>)</span>
<span id="cb60-86"><a href="#cb60-86" aria-hidden="true" tabindex="-1"></a>add_letter_to_ax(axB, <span class="st">'b'</span>)</span>
<span id="cb60-87"><a href="#cb60-87" aria-hidden="true" tabindex="-1"></a>add_letter_to_ax(axC, <span class="st">'c'</span>)</span>
<span id="cb60-88"><a href="#cb60-88" aria-hidden="true" tabindex="-1"></a>add_letter_to_ax(axD, <span class="st">'d'</span>)</span>
<span id="cb60-89"><a href="#cb60-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-90"><a href="#cb60-90" aria-hidden="true" tabindex="-1"></a><span class="co"># plt.savefig('figures/Figure 2.png', facecolor='white')</span></span>
<span id="cb60-91"><a href="#cb60-91" aria-hidden="true" tabindex="-1"></a><span class="co"># plt.savefig('figures/Figure 2.svg', facecolor='white')</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="make_plots_files/figure-html/cell-35-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-65" class="cell" data-execution_count="36">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>region_order<span class="op">=</span>[<span class="st">'not_in_segdups'</span>,<span class="st">'in_segdups'</span>]</span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Figure 2</span></span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a>fig, ((axA, axB),(axC, axD)) <span class="op">=</span> plt.subplots(<span class="dv">2</span>,<span class="dv">2</span>, figsize<span class="op">=</span>(two_col, <span class="dv">5</span>))</span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a>fig.set_layout_engine(layout<span class="op">=</span><span class="st">'constrained'</span>, h_pad<span class="op">=</span><span class="fl">0.15</span>, w_pad<span class="op">=</span><span class="fl">0.15</span>)</span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'switch_error_rate'</span>, style<span class="op">=</span><span class="st">'region'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> region_order,</span>
<span id="cb61-7"><a href="#cb61-7" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data_regions.loc[(binned_maf_data_regions.ground_truth_data_source <span class="op">==</span> <span class="st">'HPRC_samples'</span>)</span>
<span id="cb61-8"><a href="#cb61-8" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb61-9"><a href="#cb61-9" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.<span class="bu">type</span> <span class="op">==</span> <span class="st">'SNP'</span>)],</span>
<span id="cb61-10"><a href="#cb61-10" aria-hidden="true" tabindex="-1"></a>                                    <span class="co">#   &amp; (binned_maf_data_regions.syntenic != 'All')],</span></span>
<span id="cb61-11"><a href="#cb61-11" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axA, <span class="op">**</span>region_style)</span>
<span id="cb61-12"><a href="#cb61-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-13"><a href="#cb61-13" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'switch_error_rate'</span>, style<span class="op">=</span><span class="st">'region'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> region_order,</span>
<span id="cb61-14"><a href="#cb61-14" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data_regions.loc[(binned_maf_data_regions.ground_truth_data_source <span class="op">==</span> <span class="st">'HGSVC_samples_nontrios_only'</span>) <span class="co">#trios only</span></span>
<span id="cb61-15"><a href="#cb61-15" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb61-16"><a href="#cb61-16" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.<span class="bu">type</span> <span class="op">==</span> <span class="st">'SNP'</span>)],</span>
<span id="cb61-17"><a href="#cb61-17" aria-hidden="true" tabindex="-1"></a>                                    <span class="co">#   &amp; (binned_maf_data_regions.syntenic != 'All')],</span></span>
<span id="cb61-18"><a href="#cb61-18" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axB, <span class="op">**</span>region_style)</span>
<span id="cb61-19"><a href="#cb61-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-20"><a href="#cb61-20" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'switch_error_rate'</span>, style<span class="op">=</span><span class="st">'region'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> region_order,</span>
<span id="cb61-21"><a href="#cb61-21" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data_regions.loc[(binned_maf_data_regions.ground_truth_data_source <span class="op">==</span> <span class="st">'HPRC_samples'</span>)</span>
<span id="cb61-22"><a href="#cb61-22" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb61-23"><a href="#cb61-23" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.<span class="bu">type</span> <span class="op">==</span> <span class="st">'Indel'</span>)],</span>
<span id="cb61-24"><a href="#cb61-24" aria-hidden="true" tabindex="-1"></a>                                    <span class="co">#   &amp; (binned_maf_data_regions.syntenic != 'All')],</span></span>
<span id="cb61-25"><a href="#cb61-25" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axC, <span class="op">**</span>region_style)</span>
<span id="cb61-26"><a href="#cb61-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-27"><a href="#cb61-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-28"><a href="#cb61-28" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'switch_error_rate'</span>, style<span class="op">=</span><span class="st">'region'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> region_order,</span>
<span id="cb61-29"><a href="#cb61-29" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data_regions.loc[(binned_maf_data_regions.ground_truth_data_source <span class="op">==</span> <span class="st">'HGSVC_samples_nontrios_only'</span>) <span class="co">#trios only</span></span>
<span id="cb61-30"><a href="#cb61-30" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb61-31"><a href="#cb61-31" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.<span class="bu">type</span> <span class="op">==</span> <span class="st">'Indel'</span>)],</span>
<span id="cb61-32"><a href="#cb61-32" aria-hidden="true" tabindex="-1"></a>                                    <span class="co">#   &amp; (binned_maf_data_regions.syntenic != 'All')],</span></span>
<span id="cb61-33"><a href="#cb61-33" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axD, <span class="op">**</span>region_style)</span>
<span id="cb61-34"><a href="#cb61-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-35"><a href="#cb61-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-36"><a href="#cb61-36" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, ax <span class="kw">in</span> <span class="bu">enumerate</span>([axA, axB, axC, axD]):</span>
<span id="cb61-37"><a href="#cb61-37" aria-hidden="true" tabindex="-1"></a>    ax.set_yscale(<span class="st">'log'</span>)</span>
<span id="cb61-38"><a href="#cb61-38" aria-hidden="true" tabindex="-1"></a>    ax.set_xscale(<span class="st">'log'</span>)</span>
<span id="cb61-39"><a href="#cb61-39" aria-hidden="true" tabindex="-1"></a>    ax.set_ylim(<span class="fl">1e-1</span>, <span class="dv">50</span>)</span>
<span id="cb61-40"><a href="#cb61-40" aria-hidden="true" tabindex="-1"></a>    ax.set_xlim(<span class="fl">1e-2</span>, <span class="dv">55</span>)</span>
<span id="cb61-41"><a href="#cb61-41" aria-hidden="true" tabindex="-1"></a>    <span class="co"># ax.set_box_aspect(1)</span></span>
<span id="cb61-42"><a href="#cb61-42" aria-hidden="true" tabindex="-1"></a>    ax.grid()</span>
<span id="cb61-43"><a href="#cb61-43" aria-hidden="true" tabindex="-1"></a>    ax.set_axisbelow(<span class="va">True</span>)</span>
<span id="cb61-44"><a href="#cb61-44" aria-hidden="true" tabindex="-1"></a>    ax.set_xticks(MAF_xticks)</span>
<span id="cb61-45"><a href="#cb61-45" aria-hidden="true" tabindex="-1"></a>    ax.set_yticks(SER_xticks)</span>
<span id="cb61-46"><a href="#cb61-46" aria-hidden="true" tabindex="-1"></a>    ax.xaxis.set_major_formatter(perc_formatter)</span>
<span id="cb61-47"><a href="#cb61-47" aria-hidden="true" tabindex="-1"></a>    ax.yaxis.set_major_formatter(perc_formatter)</span>
<span id="cb61-48"><a href="#cb61-48" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="kw">in</span> (<span class="dv">2</span>,<span class="dv">3</span>):</span>
<span id="cb61-49"><a href="#cb61-49" aria-hidden="true" tabindex="-1"></a>        ax.set_xlabel(<span class="st">'Minor Allele Frequency (%)'</span>, labelpad<span class="op">=</span><span class="dv">3</span>)</span>
<span id="cb61-50"><a href="#cb61-50" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb61-51"><a href="#cb61-51" aria-hidden="true" tabindex="-1"></a>        ax.set_xlabel(<span class="va">None</span>)</span>
<span id="cb61-52"><a href="#cb61-52" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="kw">in</span> (<span class="dv">0</span>,<span class="dv">2</span>):</span>
<span id="cb61-53"><a href="#cb61-53" aria-hidden="true" tabindex="-1"></a>        ax.set_ylabel(<span class="st">'Switch Error Rate (%)'</span>, labelpad<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb61-54"><a href="#cb61-54" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb61-55"><a href="#cb61-55" aria-hidden="true" tabindex="-1"></a>        ax.set_ylabel(<span class="va">None</span>)</span>
<span id="cb61-56"><a href="#cb61-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-57"><a href="#cb61-57" aria-hidden="true" tabindex="-1"></a>axA.text(<span class="op">-</span><span class="fl">.2</span>,<span class="fl">0.5</span>, <span class="st">'SNPs'</span>, fontsize<span class="op">=</span><span class="dv">10</span>, transform<span class="op">=</span>axA.transAxes, weight<span class="op">=</span><span class="st">'bold'</span>, rotation<span class="op">=</span><span class="st">'vertical'</span>, horizontalalignment<span class="op">=</span><span class="st">'right'</span>, verticalalignment<span class="op">=</span><span class="st">'center'</span>)</span>
<span id="cb61-58"><a href="#cb61-58" aria-hidden="true" tabindex="-1"></a>axC.text(<span class="op">-</span><span class="fl">.2</span>,<span class="fl">0.5</span>, <span class="st">'Indels'</span>, fontsize<span class="op">=</span><span class="dv">10</span>, transform<span class="op">=</span>axC.transAxes, weight<span class="op">=</span><span class="st">'bold'</span>, rotation<span class="op">=</span><span class="st">'vertical'</span>, horizontalalignment<span class="op">=</span><span class="st">'right'</span>, verticalalignment<span class="op">=</span><span class="st">'center'</span>)</span>
<span id="cb61-59"><a href="#cb61-59" aria-hidden="true" tabindex="-1"></a>axA.set_title(<span class="st">'Mendelian phased HPRC samples'</span>, fontsize<span class="op">=</span><span class="dv">9</span>, weight<span class="op">=</span><span class="st">'bold'</span>, pad<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb61-60"><a href="#cb61-60" aria-hidden="true" tabindex="-1"></a>axB.set_title(<span class="st">'Statistically phased HGSVC samples'</span>, fontsize<span class="op">=</span><span class="dv">9</span>, weight<span class="op">=</span><span class="st">'bold'</span>, pad<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb61-61"><a href="#cb61-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-62"><a href="#cb61-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-63"><a href="#cb61-63" aria-hidden="true" tabindex="-1"></a>legend_locs<span class="op">=</span><span class="bu">dict</span>()</span>
<span id="cb61-64"><a href="#cb61-64" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, ax <span class="kw">in</span> <span class="bu">enumerate</span>([axA, axB, axC, axD]):</span>
<span id="cb61-65"><a href="#cb61-65" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="kw">not</span> axis_has_legend(ax):</span>
<span id="cb61-66"><a href="#cb61-66" aria-hidden="true" tabindex="-1"></a>        <span class="cf">continue</span></span>
<span id="cb61-67"><a href="#cb61-67" aria-hidden="true" tabindex="-1"></a>    loc <span class="op">=</span> legend_locs.get(i, <span class="st">'best'</span>)</span>
<span id="cb61-68"><a href="#cb61-68" aria-hidden="true" tabindex="-1"></a>    handles, labels <span class="op">=</span> ax.get_legend_handles_labels()</span>
<span id="cb61-69"><a href="#cb61-69" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> h_idx, handle <span class="kw">in</span> <span class="bu">enumerate</span>(handles):</span>
<span id="cb61-70"><a href="#cb61-70" aria-hidden="true" tabindex="-1"></a>        handle.set_marker(<span class="st">'o'</span>)</span>
<span id="cb61-71"><a href="#cb61-71" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> h_idx <span class="op">&gt;=</span><span class="dv">5</span>:</span>
<span id="cb61-72"><a href="#cb61-72" aria-hidden="true" tabindex="-1"></a>            handle.set_color(<span class="st">'black'</span>)</span>
<span id="cb61-73"><a href="#cb61-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-74"><a href="#cb61-74" aria-hidden="true" tabindex="-1"></a>    handles <span class="op">=</span> handles[<span class="dv">1</span>:<span class="dv">3</span>] <span class="op">+</span> handles[<span class="dv">4</span>:<span class="dv">6</span>]</span>
<span id="cb61-75"><a href="#cb61-75" aria-hidden="true" tabindex="-1"></a>    labels <span class="op">=</span> labels[<span class="dv">1</span>:<span class="dv">3</span>] <span class="op">+</span> labels[<span class="dv">4</span>:<span class="dv">6</span>]</span>
<span id="cb61-76"><a href="#cb61-76" aria-hidden="true" tabindex="-1"></a>    legend <span class="op">=</span> ax.legend(ncols<span class="op">=</span><span class="dv">1</span>, loc<span class="op">=</span>loc, handles<span class="op">=</span>handles, labels<span class="op">=</span>labels)</span>
<span id="cb61-77"><a href="#cb61-77" aria-hidden="true" tabindex="-1"></a>    legend.set_title(<span class="va">None</span>)</span>
<span id="cb61-78"><a href="#cb61-78" aria-hidden="true" tabindex="-1"></a>    legend.get_frame().<span class="bu">set</span>(</span>
<span id="cb61-79"><a href="#cb61-79" aria-hidden="true" tabindex="-1"></a>        alpha<span class="op">=</span><span class="fl">0.75</span>, boxstyle<span class="op">=</span><span class="st">'square'</span>, edgecolor<span class="op">=</span><span class="st">'grey'</span>, linewidth<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb61-80"><a href="#cb61-80" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> text <span class="kw">in</span> legend.get_texts():</span>
<span id="cb61-81"><a href="#cb61-81" aria-hidden="true" tabindex="-1"></a>        t <span class="op">=</span> text.get_text()</span>
<span id="cb61-82"><a href="#cb61-82" aria-hidden="true" tabindex="-1"></a>        new_t <span class="op">=</span> update_legend_values.get(t,t)</span>
<span id="cb61-83"><a href="#cb61-83" aria-hidden="true" tabindex="-1"></a>        text.set_text(new_t)</span>
<span id="cb61-84"><a href="#cb61-84" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-85"><a href="#cb61-85" aria-hidden="true" tabindex="-1"></a>x_pos <span class="op">=</span> <span class="fl">0.5275</span></span>
<span id="cb61-86"><a href="#cb61-86" aria-hidden="true" tabindex="-1"></a>y_pos <span class="op">=</span> <span class="fl">0.5</span></span>
<span id="cb61-87"><a href="#cb61-87" aria-hidden="true" tabindex="-1"></a>perc_of_figure<span class="op">=</span><span class="fl">0.95</span></span>
<span id="cb61-88"><a href="#cb61-88" aria-hidden="true" tabindex="-1"></a>fig.add_artist(mpl.lines.Line2D([(<span class="dv">1</span><span class="op">-</span>perc_of_figure)<span class="op">/</span><span class="dv">2</span>, <span class="dv">1</span> <span class="op">-</span> (<span class="dv">1</span><span class="op">-</span>perc_of_figure)<span class="op">/</span><span class="dv">2</span>], [y_pos, y_pos], color<span class="op">=</span><span class="st">'grey'</span>, alpha<span class="op">=</span><span class="dv">1</span>, linewidth<span class="op">=</span><span class="dv">1</span>))</span>
<span id="cb61-89"><a href="#cb61-89" aria-hidden="true" tabindex="-1"></a>fig.add_artist(mpl.lines.Line2D([x_pos, x_pos], [(<span class="dv">1</span><span class="op">-</span>perc_of_figure)<span class="op">/</span><span class="dv">2</span>, <span class="dv">1</span> <span class="op">-</span> (<span class="dv">1</span><span class="op">-</span>perc_of_figure)<span class="op">/</span><span class="dv">2</span>], color<span class="op">=</span><span class="st">'grey'</span>, alpha<span class="op">=</span><span class="dv">1</span>, linewidth<span class="op">=</span><span class="dv">1</span>))</span>
<span id="cb61-90"><a href="#cb61-90" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-91"><a href="#cb61-91" aria-hidden="true" tabindex="-1"></a>add_letter_to_ax(axA, <span class="st">'a'</span>)</span>
<span id="cb61-92"><a href="#cb61-92" aria-hidden="true" tabindex="-1"></a>add_letter_to_ax(axB, <span class="st">'b'</span>)</span>
<span id="cb61-93"><a href="#cb61-93" aria-hidden="true" tabindex="-1"></a>add_letter_to_ax(axC, <span class="st">'c'</span>)</span>
<span id="cb61-94"><a href="#cb61-94" aria-hidden="true" tabindex="-1"></a>add_letter_to_ax(axD, <span class="st">'d'</span>)</span>
<span id="cb61-95"><a href="#cb61-95" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-96"><a href="#cb61-96" aria-hidden="true" tabindex="-1"></a><span class="co"># plt.savefig('figures/Figure 2.png', facecolor='white')</span></span>
<span id="cb61-97"><a href="#cb61-97" aria-hidden="true" tabindex="-1"></a><span class="co"># plt.savefig('figures/Figure 2.svg', facecolor='white')</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="make_plots_files/figure-html/cell-36-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-66" class="cell" data-execution_count="37">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>region_order<span class="op">=</span>[<span class="st">'in_segdups'</span>,<span class="st">'not_in_segdups'</span>]</span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Figure 2</span></span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a>fig, ((axA, axB),(axC, axD)) <span class="op">=</span> plt.subplots(<span class="dv">2</span>,<span class="dv">2</span>, figsize<span class="op">=</span>(two_col, <span class="dv">5</span>))</span>
<span id="cb62-5"><a href="#cb62-5" aria-hidden="true" tabindex="-1"></a>fig.set_layout_engine(layout<span class="op">=</span><span class="st">'constrained'</span>, h_pad<span class="op">=</span><span class="fl">0.15</span>, w_pad<span class="op">=</span><span class="fl">0.15</span>)</span>
<span id="cb62-6"><a href="#cb62-6" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'gt_error_rate'</span>, style<span class="op">=</span><span class="st">'region'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> region_order,</span>
<span id="cb62-7"><a href="#cb62-7" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data_regions.loc[(binned_maf_data_regions.ground_truth_data_source <span class="op">==</span> <span class="st">'HPRC_samples'</span>)</span>
<span id="cb62-8"><a href="#cb62-8" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb62-9"><a href="#cb62-9" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.<span class="bu">type</span> <span class="op">==</span> <span class="st">'SNP'</span>)],</span>
<span id="cb62-10"><a href="#cb62-10" aria-hidden="true" tabindex="-1"></a>                                    <span class="co">#   &amp; (binned_maf_data_regions.syntenic != 'All')],</span></span>
<span id="cb62-11"><a href="#cb62-11" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axA, <span class="op">**</span>region_style)</span>
<span id="cb62-12"><a href="#cb62-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-13"><a href="#cb62-13" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'gt_error_rate'</span>, style<span class="op">=</span><span class="st">'region'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> region_order,</span>
<span id="cb62-14"><a href="#cb62-14" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data_regions.loc[(binned_maf_data_regions.ground_truth_data_source <span class="op">==</span> <span class="st">'HGSVC_samples_nontrios_only'</span>) <span class="co">#trios only</span></span>
<span id="cb62-15"><a href="#cb62-15" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb62-16"><a href="#cb62-16" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.<span class="bu">type</span> <span class="op">==</span> <span class="st">'SNP'</span>)],</span>
<span id="cb62-17"><a href="#cb62-17" aria-hidden="true" tabindex="-1"></a>                                    <span class="co">#   &amp; (binned_maf_data_regions.syntenic != 'All')],</span></span>
<span id="cb62-18"><a href="#cb62-18" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axB, <span class="op">**</span>region_style)</span>
<span id="cb62-19"><a href="#cb62-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-20"><a href="#cb62-20" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'gt_error_rate'</span>, style<span class="op">=</span><span class="st">'region'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> region_order,</span>
<span id="cb62-21"><a href="#cb62-21" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data_regions.loc[(binned_maf_data_regions.ground_truth_data_source <span class="op">==</span> <span class="st">'HPRC_samples'</span>)</span>
<span id="cb62-22"><a href="#cb62-22" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb62-23"><a href="#cb62-23" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.<span class="bu">type</span> <span class="op">==</span> <span class="st">'Indel'</span>)],</span>
<span id="cb62-24"><a href="#cb62-24" aria-hidden="true" tabindex="-1"></a>                                    <span class="co">#   &amp; (binned_maf_data_regions.syntenic != 'All')],</span></span>
<span id="cb62-25"><a href="#cb62-25" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axC, <span class="op">**</span>region_style)</span>
<span id="cb62-26"><a href="#cb62-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-27"><a href="#cb62-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-28"><a href="#cb62-28" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'gt_error_rate'</span>, style<span class="op">=</span><span class="st">'region'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> region_order,</span>
<span id="cb62-29"><a href="#cb62-29" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data_regions.loc[(binned_maf_data_regions.ground_truth_data_source <span class="op">==</span> <span class="st">'HGSVC_samples_nontrios_only'</span>) <span class="co">#trios only</span></span>
<span id="cb62-30"><a href="#cb62-30" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb62-31"><a href="#cb62-31" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.<span class="bu">type</span> <span class="op">==</span> <span class="st">'Indel'</span>)],</span>
<span id="cb62-32"><a href="#cb62-32" aria-hidden="true" tabindex="-1"></a>                                    <span class="co">#   &amp; (binned_maf_data_regions.syntenic != 'All')],</span></span>
<span id="cb62-33"><a href="#cb62-33" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axD, <span class="op">**</span>region_style)</span>
<span id="cb62-34"><a href="#cb62-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-35"><a href="#cb62-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-36"><a href="#cb62-36" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, ax <span class="kw">in</span> <span class="bu">enumerate</span>([axA, axB, axC, axD]):</span>
<span id="cb62-37"><a href="#cb62-37" aria-hidden="true" tabindex="-1"></a>    ax.set_yscale(<span class="st">'log'</span>)</span>
<span id="cb62-38"><a href="#cb62-38" aria-hidden="true" tabindex="-1"></a>    ax.set_xscale(<span class="st">'log'</span>)</span>
<span id="cb62-39"><a href="#cb62-39" aria-hidden="true" tabindex="-1"></a>    ax.set_ylim(<span class="fl">5e-2</span>, <span class="dv">105</span>)</span>
<span id="cb62-40"><a href="#cb62-40" aria-hidden="true" tabindex="-1"></a>    ax.set_xlim(<span class="fl">1e-2</span>, <span class="dv">55</span>)</span>
<span id="cb62-41"><a href="#cb62-41" aria-hidden="true" tabindex="-1"></a>    <span class="co"># ax.set_box_aspect(1)</span></span>
<span id="cb62-42"><a href="#cb62-42" aria-hidden="true" tabindex="-1"></a>    ax.grid()</span>
<span id="cb62-43"><a href="#cb62-43" aria-hidden="true" tabindex="-1"></a>    ax.set_axisbelow(<span class="va">True</span>)</span>
<span id="cb62-44"><a href="#cb62-44" aria-hidden="true" tabindex="-1"></a>    ax.set_xticks(MAF_xticks)</span>
<span id="cb62-45"><a href="#cb62-45" aria-hidden="true" tabindex="-1"></a>    ax.set_yticks(SER_xticks<span class="op">+</span>[<span class="dv">100</span>])</span>
<span id="cb62-46"><a href="#cb62-46" aria-hidden="true" tabindex="-1"></a>    ax.xaxis.set_major_formatter(perc_formatter)</span>
<span id="cb62-47"><a href="#cb62-47" aria-hidden="true" tabindex="-1"></a>    ax.yaxis.set_major_formatter(perc_formatter)</span>
<span id="cb62-48"><a href="#cb62-48" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="kw">in</span> (<span class="dv">2</span>,<span class="dv">3</span>):</span>
<span id="cb62-49"><a href="#cb62-49" aria-hidden="true" tabindex="-1"></a>        ax.set_xlabel(<span class="st">'Minor Allele Frequency (%)'</span>, labelpad<span class="op">=</span><span class="dv">3</span>)</span>
<span id="cb62-50"><a href="#cb62-50" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb62-51"><a href="#cb62-51" aria-hidden="true" tabindex="-1"></a>        ax.set_xlabel(<span class="va">None</span>)</span>
<span id="cb62-52"><a href="#cb62-52" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="kw">in</span> (<span class="dv">0</span>,<span class="dv">2</span>):</span>
<span id="cb62-53"><a href="#cb62-53" aria-hidden="true" tabindex="-1"></a>        ax.set_ylabel(<span class="st">'Genotype Error Rate (%)'</span>, labelpad<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb62-54"><a href="#cb62-54" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb62-55"><a href="#cb62-55" aria-hidden="true" tabindex="-1"></a>        ax.set_ylabel(<span class="va">None</span>)</span>
<span id="cb62-56"><a href="#cb62-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-57"><a href="#cb62-57" aria-hidden="true" tabindex="-1"></a>axA.text(<span class="op">-</span><span class="fl">.2</span>,<span class="fl">0.5</span>, <span class="st">'SNPs'</span>, fontsize<span class="op">=</span><span class="dv">10</span>, transform<span class="op">=</span>axA.transAxes, weight<span class="op">=</span><span class="st">'bold'</span>, rotation<span class="op">=</span><span class="st">'vertical'</span>, horizontalalignment<span class="op">=</span><span class="st">'right'</span>, verticalalignment<span class="op">=</span><span class="st">'center'</span>)</span>
<span id="cb62-58"><a href="#cb62-58" aria-hidden="true" tabindex="-1"></a>axC.text(<span class="op">-</span><span class="fl">.2</span>,<span class="fl">0.5</span>, <span class="st">'Indels'</span>, fontsize<span class="op">=</span><span class="dv">10</span>, transform<span class="op">=</span>axC.transAxes, weight<span class="op">=</span><span class="st">'bold'</span>, rotation<span class="op">=</span><span class="st">'vertical'</span>, horizontalalignment<span class="op">=</span><span class="st">'right'</span>, verticalalignment<span class="op">=</span><span class="st">'center'</span>)</span>
<span id="cb62-59"><a href="#cb62-59" aria-hidden="true" tabindex="-1"></a>axA.set_title(<span class="st">'Mendelian phased HPRC samples'</span>, fontsize<span class="op">=</span><span class="dv">9</span>, weight<span class="op">=</span><span class="st">'bold'</span>, pad<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb62-60"><a href="#cb62-60" aria-hidden="true" tabindex="-1"></a>axB.set_title(<span class="st">'Statistically phased HGSVC samples'</span>, fontsize<span class="op">=</span><span class="dv">9</span>, weight<span class="op">=</span><span class="st">'bold'</span>, pad<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb62-61"><a href="#cb62-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-62"><a href="#cb62-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-63"><a href="#cb62-63" aria-hidden="true" tabindex="-1"></a>legend_locs<span class="op">=</span><span class="bu">dict</span>()</span>
<span id="cb62-64"><a href="#cb62-64" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, ax <span class="kw">in</span> <span class="bu">enumerate</span>([axA, axB, axC, axD]):</span>
<span id="cb62-65"><a href="#cb62-65" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="kw">not</span> axis_has_legend(ax):</span>
<span id="cb62-66"><a href="#cb62-66" aria-hidden="true" tabindex="-1"></a>        <span class="cf">continue</span></span>
<span id="cb62-67"><a href="#cb62-67" aria-hidden="true" tabindex="-1"></a>    loc <span class="op">=</span> legend_locs.get(i, <span class="st">'best'</span>)</span>
<span id="cb62-68"><a href="#cb62-68" aria-hidden="true" tabindex="-1"></a>    handles, labels <span class="op">=</span> ax.get_legend_handles_labels()</span>
<span id="cb62-69"><a href="#cb62-69" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> h_idx, handle <span class="kw">in</span> <span class="bu">enumerate</span>(handles):</span>
<span id="cb62-70"><a href="#cb62-70" aria-hidden="true" tabindex="-1"></a>        handle.set_marker(<span class="st">'o'</span>)</span>
<span id="cb62-71"><a href="#cb62-71" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> h_idx <span class="op">&gt;=</span><span class="dv">5</span>:</span>
<span id="cb62-72"><a href="#cb62-72" aria-hidden="true" tabindex="-1"></a>            handle.set_color(<span class="st">'black'</span>)</span>
<span id="cb62-73"><a href="#cb62-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-74"><a href="#cb62-74" aria-hidden="true" tabindex="-1"></a>    handles <span class="op">=</span> handles[<span class="dv">1</span>:<span class="dv">3</span>] <span class="op">+</span> handles[<span class="dv">4</span>:<span class="dv">6</span>]</span>
<span id="cb62-75"><a href="#cb62-75" aria-hidden="true" tabindex="-1"></a>    labels <span class="op">=</span> labels[<span class="dv">1</span>:<span class="dv">3</span>] <span class="op">+</span> labels[<span class="dv">4</span>:<span class="dv">6</span>]</span>
<span id="cb62-76"><a href="#cb62-76" aria-hidden="true" tabindex="-1"></a>    legend <span class="op">=</span> ax.legend(ncols<span class="op">=</span><span class="dv">1</span>, loc<span class="op">=</span>loc, handles<span class="op">=</span>handles, labels<span class="op">=</span>labels)</span>
<span id="cb62-77"><a href="#cb62-77" aria-hidden="true" tabindex="-1"></a>    legend.set_title(<span class="va">None</span>)</span>
<span id="cb62-78"><a href="#cb62-78" aria-hidden="true" tabindex="-1"></a>    legend.get_frame().<span class="bu">set</span>(</span>
<span id="cb62-79"><a href="#cb62-79" aria-hidden="true" tabindex="-1"></a>        alpha<span class="op">=</span><span class="fl">0.75</span>, boxstyle<span class="op">=</span><span class="st">'square'</span>, edgecolor<span class="op">=</span><span class="st">'grey'</span>, linewidth<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb62-80"><a href="#cb62-80" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> text <span class="kw">in</span> legend.get_texts():</span>
<span id="cb62-81"><a href="#cb62-81" aria-hidden="true" tabindex="-1"></a>        t <span class="op">=</span> text.get_text()</span>
<span id="cb62-82"><a href="#cb62-82" aria-hidden="true" tabindex="-1"></a>        new_t <span class="op">=</span> update_legend_values.get(t,t)</span>
<span id="cb62-83"><a href="#cb62-83" aria-hidden="true" tabindex="-1"></a>        text.set_text(new_t)</span>
<span id="cb62-84"><a href="#cb62-84" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-85"><a href="#cb62-85" aria-hidden="true" tabindex="-1"></a>x_pos <span class="op">=</span> <span class="fl">0.5275</span></span>
<span id="cb62-86"><a href="#cb62-86" aria-hidden="true" tabindex="-1"></a>y_pos <span class="op">=</span> <span class="fl">0.5</span></span>
<span id="cb62-87"><a href="#cb62-87" aria-hidden="true" tabindex="-1"></a>perc_of_figure<span class="op">=</span><span class="fl">0.95</span></span>
<span id="cb62-88"><a href="#cb62-88" aria-hidden="true" tabindex="-1"></a>fig.add_artist(mpl.lines.Line2D([(<span class="dv">1</span><span class="op">-</span>perc_of_figure)<span class="op">/</span><span class="dv">2</span>, <span class="dv">1</span> <span class="op">-</span> (<span class="dv">1</span><span class="op">-</span>perc_of_figure)<span class="op">/</span><span class="dv">2</span>], [y_pos, y_pos], color<span class="op">=</span><span class="st">'grey'</span>, alpha<span class="op">=</span><span class="dv">1</span>, linewidth<span class="op">=</span><span class="dv">1</span>))</span>
<span id="cb62-89"><a href="#cb62-89" aria-hidden="true" tabindex="-1"></a>fig.add_artist(mpl.lines.Line2D([x_pos, x_pos], [(<span class="dv">1</span><span class="op">-</span>perc_of_figure)<span class="op">/</span><span class="dv">2</span>, <span class="dv">1</span> <span class="op">-</span> (<span class="dv">1</span><span class="op">-</span>perc_of_figure)<span class="op">/</span><span class="dv">2</span>], color<span class="op">=</span><span class="st">'grey'</span>, alpha<span class="op">=</span><span class="dv">1</span>, linewidth<span class="op">=</span><span class="dv">1</span>))</span>
<span id="cb62-90"><a href="#cb62-90" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-91"><a href="#cb62-91" aria-hidden="true" tabindex="-1"></a>add_letter_to_ax(axA, <span class="st">'a'</span>)</span>
<span id="cb62-92"><a href="#cb62-92" aria-hidden="true" tabindex="-1"></a>add_letter_to_ax(axB, <span class="st">'b'</span>)</span>
<span id="cb62-93"><a href="#cb62-93" aria-hidden="true" tabindex="-1"></a>add_letter_to_ax(axC, <span class="st">'c'</span>)</span>
<span id="cb62-94"><a href="#cb62-94" aria-hidden="true" tabindex="-1"></a>add_letter_to_ax(axD, <span class="st">'d'</span>)</span>
<span id="cb62-95"><a href="#cb62-95" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-96"><a href="#cb62-96" aria-hidden="true" tabindex="-1"></a><span class="co"># plt.savefig('figures/Figure 2.png', facecolor='white')</span></span>
<span id="cb62-97"><a href="#cb62-97" aria-hidden="true" tabindex="-1"></a><span class="co"># plt.savefig('figures/Figure 2.svg', facecolor='white')</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="make_plots_files/figure-html/cell-37-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-67" class="cell" data-execution_count="38">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>region_order<span class="op">=</span>[<span class="st">'biallelic'</span>,<span class="st">'multiallelic'</span>]</span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Figure 2</span></span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>fig, ((axA, axB),(axC, axD)) <span class="op">=</span> plt.subplots(<span class="dv">2</span>,<span class="dv">2</span>, figsize<span class="op">=</span>(two_col, <span class="dv">5</span>))</span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a>fig.set_layout_engine(layout<span class="op">=</span><span class="st">'constrained'</span>, h_pad<span class="op">=</span><span class="fl">0.15</span>, w_pad<span class="op">=</span><span class="fl">0.15</span>)</span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'switch_error_rate'</span>, style<span class="op">=</span><span class="st">'region'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> region_order,</span>
<span id="cb63-6"><a href="#cb63-6" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data_regions.loc[(binned_maf_data_regions.ground_truth_data_source <span class="op">==</span> <span class="st">'HPRC_samples'</span>)</span>
<span id="cb63-7"><a href="#cb63-7" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb63-8"><a href="#cb63-8" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.<span class="bu">type</span> <span class="op">==</span> <span class="st">'SNP'</span>)],</span>
<span id="cb63-9"><a href="#cb63-9" aria-hidden="true" tabindex="-1"></a>                                    <span class="co">#   &amp; (binned_maf_data_regions.syntenic != 'All')],</span></span>
<span id="cb63-10"><a href="#cb63-10" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axA, <span class="op">**</span>region_style)</span>
<span id="cb63-11"><a href="#cb63-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-12"><a href="#cb63-12" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'switch_error_rate'</span>, style<span class="op">=</span><span class="st">'region'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> region_order,</span>
<span id="cb63-13"><a href="#cb63-13" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data_regions.loc[(binned_maf_data_regions.ground_truth_data_source <span class="op">==</span> <span class="st">'HGSVC_samples_nontrios_only'</span>) <span class="co">#trios only</span></span>
<span id="cb63-14"><a href="#cb63-14" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb63-15"><a href="#cb63-15" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.<span class="bu">type</span> <span class="op">==</span> <span class="st">'SNP'</span>)],</span>
<span id="cb63-16"><a href="#cb63-16" aria-hidden="true" tabindex="-1"></a>                                    <span class="co">#   &amp; (binned_maf_data_regions.syntenic != 'All')],</span></span>
<span id="cb63-17"><a href="#cb63-17" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axB, <span class="op">**</span>region_style)</span>
<span id="cb63-18"><a href="#cb63-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-19"><a href="#cb63-19" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'switch_error_rate'</span>, style<span class="op">=</span><span class="st">'region'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> region_order,</span>
<span id="cb63-20"><a href="#cb63-20" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data_regions.loc[(binned_maf_data_regions.ground_truth_data_source <span class="op">==</span> <span class="st">'HPRC_samples'</span>)</span>
<span id="cb63-21"><a href="#cb63-21" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb63-22"><a href="#cb63-22" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.<span class="bu">type</span> <span class="op">==</span> <span class="st">'Indel'</span>)],</span>
<span id="cb63-23"><a href="#cb63-23" aria-hidden="true" tabindex="-1"></a>                                    <span class="co">#   &amp; (binned_maf_data_regions.syntenic != 'All')],</span></span>
<span id="cb63-24"><a href="#cb63-24" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axC, <span class="op">**</span>region_style)</span>
<span id="cb63-25"><a href="#cb63-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-26"><a href="#cb63-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-27"><a href="#cb63-27" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'switch_error_rate'</span>, style<span class="op">=</span><span class="st">'region'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> region_order,</span>
<span id="cb63-28"><a href="#cb63-28" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data_regions.loc[(binned_maf_data_regions.ground_truth_data_source <span class="op">==</span> <span class="st">'HGSVC_samples_nontrios_only'</span>) <span class="co">#trios only</span></span>
<span id="cb63-29"><a href="#cb63-29" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb63-30"><a href="#cb63-30" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.<span class="bu">type</span> <span class="op">==</span> <span class="st">'Indel'</span>)],</span>
<span id="cb63-31"><a href="#cb63-31" aria-hidden="true" tabindex="-1"></a>                                    <span class="co">#   &amp; (binned_maf_data_regions.syntenic != 'All')],</span></span>
<span id="cb63-32"><a href="#cb63-32" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axD, <span class="op">**</span>region_style)</span>
<span id="cb63-33"><a href="#cb63-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-34"><a href="#cb63-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-35"><a href="#cb63-35" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, ax <span class="kw">in</span> <span class="bu">enumerate</span>([axA, axB, axC, axD]):</span>
<span id="cb63-36"><a href="#cb63-36" aria-hidden="true" tabindex="-1"></a>    ax.set_yscale(<span class="st">'log'</span>)</span>
<span id="cb63-37"><a href="#cb63-37" aria-hidden="true" tabindex="-1"></a>    ax.set_xscale(<span class="st">'log'</span>)</span>
<span id="cb63-38"><a href="#cb63-38" aria-hidden="true" tabindex="-1"></a>    ax.set_ylim(<span class="fl">1e-1</span>, <span class="dv">50</span>)</span>
<span id="cb63-39"><a href="#cb63-39" aria-hidden="true" tabindex="-1"></a>    ax.set_xlim(<span class="fl">1e-2</span>, <span class="dv">55</span>)</span>
<span id="cb63-40"><a href="#cb63-40" aria-hidden="true" tabindex="-1"></a>    <span class="co"># ax.set_box_aspect(1)</span></span>
<span id="cb63-41"><a href="#cb63-41" aria-hidden="true" tabindex="-1"></a>    ax.grid()</span>
<span id="cb63-42"><a href="#cb63-42" aria-hidden="true" tabindex="-1"></a>    ax.set_axisbelow(<span class="va">True</span>)</span>
<span id="cb63-43"><a href="#cb63-43" aria-hidden="true" tabindex="-1"></a>    ax.set_xticks(MAF_xticks)</span>
<span id="cb63-44"><a href="#cb63-44" aria-hidden="true" tabindex="-1"></a>    ax.set_yticks(SER_xticks)</span>
<span id="cb63-45"><a href="#cb63-45" aria-hidden="true" tabindex="-1"></a>    ax.xaxis.set_major_formatter(perc_formatter)</span>
<span id="cb63-46"><a href="#cb63-46" aria-hidden="true" tabindex="-1"></a>    ax.yaxis.set_major_formatter(perc_formatter)</span>
<span id="cb63-47"><a href="#cb63-47" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="kw">in</span> (<span class="dv">2</span>,<span class="dv">3</span>):</span>
<span id="cb63-48"><a href="#cb63-48" aria-hidden="true" tabindex="-1"></a>        ax.set_xlabel(<span class="st">'Minor Allele Frequency (%)'</span>, labelpad<span class="op">=</span><span class="dv">3</span>)</span>
<span id="cb63-49"><a href="#cb63-49" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb63-50"><a href="#cb63-50" aria-hidden="true" tabindex="-1"></a>        ax.set_xlabel(<span class="va">None</span>)</span>
<span id="cb63-51"><a href="#cb63-51" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="kw">in</span> (<span class="dv">0</span>,<span class="dv">2</span>):</span>
<span id="cb63-52"><a href="#cb63-52" aria-hidden="true" tabindex="-1"></a>        ax.set_ylabel(<span class="st">'Switch Error Rate (%)'</span>, labelpad<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb63-53"><a href="#cb63-53" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb63-54"><a href="#cb63-54" aria-hidden="true" tabindex="-1"></a>        ax.set_ylabel(<span class="va">None</span>)</span>
<span id="cb63-55"><a href="#cb63-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-56"><a href="#cb63-56" aria-hidden="true" tabindex="-1"></a>axA.text(<span class="op">-</span><span class="fl">.2</span>,<span class="fl">0.5</span>, <span class="st">'SNPs'</span>, fontsize<span class="op">=</span><span class="dv">10</span>, transform<span class="op">=</span>axA.transAxes, weight<span class="op">=</span><span class="st">'bold'</span>, rotation<span class="op">=</span><span class="st">'vertical'</span>, horizontalalignment<span class="op">=</span><span class="st">'right'</span>, verticalalignment<span class="op">=</span><span class="st">'center'</span>)</span>
<span id="cb63-57"><a href="#cb63-57" aria-hidden="true" tabindex="-1"></a>axC.text(<span class="op">-</span><span class="fl">.2</span>,<span class="fl">0.5</span>, <span class="st">'Indels'</span>, fontsize<span class="op">=</span><span class="dv">10</span>, transform<span class="op">=</span>axC.transAxes, weight<span class="op">=</span><span class="st">'bold'</span>, rotation<span class="op">=</span><span class="st">'vertical'</span>, horizontalalignment<span class="op">=</span><span class="st">'right'</span>, verticalalignment<span class="op">=</span><span class="st">'center'</span>)</span>
<span id="cb63-58"><a href="#cb63-58" aria-hidden="true" tabindex="-1"></a>axA.set_title(<span class="st">'Mendelian phased HPRC samples'</span>, fontsize<span class="op">=</span><span class="dv">9</span>, weight<span class="op">=</span><span class="st">'bold'</span>, pad<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb63-59"><a href="#cb63-59" aria-hidden="true" tabindex="-1"></a>axB.set_title(<span class="st">'Statistically phased HGSVC samples'</span>, fontsize<span class="op">=</span><span class="dv">9</span>, weight<span class="op">=</span><span class="st">'bold'</span>, pad<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb63-60"><a href="#cb63-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-61"><a href="#cb63-61" aria-hidden="true" tabindex="-1"></a>legend_locs<span class="op">=</span><span class="bu">dict</span>()</span>
<span id="cb63-62"><a href="#cb63-62" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, ax <span class="kw">in</span> <span class="bu">enumerate</span>([axA, axB, axC, axD]):</span>
<span id="cb63-63"><a href="#cb63-63" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="kw">not</span> axis_has_legend(ax):</span>
<span id="cb63-64"><a href="#cb63-64" aria-hidden="true" tabindex="-1"></a>        <span class="cf">continue</span></span>
<span id="cb63-65"><a href="#cb63-65" aria-hidden="true" tabindex="-1"></a>    loc <span class="op">=</span> legend_locs.get(i, <span class="st">'upper right'</span>)</span>
<span id="cb63-66"><a href="#cb63-66" aria-hidden="true" tabindex="-1"></a>    handles, labels <span class="op">=</span> ax.get_legend_handles_labels()</span>
<span id="cb63-67"><a href="#cb63-67" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> h_idx, handle <span class="kw">in</span> <span class="bu">enumerate</span>(handles):</span>
<span id="cb63-68"><a href="#cb63-68" aria-hidden="true" tabindex="-1"></a>        handle.set_marker(<span class="st">'o'</span>)</span>
<span id="cb63-69"><a href="#cb63-69" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> h_idx <span class="op">&gt;=</span><span class="dv">5</span>:</span>
<span id="cb63-70"><a href="#cb63-70" aria-hidden="true" tabindex="-1"></a>            handle.set_color(<span class="st">'black'</span>)</span>
<span id="cb63-71"><a href="#cb63-71" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-72"><a href="#cb63-72" aria-hidden="true" tabindex="-1"></a>    handles <span class="op">=</span> handles[<span class="dv">1</span>:<span class="dv">3</span>] <span class="op">+</span> handles[<span class="dv">4</span>:<span class="dv">6</span>]</span>
<span id="cb63-73"><a href="#cb63-73" aria-hidden="true" tabindex="-1"></a>    labels <span class="op">=</span> labels[<span class="dv">1</span>:<span class="dv">3</span>] <span class="op">+</span> labels[<span class="dv">4</span>:<span class="dv">6</span>]</span>
<span id="cb63-74"><a href="#cb63-74" aria-hidden="true" tabindex="-1"></a>    legend <span class="op">=</span> ax.legend(ncols<span class="op">=</span><span class="dv">1</span>, loc<span class="op">=</span>loc, handles<span class="op">=</span>handles, labels<span class="op">=</span>labels)</span>
<span id="cb63-75"><a href="#cb63-75" aria-hidden="true" tabindex="-1"></a>    legend.set_title(<span class="va">None</span>)</span>
<span id="cb63-76"><a href="#cb63-76" aria-hidden="true" tabindex="-1"></a>    legend.get_frame().<span class="bu">set</span>(</span>
<span id="cb63-77"><a href="#cb63-77" aria-hidden="true" tabindex="-1"></a>        alpha<span class="op">=</span><span class="fl">0.75</span>, boxstyle<span class="op">=</span><span class="st">'square'</span>, edgecolor<span class="op">=</span><span class="st">'grey'</span>, linewidth<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb63-78"><a href="#cb63-78" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> text <span class="kw">in</span> legend.get_texts():</span>
<span id="cb63-79"><a href="#cb63-79" aria-hidden="true" tabindex="-1"></a>        t <span class="op">=</span> text.get_text()</span>
<span id="cb63-80"><a href="#cb63-80" aria-hidden="true" tabindex="-1"></a>        new_t <span class="op">=</span> update_legend_values.get(t,t)</span>
<span id="cb63-81"><a href="#cb63-81" aria-hidden="true" tabindex="-1"></a>        text.set_text(new_t)</span>
<span id="cb63-82"><a href="#cb63-82" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-83"><a href="#cb63-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-84"><a href="#cb63-84" aria-hidden="true" tabindex="-1"></a>x_pos <span class="op">=</span> <span class="fl">0.5275</span></span>
<span id="cb63-85"><a href="#cb63-85" aria-hidden="true" tabindex="-1"></a>y_pos <span class="op">=</span> <span class="fl">0.5</span></span>
<span id="cb63-86"><a href="#cb63-86" aria-hidden="true" tabindex="-1"></a>perc_of_figure<span class="op">=</span><span class="fl">0.95</span></span>
<span id="cb63-87"><a href="#cb63-87" aria-hidden="true" tabindex="-1"></a>fig.add_artist(mpl.lines.Line2D([(<span class="dv">1</span><span class="op">-</span>perc_of_figure)<span class="op">/</span><span class="dv">2</span>, <span class="dv">1</span> <span class="op">-</span> (<span class="dv">1</span><span class="op">-</span>perc_of_figure)<span class="op">/</span><span class="dv">2</span>], [y_pos, y_pos], color<span class="op">=</span><span class="st">'grey'</span>, alpha<span class="op">=</span><span class="dv">1</span>, linewidth<span class="op">=</span><span class="dv">1</span>))</span>
<span id="cb63-88"><a href="#cb63-88" aria-hidden="true" tabindex="-1"></a>fig.add_artist(mpl.lines.Line2D([x_pos, x_pos], [(<span class="dv">1</span><span class="op">-</span>perc_of_figure)<span class="op">/</span><span class="dv">2</span>, <span class="dv">1</span> <span class="op">-</span> (<span class="dv">1</span><span class="op">-</span>perc_of_figure)<span class="op">/</span><span class="dv">2</span>], color<span class="op">=</span><span class="st">'grey'</span>, alpha<span class="op">=</span><span class="dv">1</span>, linewidth<span class="op">=</span><span class="dv">1</span>))</span>
<span id="cb63-89"><a href="#cb63-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-90"><a href="#cb63-90" aria-hidden="true" tabindex="-1"></a>add_letter_to_ax(axA, <span class="st">'a'</span>)</span>
<span id="cb63-91"><a href="#cb63-91" aria-hidden="true" tabindex="-1"></a>add_letter_to_ax(axB, <span class="st">'b'</span>)</span>
<span id="cb63-92"><a href="#cb63-92" aria-hidden="true" tabindex="-1"></a>add_letter_to_ax(axC, <span class="st">'c'</span>)</span>
<span id="cb63-93"><a href="#cb63-93" aria-hidden="true" tabindex="-1"></a>add_letter_to_ax(axD, <span class="st">'d'</span>)</span>
<span id="cb63-94"><a href="#cb63-94" aria-hidden="true" tabindex="-1"></a><span class="co"># fig.suptitle('Switch Error Rates - Biallelic vs Multiallelic', fontsize=10, weight='bold', y=1.02)</span></span>
<span id="cb63-95"><a href="#cb63-95" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-96"><a href="#cb63-96" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="ss">f'</span><span class="sc">{</span>supplemental_figures_folder<span class="sc">}</span><span class="ss">/tmp/supplemental_xxx_fig2butSwitch_BivsMulti.png'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span>
<span id="cb63-97"><a href="#cb63-97" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="ss">f'</span><span class="sc">{</span>supplemental_figures_folder<span class="sc">}</span><span class="ss">/tmp/supplemental_xxx_fig2butSwitch_BivsMulti.svg'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-error">
<div class="ansi-escaped-output">
<pre><span class="ansi-red-fg">---------------------------------------------------------------------------</span>
<span class="ansi-red-fg">FileNotFoundError</span>                         Traceback (most recent call last)
<span class="ansi-cyan-fg">Cell</span><span class="ansi-cyan-fg"> </span><span class="ansi-green-fg">In[38]</span><span class="ansi-green-fg">, line 96</span>
<span class="ansi-green-fg">     93</span> add_letter_to_ax(axD, <span class="ansi-yellow-fg">'</span><span class="ansi-yellow-fg">d</span><span class="ansi-yellow-fg">'</span>)
<span class="ansi-green-fg">     94</span> <span style="font-style:italic;color:rgb(95,135,135)"># fig.suptitle('Switch Error Rates - Biallelic vs Multiallelic', fontsize=10, weight='bold', y=1.02)</span>
<span class="ansi-green-fg">---&gt; </span><span class="ansi-green-fg">96</span> <span class="ansi-yellow-bg">plt</span><span class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">savefig</span><span class="ansi-yellow-bg">(</span><span class="ansi-yellow-fg ansi-yellow-bg">f</span><span class="ansi-yellow-fg ansi-yellow-bg">'</span><span style="font-weight:bold;color:rgb(175,95,135)" class="ansi-yellow-bg">{</span><span class="ansi-yellow-bg">supplemental_figures_folder</span><span style="font-weight:bold;color:rgb(175,95,135)" class="ansi-yellow-bg">}</span><span class="ansi-yellow-fg ansi-yellow-bg">/tmp/supplemental_xxx_fig2butSwitch_BivsMulti.png</span><span class="ansi-yellow-fg ansi-yellow-bg">'</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-bg">facecolor</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-fg ansi-yellow-bg">'</span><span class="ansi-yellow-fg ansi-yellow-bg">white</span><span class="ansi-yellow-fg ansi-yellow-bg">'</span><span class="ansi-yellow-bg">)</span>
<span class="ansi-green-fg">     97</span> plt.savefig(<span class="ansi-yellow-fg">f</span><span class="ansi-yellow-fg">'</span><span style="font-weight:bold;color:rgb(175,95,135)">{</span>supplemental_figures_folder<span style="font-weight:bold;color:rgb(175,95,135)">}</span><span class="ansi-yellow-fg">/tmp/supplemental_xxx_fig2butSwitch_BivsMulti.svg</span><span class="ansi-yellow-fg">'</span>, facecolor=<span class="ansi-yellow-fg">'</span><span class="ansi-yellow-fg">white</span><span class="ansi-yellow-fg">'</span>)

<span class="ansi-cyan-fg">File </span><span class="ansi-green-fg">/home/lalli/.linuxbrew/opt/python@3.11/lib/python3.11/site-packages/matplotlib/pyplot.py:1228</span>, in <span class="ansi-cyan-fg">savefig</span><span class="ansi-blue-fg">(*args, **kwargs)</span>
<span class="ansi-green-fg">   1225</span> fig = gcf()
<span class="ansi-green-fg">   1226</span> <span style="font-style:italic;color:rgb(95,135,135)"># savefig default implementation has no return, so mypy is unhappy</span>
<span class="ansi-green-fg">   1227</span> <span style="font-style:italic;color:rgb(95,135,135)"># presumably this is here because subclasses can return?</span>
<span class="ansi-green-fg">-&gt; </span><span class="ansi-green-fg">1228</span> res = <span class="ansi-yellow-bg">fig</span><span class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">savefig</span><span class="ansi-yellow-bg">(</span><span class="ansi-yellow-bg">*</span><span class="ansi-yellow-bg">args</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-bg">*</span><span class="ansi-yellow-bg">*</span><span class="ansi-yellow-bg">kwargs</span><span class="ansi-yellow-bg">)</span>  <span style="font-style:italic;color:rgb(95,135,135)"># type: ignore[func-returns-value]</span>
<span class="ansi-green-fg">   1229</span> fig.canvas.draw_idle()  <span style="font-style:italic;color:rgb(95,135,135)"># Need this if 'transparent=True', to reset colors.</span>
<span class="ansi-green-fg">   1230</span> <span style="font-weight:bold;color:rgb(0,135,0)">return</span> res

<span class="ansi-cyan-fg">File </span><span class="ansi-green-fg">/home/lalli/.linuxbrew/opt/python@3.11/lib/python3.11/site-packages/matplotlib/figure.py:3395</span>, in <span class="ansi-cyan-fg">Figure.savefig</span><span class="ansi-blue-fg">(self, fname, transparent, **kwargs)</span>
<span class="ansi-green-fg">   3393</span>     <span style="font-weight:bold;color:rgb(0,135,0)">for</span> ax <span style="font-weight:bold;color:rgb(175,0,255)">in</span> <span style="color:rgb(0,135,0)">self</span>.axes:
<span class="ansi-green-fg">   3394</span>         _recursively_make_axes_transparent(stack, ax)
<span class="ansi-green-fg">-&gt; </span><span class="ansi-green-fg">3395</span> <span style="color:rgb(0,135,0)" class="ansi-yellow-bg">self</span><span class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">canvas</span><span class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">print_figure</span><span class="ansi-yellow-bg">(</span><span class="ansi-yellow-bg">fname</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-bg">*</span><span class="ansi-yellow-bg">*</span><span class="ansi-yellow-bg">kwargs</span><span class="ansi-yellow-bg">)</span>

<span class="ansi-cyan-fg">File </span><span class="ansi-green-fg">/home/lalli/.linuxbrew/opt/python@3.11/lib/python3.11/site-packages/matplotlib/backend_bases.py:2204</span>, in <span class="ansi-cyan-fg">FigureCanvasBase.print_figure</span><span class="ansi-blue-fg">(self, filename, dpi, facecolor, edgecolor, orientation, format, bbox_inches, pad_inches, bbox_extra_artists, backend, **kwargs)</span>
<span class="ansi-green-fg">   2200</span> <span style="font-weight:bold;color:rgb(0,135,0)">try</span>:
<span class="ansi-green-fg">   2201</span>     <span style="font-style:italic;color:rgb(95,135,135)"># _get_renderer may change the figure dpi (as vector formats</span>
<span class="ansi-green-fg">   2202</span>     <span style="font-style:italic;color:rgb(95,135,135)"># force the figure dpi to 72), so we need to set it again here.</span>
<span class="ansi-green-fg">   2203</span>     <span style="font-weight:bold;color:rgb(0,135,0)">with</span> cbook._setattr_cm(<span style="color:rgb(0,135,0)">self</span>.figure, dpi=dpi):
<span class="ansi-green-fg">-&gt; </span><span class="ansi-green-fg">2204</span>         result = <span class="ansi-yellow-bg">print_method</span><span class="ansi-yellow-bg">(</span>
<span class="ansi-green-fg">   2205</span> <span class="ansi-yellow-bg">            </span><span class="ansi-yellow-bg">filename</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">   2206</span> <span class="ansi-yellow-bg">            </span><span class="ansi-yellow-bg">facecolor</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">facecolor</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">   2207</span> <span class="ansi-yellow-bg">            </span><span class="ansi-yellow-bg">edgecolor</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">edgecolor</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">   2208</span> <span class="ansi-yellow-bg">            </span><span class="ansi-yellow-bg">orientation</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">orientation</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">   2209</span> <span class="ansi-yellow-bg">            </span><span class="ansi-yellow-bg">bbox_inches_restore</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">_bbox_inches_restore</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">   2210</span> <span class="ansi-yellow-bg">            </span><span class="ansi-yellow-bg">*</span><span class="ansi-yellow-bg">*</span><span class="ansi-yellow-bg">kwargs</span><span class="ansi-yellow-bg">)</span>
<span class="ansi-green-fg">   2211</span> <span style="font-weight:bold;color:rgb(0,135,0)">finally</span>:
<span class="ansi-green-fg">   2212</span>     <span style="font-weight:bold;color:rgb(0,135,0)">if</span> bbox_inches <span style="font-weight:bold;color:rgb(175,0,255)">and</span> restore_bbox:

<span class="ansi-cyan-fg">File </span><span class="ansi-green-fg">/home/lalli/.linuxbrew/opt/python@3.11/lib/python3.11/site-packages/matplotlib/backend_bases.py:2054</span>, in <span class="ansi-cyan-fg">FigureCanvasBase._switch_canvas_and_return_print_method.&lt;locals&gt;.&lt;lambda&gt;</span><span class="ansi-blue-fg">(*args, **kwargs)</span>
<span class="ansi-green-fg">   2050</span>     optional_kws = {  <span style="font-style:italic;color:rgb(95,135,135)"># Passed by print_figure for other renderers.</span>
<span class="ansi-green-fg">   2051</span>         <span class="ansi-yellow-fg">"</span><span class="ansi-yellow-fg">dpi</span><span class="ansi-yellow-fg">"</span>, <span class="ansi-yellow-fg">"</span><span class="ansi-yellow-fg">facecolor</span><span class="ansi-yellow-fg">"</span>, <span class="ansi-yellow-fg">"</span><span class="ansi-yellow-fg">edgecolor</span><span class="ansi-yellow-fg">"</span>, <span class="ansi-yellow-fg">"</span><span class="ansi-yellow-fg">orientation</span><span class="ansi-yellow-fg">"</span>,
<span class="ansi-green-fg">   2052</span>         <span class="ansi-yellow-fg">"</span><span class="ansi-yellow-fg">bbox_inches_restore</span><span class="ansi-yellow-fg">"</span>}
<span class="ansi-green-fg">   2053</span>     skip = optional_kws - {*inspect.signature(meth).parameters}
<span class="ansi-green-fg">-&gt; </span><span class="ansi-green-fg">2054</span>     print_method = functools.wraps(meth)(<span style="font-weight:bold;color:rgb(0,135,0)">lambda</span> *args, **kwargs: <span class="ansi-yellow-bg">meth</span><span class="ansi-yellow-bg">(</span>
<span class="ansi-green-fg">   2055</span> <span class="ansi-yellow-bg">        </span><span class="ansi-yellow-bg">*</span><span class="ansi-yellow-bg">args</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-bg">*</span><span class="ansi-yellow-bg">*</span><span class="ansi-yellow-bg">{</span><span class="ansi-yellow-bg">k</span><span class="ansi-yellow-bg">:</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-bg">v</span><span class="ansi-yellow-bg"> </span><span style="font-weight:bold;color:rgb(0,135,0)" class="ansi-yellow-bg">for</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-bg">k</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-bg">v</span><span class="ansi-yellow-bg"> </span><span style="font-weight:bold;color:rgb(175,0,255)" class="ansi-yellow-bg">in</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-bg">kwargs</span><span class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">items</span><span class="ansi-yellow-bg">(</span><span class="ansi-yellow-bg">)</span><span class="ansi-yellow-bg"> </span><span style="font-weight:bold;color:rgb(0,135,0)" class="ansi-yellow-bg">if</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-bg">k</span><span class="ansi-yellow-bg"> </span><span style="font-weight:bold;color:rgb(175,0,255)" class="ansi-yellow-bg">not</span><span class="ansi-yellow-bg"> </span><span style="font-weight:bold;color:rgb(175,0,255)" class="ansi-yellow-bg">in</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-bg">skip</span><span class="ansi-yellow-bg">}</span><span class="ansi-yellow-bg">)</span>)
<span class="ansi-green-fg">   2056</span> <span style="font-weight:bold;color:rgb(0,135,0)">else</span>:  <span style="font-style:italic;color:rgb(95,135,135)"># Let third-parties do as they see fit.</span>
<span class="ansi-green-fg">   2057</span>     print_method = meth

<span class="ansi-cyan-fg">File </span><span class="ansi-green-fg">/home/lalli/.linuxbrew/opt/python@3.11/lib/python3.11/site-packages/matplotlib/backends/backend_agg.py:496</span>, in <span class="ansi-cyan-fg">FigureCanvasAgg.print_png</span><span class="ansi-blue-fg">(self, filename_or_obj, metadata, pil_kwargs)</span>
<span class="ansi-green-fg">    449</span> <span style="font-weight:bold;color:rgb(0,135,0)">def</span><span style="color:rgb(188,188,188)"> </span><span class="ansi-blue-fg">print_png</span>(<span style="color:rgb(0,135,0)">self</span>, filename_or_obj, *, metadata=<span style="font-weight:bold;color:rgb(0,135,0)">None</span>, pil_kwargs=<span style="font-weight:bold;color:rgb(0,135,0)">None</span>):
<span class="ansi-green-fg">    450</span> <span style="color:rgb(188,188,188)">    </span><span style="font-style:italic" class="ansi-yellow-fg">"""</span>
<span class="ansi-green-fg">    451</span> <span style="font-style:italic" class="ansi-yellow-fg">    Write the figure to a PNG file.</span>
<span class="ansi-green-fg">    452</span> 
<span class="ansi-green-fg">   (...)</span><span class="ansi-green-fg">    494</span> <span style="font-style:italic" class="ansi-yellow-fg">        *metadata*, including the default 'Software' key.</span>
<span class="ansi-green-fg">    495</span> <span style="font-style:italic" class="ansi-yellow-fg">    """</span>
<span class="ansi-green-fg">--&gt; </span><span class="ansi-green-fg">496</span>     <span style="color:rgb(0,135,0)" class="ansi-yellow-bg">self</span><span class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">_print_pil</span><span class="ansi-yellow-bg">(</span><span class="ansi-yellow-bg">filename_or_obj</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-fg ansi-yellow-bg">"</span><span class="ansi-yellow-fg ansi-yellow-bg">png</span><span class="ansi-yellow-fg ansi-yellow-bg">"</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-bg">pil_kwargs</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-bg">metadata</span><span class="ansi-yellow-bg">)</span>

<span class="ansi-cyan-fg">File </span><span class="ansi-green-fg">/home/lalli/.linuxbrew/opt/python@3.11/lib/python3.11/site-packages/matplotlib/backends/backend_agg.py:445</span>, in <span class="ansi-cyan-fg">FigureCanvasAgg._print_pil</span><span class="ansi-blue-fg">(self, filename_or_obj, fmt, pil_kwargs, metadata)</span>
<span class="ansi-green-fg">    440</span> <span style="font-style:italic" class="ansi-yellow-fg">"""</span>
<span class="ansi-green-fg">    441</span> <span style="font-style:italic" class="ansi-yellow-fg">Draw the canvas, then save it using `.image.imsave` (to which</span>
<span class="ansi-green-fg">    442</span> <span style="font-style:italic" class="ansi-yellow-fg">*pil_kwargs* and *metadata* are forwarded).</span>
<span class="ansi-green-fg">    443</span> <span style="font-style:italic" class="ansi-yellow-fg">"""</span>
<span class="ansi-green-fg">    444</span> FigureCanvasAgg.draw(<span style="color:rgb(0,135,0)">self</span>)
<span class="ansi-green-fg">--&gt; </span><span class="ansi-green-fg">445</span> <span class="ansi-yellow-bg">mpl</span><span class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">image</span><span class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">imsave</span><span class="ansi-yellow-bg">(</span>
<span class="ansi-green-fg">    446</span> <span class="ansi-yellow-bg">    </span><span class="ansi-yellow-bg">filename_or_obj</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span style="color:rgb(0,135,0)" class="ansi-yellow-bg">self</span><span class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">buffer_rgba</span><span class="ansi-yellow-bg">(</span><span class="ansi-yellow-bg">)</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span style="color:rgb(0,135,0)" class="ansi-yellow-bg">format</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">fmt</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-bg">origin</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-fg ansi-yellow-bg">"</span><span class="ansi-yellow-fg ansi-yellow-bg">upper</span><span class="ansi-yellow-fg ansi-yellow-bg">"</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    447</span> <span class="ansi-yellow-bg">    </span><span class="ansi-yellow-bg">dpi</span><span class="ansi-yellow-bg">=</span><span style="color:rgb(0,135,0)" class="ansi-yellow-bg">self</span><span class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">figure</span><span class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">dpi</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-bg">metadata</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">metadata</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-bg">pil_kwargs</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">pil_kwargs</span><span class="ansi-yellow-bg">)</span>

<span class="ansi-cyan-fg">File </span><span class="ansi-green-fg">/home/lalli/.linuxbrew/opt/python@3.11/lib/python3.11/site-packages/matplotlib/image.py:1676</span>, in <span class="ansi-cyan-fg">imsave</span><span class="ansi-blue-fg">(fname, arr, vmin, vmax, cmap, format, origin, dpi, metadata, pil_kwargs)</span>
<span class="ansi-green-fg">   1674</span> pil_kwargs.setdefault(<span class="ansi-yellow-fg">"</span><span class="ansi-yellow-fg">format</span><span class="ansi-yellow-fg">"</span>, <span style="color:rgb(0,135,0)">format</span>)
<span class="ansi-green-fg">   1675</span> pil_kwargs.setdefault(<span class="ansi-yellow-fg">"</span><span class="ansi-yellow-fg">dpi</span><span class="ansi-yellow-fg">"</span>, (dpi, dpi))
<span class="ansi-green-fg">-&gt; </span><span class="ansi-green-fg">1676</span> <span class="ansi-yellow-bg">image</span><span class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">save</span><span class="ansi-yellow-bg">(</span><span class="ansi-yellow-bg">fname</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-bg">*</span><span class="ansi-yellow-bg">*</span><span class="ansi-yellow-bg">pil_kwargs</span><span class="ansi-yellow-bg">)</span>

<span class="ansi-cyan-fg">File </span><span class="ansi-green-fg">/home/lalli/.linuxbrew/opt/python@3.11/lib/python3.11/site-packages/PIL/Image.py:2600</span>, in <span class="ansi-cyan-fg">Image.save</span><span class="ansi-blue-fg">(self, fp, format, **params)</span>
<span class="ansi-green-fg">   2598</span>         fp = builtins.open(filename, <span class="ansi-yellow-fg">"</span><span class="ansi-yellow-fg">r+b</span><span class="ansi-yellow-fg">"</span>)
<span class="ansi-green-fg">   2599</span>     <span style="font-weight:bold;color:rgb(0,135,0)">else</span>:
<span class="ansi-green-fg">-&gt; </span><span class="ansi-green-fg">2600</span>         fp = builtins.open(filename, <span class="ansi-yellow-fg">"</span><span class="ansi-yellow-fg">w+b</span><span class="ansi-yellow-fg">"</span>)
<span class="ansi-green-fg">   2601</span> <span style="font-weight:bold;color:rgb(0,135,0)">else</span>:
<span class="ansi-green-fg">   2602</span>     fp = cast(IO[<span style="color:rgb(0,135,0)">bytes</span>], fp)

<span class="ansi-red-fg">FileNotFoundError</span>: [Errno 2] No such file or directory: '/mnt/ssd/lalli/phasing_T2T/phasing_T2T_project/figures/supplemental/tmp/supplemental_xxx_fig2butSwitch_BivsMulti.png'</pre>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="make_plots_files/figure-html/cell-38-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-68" class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>region_order<span class="op">=</span>[<span class="st">'biallelic'</span>,<span class="st">'multiallelic'</span>]</span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Figure 2</span></span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a>fig, ((axA, axB),(axC, axD)) <span class="op">=</span> plt.subplots(<span class="dv">2</span>,<span class="dv">2</span>, figsize<span class="op">=</span>(two_col, <span class="dv">5</span>))</span>
<span id="cb64-5"><a href="#cb64-5" aria-hidden="true" tabindex="-1"></a>fig.set_layout_engine(layout<span class="op">=</span><span class="st">'constrained'</span>, h_pad<span class="op">=</span><span class="fl">0.15</span>, w_pad<span class="op">=</span><span class="fl">0.15</span>)</span>
<span id="cb64-6"><a href="#cb64-6" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'gt_error_rate'</span>, style<span class="op">=</span><span class="st">'region'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> region_order,</span>
<span id="cb64-7"><a href="#cb64-7" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data_regions.loc[(binned_maf_data_regions.ground_truth_data_source <span class="op">==</span> <span class="st">'HPRC_samples'</span>)</span>
<span id="cb64-8"><a href="#cb64-8" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb64-9"><a href="#cb64-9" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.<span class="bu">type</span> <span class="op">==</span> <span class="st">'SNP'</span>)</span>
<span id="cb64-10"><a href="#cb64-10" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.region.isin(region_order))],</span>
<span id="cb64-11"><a href="#cb64-11" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axA, <span class="op">**</span>region_style)</span>
<span id="cb64-12"><a href="#cb64-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-13"><a href="#cb64-13" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'gt_error_rate'</span>, style<span class="op">=</span><span class="st">'region'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> region_order,</span>
<span id="cb64-14"><a href="#cb64-14" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data_regions.loc[(binned_maf_data_regions.ground_truth_data_source <span class="op">==</span> <span class="st">'HGSVC_samples_nontrios_only'</span>) <span class="co">#trios only</span></span>
<span id="cb64-15"><a href="#cb64-15" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb64-16"><a href="#cb64-16" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.<span class="bu">type</span> <span class="op">==</span> <span class="st">'SNP'</span>)</span>
<span id="cb64-17"><a href="#cb64-17" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.region.isin(region_order))],</span>
<span id="cb64-18"><a href="#cb64-18" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axB, <span class="op">**</span>region_style)</span>
<span id="cb64-19"><a href="#cb64-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-20"><a href="#cb64-20" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'gt_error_rate'</span>, style<span class="op">=</span><span class="st">'region'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> region_order,</span>
<span id="cb64-21"><a href="#cb64-21" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data_regions.loc[(binned_maf_data_regions.ground_truth_data_source <span class="op">==</span> <span class="st">'HPRC_samples'</span>)</span>
<span id="cb64-22"><a href="#cb64-22" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb64-23"><a href="#cb64-23" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.<span class="bu">type</span> <span class="op">==</span> <span class="st">'Indel'</span>)</span>
<span id="cb64-24"><a href="#cb64-24" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.region.isin(region_order))],</span>
<span id="cb64-25"><a href="#cb64-25" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axC, <span class="op">**</span>region_style)</span>
<span id="cb64-26"><a href="#cb64-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-27"><a href="#cb64-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-28"><a href="#cb64-28" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'gt_error_rate'</span>, style<span class="op">=</span><span class="st">'region'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> region_order,</span>
<span id="cb64-29"><a href="#cb64-29" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data_regions.loc[(binned_maf_data_regions.ground_truth_data_source <span class="op">==</span> <span class="st">'HGSVC_samples_nontrios_only'</span>) <span class="co">#trios only</span></span>
<span id="cb64-30"><a href="#cb64-30" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb64-31"><a href="#cb64-31" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.<span class="bu">type</span> <span class="op">==</span> <span class="st">'Indel'</span>)</span>
<span id="cb64-32"><a href="#cb64-32" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.region.isin(region_order))],</span>
<span id="cb64-33"><a href="#cb64-33" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axD, <span class="op">**</span>region_style)</span>
<span id="cb64-34"><a href="#cb64-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-35"><a href="#cb64-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-36"><a href="#cb64-36" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, ax <span class="kw">in</span> <span class="bu">enumerate</span>([axA, axB, axC, axD]):</span>
<span id="cb64-37"><a href="#cb64-37" aria-hidden="true" tabindex="-1"></a>    ax.set_yscale(<span class="st">'log'</span>)</span>
<span id="cb64-38"><a href="#cb64-38" aria-hidden="true" tabindex="-1"></a>    ax.set_xscale(<span class="st">'log'</span>)</span>
<span id="cb64-39"><a href="#cb64-39" aria-hidden="true" tabindex="-1"></a>    ax.set_ylim(<span class="fl">5e-2</span>, <span class="dv">105</span>)</span>
<span id="cb64-40"><a href="#cb64-40" aria-hidden="true" tabindex="-1"></a>    ax.set_xlim(<span class="fl">1e-2</span>, <span class="dv">55</span>)</span>
<span id="cb64-41"><a href="#cb64-41" aria-hidden="true" tabindex="-1"></a>    <span class="co"># ax.set_box_aspect(1)</span></span>
<span id="cb64-42"><a href="#cb64-42" aria-hidden="true" tabindex="-1"></a>    ax.grid()</span>
<span id="cb64-43"><a href="#cb64-43" aria-hidden="true" tabindex="-1"></a>    ax.set_axisbelow(<span class="va">True</span>)</span>
<span id="cb64-44"><a href="#cb64-44" aria-hidden="true" tabindex="-1"></a>    ax.set_xticks(MAF_xticks)</span>
<span id="cb64-45"><a href="#cb64-45" aria-hidden="true" tabindex="-1"></a>    ax.set_yticks(SER_xticks<span class="op">+</span>[<span class="dv">100</span>])</span>
<span id="cb64-46"><a href="#cb64-46" aria-hidden="true" tabindex="-1"></a>    ax.xaxis.set_major_formatter(perc_formatter)</span>
<span id="cb64-47"><a href="#cb64-47" aria-hidden="true" tabindex="-1"></a>    ax.yaxis.set_major_formatter(perc_formatter)</span>
<span id="cb64-48"><a href="#cb64-48" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="kw">in</span> (<span class="dv">2</span>,<span class="dv">3</span>):</span>
<span id="cb64-49"><a href="#cb64-49" aria-hidden="true" tabindex="-1"></a>        ax.set_xlabel(<span class="st">'Minor Allele Frequency (%)'</span>, labelpad<span class="op">=</span><span class="dv">3</span>)</span>
<span id="cb64-50"><a href="#cb64-50" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb64-51"><a href="#cb64-51" aria-hidden="true" tabindex="-1"></a>        ax.set_xlabel(<span class="va">None</span>)</span>
<span id="cb64-52"><a href="#cb64-52" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="kw">in</span> (<span class="dv">0</span>,<span class="dv">2</span>):</span>
<span id="cb64-53"><a href="#cb64-53" aria-hidden="true" tabindex="-1"></a>        ax.set_ylabel(<span class="st">'Genotype Error Rate (%)'</span>, labelpad<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb64-54"><a href="#cb64-54" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb64-55"><a href="#cb64-55" aria-hidden="true" tabindex="-1"></a>        ax.set_ylabel(<span class="va">None</span>)</span>
<span id="cb64-56"><a href="#cb64-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-57"><a href="#cb64-57" aria-hidden="true" tabindex="-1"></a>axA.text(<span class="op">-</span><span class="fl">.2</span>,<span class="fl">0.5</span>, <span class="st">'SNPs'</span>, fontsize<span class="op">=</span><span class="dv">10</span>, transform<span class="op">=</span>axA.transAxes, weight<span class="op">=</span><span class="st">'bold'</span>, rotation<span class="op">=</span><span class="st">'vertical'</span>, horizontalalignment<span class="op">=</span><span class="st">'right'</span>, verticalalignment<span class="op">=</span><span class="st">'center'</span>)</span>
<span id="cb64-58"><a href="#cb64-58" aria-hidden="true" tabindex="-1"></a>axC.text(<span class="op">-</span><span class="fl">.2</span>,<span class="fl">0.5</span>, <span class="st">'Indels'</span>, fontsize<span class="op">=</span><span class="dv">10</span>, transform<span class="op">=</span>axC.transAxes, weight<span class="op">=</span><span class="st">'bold'</span>, rotation<span class="op">=</span><span class="st">'vertical'</span>, horizontalalignment<span class="op">=</span><span class="st">'right'</span>, verticalalignment<span class="op">=</span><span class="st">'center'</span>)</span>
<span id="cb64-59"><a href="#cb64-59" aria-hidden="true" tabindex="-1"></a>axA.set_title(<span class="st">'Mendelian phased HPRC samples'</span>, fontsize<span class="op">=</span><span class="dv">9</span>, weight<span class="op">=</span><span class="st">'bold'</span>, pad<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb64-60"><a href="#cb64-60" aria-hidden="true" tabindex="-1"></a>axB.set_title(<span class="st">'Statistically phased HGSVC samples'</span>, fontsize<span class="op">=</span><span class="dv">9</span>, weight<span class="op">=</span><span class="st">'bold'</span>, pad<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb64-61"><a href="#cb64-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-62"><a href="#cb64-62" aria-hidden="true" tabindex="-1"></a>legend_locs <span class="op">=</span> {<span class="dv">0</span>:<span class="st">'upper right'</span>,<span class="dv">1</span>:<span class="st">'upper right'</span>,<span class="dv">2</span>:<span class="st">'upper center'</span>,<span class="dv">3</span>:<span class="st">'lower left'</span>}</span>
<span id="cb64-63"><a href="#cb64-63" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, ax <span class="kw">in</span> <span class="bu">enumerate</span>([axA, axB, axC, axD]):</span>
<span id="cb64-64"><a href="#cb64-64" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="kw">not</span> axis_has_legend(ax):</span>
<span id="cb64-65"><a href="#cb64-65" aria-hidden="true" tabindex="-1"></a>        <span class="cf">continue</span></span>
<span id="cb64-66"><a href="#cb64-66" aria-hidden="true" tabindex="-1"></a>    loc <span class="op">=</span> legend_locs.get(i, <span class="st">'upper right'</span>)</span>
<span id="cb64-67"><a href="#cb64-67" aria-hidden="true" tabindex="-1"></a>    handles, labels <span class="op">=</span> ax.get_legend_handles_labels()</span>
<span id="cb64-68"><a href="#cb64-68" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> h_idx, handle <span class="kw">in</span> <span class="bu">enumerate</span>(handles):</span>
<span id="cb64-69"><a href="#cb64-69" aria-hidden="true" tabindex="-1"></a>        handle.set_marker(<span class="st">'o'</span>)</span>
<span id="cb64-70"><a href="#cb64-70" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> h_idx <span class="op">&gt;=</span><span class="dv">5</span>:</span>
<span id="cb64-71"><a href="#cb64-71" aria-hidden="true" tabindex="-1"></a>            handle.set_color(<span class="st">'black'</span>)</span>
<span id="cb64-72"><a href="#cb64-72" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-73"><a href="#cb64-73" aria-hidden="true" tabindex="-1"></a>    handles <span class="op">=</span> handles[<span class="dv">1</span>:<span class="dv">3</span>] <span class="op">+</span> handles[<span class="dv">4</span>:<span class="dv">6</span>]</span>
<span id="cb64-74"><a href="#cb64-74" aria-hidden="true" tabindex="-1"></a>    labels <span class="op">=</span> labels[<span class="dv">1</span>:<span class="dv">3</span>] <span class="op">+</span> labels[<span class="dv">4</span>:<span class="dv">6</span>]</span>
<span id="cb64-75"><a href="#cb64-75" aria-hidden="true" tabindex="-1"></a>    <span class="co"># print (handles)</span></span>
<span id="cb64-76"><a href="#cb64-76" aria-hidden="true" tabindex="-1"></a>    legend <span class="op">=</span> ax.legend(ncols<span class="op">=</span><span class="dv">1</span>, loc<span class="op">=</span>loc, handles<span class="op">=</span>handles, labels<span class="op">=</span>labels)</span>
<span id="cb64-77"><a href="#cb64-77" aria-hidden="true" tabindex="-1"></a>    legend.set_title(<span class="va">None</span>)</span>
<span id="cb64-78"><a href="#cb64-78" aria-hidden="true" tabindex="-1"></a>    legend.get_frame().<span class="bu">set</span>(</span>
<span id="cb64-79"><a href="#cb64-79" aria-hidden="true" tabindex="-1"></a>        alpha<span class="op">=</span><span class="fl">0.75</span>, boxstyle<span class="op">=</span><span class="st">'square'</span>, edgecolor<span class="op">=</span><span class="st">'grey'</span>, linewidth<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb64-80"><a href="#cb64-80" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> text <span class="kw">in</span> legend.get_texts():</span>
<span id="cb64-81"><a href="#cb64-81" aria-hidden="true" tabindex="-1"></a>        t <span class="op">=</span> text.get_text()</span>
<span id="cb64-82"><a href="#cb64-82" aria-hidden="true" tabindex="-1"></a>        new_t <span class="op">=</span> update_legend_values.get(t,t)</span>
<span id="cb64-83"><a href="#cb64-83" aria-hidden="true" tabindex="-1"></a>        text.set_text(new_t)</span>
<span id="cb64-84"><a href="#cb64-84" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-85"><a href="#cb64-85" aria-hidden="true" tabindex="-1"></a>x_pos <span class="op">=</span> <span class="fl">0.5275</span></span>
<span id="cb64-86"><a href="#cb64-86" aria-hidden="true" tabindex="-1"></a>y_pos <span class="op">=</span> <span class="fl">0.5</span></span>
<span id="cb64-87"><a href="#cb64-87" aria-hidden="true" tabindex="-1"></a>perc_of_figure<span class="op">=</span><span class="fl">0.95</span></span>
<span id="cb64-88"><a href="#cb64-88" aria-hidden="true" tabindex="-1"></a>fig.add_artist(mpl.lines.Line2D([(<span class="dv">1</span><span class="op">-</span>perc_of_figure)<span class="op">/</span><span class="dv">2</span>, <span class="dv">1</span> <span class="op">-</span> (<span class="dv">1</span><span class="op">-</span>perc_of_figure)<span class="op">/</span><span class="dv">2</span>], [y_pos, y_pos], color<span class="op">=</span><span class="st">'grey'</span>, alpha<span class="op">=</span><span class="dv">1</span>, linewidth<span class="op">=</span><span class="dv">1</span>))</span>
<span id="cb64-89"><a href="#cb64-89" aria-hidden="true" tabindex="-1"></a>fig.add_artist(mpl.lines.Line2D([x_pos, x_pos], [(<span class="dv">1</span><span class="op">-</span>perc_of_figure)<span class="op">/</span><span class="dv">2</span>, <span class="dv">1</span> <span class="op">-</span> (<span class="dv">1</span><span class="op">-</span>perc_of_figure)<span class="op">/</span><span class="dv">2</span>], color<span class="op">=</span><span class="st">'grey'</span>, alpha<span class="op">=</span><span class="dv">1</span>, linewidth<span class="op">=</span><span class="dv">1</span>))</span>
<span id="cb64-90"><a href="#cb64-90" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-91"><a href="#cb64-91" aria-hidden="true" tabindex="-1"></a>add_letter_to_ax(axA, <span class="st">'a'</span>)</span>
<span id="cb64-92"><a href="#cb64-92" aria-hidden="true" tabindex="-1"></a>add_letter_to_ax(axB, <span class="st">'b'</span>)</span>
<span id="cb64-93"><a href="#cb64-93" aria-hidden="true" tabindex="-1"></a>add_letter_to_ax(axC, <span class="st">'c'</span>)</span>
<span id="cb64-94"><a href="#cb64-94" aria-hidden="true" tabindex="-1"></a>add_letter_to_ax(axD, <span class="st">'d'</span>)</span>
<span id="cb64-95"><a href="#cb64-95" aria-hidden="true" tabindex="-1"></a><span class="co"># fig.suptitle('Genotyping Error Rates - Biallelic vs Multiallelic', fontsize=10, weight='bold', y=1.02)</span></span>
<span id="cb64-96"><a href="#cb64-96" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="ss">f'</span><span class="sc">{</span>supplemental_figures_folder<span class="sc">}</span><span class="ss">/tmp/supplemental_xxx_fig2butGT_BivsMulti.png'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span>
<span id="cb64-97"><a href="#cb64-97" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="ss">f'</span><span class="sc">{</span>supplemental_figures_folder<span class="sc">}</span><span class="ss">/tmp/supplemental_xxx_fig2butGT_BivsMulti.svg'</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="make_plots_files/figure-html/cell-39-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-69" class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>region_order<span class="op">=</span>[<span class="st">'not_in_platinum_STRs'</span>,<span class="st">'in_platinum_STRs'</span>]</span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Figure 2</span></span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a>fig, ((axA, axB),(axC, axD)) <span class="op">=</span> plt.subplots(<span class="dv">2</span>,<span class="dv">2</span>, figsize<span class="op">=</span>(two_col, <span class="dv">5</span>))</span>
<span id="cb65-4"><a href="#cb65-4" aria-hidden="true" tabindex="-1"></a>fig.set_layout_engine(layout<span class="op">=</span><span class="st">'constrained'</span>, h_pad<span class="op">=</span><span class="fl">0.15</span>, w_pad<span class="op">=</span><span class="fl">0.15</span>)</span>
<span id="cb65-5"><a href="#cb65-5" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'switch_error_rate'</span>, style<span class="op">=</span><span class="st">'region'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> region_order,</span>
<span id="cb65-6"><a href="#cb65-6" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data_regions.loc[(binned_maf_data_regions.ground_truth_data_source <span class="op">==</span> <span class="st">'HPRC_samples'</span>)</span>
<span id="cb65-7"><a href="#cb65-7" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb65-8"><a href="#cb65-8" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.<span class="bu">type</span> <span class="op">==</span> <span class="st">'SNP'</span>)],</span>
<span id="cb65-9"><a href="#cb65-9" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axA, <span class="op">**</span>region_style)</span>
<span id="cb65-10"><a href="#cb65-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-11"><a href="#cb65-11" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'switch_error_rate'</span>, style<span class="op">=</span><span class="st">'region'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> region_order,</span>
<span id="cb65-12"><a href="#cb65-12" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data_regions.loc[(binned_maf_data_regions.ground_truth_data_source <span class="op">==</span> <span class="st">'HGSVC_samples_nontrios_only'</span>) <span class="co">#trios only</span></span>
<span id="cb65-13"><a href="#cb65-13" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb65-14"><a href="#cb65-14" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.<span class="bu">type</span> <span class="op">==</span> <span class="st">'SNP'</span>)],</span>
<span id="cb65-15"><a href="#cb65-15" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axB, <span class="op">**</span>region_style)</span>
<span id="cb65-16"><a href="#cb65-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-17"><a href="#cb65-17" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'switch_error_rate'</span>, style<span class="op">=</span><span class="st">'region'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> region_order,</span>
<span id="cb65-18"><a href="#cb65-18" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data_regions.loc[(binned_maf_data_regions.ground_truth_data_source <span class="op">==</span> <span class="st">'HPRC_samples'</span>)</span>
<span id="cb65-19"><a href="#cb65-19" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb65-20"><a href="#cb65-20" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.<span class="bu">type</span> <span class="op">==</span> <span class="st">'Indel'</span>)],</span>
<span id="cb65-21"><a href="#cb65-21" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axC, <span class="op">**</span>region_style)</span>
<span id="cb65-22"><a href="#cb65-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-23"><a href="#cb65-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-24"><a href="#cb65-24" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'switch_error_rate'</span>, style<span class="op">=</span><span class="st">'region'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> region_order,</span>
<span id="cb65-25"><a href="#cb65-25" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data_regions.loc[(binned_maf_data_regions.ground_truth_data_source <span class="op">==</span> <span class="st">'HGSVC_samples_nontrios_only'</span>) <span class="co">#trios only</span></span>
<span id="cb65-26"><a href="#cb65-26" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb65-27"><a href="#cb65-27" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.<span class="bu">type</span> <span class="op">==</span> <span class="st">'Indel'</span>)],</span>
<span id="cb65-28"><a href="#cb65-28" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axD, <span class="op">**</span>region_style)</span>
<span id="cb65-29"><a href="#cb65-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-30"><a href="#cb65-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-31"><a href="#cb65-31" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, ax <span class="kw">in</span> <span class="bu">enumerate</span>([axA, axB, axC, axD]):</span>
<span id="cb65-32"><a href="#cb65-32" aria-hidden="true" tabindex="-1"></a>    ax.set_yscale(<span class="st">'log'</span>)</span>
<span id="cb65-33"><a href="#cb65-33" aria-hidden="true" tabindex="-1"></a>    ax.set_xscale(<span class="st">'log'</span>)</span>
<span id="cb65-34"><a href="#cb65-34" aria-hidden="true" tabindex="-1"></a>    ax.set_ylim(<span class="fl">1e-1</span>, <span class="dv">105</span>)</span>
<span id="cb65-35"><a href="#cb65-35" aria-hidden="true" tabindex="-1"></a>    ax.set_xlim(<span class="fl">1e-2</span>, <span class="dv">55</span>)</span>
<span id="cb65-36"><a href="#cb65-36" aria-hidden="true" tabindex="-1"></a>    <span class="co"># ax.set_box_aspect(1)</span></span>
<span id="cb65-37"><a href="#cb65-37" aria-hidden="true" tabindex="-1"></a>    ax.grid()</span>
<span id="cb65-38"><a href="#cb65-38" aria-hidden="true" tabindex="-1"></a>    ax.set_axisbelow(<span class="va">True</span>)</span>
<span id="cb65-39"><a href="#cb65-39" aria-hidden="true" tabindex="-1"></a>    ax.set_xticks(MAF_xticks)</span>
<span id="cb65-40"><a href="#cb65-40" aria-hidden="true" tabindex="-1"></a>    ax.set_yticks(SER_xticks<span class="op">+</span>[<span class="dv">100</span>])</span>
<span id="cb65-41"><a href="#cb65-41" aria-hidden="true" tabindex="-1"></a>    ax.xaxis.set_major_formatter(perc_formatter)</span>
<span id="cb65-42"><a href="#cb65-42" aria-hidden="true" tabindex="-1"></a>    ax.yaxis.set_major_formatter(perc_formatter)</span>
<span id="cb65-43"><a href="#cb65-43" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="kw">in</span> (<span class="dv">2</span>,<span class="dv">3</span>):</span>
<span id="cb65-44"><a href="#cb65-44" aria-hidden="true" tabindex="-1"></a>        ax.set_xlabel(<span class="st">'Minor Allele Frequency (%)'</span>, labelpad<span class="op">=</span><span class="dv">3</span>)</span>
<span id="cb65-45"><a href="#cb65-45" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb65-46"><a href="#cb65-46" aria-hidden="true" tabindex="-1"></a>        ax.set_xlabel(<span class="va">None</span>)</span>
<span id="cb65-47"><a href="#cb65-47" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="kw">in</span> (<span class="dv">0</span>,<span class="dv">2</span>):</span>
<span id="cb65-48"><a href="#cb65-48" aria-hidden="true" tabindex="-1"></a>        ax.set_ylabel(<span class="st">'Switch Error Rate (%)'</span>, labelpad<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb65-49"><a href="#cb65-49" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb65-50"><a href="#cb65-50" aria-hidden="true" tabindex="-1"></a>        ax.set_ylabel(<span class="va">None</span>)</span>
<span id="cb65-51"><a href="#cb65-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-52"><a href="#cb65-52" aria-hidden="true" tabindex="-1"></a>axA.text(<span class="op">-</span><span class="fl">.2</span>,<span class="fl">0.5</span>, <span class="st">'SNPs'</span>, fontsize<span class="op">=</span><span class="dv">10</span>, transform<span class="op">=</span>axA.transAxes, weight<span class="op">=</span><span class="st">'bold'</span>, rotation<span class="op">=</span><span class="st">'vertical'</span>, horizontalalignment<span class="op">=</span><span class="st">'right'</span>, verticalalignment<span class="op">=</span><span class="st">'center'</span>)</span>
<span id="cb65-53"><a href="#cb65-53" aria-hidden="true" tabindex="-1"></a>axC.text(<span class="op">-</span><span class="fl">.2</span>,<span class="fl">0.5</span>, <span class="st">'Indels'</span>, fontsize<span class="op">=</span><span class="dv">10</span>, transform<span class="op">=</span>axC.transAxes, weight<span class="op">=</span><span class="st">'bold'</span>, rotation<span class="op">=</span><span class="st">'vertical'</span>, horizontalalignment<span class="op">=</span><span class="st">'right'</span>, verticalalignment<span class="op">=</span><span class="st">'center'</span>)</span>
<span id="cb65-54"><a href="#cb65-54" aria-hidden="true" tabindex="-1"></a>axA.set_title(<span class="st">'Mendelian phased HPRC samples'</span>, fontsize<span class="op">=</span><span class="dv">9</span>, weight<span class="op">=</span><span class="st">'bold'</span>, pad<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb65-55"><a href="#cb65-55" aria-hidden="true" tabindex="-1"></a>axB.set_title(<span class="st">'Statistically phased HGSVC samples'</span>, fontsize<span class="op">=</span><span class="dv">9</span>, weight<span class="op">=</span><span class="st">'bold'</span>, pad<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb65-56"><a href="#cb65-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-57"><a href="#cb65-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-58"><a href="#cb65-58" aria-hidden="true" tabindex="-1"></a>legend_locs<span class="op">=</span><span class="bu">dict</span>()</span>
<span id="cb65-59"><a href="#cb65-59" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, ax <span class="kw">in</span> <span class="bu">enumerate</span>([axA, axB, axC, axD]):</span>
<span id="cb65-60"><a href="#cb65-60" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="kw">not</span> axis_has_legend(ax):</span>
<span id="cb65-61"><a href="#cb65-61" aria-hidden="true" tabindex="-1"></a>        <span class="cf">continue</span></span>
<span id="cb65-62"><a href="#cb65-62" aria-hidden="true" tabindex="-1"></a>    loc <span class="op">=</span> legend_locs.get(i, <span class="st">'best'</span>)</span>
<span id="cb65-63"><a href="#cb65-63" aria-hidden="true" tabindex="-1"></a>    handles, labels <span class="op">=</span> ax.get_legend_handles_labels()</span>
<span id="cb65-64"><a href="#cb65-64" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> h_idx, handle <span class="kw">in</span> <span class="bu">enumerate</span>(handles):</span>
<span id="cb65-65"><a href="#cb65-65" aria-hidden="true" tabindex="-1"></a>        handle.set_marker(<span class="st">'o'</span>)</span>
<span id="cb65-66"><a href="#cb65-66" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> h_idx <span class="op">&gt;=</span><span class="dv">5</span>:</span>
<span id="cb65-67"><a href="#cb65-67" aria-hidden="true" tabindex="-1"></a>            handle.set_color(<span class="st">'black'</span>)</span>
<span id="cb65-68"><a href="#cb65-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-69"><a href="#cb65-69" aria-hidden="true" tabindex="-1"></a>    handles <span class="op">=</span> handles[<span class="dv">1</span>:<span class="dv">3</span>] <span class="op">+</span> handles[<span class="dv">4</span>:<span class="dv">6</span>]</span>
<span id="cb65-70"><a href="#cb65-70" aria-hidden="true" tabindex="-1"></a>    labels <span class="op">=</span> labels[<span class="dv">1</span>:<span class="dv">3</span>] <span class="op">+</span> labels[<span class="dv">4</span>:<span class="dv">6</span>]</span>
<span id="cb65-71"><a href="#cb65-71" aria-hidden="true" tabindex="-1"></a>    legend <span class="op">=</span> ax.legend(ncols<span class="op">=</span><span class="dv">1</span>, loc<span class="op">=</span>loc, handles<span class="op">=</span>handles, labels<span class="op">=</span>labels)</span>
<span id="cb65-72"><a href="#cb65-72" aria-hidden="true" tabindex="-1"></a>    legend.set_title(<span class="va">None</span>)</span>
<span id="cb65-73"><a href="#cb65-73" aria-hidden="true" tabindex="-1"></a>    legend.get_frame().<span class="bu">set</span>(</span>
<span id="cb65-74"><a href="#cb65-74" aria-hidden="true" tabindex="-1"></a>        alpha<span class="op">=</span><span class="fl">0.75</span>, boxstyle<span class="op">=</span><span class="st">'square'</span>, edgecolor<span class="op">=</span><span class="st">'grey'</span>, linewidth<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb65-75"><a href="#cb65-75" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> text <span class="kw">in</span> legend.get_texts():</span>
<span id="cb65-76"><a href="#cb65-76" aria-hidden="true" tabindex="-1"></a>        t <span class="op">=</span> text.get_text()</span>
<span id="cb65-77"><a href="#cb65-77" aria-hidden="true" tabindex="-1"></a>        new_t <span class="op">=</span> update_legend_values.get(t,t)</span>
<span id="cb65-78"><a href="#cb65-78" aria-hidden="true" tabindex="-1"></a>        text.set_text(new_t)</span>
<span id="cb65-79"><a href="#cb65-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-80"><a href="#cb65-80" aria-hidden="true" tabindex="-1"></a>x_pos <span class="op">=</span> <span class="fl">0.5275</span></span>
<span id="cb65-81"><a href="#cb65-81" aria-hidden="true" tabindex="-1"></a>y_pos <span class="op">=</span> <span class="fl">0.5</span></span>
<span id="cb65-82"><a href="#cb65-82" aria-hidden="true" tabindex="-1"></a>perc_of_figure<span class="op">=</span><span class="fl">0.95</span></span>
<span id="cb65-83"><a href="#cb65-83" aria-hidden="true" tabindex="-1"></a>fig.add_artist(mpl.lines.Line2D([(<span class="dv">1</span><span class="op">-</span>perc_of_figure)<span class="op">/</span><span class="dv">2</span>, <span class="dv">1</span> <span class="op">-</span> (<span class="dv">1</span><span class="op">-</span>perc_of_figure)<span class="op">/</span><span class="dv">2</span>], [y_pos, y_pos], color<span class="op">=</span><span class="st">'grey'</span>, alpha<span class="op">=</span><span class="dv">1</span>, linewidth<span class="op">=</span><span class="dv">1</span>))</span>
<span id="cb65-84"><a href="#cb65-84" aria-hidden="true" tabindex="-1"></a>fig.add_artist(mpl.lines.Line2D([x_pos, x_pos], [(<span class="dv">1</span><span class="op">-</span>perc_of_figure)<span class="op">/</span><span class="dv">2</span>, <span class="dv">1</span> <span class="op">-</span> (<span class="dv">1</span><span class="op">-</span>perc_of_figure)<span class="op">/</span><span class="dv">2</span>], color<span class="op">=</span><span class="st">'grey'</span>, alpha<span class="op">=</span><span class="dv">1</span>, linewidth<span class="op">=</span><span class="dv">1</span>))</span>
<span id="cb65-85"><a href="#cb65-85" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-86"><a href="#cb65-86" aria-hidden="true" tabindex="-1"></a><span class="co"># fig.suptitle('Phasing accuracy stratified by presence in Platinum repeat regions', fontsize=11, weight='bold', y=1.02)</span></span>
<span id="cb65-87"><a href="#cb65-87" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-88"><a href="#cb65-88" aria-hidden="true" tabindex="-1"></a>add_letter_to_ax(axA, <span class="st">'a'</span>)</span>
<span id="cb65-89"><a href="#cb65-89" aria-hidden="true" tabindex="-1"></a>add_letter_to_ax(axB, <span class="st">'b'</span>)</span>
<span id="cb65-90"><a href="#cb65-90" aria-hidden="true" tabindex="-1"></a>add_letter_to_ax(axC, <span class="st">'c'</span>)</span>
<span id="cb65-91"><a href="#cb65-91" aria-hidden="true" tabindex="-1"></a>add_letter_to_ax(axD, <span class="st">'d'</span>)</span>
<span id="cb65-92"><a href="#cb65-92" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-93"><a href="#cb65-93" aria-hidden="true" tabindex="-1"></a><span class="co"># plt.savefig('figures/Figure 2.png', facecolor='white')</span></span>
<span id="cb65-94"><a href="#cb65-94" aria-hidden="true" tabindex="-1"></a><span class="co"># plt.savefig('figures/Figure 2.svg', facecolor='white')</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="make_plots_files/figure-html/cell-40-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-70" class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>region_order<span class="op">=</span>[<span class="st">'not_in_platinum_STRs'</span>,<span class="st">'in_platinum_STRs'</span>]</span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Figure 2</span></span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a>fig, ((axA, axB),(axC, axD)) <span class="op">=</span> plt.subplots(<span class="dv">2</span>,<span class="dv">2</span>, figsize<span class="op">=</span>(two_col, <span class="dv">5</span>))</span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a>fig.set_layout_engine(layout<span class="op">=</span><span class="st">'constrained'</span>, h_pad<span class="op">=</span><span class="fl">0.15</span>, w_pad<span class="op">=</span><span class="fl">0.15</span>)</span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'gt_error_rate'</span>, style<span class="op">=</span><span class="st">'region'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> region_order,</span>
<span id="cb66-6"><a href="#cb66-6" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data_regions.loc[(binned_maf_data_regions.ground_truth_data_source <span class="op">==</span> <span class="st">'HPRC_samples'</span>)</span>
<span id="cb66-7"><a href="#cb66-7" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb66-8"><a href="#cb66-8" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.<span class="bu">type</span> <span class="op">==</span> <span class="st">'SNP'</span>)],</span>
<span id="cb66-9"><a href="#cb66-9" aria-hidden="true" tabindex="-1"></a>                                    <span class="co">#   &amp; (binned_maf_data_regions.syntenic != 'All')],</span></span>
<span id="cb66-10"><a href="#cb66-10" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axA, <span class="op">**</span>region_style)</span>
<span id="cb66-11"><a href="#cb66-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-12"><a href="#cb66-12" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'gt_error_rate'</span>, style<span class="op">=</span><span class="st">'region'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> region_order,</span>
<span id="cb66-13"><a href="#cb66-13" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data_regions.loc[(binned_maf_data_regions.ground_truth_data_source <span class="op">==</span> <span class="st">'HGSVC_samples_nontrios_only'</span>) <span class="co">#trios only</span></span>
<span id="cb66-14"><a href="#cb66-14" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb66-15"><a href="#cb66-15" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.<span class="bu">type</span> <span class="op">==</span> <span class="st">'SNP'</span>)],</span>
<span id="cb66-16"><a href="#cb66-16" aria-hidden="true" tabindex="-1"></a>                                    <span class="co">#   &amp; (binned_maf_data_regions.syntenic != 'All')],</span></span>
<span id="cb66-17"><a href="#cb66-17" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axB, <span class="op">**</span>region_style)</span>
<span id="cb66-18"><a href="#cb66-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-19"><a href="#cb66-19" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'gt_error_rate'</span>, style<span class="op">=</span><span class="st">'region'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> region_order,</span>
<span id="cb66-20"><a href="#cb66-20" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data_regions.loc[(binned_maf_data_regions.ground_truth_data_source <span class="op">==</span> <span class="st">'HPRC_samples'</span>)</span>
<span id="cb66-21"><a href="#cb66-21" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb66-22"><a href="#cb66-22" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.<span class="bu">type</span> <span class="op">==</span> <span class="st">'Indel'</span>)],</span>
<span id="cb66-23"><a href="#cb66-23" aria-hidden="true" tabindex="-1"></a>                                    <span class="co">#   &amp; (binned_maf_data_regions.syntenic != 'All')],</span></span>
<span id="cb66-24"><a href="#cb66-24" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axC, <span class="op">**</span>region_style)</span>
<span id="cb66-25"><a href="#cb66-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-26"><a href="#cb66-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-27"><a href="#cb66-27" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'MAF'</span>,y<span class="op">=</span><span class="st">'gt_error_rate'</span>, style<span class="op">=</span><span class="st">'region'</span>, hue<span class="op">=</span><span class="st">'genome'</span>, hue_order <span class="op">=</span> genome_order, style_order <span class="op">=</span> region_order,</span>
<span id="cb66-28"><a href="#cb66-28" aria-hidden="true" tabindex="-1"></a>             data<span class="op">=</span>binned_maf_data_regions.loc[(binned_maf_data_regions.ground_truth_data_source <span class="op">==</span> <span class="st">'HGSVC_samples_nontrios_only'</span>) <span class="co">#trios only</span></span>
<span id="cb66-29"><a href="#cb66-29" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.method_of_phasing <span class="op">==</span> <span class="st">'phased_with_parents_and_pedigree'</span>)</span>
<span id="cb66-30"><a href="#cb66-30" aria-hidden="true" tabindex="-1"></a>                                      <span class="op">&amp;</span> (binned_maf_data_regions.<span class="bu">type</span> <span class="op">==</span> <span class="st">'Indel'</span>)],</span>
<span id="cb66-31"><a href="#cb66-31" aria-hidden="true" tabindex="-1"></a>                                    <span class="co">#   &amp; (binned_maf_data_regions.syntenic != 'All')],</span></span>
<span id="cb66-32"><a href="#cb66-32" aria-hidden="true" tabindex="-1"></a>             legend<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axD, <span class="op">**</span>region_style)</span>
<span id="cb66-33"><a href="#cb66-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-34"><a href="#cb66-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-35"><a href="#cb66-35" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, ax <span class="kw">in</span> <span class="bu">enumerate</span>([axA, axB, axC, axD]):</span>
<span id="cb66-36"><a href="#cb66-36" aria-hidden="true" tabindex="-1"></a>    ax.set_yscale(<span class="st">'log'</span>)</span>
<span id="cb66-37"><a href="#cb66-37" aria-hidden="true" tabindex="-1"></a>    ax.set_xscale(<span class="st">'log'</span>)</span>
<span id="cb66-38"><a href="#cb66-38" aria-hidden="true" tabindex="-1"></a>    ax.set_ylim(<span class="fl">5e-2</span>, <span class="dv">105</span>)</span>
<span id="cb66-39"><a href="#cb66-39" aria-hidden="true" tabindex="-1"></a>    ax.set_xlim(<span class="fl">1e-2</span>, <span class="dv">55</span>)</span>
<span id="cb66-40"><a href="#cb66-40" aria-hidden="true" tabindex="-1"></a>    <span class="co"># ax.set_box_aspect(1)</span></span>
<span id="cb66-41"><a href="#cb66-41" aria-hidden="true" tabindex="-1"></a>    ax.grid()</span>
<span id="cb66-42"><a href="#cb66-42" aria-hidden="true" tabindex="-1"></a>    ax.set_axisbelow(<span class="va">True</span>)</span>
<span id="cb66-43"><a href="#cb66-43" aria-hidden="true" tabindex="-1"></a>    ax.set_xticks(MAF_xticks)</span>
<span id="cb66-44"><a href="#cb66-44" aria-hidden="true" tabindex="-1"></a>    ax.set_yticks(SER_xticks<span class="op">+</span>[<span class="dv">100</span>])</span>
<span id="cb66-45"><a href="#cb66-45" aria-hidden="true" tabindex="-1"></a>    ax.xaxis.set_major_formatter(perc_formatter)</span>
<span id="cb66-46"><a href="#cb66-46" aria-hidden="true" tabindex="-1"></a>    ax.yaxis.set_major_formatter(perc_formatter)</span>
<span id="cb66-47"><a href="#cb66-47" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="kw">in</span> (<span class="dv">2</span>,<span class="dv">3</span>):</span>
<span id="cb66-48"><a href="#cb66-48" aria-hidden="true" tabindex="-1"></a>        ax.set_xlabel(<span class="st">'Minor Allele Frequency (%)'</span>, labelpad<span class="op">=</span><span class="dv">3</span>)</span>
<span id="cb66-49"><a href="#cb66-49" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb66-50"><a href="#cb66-50" aria-hidden="true" tabindex="-1"></a>        ax.set_xlabel(<span class="va">None</span>)</span>
<span id="cb66-51"><a href="#cb66-51" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="kw">in</span> (<span class="dv">0</span>,<span class="dv">2</span>):</span>
<span id="cb66-52"><a href="#cb66-52" aria-hidden="true" tabindex="-1"></a>        ax.set_ylabel(<span class="st">'Genotype Error Rate (%)'</span>, labelpad<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb66-53"><a href="#cb66-53" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb66-54"><a href="#cb66-54" aria-hidden="true" tabindex="-1"></a>        ax.set_ylabel(<span class="va">None</span>)</span>
<span id="cb66-55"><a href="#cb66-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-56"><a href="#cb66-56" aria-hidden="true" tabindex="-1"></a>axA.text(<span class="op">-</span><span class="fl">.2</span>,<span class="fl">0.5</span>, <span class="st">'SNPs'</span>, fontsize<span class="op">=</span><span class="dv">10</span>, transform<span class="op">=</span>axA.transAxes, weight<span class="op">=</span><span class="st">'bold'</span>, rotation<span class="op">=</span><span class="st">'vertical'</span>, horizontalalignment<span class="op">=</span><span class="st">'right'</span>, verticalalignment<span class="op">=</span><span class="st">'center'</span>)</span>
<span id="cb66-57"><a href="#cb66-57" aria-hidden="true" tabindex="-1"></a>axC.text(<span class="op">-</span><span class="fl">.2</span>,<span class="fl">0.5</span>, <span class="st">'Indels'</span>, fontsize<span class="op">=</span><span class="dv">10</span>, transform<span class="op">=</span>axC.transAxes, weight<span class="op">=</span><span class="st">'bold'</span>, rotation<span class="op">=</span><span class="st">'vertical'</span>, horizontalalignment<span class="op">=</span><span class="st">'right'</span>, verticalalignment<span class="op">=</span><span class="st">'center'</span>)</span>
<span id="cb66-58"><a href="#cb66-58" aria-hidden="true" tabindex="-1"></a>axA.set_title(<span class="st">'Mendelian phased HPRC samples'</span>, fontsize<span class="op">=</span><span class="dv">9</span>, weight<span class="op">=</span><span class="st">'bold'</span>, pad<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb66-59"><a href="#cb66-59" aria-hidden="true" tabindex="-1"></a>axB.set_title(<span class="st">'Statistically phased HGSVC samples'</span>, fontsize<span class="op">=</span><span class="dv">9</span>, weight<span class="op">=</span><span class="st">'bold'</span>, pad<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb66-60"><a href="#cb66-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-61"><a href="#cb66-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-62"><a href="#cb66-62" aria-hidden="true" tabindex="-1"></a>legend_locs<span class="op">=</span><span class="bu">dict</span>()</span>
<span id="cb66-63"><a href="#cb66-63" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, ax <span class="kw">in</span> <span class="bu">enumerate</span>([axA, axB, axC, axD]):</span>
<span id="cb66-64"><a href="#cb66-64" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="kw">not</span> axis_has_legend(ax):</span>
<span id="cb66-65"><a href="#cb66-65" aria-hidden="true" tabindex="-1"></a>        <span class="cf">continue</span></span>
<span id="cb66-66"><a href="#cb66-66" aria-hidden="true" tabindex="-1"></a>    loc <span class="op">=</span> legend_locs.get(i, <span class="st">'best'</span>)</span>
<span id="cb66-67"><a href="#cb66-67" aria-hidden="true" tabindex="-1"></a>    handles, labels <span class="op">=</span> ax.get_legend_handles_labels()</span>
<span id="cb66-68"><a href="#cb66-68" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> h_idx, handle <span class="kw">in</span> <span class="bu">enumerate</span>(handles):</span>
<span id="cb66-69"><a href="#cb66-69" aria-hidden="true" tabindex="-1"></a>        handle.set_marker(<span class="st">'o'</span>)</span>
<span id="cb66-70"><a href="#cb66-70" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> h_idx <span class="op">&gt;=</span><span class="dv">5</span>:</span>
<span id="cb66-71"><a href="#cb66-71" aria-hidden="true" tabindex="-1"></a>            handle.set_color(<span class="st">'black'</span>)</span>
<span id="cb66-72"><a href="#cb66-72" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-73"><a href="#cb66-73" aria-hidden="true" tabindex="-1"></a>    handles <span class="op">=</span> handles[<span class="dv">1</span>:<span class="dv">3</span>] <span class="op">+</span> handles[<span class="dv">4</span>:<span class="dv">6</span>]</span>
<span id="cb66-74"><a href="#cb66-74" aria-hidden="true" tabindex="-1"></a>    labels <span class="op">=</span> labels[<span class="dv">1</span>:<span class="dv">3</span>] <span class="op">+</span> labels[<span class="dv">4</span>:<span class="dv">6</span>]</span>
<span id="cb66-75"><a href="#cb66-75" aria-hidden="true" tabindex="-1"></a>    legend <span class="op">=</span> ax.legend(ncols<span class="op">=</span><span class="dv">1</span>, loc<span class="op">=</span>loc, handles<span class="op">=</span>handles, labels<span class="op">=</span>labels)</span>
<span id="cb66-76"><a href="#cb66-76" aria-hidden="true" tabindex="-1"></a>    legend.set_title(<span class="va">None</span>)</span>
<span id="cb66-77"><a href="#cb66-77" aria-hidden="true" tabindex="-1"></a>    legend.get_frame().<span class="bu">set</span>(</span>
<span id="cb66-78"><a href="#cb66-78" aria-hidden="true" tabindex="-1"></a>        alpha<span class="op">=</span><span class="fl">0.75</span>, boxstyle<span class="op">=</span><span class="st">'square'</span>, edgecolor<span class="op">=</span><span class="st">'grey'</span>, linewidth<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb66-79"><a href="#cb66-79" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> text <span class="kw">in</span> legend.get_texts():</span>
<span id="cb66-80"><a href="#cb66-80" aria-hidden="true" tabindex="-1"></a>        t <span class="op">=</span> text.get_text()</span>
<span id="cb66-81"><a href="#cb66-81" aria-hidden="true" tabindex="-1"></a>        new_t <span class="op">=</span> update_legend_values.get(t,t)</span>
<span id="cb66-82"><a href="#cb66-82" aria-hidden="true" tabindex="-1"></a>        text.set_text(new_t)</span>
<span id="cb66-83"><a href="#cb66-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-84"><a href="#cb66-84" aria-hidden="true" tabindex="-1"></a>x_pos <span class="op">=</span> <span class="fl">0.5275</span></span>
<span id="cb66-85"><a href="#cb66-85" aria-hidden="true" tabindex="-1"></a>y_pos <span class="op">=</span> <span class="fl">0.5</span></span>
<span id="cb66-86"><a href="#cb66-86" aria-hidden="true" tabindex="-1"></a>perc_of_figure<span class="op">=</span><span class="fl">0.95</span></span>
<span id="cb66-87"><a href="#cb66-87" aria-hidden="true" tabindex="-1"></a>fig.add_artist(mpl.lines.Line2D([(<span class="dv">1</span><span class="op">-</span>perc_of_figure)<span class="op">/</span><span class="dv">2</span>, <span class="dv">1</span> <span class="op">-</span> (<span class="dv">1</span><span class="op">-</span>perc_of_figure)<span class="op">/</span><span class="dv">2</span>], [y_pos, y_pos], color<span class="op">=</span><span class="st">'grey'</span>, alpha<span class="op">=</span><span class="dv">1</span>, linewidth<span class="op">=</span><span class="dv">1</span>))</span>
<span id="cb66-88"><a href="#cb66-88" aria-hidden="true" tabindex="-1"></a>fig.add_artist(mpl.lines.Line2D([x_pos, x_pos], [(<span class="dv">1</span><span class="op">-</span>perc_of_figure)<span class="op">/</span><span class="dv">2</span>, <span class="dv">1</span> <span class="op">-</span> (<span class="dv">1</span><span class="op">-</span>perc_of_figure)<span class="op">/</span><span class="dv">2</span>], color<span class="op">=</span><span class="st">'grey'</span>, alpha<span class="op">=</span><span class="dv">1</span>, linewidth<span class="op">=</span><span class="dv">1</span>))</span>
<span id="cb66-89"><a href="#cb66-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-90"><a href="#cb66-90" aria-hidden="true" tabindex="-1"></a><span class="co"># fig.suptitle('Phasing accuracy stratified by presence in Platinum repeat regions', fontsize=11, weight='bold', y=1.02)</span></span>
<span id="cb66-91"><a href="#cb66-91" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-92"><a href="#cb66-92" aria-hidden="true" tabindex="-1"></a>add_letter_to_ax(axA, <span class="st">'a'</span>)</span>
<span id="cb66-93"><a href="#cb66-93" aria-hidden="true" tabindex="-1"></a>add_letter_to_ax(axB, <span class="st">'b'</span>)</span>
<span id="cb66-94"><a href="#cb66-94" aria-hidden="true" tabindex="-1"></a>add_letter_to_ax(axC, <span class="st">'c'</span>)</span>
<span id="cb66-95"><a href="#cb66-95" aria-hidden="true" tabindex="-1"></a>add_letter_to_ax(axD, <span class="st">'d'</span>)</span>
<span id="cb66-96"><a href="#cb66-96" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-97"><a href="#cb66-97" aria-hidden="true" tabindex="-1"></a><span class="co"># plt.savefig('figures/Figure 2.png', facecolor='white')</span></span>
<span id="cb66-98"><a href="#cb66-98" aria-hidden="true" tabindex="-1"></a><span class="co"># plt.savefig('figures/Figure 2.svg', facecolor='white')</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="make_plots_files/figure-html/cell-41-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>